{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from operator import itemgetter\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import spacy\n",
    "nlp = spacy.load('en_core_web_lg')\n",
    "stopwords = nlp.Defaults.stop_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "#os.chdir(\"structural_optimization\")\n",
    "#os.chdir(\"smo\")\n",
    "#os.chdir(\"anja3\")\n",
    "#os.chdir(\"driver_glance_behaviour\")\n",
    "os.chdir(\"cov\")\n",
    "lvl = 0.5\n",
    "\n",
    "work_dict = pickle.load(open(\"work_dict_lvl{}.p\".format(lvl), \"rb\"))\n",
    "#work_dict = pickle.load(open(\"work_dict_test_lvl1.p\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from requestWorksV3 import *\n",
    "import json\n",
    "\n",
    "db_engine = create_engine('sqlite:///COV.sqllite', echo=False)\n",
    "Session.configure(bind=db_engine)\n",
    "db_session = Session()\n",
    "\n",
    "work_dict = {}\n",
    "\n",
    "for res in db_session.query(CrossRefWork).yield_per(1000):\n",
    "    #print(work)\n",
    "    work = json.loads(res.json_content)\n",
    "    doi = res.doi\n",
    "    \n",
    "    work_dict[doi] = work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "292731"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#work_dict = pickle.load(open(\"work_dict_lvl1.p\", \"rb\"))\n",
    "#work_dict = pickle.load(open(\"work_dict_lvl2.p\", \"rb\"))\n",
    "#work_dict = pickle.load(open(\"crashworthiness/work_dict_lvl3.p\", \"rb\"))\n",
    "len(work_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gensim.models import Word2Vec, KeyedVectors\n",
    "import logging\n",
    "logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reference_dois(work):\n",
    "    result = []\n",
    "    if \"reference\" in work:\n",
    "        for ref in work[\"reference\"]:\n",
    "            if \"DOI\" in ref:\n",
    "                result.append(ref[\"DOI\"])\n",
    "        #if \"DOI\" in work:\n",
    "            #result.append(work[\"DOI\"])\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "issn_counts = defaultdict(int)\n",
    "issn2journalTitle = {}\n",
    "\n",
    "for work in work_dict.values():\n",
    "    if \"ISSN\" in work:\n",
    "        for issn in work[\"ISSN\"]:\n",
    "            issn_counts[issn] += 1\n",
    "            if not issn in issn2journalTitle:\n",
    "                if \"container-title\" in work:\n",
    "                    issn2journalTitle[issn] = \";\".join(work[\"container-title\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#pd.DataFrame(index=issn_counts.keys(), data=issn_counts)\n",
    "#pd.DataFrame(index=issn_counts.keys(), data={count: issn for issn, count in issn_counts.items()})\n",
    "df = pd.DataFrame(index=issn_counts.keys(), columns=[\"Titel\", \"Publikationen\"])\n",
    "for issn, count in sorted(issn_counts.items(), key=itemgetter(1), reverse=True):\n",
    "    df.at[issn, \"Titel\"] = issn2journalTitle[issn]\n",
    "    df.at[issn, \"Publikationen\"] = count\n",
    "\n",
    "df.to_excel(\"journal_list.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#jt = set()\n",
    "#f = open(\"journal_list.csv\", \"w\", encoding=\"utf-8\")\n",
    "#f.write(\"Count\\tJournalTitle\\tISSN\\n\")\n",
    "#for issn, count in sorted(issn_counts.items(), key=itemgetter(1), reverse=True):\n",
    "#    if not (count, issn2journalTitle[issn]) in jt:\n",
    "#        print(count, issn, issn2journalTitle[issn])\n",
    "#        jt.add((count, issn2journalTitle[issn]))\n",
    "#        f.write(\"{}\\t{}\\t{}\\n\".format(count, issn2journalTitle[issn], issn))\n",
    "#f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "158370"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_list = []\n",
    "ref_counts = defaultdict(int)\n",
    "for work in work_dict.values():\n",
    "    #print(work)\n",
    "    work_refs = get_reference_dois(work)\n",
    "    #print(work_refs)\n",
    "    #break\n",
    "    for doi in work_refs:\n",
    "        ref_counts[doi] += 1\n",
    "    work_refs.append(work[\"DOI\"])\n",
    "    if len(work_refs) > 2:\n",
    "        references_list.append(work_refs)\n",
    "len(references_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_counts_by_year = defaultdict(list)\n",
    "for doi, count in ref_counts.items():\n",
    "    #work = work_dict[doi]\n",
    "    try:        \n",
    "        y = work_dict[doi][\"issued\"][\"date-parts\"][0][0]\n",
    "        if y:\n",
    "            ref_counts_by_year[y].append(count)\n",
    "    except KeyError:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([(2020, 2.72962962962963, 540, 6.285204494580043, 0.36635006784260515),\n",
       "  (2019, 1.3628203272042856, 6907, 3.138009769849855, 0.7337724423669394),\n",
       "  (2018, 1.7233879222108495, 11724, 3.9682473391286837, 0.5802524127691165),\n",
       "  (2017, 2.100274254370929, 11668, 4.8360601893136845, 0.4761282951114013),\n",
       "  (2016, 2.450482879719052, 11390, 5.642445349478209, 0.4080828347246605),\n",
       "  (2015, 2.8615443498229833, 10451, 6.588949362843739, 0.3494616464923427),\n",
       "  (2014, 3.2079062957540265, 9562, 7.386477216324969, 0.3117298037425833),\n",
       "  (2013, 3.3266011448320554, 9259, 7.659782206427216, 0.30060712314535243),\n",
       "  (2012, 3.5453888755116783, 8306, 8.16355957362011, 0.28205650638413476),\n",
       "  (2011, 3.6874177415109237, 7598, 8.490593123244825, 0.27119249027376235),\n",
       "  (2010, 3.9809588116137746, 7405, 9.166496415445168, 0.25119576647783165),\n",
       "  (2009, 4.141344778254649, 6990, 9.535798751357888, 0.24146745889180599),\n",
       "  (2008, 4.189863726841219, 6531, 9.647517759101065, 0.23867124689372898),\n",
       "  (2007, 5.257680250783699, 3190, 12.10625616918374, 0.19019794896255665),\n",
       "  (2006, 5.722778891115564, 2994, 13.177185365183691, 0.17474028247928097),\n",
       "  (2005, 5.863930064614215, 2631, 13.502197953140302, 0.17053409385532797),\n",
       "  (2004, 5.60250521920668, 2395, 12.90024500116664, 0.17849157847667313),\n",
       "  (2003, 6.153067301965455, 1679, 14.16796104569475, 0.16252056916077823),\n",
       "  (2002, 5.955405405405405, 1480, 13.712827709222646, 0.16791468118901748),\n",
       "  (2001, 6.5807609475951185, 1393, 15.15276205848989, 0.15195811061415948),\n",
       "  (2000, 6.267657992565056, 1345, 14.43181586166528, 0.15954922894424675),\n",
       "  (1999, 6.563106796116505, 1030, 15.112111872465775, 0.15236686390532544),\n",
       "  (1998, 6.77928870292887, 956, 15.609889108466955, 0.14750810060175898),\n",
       "  (1997, 6.900124843945068, 801, 15.88812460546578, 0.14492491405825944),\n",
       "  (1996, 5.521472392638037, 815, 12.713660022666508, 0.1811111111111111),\n",
       "  (1995, 6.175641025641026, 780, 14.219938965323482, 0.16192651027610547),\n",
       "  (1994, 5.757823129251701, 735, 13.257877705511294, 0.17367674858223062),\n",
       "  (1993, 6.479846449136277, 521, 14.920397838671589, 0.15432464454976302),\n",
       "  (1992, 6.932504440497335, 563, 15.96268138180419, 0.1442480143479375),\n",
       "  (1991, 6.813278008298755, 482, 15.688152376332875, 0.14677222898903775),\n",
       "  (1990, 7.197777777777778, 450, 16.573495813794917, 0.1389317690645261),\n",
       "  (1989, 7.002358490566038, 424, 16.123526276177643, 0.14280902660828562),\n",
       "  (1988, 7.20682302771855, 469, 16.594323271470945, 0.13875739644970414),\n",
       "  (1987, 5.826666666666667, 375, 13.416395808511972, 0.17162471395881007),\n",
       "  (1986, 8.36196319018405, 326, 19.254131789882724, 0.11958914159941306),\n",
       "  (1985, 6.783536585365853, 328, 15.619670219243142, 0.14741573033707867),\n",
       "  (1984, 5.86896551724138, 290, 13.513792511296089, 0.17038777908343125),\n",
       "  (1983, 7.640316205533597, 253, 17.59247820062249, 0.13088463528194516),\n",
       "  (1982, 6.257575757575758, 264, 14.40860065767486, 0.15980629539951574),\n",
       "  (1981, 7.3875, 240, 17.01034737449351, 0.1353637901861252),\n",
       "  (1980, 6.016260162601626, 246, 13.852950965980437, 0.1662162162162162),\n",
       "  (1979, 8.246305418719212, 203, 18.987819929418876, 0.12126642771804062),\n",
       "  (1978, 5.0, 190, 11.512925464970227, 0.2),\n",
       "  (1977, 10.398907103825136, 183, 23.944368480697644, 0.09616395165528113),\n",
       "  (1976, 5.093167701863354, 161, 11.7274520264293, 0.19634146341463415),\n",
       "  (1975, 4.509803921568627, 153, 10.384207282130008, 0.2217391304347826),\n",
       "  (1974, 5.056737588652482, 141, 11.643568590813862, 0.19775596072931276),\n",
       "  (1973, 9.315384615384616, 130, 21.449465750890685, 0.10734929810074319),\n",
       "  (1972, 6.684210526315789, 133, 15.390963516328618, 0.14960629921259844),\n",
       "  (1971, 6.619834710743802, 121, 15.242732723043225, 0.15106117353308365),\n",
       "  (1970, 8.0, 122, 18.420680743952364, 0.125),\n",
       "  (1969, 5.631578947368421, 114, 12.967189734229626, 0.17757009345794392),\n",
       "  (1968, 6.247706422018348, 109, 14.385875672742614, 0.16005873715124816),\n",
       "  (1967, 5.182926829268292, 82, 11.934130055152066, 0.19294117647058823),\n",
       "  (1966, 4.9879518072289155, 83, 11.48518347589801, 0.20048309178743962),\n",
       "  (1965, 7.048192771084337, 83, 16.229063607247188, 0.14188034188034188),\n",
       "  (1964, 6.052631578947368, 76, 13.936699247069223, 0.16521739130434782),\n",
       "  (1963, 8.485294117647058, 68, 19.53811174496418, 0.11785095320623917),\n",
       "  (1962, 8.018181818181818, 55, 18.462545927461345, 0.12471655328798185),\n",
       "  (1961, 5.923076923076923, 65, 13.638388627733962, 0.16883116883116883),\n",
       "  (1960, 14.288461538461538, 52, 32.90039854028031, 0.06998654104979811),\n",
       "  (1959, 7.098360655737705, 61, 16.344579430597076, 0.14087759815242495),\n",
       "  (1958, 7.051724137931035, 58, 16.237194879906287, 0.14180929095354522),\n",
       "  (1957, 6.555555555555555, 45, 15.09472449851652, 0.15254237288135594),\n",
       "  (1956, 11.959183673469388, 49, 27.537038050908382, 0.08361774744027303),\n",
       "  (1955, 6.780487804878049, 41, 15.612650142740113, 0.1474820143884892),\n",
       "  (1954, 10.393939393939394, 33, 23.93292990596841, 0.09620991253644315),\n",
       "  (1953, 5.935483870967742, 31, 13.666956680996915, 0.16847826086956522),\n",
       "  (1952, 9.71875, 32, 22.37824887253588, 0.10289389067524116),\n",
       "  (1951, 6.0, 24, 13.815510557964274, 0.16666666666666666),\n",
       "  (1950, 3.909090909090909, 22, 9.001014454431267, 0.2558139534883721),\n",
       "  (1949, 5.583333333333333, 24, 12.856100102550087, 0.1791044776119403),\n",
       "  (1948, 14.454545454545455, 22, 33.2828208896412, 0.06918238993710692)],\n",
       " 0.1485571999071787)"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_ref_count = defaultdict(int)\n",
    "median_ref_count = defaultdict(int)\n",
    "var_ref_count = defaultdict(int)\n",
    "n_ref_count = defaultdict(int)\n",
    "lambda_mle_ref_count = defaultdict(int)\n",
    "percentile_ref_count = defaultdict(int)\n",
    "for year, counts in ref_counts_by_year.items():\n",
    "    if len(counts) > 20:\n",
    "        avg_ref_count[year] = np.average(counts)\n",
    "        median_ref_count[year] = np.median(counts)\n",
    "        var_ref_count[year] = np.var(counts)\n",
    "        n_ref_count[year] = len(counts)\n",
    "        lambda_mle_ref_count[year] = len(counts) / sum(counts)\n",
    "        percentile_ref_count[year] = - np.log(0.1) / lambda_mle_ref_count[year]\n",
    "        \n",
    "#np.average(ref_counts_by_year[2017])\n",
    "\n",
    "sorted(avg_ref_count.items(), reverse=True)\n",
    "sorted(median_ref_count.items(), reverse=True)\n",
    "sorted(var_ref_count.items(), reverse=True)\n",
    "\n",
    "lambda_mle_ref_count_default = np.median([l for y, l in lambda_mle_ref_count.items() if y < 2000])\n",
    "\n",
    "x = sorted(zip(avg_ref_count.keys(), avg_ref_count.values(), median_ref_count.values(), var_ref_count.values(), n_ref_count.values()), reverse=True)\n",
    "#pprint(x)\n",
    "x = sorted(zip(avg_ref_count.keys(), avg_ref_count.values(), n_ref_count.values(), percentile_ref_count.values(), lambda_mle_ref_count.values()), reverse=True)\n",
    "#print(x)\n",
    "x, lambda_mle_ref_count_default\n",
    "\n",
    "\n",
    "#lambda_mle_ref_count_default"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3114814, 139496)"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_count_normalized = defaultdict(float)\n",
    "\n",
    "for doi, count in ref_counts.items():\n",
    "    try:\n",
    "        year = work_dict[doi][\"issued\"][\"date-parts\"][0][0]\n",
    "        if year in lambda_mle_ref_count:\n",
    "            if year >= 2000:\n",
    "                l = lambda_mle_ref_count[year]\n",
    "            else:\n",
    "                l = lambda_mle_ref_count_default\n",
    "        else:\n",
    "            l = lambda_mle_ref_count_default\n",
    "        if l < lambda_mle_ref_count_default:\n",
    "            l = lambda_mle_ref_count_default\n",
    "        ref_count_normalized[doi] = l*count\n",
    "    except KeyError:\n",
    "        continue\n",
    "\n",
    "len(ref_counts), len(ref_count_normalized)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 16:13:18,262 : INFO : loading Word2VecKeyedVectors object from vectors_db.rv\n",
      "2020-03-18 16:13:18,481 : INFO : loading vectors from vectors_db.rv.vectors.npy with mmap=None\n",
      "2020-03-18 16:13:18,965 : INFO : setting ignored attribute vectors_norm to None\n",
      "2020-03-18 16:13:18,965 : INFO : loaded vectors_db.rv\n"
     ]
    }
   ],
   "source": [
    "#model = Word2Vec(iter=1, size=300, window=1000, min_count=5, workers=4, sg=1)  # an empty model, no training yet\n",
    "#model.build_vocab(references_list)\n",
    "#model.train(references_list)\n",
    "\n",
    "load_model = True\n",
    "\n",
    "fname = \"vectors_db.rv\"\n",
    "#fname = \"testRVnew.rv\"\n",
    "\n",
    "\n",
    "#load_model = True\n",
    "\n",
    "#fname = \"vectors_db.rv\"\n",
    "\n",
    "if not load_model or not fname in os.listdir():\n",
    "    model = Word2Vec(references_list, size=300, window=1000, min_count=3, workers=16, sg=1, iter=20)\n",
    "    \n",
    "    reference_vectors = model.wv\n",
    "    reference_vectors.save(fname)\n",
    "    del model\n",
    "else:\n",
    "    reference_vectors = KeyedVectors.load(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164535"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reference_vectors.vocab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reference_vectors = model.wv\n",
    "#del model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fname = \"vectors.rv\"\n",
    "#reference_vectors.save(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of X_dict: 164535\n"
     ]
    }
   ],
   "source": [
    "X_dict = {}\n",
    "\n",
    "for doi in reference_vectors.vocab.keys():\n",
    "    newX = {}\n",
    "    newX[\"lvl\"] = 0\n",
    "    newX[\"vec\"] = reference_vectors[doi]\n",
    "    X_dict[doi] = newX\n",
    "print(\"Size of X_dict:\", len(X_dict))\n",
    "if 0:\n",
    "    for lvl in range(1,2):\n",
    "        print (\"Level:\", lvl)\n",
    "        dois = set(work_dict.keys() - X_dict.keys())\n",
    "        print(\"No. DOIs:\", len(dois), \"/\", len(work_dict))\n",
    "        for doi in dois:\n",
    "            work = work_dict[doi]\n",
    "        #for doi, work in work_dict.items():\n",
    "            if not doi in X_dict:\n",
    "                ref_dois = get_reference_dois(work)\n",
    "                ref_dois = set(ref_dois) & X_dict.keys()            \n",
    "                if len(ref_dois) >= 3:\n",
    "                    vec = np.zeros(300)\n",
    "                    for ref_doi in ref_dois:\n",
    "                        #print(work_dict[ref_doi])\n",
    "                        vec = vec + X_dict[ref_doi][\"vec\"]\n",
    "                    vec = vec / len(ref_dois)\n",
    "                    newX = {}\n",
    "                    newX[\"lvl\"] = lvl\n",
    "                    newX[\"vec\"] = vec\n",
    "                    X_dict[doi] = newX\n",
    "                pass\n",
    "        print(\"Size of X_dict:\", len(X_dict))\n",
    "\n",
    "    \n",
    "    work_dict0 = pickle.load(open(\"work_dict_lvl0.p\", \"rb\"))\n",
    "    len(X_dict), len(work_dict0), len(work_dict), len(X_dict.keys() & work_dict.keys()), len(X_dict.keys() & work_dict0.keys()), len(reference_vectors.vocab.keys() & work_dict0.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.manifold import TSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164535"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reference_vectors.vocab.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31261 possible\n",
      "['A pneumonia outbreak associated with a new coronavirus of probable bat origin']\n",
      "2303 selected\n",
      "2303 selected\n"
     ]
    }
   ],
   "source": [
    "dois_to_plot = list(reference_vectors.vocab.keys() & work_dict.keys())\n",
    "#dois_to_plot = list(reference_vectors.vocab.keys())\n",
    "print(len(dois_to_plot), \"possible\")\n",
    "\n",
    "def doi_plot_filter(doi):\n",
    "    if not \"title\" in work_dict[doi]:\n",
    "        return False\n",
    "    if len(work_dict[doi][\"title\"]) == 0:\n",
    "        return False\n",
    "    if not work_dict[doi][\"issued\"][\"date-parts\"][0][0]:\n",
    "        return False\n",
    "    if work_dict[doi][\"issued\"][\"date-parts\"][0][0] >= 2014:\n",
    "        return True\n",
    "    elif ref_counts[doi] >= 20:\n",
    "        return True\n",
    "    elif work_dict[doi][\"issued\"][\"date-parts\"][0][0] >= 2008 and ref_counts[doi] >= 10:\n",
    "        return True\n",
    "    elif work_dict[doi][\"issued\"][\"date-parts\"][0][0] >= 2012 and ref_counts[doi] >= 5:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1065/lca2008.02.376\",\n",
    "    \"10.1016/j.cirp.2010.05.004\",\n",
    "    #\"10.1007/s11367-014-0788-0\",\n",
    "    \"10.1007/978-3-642-01421-5\",\n",
    "    \"10.1108/09564230310474138\",\n",
    "    \"10.1080/17517575.2012.683812\",\n",
    "    \"10.1002/bse.414\",\n",
    "    \"10.1111/j.1468-2370.2007.00202.x\"\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1371/journal.pone.0031092\",\n",
    "    \"10.1093/nar/gks017\",\n",
    "    \"10.1016/j.chembiol.2012.01.015\",\n",
    "    \"10.1111/j.1462-2920.2012.02719.x\",\n",
    "    \"10.1021/ja3072397\",\n",
    "    \"10.1371/journal.ppat.1002760\",\n",
    "    \"10.1111/1462-2920.12024\",\n",
    "    \"10.1371/journal.ppat.1004744\",\n",
    "    \"10.1186/1475-2859-13-88\",\n",
    "    \"10.1016/j.mimet.2011.11.015\",\n",
    "    \"10.1099/jmm.0.05069-0\"\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1007/s001580050176\", # A 99 line topology optimization code written in Matlab\n",
    "    \"10.1038/451652a\" # Building better batteries\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1038/nature14539\", # Deep learning\n",
    "    \"10.1038/nature16961\",\n",
    "    \"10.1038/nature21056\"\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1016/j.tcs.2014.11.028\",\n",
    "    \"10.1007/978-3-662-05094-1\",\n",
    "    \"10.1038/nature14422\",\n",
    "    \"10.1126/science.1165893\",\n",
    "    \"10.1109/4235.996017\",\n",
    "    \"10.1007/978-3-642-17339-4\"\n",
    "]\n",
    "\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1016/j.addma.2016.06.010\", # Topology optimization of 3D self-supporting structures for additive manufacturing\n",
    "    \"10.1007/s00158-015-1261-9\" #Topology and shape optimization methods using evolutionary algorithms: a review\n",
    "]\n",
    "\n",
    "#test_dois = [\n",
    "#    \"10.1016/j.actamat.2016.07.019\" # Additive manufacturing of metals\n",
    "#]\n",
    "\n",
    "#test_dois = [\n",
    "#    \"10.1115/1.4004465\", # Multidisciplinary Design Optimization for Complex Engineered Systems: Report From a National Science Foundation Workshop\n",
    "#    \"10.1007/s00158-012-0763-y\", # Extensions to the design structure matrix for the description of multidisciplinary design, analysis, and optimization processes\n",
    "#    \"10.1016/j.paerosci.2011.05.001\"\n",
    "#    \n",
    "#]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1016/j.cad.2015.04.001\",\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1561/1500000011\",\n",
    "    \"10.1145/1014052.1014073\",\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1016/j.trf.2014.06.016\"   # Effects of adaptive cruise control and highly automated driving on workload and situation awareness: A review of the empirical evidence\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1126/sciadv.1602614\", #Data-driven discovery of partial differential equations\n",
    "    #\"10.1146/annurev-fluid-010816-060042\",\n",
    "    \"10.1063/1.4927765\",\n",
    "    \"10.1017/jfm.2016.615\"\n",
    "]\n",
    "\n",
    "test_dois = [\n",
    "    \"10.1038/s41586-020-2012-7\" #cov\n",
    "]\n",
    "\n",
    "for doi in test_dois:\n",
    "    if doi in work_dict:\n",
    "        print(work_dict[doi][\"title\"])\n",
    "\n",
    "def doi_plot_filter2(doi, min_sim=0.35):\n",
    "    \n",
    "    \n",
    "    #min_sim = 0.35\n",
    "    \n",
    "    for test_doi in test_dois:\n",
    "        if reference_vectors.similarity(doi, test_doi) > min_sim:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "#dois_to_plot = [doi for doi, count in sorted(ref_count_normalized.items(), key=itemgetter(1), reverse=True)][:3000]\n",
    "        \n",
    "#dois_to_plot = [doi for doi in dois_to_plot if doi_plot_filter(doi)]\n",
    "dois_to_plot = [doi for doi in dois_to_plot if doi_plot_filter2(doi, 0.485)]\n",
    "\n",
    "dois_to_plot = [doi for doi in dois_to_plot if \"title\" in work_dict[doi]]\n",
    "dois_to_plot = [doi for doi in dois_to_plot if len(work_dict[doi][\"title\"]) > 0]\n",
    "dois_to_plot = [doi for doi in dois_to_plot if work_dict[doi][\"issued\"][\"date-parts\"][0][0]]\n",
    "dois_to_plot = [doi for doi in dois_to_plot if work_dict[doi][\"issued\"][\"date-parts\"][0][0] >= 2008]\n",
    "#dois_to_plot = [doi for doi in dois_to_plot if ref_counts[doi] >= 4]\n",
    "\n",
    "#filter_issn = \"2071-1050\" # Sustainability\n",
    "#filter_issn = \"1615-1488\" # SMO\n",
    "#filter_issn = \"0268-3768\"\n",
    "#dois_to_plot = [doi for doi in dois_to_plot if \"ISSN\" in work_dict[doi]]\n",
    "#dois_to_plot = [doi for doi in dois_to_plot if filter_issn in work_dict[doi][\"ISSN\"]]\n",
    "\n",
    "#test_dois = [\"10.1016/j.actamat.2016.02.014\"] # Laser powder-bed fusion additive manufacturing: Physics of complex melt flow and formation mechanisms of pores, spatter, and denudation zones\n",
    "\n",
    "#dois_to_plot = [doi for doi in dois_to_plot if doi_plot_filter2(doi, 0.8)]\n",
    "\n",
    "print(len(dois_to_plot), \"selected\")\n",
    "\n",
    "dois_to_plot = [doi for doi in dois_to_plot if doi in ref_count_normalized]\n",
    "#dois_to_plot = [doi for doi in dois_to_plot if ref_count_normalized[doi] >= 1.5]\n",
    "\n",
    "#dois_to_plot = [doi for doi, count in sorted([doi, ref_count_normalized[doi] for doi in dois_to_plot], key=itemgetter(1), reverse=True)][:3000]\n",
    "\n",
    "a = [(doi, ref_count_normalized[doi]) for doi in dois_to_plot]\n",
    "b = sorted(a, key=itemgetter(1), reverse=True)[:2500]\n",
    "dois_to_plot = [doi for doi, count in b]\n",
    "\n",
    "print(len(dois_to_plot), \"selected\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = reference_vectors[dois_to_plot]\n",
    "\n",
    "tsne = TSNE(n_components=2, metric=\"cosine\")\n",
    "X_embedded = tsne.fit_transform(X)\n",
    "#X_embedded = TSNE(n_components=2).fit_transform(X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 0:\n",
    "    dois_to_plot = list(X_dict.keys() & work_dict0.keys())\n",
    "    print(len(dois_to_plot))\n",
    "    X = np.zeros((len(dois_to_plot), 300))\n",
    "    for i, doi in enumerate(dois_to_plot):\n",
    "        X[i, :] = X_dict[doi][\"vec\"]\n",
    "\n",
    "    X_embedded = TSNE(n_components=2).fit_transform(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "C_embedded = TSNE(n_components=3, metric=\"cosine\").fit_transform(X)\n",
    "#C_embedded = TSNE(n_components=3).fit_transform(X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "#X_embedded *= 3.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-18.861486, 21.46803)"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "#colors_r = np.floor(MinMaxScaler().fit_transform(C_embedded[:,0])*256)\n",
    "\n",
    "cmin = np.min(C_embedded)\n",
    "cmax = np.max(C_embedded)\n",
    "cmin, cmax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[107.,  95.,   4.],\n",
       "       [148., 126., 170.],\n",
       "       [115.,  83.,   2.],\n",
       "       ...,\n",
       "       [ 64.,  42.,  94.],\n",
       "       [104.,  46.,  25.],\n",
       "       [100.,  69., 120.]], dtype=float32)"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colors = (C_embedded - cmin) / (cmax - cmin) * 256\n",
    "colors = np.floor(colors)\n",
    "colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-26.923471,  31.237288], dtype=float32)"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_embedded[0,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\"doi\", \"title\", \"year\", \"link\", \"authors\", \"containerTitles\", \"refCount\", \"xval\", \"yval\", \"rgb\", \"clusterNo\"]\n",
    "df = pd.DataFrame(index=dois_to_plot, columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from scipy.cluster.hierarchy import dendrogram, linkage, fcluster\n",
    "#\n",
    "#max_d = 0.2\n",
    "#\n",
    "#Z = linkage(X, method='average', metric='cosine')\n",
    "##Z = linkage(X, method='complete', metric='cosine')\n",
    "#clLabels = fcluster(Z, max_d, criterion='distance') - 1\n",
    "##clLabels = fcluster(Z, 50, criterion='maxclust') - 1\n",
    "#\n",
    "#cluster_ids = sorted(np.unique(clLabels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from sklearn.cluster import DBSCAN\n",
    "#\n",
    "#clLabels = DBSCAN(eps=0.1, min_samples=5, metric=\"cosine\").fit_predict(X)\n",
    "#cluster_ids = sorted(np.unique(clLabels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int(np.sqrt(len(dois_to_plot)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47\n",
      "Initialization 0\n",
      "Initialization converged: True\n"
     ]
    }
   ],
   "source": [
    "from sklearn.mixture import GaussianMixture\n",
    "\n",
    "n = int(np.sqrt(len(dois_to_plot)))\n",
    "print(n)\n",
    "\n",
    "gmm = GaussianMixture(n_components=n, verbose=True)\n",
    "\n",
    "clLabels = gmm.fit_predict(X)\n",
    "cluster_ids = sorted(np.unique(clLabels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 73),\n",
       " (1, 46),\n",
       " (2, 104),\n",
       " (3, 51),\n",
       " (4, 18),\n",
       " (5, 21),\n",
       " (6, 87),\n",
       " (7, 132),\n",
       " (8, 19),\n",
       " (9, 88),\n",
       " (10, 22),\n",
       " (11, 101),\n",
       " (12, 31),\n",
       " (13, 18),\n",
       " (14, 48),\n",
       " (15, 37),\n",
       " (16, 39),\n",
       " (17, 65),\n",
       " (18, 46),\n",
       " (19, 46),\n",
       " (20, 92),\n",
       " (21, 53),\n",
       " (22, 10),\n",
       " (23, 86),\n",
       " (24, 61),\n",
       " (25, 55),\n",
       " (26, 62),\n",
       " (27, 54),\n",
       " (28, 15),\n",
       " (29, 61),\n",
       " (30, 16),\n",
       " (31, 25),\n",
       " (32, 18),\n",
       " (33, 28),\n",
       " (34, 17),\n",
       " (35, 61),\n",
       " (36, 44),\n",
       " (37, 63),\n",
       " (38, 45),\n",
       " (39, 81),\n",
       " (40, 47),\n",
       " (41, 7),\n",
       " (42, 17),\n",
       " (43, 70),\n",
       " (44, 41),\n",
       " (45, 72),\n",
       " (46, 10)]"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(cl, np.sum(clLabels==cl)) for cl in cluster_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([35,  9, 35, ..., 43, 43,  6], dtype=int64), 46)"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clLabels, np.max(cluster_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gensim.models.phrases import *\n",
    "from pprint import pprint\n",
    "import operator\n",
    "\n",
    "def phrase_lines(texts_temp):\n",
    "\n",
    "    phrases = Phrases(texts_temp, min_count=10, threshold=10)\n",
    "    bigram = Phraser(phrases)\n",
    "    texts_phrased = [bigram[t] for t in texts_temp]\n",
    "\n",
    "    phrase_scores = {}\n",
    "    for phrase, score in phrases.export_phrases(texts_temp):\n",
    "        phrase_scores[phrase] = score\n",
    "    print(\"Bigramme\")\n",
    "    pprint(sorted(phrase_scores.items(), key=operator.itemgetter(1), reverse=True)[:50])\n",
    "\n",
    "    texts_temp = texts_phrased\n",
    "    phrases = Phrases(texts_temp, min_count=5, threshold=7)\n",
    "    trigram = Phraser(phrases)\n",
    "    texts_phrased = [trigram[t] for t in texts_temp]\n",
    "\n",
    "    phrase_scores = {}\n",
    "    for phrase, score in phrases.export_phrases(texts_temp):\n",
    "        phrase_scores[phrase] = score\n",
    "    print(\"Trigramme\")\n",
    "    pprint(sorted(phrase_scores.items(), key=operator.itemgetter(1), reverse=True)[:50])\n",
    "\n",
    "    return texts_phrased, bigram, trigram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 16:16:24,342 : INFO : collecting all words and their counts\n",
      "2020-03-18 16:16:24,342 : INFO : PROGRESS: at sentence #0, processed 0 words and 0 word types\n",
      "2020-03-18 16:16:24,373 : INFO : collected 21469 word types from a corpus of 20026 words (unigram + bigrams) and 2303 sentences\n",
      "2020-03-18 16:16:24,373 : INFO : using 21469 counts as vocab in Phrases<0 vocab, min_count=10, threshold=10, max_vocab_size=40000000>\n",
      "2020-03-18 16:16:24,373 : INFO : source_vocab length 21469\n",
      "2020-03-18 16:16:24,514 : INFO : Phraser built with 43 phrasegrams\n",
      "2020-03-18 16:16:24,623 : INFO : collecting all words and their counts\n",
      "2020-03-18 16:16:24,623 : INFO : PROGRESS: at sentence #0, processed 0 words and 0 word types\n",
      "2020-03-18 16:16:24,654 : INFO : collected 21614 word types from a corpus of 19246 words (unigram + bigrams) and 2303 sentences\n",
      "2020-03-18 16:16:24,654 : INFO : using 21614 counts as vocab in Phrases<0 vocab, min_count=5, threshold=7, max_vocab_size=40000000>\n",
      "2020-03-18 16:16:24,654 : INFO : source_vocab length 21614\n",
      "2020-03-18 16:16:24,795 : INFO : Phraser built with 112 phrasegrams\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bigramme\n",
      "[(b'body area', 207.42995169082124),\n",
      " (b'middle east', 195.38673245614035),\n",
      " (b'severe acute', 182.16121212121215),\n",
      " (b'hepatitis c', 170.38888888888889),\n",
      " (b'respiratory syndrome', 128.161053160739),\n",
      " (b'east respiratory', 126.92559480894016),\n",
      " (b'receptor-binding domain', 120.27450980392157),\n",
      " (b'routing protocol', 108.33809924306138),\n",
      " (b'indoor positioning', 103.47310323383086),\n",
      " (b'sensor networks', 102.5947237496727),\n",
      " (b'wireless sensor', 95.78155418559378),\n",
      " (b'kalman filter', 85.19444444444444),\n",
      " (b'underwater acoustic', 84.33389450056117),\n",
      " (b'2019 novel', 78.64102564102564),\n",
      " (b'genetic programming', 66.46749226006192),\n",
      " (b'acute lymphoblastic', 66.05846153846154),\n",
      " (b'acute respiratory', 61.76013698630137),\n",
      " (b'spike protein', 57.836745689655174),\n",
      " (b'public sector', 45.87393162393162),\n",
      " (b'systematic review', 43.37171717171717),\n",
      " (b'acoustic sensor', 35.60364842454394),\n",
      " (b'crystal structure', 33.13117283950617),\n",
      " (b'indoor localization', 32.72713414634146),\n",
      " (b'sensor networks:', 32.04328358208955),\n",
      " (b'underwater wireless', 31.82164031620553),\n",
      " (b'ebola virus', 28.18933823529412),\n",
      " (b'sars coronavirus', 28.185108777194298),\n",
      " (b'positioning system', 27.078831196132015),\n",
      " (b'coronavirus (mers-cov)', 27.05267137096774),\n",
      " (b'novel coronavirus', 23.7825682382134),\n",
      " (b'sensor network', 22.888059701492537),\n",
      " (b'coronavirus (2019-ncov)', 21.642137096774192),\n",
      " (b'wireless body', 20.292060491493384),\n",
      " (b'underwater sensor', 19.42017186793306),\n",
      " (b'virus replication', 18.041176470588233),\n",
      " (b'influenza virus', 17.00033936651584),\n",
      " (b'mers coronavirus', 16.031212664277177),\n",
      " (b'coronavirus spike', 14.428091397849462),\n",
      " (b'zika virus', 14.350935828877008),\n",
      " (b'mers-cov infection', 13.818279339197597),\n",
      " (b'virus infection', 11.989389426656741)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 16:16:24,904 : INFO : adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2020-03-18 16:16:24,935 : INFO : built Dictionary(6964 unique tokens: ['data', 'flexible', 'illumina', 'sequence', 'trimmer']...) from 2303 documents (total 18687 corpus positions)\n",
      "2020-03-18 16:16:24,951 : INFO : collecting document frequencies\n",
      "2020-03-18 16:16:24,951 : INFO : PROGRESS: processing document #0\n",
      "2020-03-18 16:16:24,967 : INFO : calculating IDF weights for 2303 documents and 6963 features (18517 matrix non-zeros)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trigramme\n",
      "[(b'de novo', 982.4545454545455),\n",
      " (b'wireless_body area', 720.4666666666667),\n",
      " (b'escherichia coli', 686.1587301587301),\n",
      " (b'saudi arabia', 635.7058823529412),\n",
      " (b'critically ill', 600.3888888888888),\n",
      " (b'acute_lymphoblastic leukemia', 540.35),\n",
      " (b'hereditary angioedema', 514.6190476190476),\n",
      " (b'carboxylesterase 1', 415.6538461538462),\n",
      " (b'novel_coronavirus (2019-ncov)', 392.98181818181826),\n",
      " (b'monoclonal antibody', 377.8671328671329),\n",
      " (b'mhc class', 350.8766233766234),\n",
      " (b'rhesus macaques', 343.07936507936506),\n",
      " (b'ion channel', 308.7714285714286),\n",
      " (b'received signal', 280.7012987012987),\n",
      " (b'underwater_wireless sensor_networks', 267.66563467492256),\n",
      " (b'hepatitis e', 240.15555555555554),\n",
      " (b'underwater_acoustic sensor_networks', 224.36678200692043),\n",
      " (b'structural basis', 186.32758620689654),\n",
      " (b'middle_east respiratory_syndrome', 181.01054590570718),\n",
      " (b'wireless_sensor networks', 178.20053191489362),\n",
      " (b'routing protocols', 157.19272727272727),\n",
      " (b'health care', 140.35064935064935),\n",
      " (b'spike glycoprotein', 135.0875),\n",
      " (b'doa estimation', 133.41975308641975),\n",
      " (b'quorum sensing', 127.14117647058823),\n",
      " (b'underwater_sensor networks', 120.07777777777777),\n",
      " (b'innate immune', 108.07000000000001),\n",
      " (b'rna polymerase', 101.7927786499215),\n",
      " (b'inertial sensors', 95.55260831122898),\n",
      " (b'severe_acute respiratory_syndrome', 95.25400641025641),\n",
      " (b'respiratory_syndrome coronavirus', 82.32105394605395),\n",
      " (b'respiratory tract', 81.87121212121212),\n",
      " (b'nmda receptor', 75.0486111111111),\n",
      " (b'fruit bats', 67.54375),\n",
      " (b'envelope protein', 65.4969696969697),\n",
      " (b'energy efficient', 63.57058823529412),\n",
      " (b'immune responses', 63.57058823529412),\n",
      " (b'neutralizing antibody', 61.40340909090909),\n",
      " (b'hepatitis_c virus', 55.515410958904106),\n",
      " (b'severe_acute respiratory', 54.580808080808076),\n",
      " (b'cryo-em structure', 50.265116279069765),\n",
      " (b'structure determination', 50.265116279069765),\n",
      " (b'indoor_positioning system', 47.92461197339246),\n",
      " (b'coronavirus_spike protein', 43.66464646464647),\n",
      " (b'2019_novel coronavirus', 41.27960275019099),\n",
      " (b'navigation system', 39.695133149678604),\n",
      " (b'human monoclonal', 34.39893899204244),\n",
      " (b'pedestrian navigation', 32.74848484848485),\n",
      " (b'influenza viruses', 31.054597701149426),\n",
      " (b'small rna', 29.406802721088432)]\n"
     ]
    }
   ],
   "source": [
    "from gensim import corpora, models, similarities\n",
    "\n",
    "titles = []\n",
    "for doi in dois_to_plot:\n",
    "    pub = work_dict[doi]\n",
    "    if \"title\" in pub:\n",
    "        if len(pub[\"title\"]) > 0:\n",
    "            pub_title = [s for s in pub[\"title\"][0].lower().split() if not s in stopwords]\n",
    "            titles.append(pub_title)\n",
    "\n",
    "phrased_lines, bigram, trigram = phrase_lines(titles)\n",
    "\n",
    "frequency = defaultdict(int)\n",
    "\n",
    "for line in phrased_lines:\n",
    "    for term in line:\n",
    "        frequency[term] += 1\n",
    "\n",
    "dictionary = corpora.Dictionary(phrased_lines)\n",
    "\n",
    "corpus = [dictionary.doc2bow(line) for line in phrased_lines]\n",
    "\n",
    "tfidf = models.TfidfModel(corpus, normalize=True) # step 1 -- initialize a model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 0\n",
      "Size: 73\n",
      "\n",
      "Cluster No. 1\n",
      "Size: 46\n",
      "\n",
      "Cluster No. 2\n",
      "Size: 104\n",
      "\n",
      "Cluster No. 3\n",
      "Size: 51\n",
      "\n",
      "Cluster No. 4\n",
      "Size: 18\n",
      "\n",
      "Cluster No. 5\n",
      "Size: 21\n",
      "\n",
      "Cluster No. 6\n",
      "Size: 87\n",
      "\n",
      "Cluster No. 7\n",
      "Size: 132\n",
      "\n",
      "Cluster No. 8\n",
      "Size: 19\n",
      "\n",
      "Cluster No. 9\n",
      "Size: 88\n",
      "\n",
      "Cluster No. 10\n",
      "Size: 22\n",
      "\n",
      "Cluster No. 11\n",
      "Size: 101\n",
      "\n",
      "Cluster No. 12\n",
      "Size: 31\n",
      "\n",
      "Cluster No. 13\n",
      "Size: 18\n",
      "\n",
      "Cluster No. 14\n",
      "Size: 48\n",
      "\n",
      "Cluster No. 15\n",
      "Size: 37\n",
      "\n",
      "Cluster No. 16\n",
      "Size: 39\n",
      "\n",
      "Cluster No. 17\n",
      "Size: 65\n",
      "\n",
      "Cluster No. 18\n",
      "Size: 46\n",
      "\n",
      "Cluster No. 19\n",
      "Size: 46\n",
      "\n",
      "Cluster No. 20\n",
      "Size: 92\n",
      "\n",
      "Cluster No. 21\n",
      "Size: 53\n",
      "\n",
      "Cluster No. 22\n",
      "Size: 10\n",
      "\n",
      "Cluster No. 23\n",
      "Size: 86\n",
      "\n",
      "Cluster No. 24\n",
      "Size: 61\n",
      "\n",
      "Cluster No. 25\n",
      "Size: 55\n",
      "\n",
      "Cluster No. 26\n",
      "Size: 62\n",
      "\n",
      "Cluster No. 27\n",
      "Size: 54\n",
      "\n",
      "Cluster No. 28\n",
      "Size: 15\n",
      "\n",
      "Cluster No. 29\n",
      "Size: 61\n",
      "\n",
      "Cluster No. 30\n",
      "Size: 16\n",
      "\n",
      "Cluster No. 31\n",
      "Size: 25\n",
      "\n",
      "Cluster No. 32\n",
      "Size: 18\n",
      "\n",
      "Cluster No. 33\n",
      "Size: 28\n",
      "\n",
      "Cluster No. 34\n",
      "Size: 17\n",
      "\n",
      "Cluster No. 35\n",
      "Size: 61\n",
      "\n",
      "Cluster No. 36\n",
      "Size: 44\n",
      "\n",
      "Cluster No. 37\n",
      "Size: 63\n",
      "\n",
      "Cluster No. 38\n",
      "Size: 45\n",
      "\n",
      "Cluster No. 39\n",
      "Size: 81\n",
      "\n",
      "Cluster No. 40\n",
      "Size: 47\n",
      "\n",
      "Cluster No. 41\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 42\n",
      "Size: 17\n",
      "\n",
      "Cluster No. 43\n",
      "Size: 70\n",
      "\n",
      "Cluster No. 44\n",
      "Size: 41\n",
      "\n",
      "Cluster No. 45\n",
      "Size: 72\n",
      "\n",
      "Cluster No. 46\n",
      "Size: 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clSizes = []\n",
    "clTitles = defaultdict(list)\n",
    "\n",
    "for cl in cluster_ids:    \n",
    "    print(\"Cluster No.\", cl)\n",
    "    cl_dois = [doi for i, doi in enumerate(dois_to_plot) if clLabels[i] == cl]\n",
    "    print(\"Size:\", len(cl_dois))\n",
    "    for doi in cl_dois:\n",
    "        #pprint(work_dict[doi][\"title\"])\n",
    "        pub = work_dict[doi]\n",
    "        if \"title\" in pub:\n",
    "            if len(pub[\"title\"]) > 0:\n",
    "                pub_title = [s for s in pub[\"title\"][0].lower().split() if not s in stopwords]\n",
    "                clTitles[cl].append(pub_title)        \n",
    "    print()\n",
    "    clSizes.append(len(cl_dois))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 0\n",
      "Size: 73\n",
      "Terms: \n",
      "wireless_sensor_networks (31, 25, 0.41), networks (36, 16, 0.25), routing (18, 13, 0.24), wireless_body_area (11, 11, 0.22), mobile (26, 12, 0.21), energy (27, 11, 0.19), scheme (19, 10, 0.18), mac (9, 8, 0.17), data (64, 12, 0.17), survey (32, 10, 0.16), protocol (13, 8, 0.16), sensor_networks (10, 7, 0.15), clustering (14, 7, 0.14), routing_protocol (16, 7, 0.13), wsns (10, 6, 0.12), routing_protocols (7, 5, 0.11), energy-efficient (10, 5, 0.10), gathering (4, 4, 0.10), based (75, 7, 0.09), secure (6, 4, 0.09)\n",
      "top_authors: [('Liu', 21), ('Wang', 18), ('Kim', 7), ('Zhang', 7), ('Park', 6), ('Xu', 6), ('Zhao', 6), ('Gao', 5), ('Xiong', 5), ('Chen', 5)]\n",
      "\n",
      "Cluster No. 1\n",
      "Size: 46\n",
      "Terms: \n",
      "porcine (7, 7, 0.28), coronavirus (126, 10, 0.20), protein (83, 7, 0.16), sars_coronavirus (24, 5, 0.16), protease (10, 4, 0.15), deltacoronavirus (3, 3, 0.14), pathogenesis (18, 4, 0.13), 3c-like (4, 3, 0.13), nucleocapsid (6, 3, 0.12), sars-cov (27, 4, 0.12), activity (27, 4, 0.12), rna (35, 4, 0.11), molecular (42, 4, 0.11), sars (19, 3, 0.10), virus (65, 4, 0.10), dimerization (2, 2, 0.10), replication (25, 3, 0.09), diarrhoea (3, 2, 0.09), substrate (3, 2, 0.09), c-terminal (3, 2, 0.09)\n",
      "top_authors: [('Chen', 10), ('Huang', 5), ('Zhang', 5), ('Li', 4), ('Chang', 4), ('Wang', 4), ('Perlman', 3), ('Jiang', 3), ('Baker', 3), ('Yuen', 2)]\n",
      "\n",
      "Cluster No. 2\n",
      "Size: 104\n",
      "Terms: \n",
      "frankia (11, 11, 0.22), sp. (13, 9, 0.17), isolated (11, 8, 0.15), ambrosia (7, 7, 0.15), production (13, 8, 0.15), bacteria (14, 8, 0.14), species (31, 9, 0.14), quorum_sensing (6, 6, 0.13), ord. (4, 4, 0.11), spontaneous (5, 5, 0.11), harmonia (5, 5, 0.11), nov. (4, 4, 0.10), quorum (4, 4, 0.09), soil (4, 4, 0.09), nov., (8, 4, 0.09), isobutanol (4, 4, 0.09), rrna (5, 4, 0.09), gene (39, 6, 0.09), strain (6, 4, 0.08), pseudomonas (6, 4, 0.08)\n",
      "top_authors: [('Chan', 13), ('Yin', 12), ('Gtari', 8), ('Nouioui', 8), ('Koh', 7), ('Sam', 7), ('Ghodhbane-Gtari', 7), ('Lee', 6), ('Tisa', 6), ('Li', 6)]\n",
      "\n",
      "Cluster No. 3\n",
      "Size: 51\n",
      "Terms: \n",
      "middle_east_respiratory_syndrome (86, 18, 0.32), coronavirus (126, 14, 0.22), treatment (25, 7, 0.17), sars (19, 5, 0.13), convalescent (6, 4, 0.13), infection (48, 6, 0.13), orf (3, 3, 0.13), ribavirin (7, 4, 0.13), interferon (9, 4, 0.12), patients (30, 5, 0.12), clinical (35, 5, 0.11), pathogenesis (18, 4, 0.11), induction (4, 3, 0.10), response (23, 4, 0.10), replication (25, 4, 0.10), emerging (26, 4, 0.10), combination (6, 3, 0.10), t (7, 3, 0.10), viral (33, 4, 0.09), novel_human (9, 3, 0.09)\n",
      "top_authors: [('Chan', 28), ('Yuen', 14), ('Li', 11), ('To', 10), ('Lau', 10), ('Cheng', 9), ('Chen', 9), ('Feldmann', 7), ('Liu', 6), ('Zhou', 6)]\n",
      "\n",
      "Cluster No. 4\n",
      "Size: 18\n",
      "Terms: \n",
      "toll-like (5, 4, 0.28), innate_immune (6, 3, 0.20), receptor (34, 4, 0.19), receptors (16, 3, 0.17), immunodeficiency, (2, 2, 0.16), lubac (2, 2, 0.16), nlrx1 (2, 2, 0.16), tlr7 (2, 2, 0.16), 8 (3, 2, 0.15), deficiency (3, 2, 0.15), mitochondrial (3, 2, 0.15), signaling (7, 2, 0.13), antiviral (17, 2, 0.11), disease (33, 2, 0.10), viral (33, 2, 0.10), mtor (1, 1, 0.09), immunity: (1, 1, 0.09), pattern-recognition (1, 1, 0.09), adaptor (1, 1, 0.09), sting (1, 1, 0.09)\n",
      "top_authors: [('Chen', 5), ('Wang', 3), ('Zhang', 3), ('Tanji', 2), ('Ohto', 2), ('Shibata', 2), ('Miyake', 2), ('Shimizu', 2), ('Yang', 2), ('Yu', 2)]\n",
      "\n",
      "Cluster No. 5\n",
      "Size: 21\n",
      "Terms: \n",
      "carboxylesterase_1 (7, 7, 0.38), hereditary_angioedema (6, 6, 0.34), gene (39, 4, 0.15), mutation (13, 3, 0.15), defective (2, 2, 0.15), mutations (20, 3, 0.13), attacks (2, 2, 0.13), candidiasis (2, 2, 0.13), mucocutaneous (2, 2, 0.13), stat1 (3, 2, 0.13), variants (4, 2, 0.12), oseltamivir (4, 2, 0.12), chronic (4, 2, 0.12), polymorphism (6, 2, 0.11), genetics (7, 2, 0.11), mrna (7, 2, 0.11), natural (9, 2, 0.10), activation (16, 2, 0.09), activity (27, 2, 0.08), patients (30, 2, 0.08)\n",
      "top_authors: [('Zhu', 5), ('Markowitz', 5), ('Bork', 4), ('Cicardi', 2), ('Wang', 2), ('Johnson', 2), ('Maurer', 2), ('Longhurst', 2), ('Witzke', 2), ('Hardt', 2)]\n",
      "\n",
      "Cluster No. 6\n",
      "Size: 87\n",
      "Terms: \n",
      "chlorophyll (7, 7, 0.18), nmda_receptor (7, 7, 0.17), protein (83, 11, 0.16), receptors (16, 7, 0.15), allosteric (7, 6, 0.15), binding (24, 7, 0.14), nmda (5, 5, 0.13), serotonin (6, 5, 0.13), photosystem (7, 5, 0.12), mechanism (14, 5, 0.11), mycobacterium (4, 4, 0.11), molecular (42, 6, 0.10), domain (20, 5, 0.10), force (8, 4, 0.10), plant (9, 4, 0.09), synthesis (11, 4, 0.09), light (14, 4, 0.09), properties (15, 4, 0.09), structures (15, 4, 0.09), glutamate (3, 3, 0.08)\n",
      "top_authors: [('Reynisson', 6), ('Chen', 5), ('Hansen', 5), ('Paoletti', 4), ('Shen', 4), ('Li', 3), ('Gouaux', 3), ('Bennett', 3), ('Traynelis', 3), ('Karakas', 3)]\n",
      "\n",
      "Cluster No. 7\n",
      "Size: 132\n",
      "Terms: \n",
      "indoor_positioning (30, 30, 0.36), indoor (33, 24, 0.28), indoor_localization (22, 19, 0.25), wi-fi (14, 14, 0.20), based (75, 19, 0.18), smartphone (11, 11, 0.16), pedestrian (14, 11, 0.16), fingerprinting (11, 10, 0.15), indoor_positioning_system (11, 10, 0.15), positioning (21, 11, 0.14), location (15, 10, 0.14), strength (9, 8, 0.12), algorithm (48, 11, 0.12), magnetic (16, 8, 0.12), bluetooth (7, 7, 0.11), robust (17, 8, 0.11), system (48, 10, 0.11), wireless (14, 7, 0.10), received_signal (6, 6, 0.10), wifi (6, 6, 0.10)\n",
      "top_authors: [('Liu', 21), ('Zhang', 20), ('Wang', 20), ('Li', 19), ('Chen', 16), ('Zhou', 11), ('Park', 9), ('Ma', 9), ('Xu', 9), ('Yang', 8)]\n",
      "\n",
      "Cluster No. 8\n",
      "Size: 19\n",
      "Terms: \n",
      "police (8, 5, 0.33), theatre (2, 2, 0.18), effect (17, 3, 0.17), representation (2, 2, 0.17), model (35, 3, 0.15), force (8, 2, 0.14), findings (10, 2, 0.13), policy (25, 2, 0.11), use (19, 2, 0.11), fund: (1, 1, 0.09), monetary (1, 1, 0.09), calea, (1, 1, 0.09), enforcement: (1, 1, 0.09), incentives, (1, 1, 0.09), policing, (1, 1, 0.09), screening, (1, 1, 0.09), unions, (1, 1, 0.09), evening (1, 1, 0.09), experiments (1, 1, 0.09), productions (1, 1, 0.09)\n",
      "top_authors: [('Grisolía', 2), ('Willis', 2), ('Terrill', 2), ('Cordner', 2)]\n",
      "\n",
      "Cluster No. 9\n",
      "Size: 88\n",
      "Terms: \n",
      "middle_east_respiratory_syndrome (86, 37, 0.45), mers-cov (46, 19, 0.27), coronavirus (126, 19, 0.20), transmission (52, 14, 0.20), mers (12, 8, 0.15), saudi (7, 7, 0.15), saudi_arabia (10, 7, 0.14), south (10, 7, 0.14), dromedary (11, 7, 0.14), 2015 (11, 7, 0.14), coronavirus_(mers-cov) (15, 7, 0.13), camels, (7, 6, 0.13), korea (12, 6, 0.12), infection (48, 8, 0.11), arabia, (5, 5, 0.11), korea, (7, 5, 0.11), risk (24, 6, 0.10), mers-cov_infection (13, 5, 0.10), mers_coronavirus (15, 5, 0.09), camels (6, 4, 0.09)\n",
      "top_authors: [('Kim', 22), ('Lee', 16), ('Drosten', 11), ('Park', 10), ('Müller', 9), ('Peiris', 9), ('Haagmans', 8), ('Poon', 8), ('Zhou', 8), ('Gerber', 8)]\n",
      "\n",
      "Cluster No. 10\n",
      "Size: 22\n",
      "Terms: \n",
      "mhc_class (10, 7, 0.33), birds (6, 4, 0.21), variation (11, 4, 0.19), ii (14, 4, 0.18), genes (17, 4, 0.17), histocompatibility (4, 3, 0.17), mhc (4, 3, 0.17), genetic (26, 4, 0.16), major (8, 3, 0.15), selection (15, 3, 0.13), diversity (17, 3, 0.13), characterization (19, 3, 0.13), diversifying (2, 2, 0.12), bottlenecked (2, 2, 0.12), loss (4, 2, 0.12), software (4, 2, 0.11), polymorphism (6, 2, 0.10), b (8, 2, 0.10), avian (8, 2, 0.10), comprehensive (11, 2, 0.09)\n",
      "top_authors: [('Braun', 2), ('Sheldon', 2), ('Murrell', 2), ('Moola', 2), ('Weighill', 2), ('Scheffler', 2), ('Kosakovsky Pond', 2), ('Burri', 2), ('Fumagalli', 2), ('Bollmer', 2)]\n",
      "\n",
      "Cluster No. 11\n",
      "Size: 101\n",
      "Terms: \n",
      "china (27, 17, 0.26), 2019-ncov (14, 13, 0.23), outbreak (32, 15, 0.22), wuhan, (12, 12, 0.22), pneumonia (19, 13, 0.22), novel_coronavirus (11, 10, 0.19), 2019_novel_coronavirus (10, 9, 0.17), novel_coronavirus_(2019-ncov) (9, 8, 0.16), coronavirus (126, 14, 0.14), 2019_novel (7, 7, 0.14), coronaviruses (34, 9, 0.13), health (8, 6, 0.12), transmission (52, 9, 0.12), clinical (35, 8, 0.12), characteristics (12, 6, 0.11), wuhan (5, 5, 0.11), coronavirus_(2019-ncov) (5, 5, 0.11), early (18, 6, 0.10), china: (7, 5, 0.10), potential (20, 6, 0.10)\n",
      "top_authors: [('Wang', 51), ('Li', 29), ('Zhang', 27), ('Liu', 26), ('Chen', 22), ('Yang', 17), ('Hu', 15), ('Zhao', 15), ('Zhu', 13), ('Xu', 13)]\n",
      "\n",
      "Cluster No. 12\n",
      "Size: 31\n",
      "Terms: \n",
      "rna (35, 8, 0.26), sars-coronavirus (11, 6, 0.25), complex (28, 7, 0.24), helicase (5, 4, 0.20), nonstructural (6, 4, 0.18), sars-cov (27, 5, 0.17), protein (83, 6, 0.16), cap (3, 3, 0.15), sars_coronavirus (24, 4, 0.14), coronavirus (126, 6, 0.13), coronaviruses (34, 4, 0.13), rna_polymerase (8, 3, 0.13), genome (63, 4, 0.11), exoribonuclease (2, 2, 0.11), nsp12 (2, 2, 0.11), primer (2, 2, 0.11), fidelity (2, 2, 0.11), proteins (25, 3, 0.10), functional (26, 3, 0.10), activity (27, 3, 0.10)\n",
      "top_authors: [('Snijder', 11), ('Decroly', 7), ('Denison', 6), ('Imbert', 6), ('Canard', 6), ('Gorbalenya', 6), ('Zhang', 5), ('Chen', 5), ('Graham', 4), ('Feng', 4)]\n",
      "\n",
      "Cluster No. 13\n",
      "Size: 18\n",
      "Terms: \n",
      "policy (25, 16, 0.63), systems: (5, 4, 0.20), advice (5, 4, 0.20), capacity (4, 3, 0.16), consultants (3, 3, 0.16), evidence-based (5, 3, 0.15), governance (6, 3, 0.15), government (7, 3, 0.14), big (3, 2, 0.12), federal (2, 2, 0.12), analysts (2, 2, 0.12), westminster (2, 2, 0.12), departments (2, 2, 0.12), management (26, 3, 0.11), multi-level (3, 2, 0.11), canadian (4, 2, 0.10), work (6, 2, 0.10), political (8, 2, 0.10), accounting (14, 2, 0.08), systems (25, 2, 0.07)\n",
      "top_authors: [('Howlett', 6), ('Migone', 2)]\n",
      "\n",
      "Cluster No. 14\n",
      "Size: 48\n",
      "Terms: \n",
      "patients (30, 8, 0.20), lung (11, 6, 0.18), distress (4, 4, 0.15), influenza (35, 6, 0.14), angiotensin (5, 4, 0.14), acute_respiratory (8, 4, 0.13), mortality (9, 4, 0.13), 2 (11, 4, 0.13), systematic_review (11, 4, 0.12), sepsis (3, 3, 0.12), adipocyte (3, 3, 0.12), ii (14, 4, 0.12), insulin (3, 3, 0.11), type (19, 4, 0.11), scoring (4, 3, 0.11), intensive (4, 3, 0.11), acute (6, 3, 0.10), meta-analysis (6, 3, 0.10), injury (6, 3, 0.10), critically_ill (6, 3, 0.10)\n",
      "top_authors: [('Li', 9), ('Liu', 6), ('Zhao', 4), ('Wang', 4), ('Yang', 4), ('Jones', 3), ('Zhou', 3), ('Cook', 3), ('Zou', 3), ('Gao', 3)]\n",
      "\n",
      "Cluster No. 15\n",
      "Size: 37\n",
      "Terms: \n",
      "welfare (7, 7, 0.28), europe (9, 8, 0.28), policy (25, 7, 0.21), preferences (10, 6, 0.21), insecurity (5, 5, 0.19), social (10, 5, 0.17), childcare (4, 4, 0.17), institutions (5, 4, 0.15), political (8, 4, 0.15), oecd (3, 3, 0.13), germany (4, 3, 0.12), explaining (4, 3, 0.12), market (4, 3, 0.12), european (7, 3, 0.11), opinion (2, 2, 0.09), regulations (2, 2, 0.09), western (2, 2, 0.09), policy: (2, 2, 0.09), worlds (2, 2, 0.09), vocational (2, 2, 0.09)\n",
      "top_authors: [('Marx', 3), ('Chung', 3), ('Soroka', 2), ('Wlezien', 2), ('Busemeyer', 2), ('Häusermann', 2)]\n",
      "\n",
      "Cluster No. 16\n",
      "Size: 39\n",
      "Terms: \n",
      "coronavirus (126, 18, 0.35), protein (83, 13, 0.29), envelope_protein (8, 7, 0.26), severe_acute_respiratory_syndrome (16, 7, 0.23), e (8, 6, 0.22), ion_channel (7, 5, 0.19), intracellular (3, 3, 0.13), inflammasome (3, 3, 0.13), viroporin (3, 3, 0.13), activates (5, 3, 0.12), virus (65, 5, 0.12), sars-cov (27, 4, 0.12), membrane (10, 3, 0.11), infections (19, 3, 0.10), respiratory (20, 3, 0.09), attenuated (2, 2, 0.09), nlrp3 (2, 2, 0.09), autophagy (2, 2, 0.09), pdz (2, 2, 0.09), promiscuity (2, 2, 0.09)\n",
      "top_authors: [('Enjuanes', 12), ('DeDiego', 11), ('Nieto-Torres', 11), ('Regla-Nava', 7), ('Jimenez-Guardeño', 5), ('Fernandez-Delgado', 5), ('Castaño-Rodriguez', 5), ('Torres', 5), ('Álvarez', 4), ('Perlman', 4)]\n",
      "\n",
      "Cluster No. 17\n",
      "Size: 65\n",
      "Terms: \n",
      "mers-cov (46, 21, 0.32), middle_east_respiratory_syndrome (86, 22, 0.28), coronavirus (126, 22, 0.25), vaccine (27, 13, 0.23), receptor-binding_domain (14, 10, 0.20), vaccines (14, 9, 0.18), human (111, 13, 0.15), spike_protein (17, 8, 0.15), neutralizing (10, 7, 0.15), mers-cov_infection (13, 7, 0.14), mers_coronavirus (15, 7, 0.14), subunit (9, 6, 0.13), mice (20, 7, 0.13), antibody (12, 6, 0.12), recombinant (15, 6, 0.12), receptor-binding (7, 5, 0.11), human_monoclonal (8, 5, 0.11), immunity (11, 5, 0.10), coronavirus_(mers-cov) (15, 5, 0.10), entry (17, 5, 0.09)\n",
      "top_authors: [('Wang', 38), ('Jiang', 26), ('Li', 24), ('Du', 23), ('Zhang', 16), ('Zhou', 16), ('Lu', 15), ('Zhao', 15), ('Chen', 12), ('Tseng', 12)]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 18\n",
      "Size: 46\n",
      "Terms: \n",
      "zika_virus (11, 8, 0.28), concentrates (5, 5, 0.20), platelet (6, 5, 0.19), dengue (7, 5, 0.19), reduction (4, 4, 0.16), light (14, 4, 0.13), ebola (13, 4, 0.13), detection (38, 5, 0.13), ultraviolet (3, 3, 0.13), viruses (22, 4, 0.12), virus (65, 5, 0.12), disease (33, 4, 0.11), plasma (9, 3, 0.11), pathogen (10, 3, 0.11), notifiable (2, 2, 0.10), acid (16, 3, 0.10), influenza_virus (17, 3, 0.10), humans (20, 3, 0.09), establishment (2, 2, 0.09), characterization, (2, 2, 0.09)\n",
      "top_authors: [('Zhang', 7), ('Chen', 6), ('Seltsam', 4), ('Wang', 4), ('Holmes', 3), ('Lee', 3), ('Gravemann', 3), ('Huang', 3), ('Tang', 3), ('Müller', 3)]\n",
      "\n",
      "Cluster No. 19\n",
      "Size: 46\n",
      "Terms: \n",
      "acute_lymphoblastic_leukemia (8, 8, 0.30), sequencing (24, 6, 0.18), variation (11, 5, 0.18), acute_lymphoblastic (4, 4, 0.17), leukaemia (4, 4, 0.17), mutations (20, 5, 0.16), schizophrenia (3, 3, 0.13), precursor (4, 3, 0.13), cytokine (4, 3, 0.13), data (64, 5, 0.12), genetic (26, 4, 0.12), genome (63, 5, 0.12), macrophages (6, 3, 0.12), single-cell (8, 3, 0.11), next-generation (8, 3, 0.11), receptor (34, 4, 0.11), human (111, 5, 0.10), single (15, 3, 0.10), humans (20, 3, 0.09), neoantigens (2, 2, 0.09)\n",
      "top_authors: [('Chen', 11), ('Ma', 10), ('Zhang', 8), ('Mullighan', 8), ('Li', 7), ('Downing', 7), ('Lu', 6), ('Carroll', 6), ('Payne-Turner', 6), ('Hunger', 6)]\n",
      "\n",
      "Cluster No. 20\n",
      "Size: 92\n",
      "Terms: \n",
      "genome (63, 17, 0.26), genomes (21, 12, 0.24), microbial (11, 9, 0.20), metagenomic (6, 6, 0.15), macrococcus (5, 5, 0.15), sequence (30, 8, 0.15), prokaryotic (7, 6, 0.14), fast (13, 6, 0.13), analysis (107, 10, 0.13), comparison (17, 6, 0.12), annotation (7, 5, 0.12), gene (39, 7, 0.12), database (23, 6, 0.11), nov., (8, 4, 0.11), bacterial (18, 5, 0.10), sequencing (24, 5, 0.09), identification (27, 5, 0.09), genomics (13, 4, 0.09), metagenome (3, 3, 0.08), metabolic (3, 3, 0.08)\n",
      "top_authors: [('Li', 9), ('Chen', 5), ('Wu', 4), ('Nielsen', 4), ('Sullivan', 4), ('Gurevich', 3), ('Sirotkin', 3), ('Tesler', 3), ('Disz', 3), ('Edwards', 3)]\n",
      "\n",
      "Cluster No. 21\n",
      "Size: 53\n",
      "Terms: \n",
      "public (28, 14, 0.33), performance (34, 14, 0.32), accounting (14, 11, 0.31), public_sector (11, 10, 0.29), management (26, 9, 0.22), local (14, 6, 0.17), government (7, 4, 0.12), new (62, 6, 0.12), research (30, 5, 0.12), government: (3, 3, 0.11), institutional (3, 3, 0.11), reforms: (3, 3, 0.11), information (17, 4, 0.11), use (19, 4, 0.10), nonprofit (4, 3, 0.10), leadership (4, 3, 0.10), perspectives (7, 3, 0.09), international (14, 3, 0.08), future (14, 3, 0.08), npm (2, 2, 0.08)\n",
      "top_authors: [('Lapsley', 4), ('Mitchell', 3), ('Steccolini', 2), ('Johansson', 2), ('Hyndman', 2), ('Jacobs', 2), ('Lee', 2), ('Skaerbaek', 2), ('Liguori', 2)]\n",
      "\n",
      "Cluster No. 22\n",
      "Size: 10\n",
      "Terms: \n",
      "thermobifida (4, 4, 0.32), polyethylene (4, 4, 0.32), terephthalate (3, 3, 0.25), structural (20, 3, 0.18), hydrolases (2, 2, 0.18), polyester-degrading (2, 2, 0.18), biodegradation (2, 2, 0.18), plastic-eating (2, 2, 0.18), esterase (2, 2, 0.18), hydrolysis (3, 2, 0.17), diversity (17, 2, 0.12), bacterial (18, 2, 0.12), characterization (19, 2, 0.12), functional (26, 2, 0.11), cutinases (1, 1, 0.10), pet: (1, 1, 0.10), degrading (1, 1, 0.10), fusca (1, 1, 0.10), hydrolase (1, 1, 0.10), thermostable (1, 1, 0.10)\n",
      "top_authors: [('Zimmermann', 5), ('Yang', 5), ('Wei', 4), ('Herrero Acero', 3), ('Ribitsch', 3), ('Greimel', 3), ('Schwab', 3), ('Guebitz', 3), ('Oeser', 3), ('Then', 3)]\n",
      "\n",
      "Cluster No. 23\n",
      "Size: 86\n",
      "Terms: \n",
      "society (5, 4, 0.13), research (30, 6, 0.13), life (14, 5, 0.13), annual (5, 4, 0.12), surveillance (8, 4, 0.11), korea (12, 4, 0.10), influenza (35, 5, 0.10), malaria (4, 3, 0.10), review (22, 4, 0.09), plasmodium (5, 3, 0.09), human (111, 6, 0.09), virus (65, 5, 0.09), global (27, 4, 0.09), clinical (35, 4, 0.08), diseases (10, 3, 0.08), vector (9, 3, 0.08), report (10, 3, 0.08), c (11, 3, 0.08), infection (48, 4, 0.08), facility (2, 2, 0.08)\n",
      "top_authors: [('Kim', 14), ('Lee', 12), ('Zhang', 6), ('Wang', 4), ('Park', 4), ('Hemp', 4), ('Dey', 3), ('Beard', 3), ('Chu', 3), ('Preschern', 3)]\n",
      "\n",
      "Cluster No. 24\n",
      "Size: 61\n",
      "Terms: \n",
      "inertial (29, 18, 0.32), gait (12, 12, 0.26), wearable (10, 9, 0.20), motion (15, 9, 0.19), kinematics (7, 7, 0.17), magnetic (16, 7, 0.15), measurement (14, 7, 0.15), joint (7, 6, 0.15), system (48, 9, 0.14), inertial_sensors (10, 6, 0.13), estimation (46, 8, 0.13), ski (5, 5, 0.13), running (5, 5, 0.13), body (8, 5, 0.12), sensor (21, 6, 0.12), based (75, 8, 0.11), human (111, 9, 0.11), orientation (10, 5, 0.11), robotic (5, 4, 0.10), sensors: (5, 4, 0.10)\n",
      "top_authors: [('Aminian', 6), ('Sabatini', 4), ('Chardonnens', 4), ('Tadano', 3), ('Müller', 3), ('Spörri', 3), ('Favre', 3), ('Delp', 3), ('Miezal', 2), ('Bleser', 2)]\n",
      "\n",
      "Cluster No. 25\n",
      "Size: 55\n",
      "Terms: \n",
      "phylogenomic (8, 7, 0.23), phylogeny (11, 6, 0.18), phylogenetic (14, 6, 0.18), phylogenetics (6, 5, 0.17), new (62, 7, 0.15), bayesian (14, 5, 0.15), relationships (9, 4, 0.13), data (64, 6, 0.12), evolution (39, 5, 0.12), molecular (42, 5, 0.11), raxml (3, 3, 0.11), annelid (3, 3, 0.11), tylenchidae) (3, 3, 0.11), nuclear (7, 3, 0.10), species (31, 4, 0.10), tree (8, 3, 0.10), sequences (10, 3, 0.09), malenchus (2, 2, 0.09), usnea (2, 2, 0.09), evolutionary (17, 3, 0.08)\n",
      "top_authors: [('Li', 14), ('Wang', 8), ('Giribet', 7), ('Zhang', 7), ('Suchard', 6), ('Rambaut', 6), ('Dunn', 6), ('Yang', 6), ('Drummond', 5), ('Hejnol', 5)]\n",
      "\n",
      "Cluster No. 26\n",
      "Size: 62\n",
      "Terms: \n",
      "prediction (22, 13, 0.31), peptide (8, 8, 0.23), protein (83, 13, 0.23), database (23, 8, 0.19), peptides (6, 6, 0.18), docking (7, 6, 0.18), structure (34, 8, 0.17), web (11, 6, 0.16), epitope (6, 5, 0.15), server (9, 5, 0.14), predictions (6, 4, 0.12), neural (7, 4, 0.12), screening (7, 4, 0.12), improving (10, 4, 0.11), epitopes (3, 3, 0.10), modeling (16, 4, 0.10), virtual (4, 3, 0.10), binding (24, 4, 0.09), resource (9, 3, 0.09), immune (9, 3, 0.09)\n",
      "top_authors: [('Wang', 9), ('Nielsen', 9), ('Li', 6), ('Peters', 6), ('Lund', 5), ('Zhang', 4), ('Sette', 4), ('Lundegaard', 4), ('Kim', 3), ('Xu', 3)]\n",
      "\n",
      "Cluster No. 27\n",
      "Size: 54\n",
      "Terms: \n",
      "localization (53, 14, 0.31), localization_wireless_sensor (7, 7, 0.24), tracking (18, 7, 0.21), algorithm (48, 8, 0.18), networks (36, 7, 0.17), target (16, 5, 0.15), distributed (10, 4, 0.13), node (10, 4, 0.13), wireless_sensor_networks (31, 5, 0.13), optimization (12, 4, 0.12), measurements (13, 4, 0.12), based (75, 6, 0.12), sound (3, 3, 0.12), estimation (46, 5, 0.12), aoa (4, 3, 0.11), positioning (21, 4, 0.11), rss (6, 3, 0.11), hybrid (8, 3, 0.10), networks: (10, 3, 0.10), sensor_networks (10, 3, 0.10)\n",
      "top_authors: [('Tomic', 5), ('Beko', 5), ('Zhang', 5), ('Dinis', 4), ('Liu', 4), ('Li', 3), ('Chen', 3), ('Kim', 3), ('Tuba', 2), ('Bacanin', 2)]\n",
      "\n",
      "Cluster No. 28\n",
      "Size: 15\n",
      "Terms: \n",
      "leisure (10, 10, 0.53), constraints (8, 7, 0.39), specialization (7, 5, 0.28), recreation (4, 4, 0.24), negotiation (3, 3, 0.19), fishing (2, 2, 0.14), understand (4, 2, 0.12), chinese (5, 2, 0.12), role (23, 2, 0.09), activity (27, 2, 0.09), model (35, 2, 0.08), constraints: (1, 1, 0.07), intervening (1, 1, 0.07), self-construal (1, 1, 0.07), students (1, 1, 0.07), university (1, 1, 0.07), alternate (1, 1, 0.07), blowing (1, 1, 0.07), dichotomy: (1, 1, 0.07), leisure-casual (1, 1, 0.07)\n",
      "top_authors: [('Oh', 4), ('Lyu', 2)]\n",
      "\n",
      "Cluster No. 29\n",
      "Size: 61\n",
      "Terms: \n",
      "underwater_wireless_sensor_networks (13, 13, 0.33), underwater_sensor_networks (13, 12, 0.30), underwater_acoustic_sensor_networks (11, 11, 0.29), routing_protocol (16, 8, 0.19), localization (53, 10, 0.18), underwater (12, 7, 0.18), underwater_acoustic (6, 6, 0.17), underwater_wireless (6, 6, 0.17), based (75, 10, 0.17), energy (27, 7, 0.15), underwater_sensor (5, 5, 0.15), sensor_network (6, 5, 0.14), communications (7, 5, 0.14), node (10, 5, 0.13), routing (18, 5, 0.12), acoustic (7, 4, 0.11), self-deployment (3, 3, 0.10), deployment (3, 3, 0.10), scheme (19, 4, 0.09), reliable (4, 3, 0.09)\n",
      "top_authors: [('Khan', 12), ('Javaid', 8), ('Wang', 7), ('Lloret', 6), ('Jiang', 6), ('Zhang', 6), ('Chen', 5), ('Liu', 5), ('Ahmad', 5), ('Wu', 4)]\n",
      "\n",
      "Cluster No. 30\n",
      "Size: 16\n",
      "Terms: \n",
      "doa_estimation (8, 7, 0.36), array (8, 6, 0.31), coarray (4, 4, 0.23), direction (4, 4, 0.23), mimo (3, 3, 0.18), coupling (3, 3, 0.18), mutual (3, 3, 0.18), uniform (3, 3, 0.18), estimation (46, 5, 0.18), real-valued (4, 3, 0.17), radar (4, 3, 0.17), unknown (5, 3, 0.17), direction-of-arrival (2, 2, 0.13), esprit (2, 2, 0.13), arrays (2, 2, 0.13), noncircular (2, 2, 0.13), (doa) (2, 2, 0.13), interpolation (3, 2, 0.12), arrival (3, 2, 0.12), approach (32, 3, 0.12)\n",
      "top_authors: [('Wang', 13), ('Chen', 6), ('Li', 4), ('Guo', 3), ('Liu', 3), ('Ren', 3), ('Ding', 2), ('Wu', 2), ('Zhao', 2)]\n",
      "\n",
      "Cluster No. 31\n",
      "Size: 25\n",
      "Terms: \n",
      "small_rna (6, 6, 0.30), salmonella (5, 5, 0.26), regulatory (9, 4, 0.19), hfq (3, 3, 0.17), typhimurium (3, 3, 0.17), formation (7, 3, 0.15), escherichia_coli (7, 3, 0.15), rna (35, 4, 0.14), mapping (16, 3, 0.13), enterica (2, 2, 0.12), serovar (2, 2, 0.12), flhdc (2, 2, 0.12), operon (2, 2, 0.12), ppgpp (2, 2, 0.12), response (23, 3, 0.12), global (27, 3, 0.11), chaperone (3, 2, 0.11), initiation (4, 2, 0.11), mrna (7, 2, 0.10), stress (7, 2, 0.10)\n",
      "top_authors: [('Vogel', 6), ('Lee', 3), ('Luisi', 2), ('Colgan', 2), ('Sivasankaran', 2), ('Hokamp', 2), ('Hinton', 2), ('Papenfort', 2), ('Chao', 2), ('Thomson', 2)]\n",
      "\n",
      "Cluster No. 32\n",
      "Size: 18\n",
      "Terms: \n",
      "education (11, 7, 0.38), physical (8, 4, 0.22), job (10, 4, 0.20), teacher (3, 3, 0.20), hospitality (3, 3, 0.19), family (9, 3, 0.16), police (8, 3, 0.16), organizational (11, 3, 0.15), burnout (2, 2, 0.13), labor (2, 2, 0.13), officer (2, 2, 0.13), turnover (2, 2, 0.13), employees (2, 2, 0.13), frontline (2, 2, 0.13), emotional (3, 2, 0.12), satisfaction (3, 2, 0.12), understanding (8, 2, 0.11), support (14, 2, 0.10), impact (15, 2, 0.10), study (41, 2, 0.08)\n",
      "top_authors: [('Richards', 3), ('Templin', 3), ('Hsieh', 2)]\n",
      "\n",
      "Cluster No. 33\n",
      "Size: 28\n",
      "Terms: \n",
      "genetic_programming (12, 11, 0.47), semantic (9, 9, 0.40), geometric (7, 5, 0.23), bloat (4, 4, 0.23), crossover (3, 3, 0.16), analysis (107, 5, 0.12), algorithms (14, 3, 0.12), evolutionary (17, 3, 0.12), gp (2, 2, 0.11), cartesian (2, 2, 0.11), locality (2, 2, 0.11), genetic (26, 3, 0.11), regression (3, 2, 0.11), roles (3, 2, 0.11), survey (32, 3, 0.10), methods (8, 2, 0.09), problem (9, 2, 0.09), new (62, 3, 0.09), application (12, 2, 0.08), control (15, 2, 0.08)\n",
      "top_authors: [('Silva', 4), ('O’Neill', 3), ('Castelli', 3), ('Vanneschi', 3), ('Krawiec', 3), ('Doerr', 2), ('Uy', 2), ('Hoai', 2), ('McKay', 2), ('Galván-López', 2)]\n",
      "\n",
      "Cluster No. 34\n",
      "Size: 17\n",
      "Terms: \n",
      "femtosecond (6, 6, 0.34), x-ray (9, 6, 0.32), laser (8, 5, 0.27), crystallography (4, 4, 0.25), serial (4, 4, 0.25), structure (34, 6, 0.24), determined (4, 3, 0.18), free-electron (2, 2, 0.14), microed (2, 2, 0.14), structure_determination (3, 2, 0.13), atomic (3, 2, 0.13), reaction (5, 2, 0.12), native (6, 2, 0.11), photosystem (7, 2, 0.11), de_novo (10, 2, 0.10), resolution (10, 2, 0.10), crystal_structure (11, 2, 0.10), ii (14, 2, 0.10), receptors (16, 2, 0.10), structural (20, 2, 0.09)\n",
      "top_authors: [('Fromme', 13), ('Boutet', 11), ('Wang', 11), ('Williams', 9), ('Messerschmidt', 9), ('Barty', 8), ('Spence', 8), ('White', 7), ('Weierstall', 7), ('Zatsepin', 7)]\n",
      "\n",
      "Cluster No. 35\n",
      "Size: 61\n",
      "Terms: \n",
      "genome (63, 17, 0.33), assembly (20, 10, 0.25), fungal (10, 8, 0.23), pathogen (10, 6, 0.17), wheat (5, 5, 0.16), tritici (4, 4, 0.14), genomes (21, 5, 0.13), tool (9, 4, 0.12), sequence (30, 5, 0.12), data (64, 6, 0.12), sequences (10, 4, 0.12), genomics (13, 4, 0.11), long-read (3, 3, 0.11), eukaryotic (3, 3, 0.11), zymoseptoria (3, 3, 0.11), rna-seq (5, 3, 0.10), short (5, 3, 0.10), finding (5, 3, 0.10), chromosomal (6, 3, 0.10), saccharomyces (7, 3, 0.09)\n",
      "top_authors: [('Li', 13), ('Kim', 9), ('Wang', 8), ('Zhang', 8), ('Lee', 7), ('Chen', 5), ('Croll', 5), ('McDonald', 5), ('Smith', 4), ('Robinson', 3)]\n",
      "\n",
      "Cluster No. 36\n",
      "Size: 44\n",
      "Terms: \n",
      "bats (26, 15, 0.38), bat (38, 16, 0.38), african (9, 6, 0.19), viruses (22, 7, 0.18), fruit_bats (6, 5, 0.17), zoonotic (7, 5, 0.16), virus (65, 8, 0.16), fruit (4, 4, 0.14), marburg (5, 4, 0.14), west (6, 4, 0.13), emerging (26, 5, 0.12), spillover (3, 3, 0.11), flight (3, 3, 0.11), aegyptiacus (3, 3, 0.11), rousettus (3, 3, 0.11), henipavirus (3, 3, 0.11), lagos (3, 3, 0.11), africa (6, 3, 0.10), viruses: (6, 3, 0.10), ebola (13, 3, 0.09)\n",
      "top_authors: [('Wang', 17), ('Hayman', 9), ('Cunningham', 9), ('Wood', 9), ('Towner', 9), ('Nichol', 7), ('Tachedjian', 6), ('Smith', 6), ('Suu-Ire', 6), ('Fooks', 5)]\n",
      "\n",
      "Cluster No. 37\n",
      "Size: 63\n",
      "Terms: \n",
      "influenza (35, 17, 0.33), pandemic (10, 8, 0.20), transmission (52, 10, 0.18), masks (6, 6, 0.17), health_care (6, 6, 0.17), infectious (20, 7, 0.16), trial (8, 5, 0.13), workers (9, 5, 0.13), epidemic (13, 5, 0.12), respirator (4, 4, 0.12), modeling (16, 5, 0.12), analysis (107, 8, 0.11), pandemic_influenza (6, 4, 0.11), infection (48, 6, 0.11), randomized (7, 4, 0.11), 2009 (8, 4, 0.11), spread (9, 4, 0.10), virus (65, 6, 0.10), control (15, 4, 0.09), n95 (3, 3, 0.09)\n",
      "top_authors: [('Nishiura', 7), ('Ferguson', 6), ('Colizza', 6), ('Cowling', 6), ('MacIntyre', 5), ('Yang', 5), ('Cauchemez', 4), ('Vespignani', 4), ('Seale', 4), ('Chiu', 4)]\n",
      "\n",
      "Cluster No. 38\n",
      "Size: 45\n",
      "Terms: \n",
      "bat (38, 15, 0.38), sars-like (12, 9, 0.28), coronaviruses (34, 10, 0.25), bats (26, 9, 0.24), coronavirus (126, 12, 0.21), bats, (6, 5, 0.18), origin (19, 5, 0.14), bat_coronaviruses (6, 4, 0.14), viruses (22, 5, 0.14), human_coronavirus (9, 4, 0.13), human (111, 6, 0.11), betacoronavirus (7, 3, 0.10), human_coronaviruses (8, 3, 0.10), detection (38, 4, 0.10), novel (45, 4, 0.09), relatives (2, 2, 0.09), diversity (17, 3, 0.09), evidence (18, 3, 0.09), sars-related (2, 2, 0.08), close (2, 2, 0.08)\n",
      "top_authors: [('Zhang', 26), ('Wang', 22), ('Li', 21), ('Yang', 16), ('Shi', 13), ('Zhu', 10), ('Chen', 10), ('Zhou', 9), ('Daszak', 7), ('Wu', 7)]\n",
      "\n",
      "Cluster No. 39\n",
      "Size: 81\n",
      "Terms: \n",
      "kalman_filter (11, 10, 0.20), localization (53, 14, 0.20), navigation (20, 11, 0.20), slam (10, 9, 0.19), vehicle (9, 8, 0.17), odometry (7, 7, 0.15), mapping (16, 8, 0.15), rgb-d (8, 7, 0.15), system (48, 10, 0.15), inertial (29, 8, 0.13), navigation_system (7, 6, 0.13), alignment (18, 7, 0.13), based (75, 10, 0.13), visual (8, 6, 0.13), sensor (21, 7, 0.12), monocular (5, 5, 0.12), lidar (5, 5, 0.12), matching (6, 5, 0.11), extended (6, 5, 0.11), sins (6, 5, 0.11)\n",
      "top_authors: [('Liu', 13), ('Li', 11), ('Wang', 11), ('Gao', 6), ('Zhang', 6), ('Chen', 5), ('Siegwart', 4), ('Ravankar', 4), ('Reinoso', 4), ('Tang', 4)]\n",
      "\n",
      "Cluster No. 40\n",
      "Size: 47\n",
      "Terms: \n",
      "microscopy (6, 6, 0.22), cryo-electron (5, 5, 0.19), cryo-em_structure (6, 5, 0.18), coronavirus_spike_protein (8, 5, 0.17), coronavirus (126, 9, 0.16), receptor (34, 6, 0.16), structures (15, 5, 0.15), spike_protein (17, 5, 0.15), spike_glycoprotein (6, 4, 0.15), coronavirus_spike (7, 4, 0.14), determination (7, 4, 0.14), spike (10, 4, 0.13), changes (10, 4, 0.13), electron (3, 3, 0.12), entry (17, 4, 0.12), cryo-em (4, 3, 0.12), fusion (21, 4, 0.12), sars_coronavirus (24, 4, 0.11), conformational (6, 3, 0.11), sars-cov (27, 4, 0.11)\n",
      "top_authors: [('Li', 12), ('Wang', 11), ('Bosch', 8), ('Walls', 5), ('Tortorici', 5), ('Rey', 5), ('Veesler', 5), ('Zhang', 4), ('Zheng', 4), ('van Kuppeveld', 4)]\n",
      "\n",
      "Cluster No. 41\n",
      "Size: 7\n",
      "Terms: \n",
      "incorporating (3, 2, 0.25), proteins (25, 2, 0.17), structure (34, 2, 0.16), sialyltransferase (1, 1, 0.14), st8sia (1, 1, 0.14), multi-location (1, 1, 0.14), ploc-mvirus: (1, 1, 0.14), pseaac (1, 1, 0.14), driven (1, 1, 0.14), medicinal (1, 1, 0.14), unprecedented (1, 1, 0.14), diagram (1, 1, 0.14), helix-helix (1, 1, 0.14), nmr (1, 1, 0.14), progresses (1, 1, 0.14), studying (1, 1, 0.14), wenxiang (1, 1, 0.14), brain: (1, 1, 0.14), development, (1, 1, 0.14), gangliosides (1, 1, 0.14)\n",
      "top_authors: [('Huang', 3), ('Zhou', 3), ('Chou', 3), ('Cheng', 2), ('Liao', 2)]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 42\n",
      "Size: 17\n",
      "Terms: \n",
      "kinetochore (6, 6, 0.41), unconventional (2, 2, 0.16), eukaryotes (2, 2, 0.16), trypanosome (2, 2, 0.16), checkpoint (2, 2, 0.16), spindle (2, 2, 0.16), toolkit (3, 2, 0.15), brucei (3, 2, 0.15), kinetoplastid (3, 2, 0.15), cell (34, 3, 0.15), chromosome (5, 2, 0.14), evolutionary (17, 2, 0.11), dynamics (17, 2, 0.11), assembly (20, 2, 0.11), completely (1, 1, 0.09), hhpred (1, 1, 0.09), mpi (1, 1, 0.09), reimplemented (1, 1, 0.09), hmmer (1, 1, 0.09), 2019: (1, 1, 0.09)\n",
      "top_authors: [('Akiyoshi', 5), ('Gull', 4), ('Hayashi', 2)]\n",
      "\n",
      "Cluster No. 43\n",
      "Size: 70\n",
      "Terms: \n",
      "human (111, 14, 0.20), gene (39, 10, 0.20), htlv-1 (6, 6, 0.18), drosophila (6, 6, 0.17), expression (19, 7, 0.16), noncoding (6, 5, 0.14), ribosomal (6, 4, 0.12), cells (26, 5, 0.11), genome (63, 6, 0.10), rna (35, 5, 0.10), protein (83, 6, 0.10), chromatin (3, 3, 0.10), retinal (3, 3, 0.10), dystrophin (3, 3, 0.10), splicing (3, 3, 0.10), resistance (17, 4, 0.09), origin (19, 4, 0.09), intergenic (4, 3, 0.09), epithelium (4, 3, 0.09), pigment (4, 3, 0.09)\n",
      "top_authors: [('Li', 4), ('Wang', 4), ('Nguyen', 3), ('Zhang', 3), ('Liu', 3), ('Lee', 3), ('Grummt', 3), ('Miller', 3), ('Weissman', 2), ('Ruan', 2)]\n",
      "\n",
      "Cluster No. 44\n",
      "Size: 41\n",
      "Terms: \n",
      "management (26, 9, 0.29), conflict (8, 7, 0.28), organizational (11, 5, 0.19), multilevel (4, 3, 0.14), performance (34, 4, 0.12), theory (9, 3, 0.12), job (10, 3, 0.12), impact (15, 3, 0.11), latent (2, 2, 0.10), dimensionality (2, 2, 0.10), mediation (2, 2, 0.10), moderation (2, 2, 0.10), item (2, 2, 0.10), irt (2, 2, 0.10), mixed-measurement (2, 2, 0.10), equivalence (2, 2, 0.10), construct (2, 2, 0.10), predictors (2, 2, 0.10), styles (2, 2, 0.10), validity (3, 2, 0.09)\n",
      "top_authors: [('Liu', 4), ('Tay', 4), ('Li', 3), ('Drasgow', 3), ('Vermunt', 3), ('Podsakoff', 2), ('Smith', 2), ('Newman', 2), ('Morin', 2), ('Bowling', 2)]\n",
      "\n",
      "Cluster No. 45\n",
      "Size: 72\n",
      "Terms: \n",
      "ebola_virus (15, 11, 0.27), virus (65, 12, 0.21), hepatitis_e (7, 7, 0.20), h7n9 (8, 7, 0.19), influenza_virus (17, 7, 0.17), hepatitis_c_virus (8, 6, 0.17), hepatitis_c (8, 6, 0.17), virus_replication (10, 5, 0.13), replication (25, 6, 0.13), disease (33, 6, 0.12), influenza (35, 6, 0.12), antiviral (17, 5, 0.12), infection (48, 6, 0.11), influenza_viruses (7, 4, 0.11), ebola (13, 4, 0.10), inhibitors (14, 4, 0.10), sofosbuvir (3, 3, 0.10), novel (45, 5, 0.10), nonhuman (5, 3, 0.09), primates (5, 3, 0.09)\n",
      "top_authors: [('Wang', 10), ('Zhang', 10), ('Liu', 8), ('Xu', 7), ('Gao', 6), ('Yang', 6), ('Smith', 5), ('Geisbert', 5), ('Rice', 5), ('Li', 5)]\n",
      "\n",
      "Cluster No. 46\n",
      "Size: 10\n",
      "Terms: \n",
      "free (11, 7, 0.49), groups (5, 5, 0.40), laminations (4, 4, 0.35), ℝ-trees (3, 3, 0.26), ℝ-tree (2, 2, 0.18), complex (28, 3, 0.17), dual (6, 2, 0.16), factors (13, 2, 0.14), hyperbolicity (1, 1, 0.10), ii: (1, 1, 0.10), lamination (1, 1, 0.10), curve (1, 1, 0.10), intersection (1, 1, 0.10), \\mathbb{r}-trees (1, 1, 0.10), actions (1, 1, 0.10), finitely (1, 1, 0.10), algebraic (1, 1, 0.10), i: (1, 1, 0.10), co-surface (1, 1, 0.10), geometry (1, 1, 0.10)\n",
      "top_authors: [('Lustig', 4), ('Bestvina', 3), ('Coulbois', 3), ('Hilion', 3), ('Feighn', 2)]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clDict = {}\n",
    "\n",
    "for cl in cluster_ids[:]:\n",
    "    #cl_titles_phrased = []\n",
    "    clTokens = []\n",
    "    if len(clTitles[cl]) < 3:\n",
    "        continue\n",
    "    for title in clTitles[cl]:\n",
    "        title_phrased = trigram[bigram[title]]\n",
    "        #pprint(title_phrased)\n",
    "        #cl_titles_phrased.append(title_phrased)\n",
    "        for t in title_phrased:\n",
    "            clTokens.append(t)\n",
    "    #print(clTokens)\n",
    "    \n",
    "    cl_dois = [doi for i, doi in enumerate(dois_to_plot) if clLabels[i] == cl]\n",
    "    \n",
    "    clBow = dictionary.doc2bow(clTokens)\n",
    "    clFrequency = {}\n",
    "    for i, freq in clBow:\n",
    "        clFrequency[i] = freq\n",
    "    #print(clBow)\n",
    "    vec = sorted(tfidf[clBow], key=itemgetter(1), reverse=True)\n",
    "    \n",
    "    print(\"Cluster No.\", cl)\n",
    "    print(\"Size:\", len(cl_dois))\n",
    "    print(\"Terms:\", \", \".join([\"{}\".format(dictionary[i]) for i, v in vec[:20] if v >= 100 and clFrequency[i] >= 5]))\n",
    "    print(\", \".join([\"{} ({}, {}, {:.2f})\".format(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec[:20]]))\n",
    "    #print([(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec])\n",
    "    \n",
    "    author_counts = defaultdict(int)\n",
    "    for doi in cl_dois:\n",
    "        work = work_dict[doi]\n",
    "        if \"author\" in work:\n",
    "            for author in work[\"author\"]:\n",
    "                if \"family\" in author:\n",
    "                    author_counts[author[\"family\"]] += 1\n",
    "    top_authors = sorted(author_counts.items(), key=itemgetter(1), reverse = True)\n",
    "    top_authors = [(a, n) for a,n in top_authors if n >= 2][:10]\n",
    "    print(\"top_authors:\", top_authors)\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    \n",
    "    clDict[cl] = {}\n",
    "    clDict[cl][\"id\"] = cl\n",
    "    clDict[cl][\"size\"] = len(cl_dois)\n",
    "    clDict[cl][\"dois\"] = cl_dois\n",
    "    clDict[cl][\"vec\"] =  [(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec]\n",
    "    clDict[cl][\"top_authors\"] = top_authors\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1.0,\n",
       " 0.95004162504212,\n",
       " 0.802624679775096,\n",
       " 0.6200510377447751,\n",
       " 0.28370212980097553,\n",
       " 0.1]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bokeh.colors import RGB\n",
    "\n",
    "#year2alpha = lambda year: 1.5 / (1. + np.exp(0.2*(2018 - year))) + 0.25\n",
    "year2alpha = lambda year: np.clip(2. / (1. + np.exp(0.10*(2018 - year))), a_min=0.1, a_max=None)\n",
    "\n",
    "\n",
    "for i, doi in enumerate(dois_to_plot):\n",
    "    pub = work_dict[doi]\n",
    "    df.at[doi, \"doi\"] = doi\n",
    "    df.at[doi, \"title\"] = pub[\"title\"][0]\n",
    "    if \"container-title\" in pub:\n",
    "        df.at[doi, \"containerTitle\"] = \";\".join(pub[\"container-title\"])\n",
    "    #line = pub[\"title\"][0]\n",
    "    #n = 8\n",
    "    #df.at[doi, \"title\"] = \"\\n\".join([line[j:j+n] for j in range(0, len(line), n)])\n",
    "    df.at[doi, \"issued\"] = \"-\".join([str(x) for x in pub[\"issued\"][\"date-parts\"][0]])\n",
    "    df.at[doi, \"refCount\"] = ref_counts[doi]\n",
    "    df.at[doi, \"refCountNormalized\"] = ref_count_normalized[doi]\n",
    "    #df.at[doi, \"scatterSize\"] = 6.+ np.power(ref_counts[doi], 0.8)\n",
    "    df.at[doi, \"scatterSize\"] = 4+2.5*np.power(ref_count_normalized[doi], 1.)\n",
    "    df.at[doi, \"xval\"] = X_embedded[i,0]\n",
    "    df.at[doi, \"yval\"] = X_embedded[i,1]\n",
    "    df.at[doi, \"rgb\"] = RGB(colors[i,0], colors[i,1], colors[i,2])\n",
    "    df.at[doi, \"alpha\"] = year2alpha(pub[\"issued\"][\"date-parts\"][0][0])\n",
    "    try:\n",
    "        pub_authors = [a[\"family\"] for a in pub[\"author\"]]\n",
    "        df.at[doi, \"author\"] = \", \".join(pub_authors)\n",
    "    except KeyError:\n",
    "        df.at[doi, \"author\"] = \"unavailable\"\n",
    "    \n",
    "    df.at[doi, \"clusterNo\"] = clLabels[i]\n",
    "    \n",
    "\n",
    "[year2alpha(y) for y in (2018, 2017, 2014, 2010, 2000, 1980)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df[\"scatterSize\"] *= 0.75\n",
    "df[\"scatterSize\"] *= 1.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"xval\"].max()\n",
    "df[\"xval\"].min()\n",
    "x_span = df[\"xval\"].max() - df[\"xval\"].min()\n",
    "y_span = df[\"yval\"].max() - df[\"yval\"].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.zeros(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tsne.transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0\n",
      "wireless_sensor_networks, networks, routing, wireless_body_area, mobile\n",
      "40.25177001953125 -13.080853462219238 73\n",
      "0.022296816984187142 0.025984917730798033 0.03423980135898394\n",
      "230.57807060872415\n",
      "\n",
      "1\n",
      "porcine, coronavirus, protein, sars_coronavirus, protease\n",
      "-1.8652465343475342 -10.985311508178711 46\n",
      "0.050812303212273646 0.049699567775072825 0.07107698076568174\n",
      "117.17963987859467\n",
      "\n",
      "2\n",
      "frankia, sp., isolated, ambrosia, production\n",
      "-32.83207130432129 39.02933120727539 104\n",
      "0.07741316339817451 0.07567859122433368 0.10825916606464443\n",
      "256.3918230790151\n",
      "std too big\n",
      "\n",
      "3\n",
      "middle_east_respiratory_syndrome, coronavirus, treatment, sars, convalescent\n",
      "-9.132807731628418 -23.198823928833008 51\n",
      "0.025852719300624456 0.04831347866287366 0.05479557751994981\n",
      "277.4351149965737\n",
      "\n",
      "4\n",
      "toll-like, innate_immune, receptor, receptors, 8\n",
      "-9.87115478515625 8.040507316589355 18\n",
      "0.02849485626647302 0.04899496090338663 0.056678594086051054\n",
      "36.80951512658132\n",
      "\n",
      "5\n",
      "carboxylesterase_1, hereditary_angioedema, gene, mutation, mutations\n",
      "-22.789487838745117 19.44241714477539 21\n",
      "0.04482958688561366 0.04235696229204996 0.06167498775796333\n",
      "48.06005108059593\n",
      "\n",
      "6\n",
      "chlorophyll, nmda_receptor, protein, receptors, allosteric\n",
      "9.903870582580566 24.850072860717773 87\n",
      "0.07297111878049109 0.11210450142500394 0.13376174122605183\n",
      "165.44736665295162\n",
      "std too big\n",
      "\n",
      "7\n",
      "indoor_positioning, indoor, indoor_localization, wi-fi, based\n",
      "52.99466133117676 8.548147201538086 132\n",
      "0.03763253396294586 0.023461283225581334 0.04434680848790832\n",
      "415.4773629882837\n",
      "\n",
      "8\n",
      "police, effect, model, force, findings\n",
      "9.93327522277832 -43.940181732177734 19\n",
      "0.06418975089453692 0.04813445490063333 0.08023247390230347\n",
      "30.219340580749062\n",
      "\n",
      "9\n",
      "middle_east_respiratory_syndrome, mers-cov, coronavirus, transmission, mers\n",
      "-9.291117191314697 -30.684757232666016 88\n",
      "0.04016110810357187 0.03559744724435826 0.05366649657301709\n",
      "410.57116599174503\n",
      "\n",
      "10\n",
      "mhc_class, birds, variation, ii, genes\n",
      "-41.36316108703613 18.498282432556152 22\n",
      "0.0065049717630811296 0.012224069767446024 0.013847113031887575\n",
      "59.82444811723577\n",
      "\n",
      "11\n",
      "china, 2019-ncov, outbreak, wuhan,, pneumonia\n",
      "-30.508604049682617 -13.98934268951416 101\n",
      "0.04981229494761281 0.058359859418569957 0.0767276867845188\n",
      "439.36282838489166\n",
      "\n",
      "12\n",
      "rna, sars-coronavirus, complex, helicase, nonstructural\n",
      "3.439861536026001 -8.146056175231934 31\n",
      "0.039222922208211275 0.008796807238205827 0.04019728155158684\n",
      "122.6314768636618\n",
      "\n",
      "13\n",
      "policy, systems:, advice, capacity, consultants\n",
      "26.29372501373291 -51.387237548828125 18\n",
      "0.00584899271503022 0.005453890564071207 0.007997226898453084\n",
      "28.853840492962878\n",
      "\n",
      "14\n",
      "patients, lung, distress, influenza, angiotensin\n",
      "-25.563318252563477 -2.225185990333557 48\n",
      "0.12138619640637198 0.12726351118030482 0.17587100373838174\n",
      "89.61354316742262\n",
      "std too big\n",
      "\n",
      "15\n",
      "welfare, europe, policy, preferences, insecurity\n",
      "26.817556381225586 -64.12828826904297 37\n",
      "0.00949668863384638 0.012045806184525939 0.01533911802038795\n",
      "78.32303962450416\n",
      "\n",
      "16\n",
      "coronavirus, protein, envelope_protein, severe_acute_respiratory_syndrome, e\n",
      "-2.059544086456299 -0.45696789026260376 39\n",
      "0.015075725429748664 0.023673750610056003 0.028066420597936913\n",
      "94.49341918052586\n",
      "\n",
      "17\n",
      "mers-cov, middle_east_respiratory_syndrome, coronavirus, vaccine, receptor-binding_domain\n",
      "-1.4798554182052612 -28.198047637939453 65\n",
      "0.021673271012271424 0.026079376008919197 0.033909652451565395\n",
      "405.40665378475376\n",
      "\n",
      "18\n",
      "zika_virus, concentrates, platelet, dengue, reduction\n",
      "-21.88584041595459 2.832048177719116 46\n",
      "0.06180260359959118 0.08348198101352179 0.1038691627271069\n",
      "94.59697404590753\n",
      "std too big\n",
      "\n",
      "19\n",
      "acute_lymphoblastic_leukemia, sequencing, variation, acute_lymphoblastic, leukaemia\n",
      "-14.739650249481201 23.98355484008789 46\n",
      "0.026370133407566925 0.024312005260262115 0.03586722090862198\n",
      "142.26533825100233\n",
      "\n",
      "20\n",
      "genome, genomes, microbial, metagenomic, macrococcus\n",
      "-24.3460054397583 38.47444152832031 92\n",
      "0.03563169917597907 0.0772614121730653 0.08508198279978993\n",
      "400.3167070908694\n",
      "\n",
      "21\n",
      "public, performance, accounting, public_sector, management\n",
      "22.1043758392334 -49.870521545410156 53\n",
      "0.029985660601801887 0.02649108484749611 0.04001146608315786\n",
      "122.66833773788842\n",
      "\n",
      "22\n",
      "thermobifida, polyethylene, terephthalate, structural, hydrolysis\n",
      "4.819597005844116 35.27365493774414 10\n",
      "0.0017549789199350926 0.001917179286918533 0.002599139747609966\n",
      "16.062153404369617\n",
      "\n",
      "23\n",
      "society, research, life, annual, surveillance\n",
      "-28.18621253967285 16.320935249328613 86\n",
      "0.24439689474349235 0.18343566701478897 0.30557893594523533\n",
      "144.71451742208507\n",
      "std too big\n",
      "\n",
      "24\n",
      "inertial, gait, wearable, motion, kinematics\n",
      "70.88712310791016 2.4808733463287354 61\n",
      "0.020703359723273116 0.02706253675014583 0.03407359680433284\n",
      "190.06366374447225\n",
      "\n",
      "25\n",
      "phylogenomic, phylogeny, phylogenetic, phylogenetics, new\n",
      "-43.82094955444336 24.053661346435547 55\n",
      "0.05250655063449071 0.05715196441187574 0.0776098247367474\n",
      "292.60921241484135\n",
      "\n",
      "26\n",
      "prediction, peptide, protein, database, peptides\n",
      "-0.11516466736793518 21.02427101135254 62\n",
      "0.04576077679214866 0.044400812989738345 0.0637611236316501\n",
      "166.58199683210904\n",
      "\n",
      "27\n",
      "localization, localization_wireless_sensor, tracking, algorithm, networks\n",
      "46.96911811828613 -1.7381477355957031 54\n",
      "0.03891838766688007 0.04079450940673546 0.05638113954440608\n",
      "116.97271803065239\n",
      "\n",
      "28\n",
      "leisure, constraints, specialization, recreation, negotiation\n",
      "-7.542612552642822 -46.46993637084961 15\n",
      "0.005357466665588628 0.004476928328188981 0.0069817860414534555\n",
      "27.883144467259875\n",
      "\n",
      "29\n",
      "underwater_wireless_sensor_networks, underwater_sensor_networks, underwater_acoustic_sensor_networks, routing_protocol, localization\n",
      "29.72447395324707 -0.706668496131897 61\n",
      "0.015101363791269891 0.013373577933655475 0.020171856015345588\n",
      "183.08776688306196\n",
      "\n",
      "30\n",
      "doa_estimation, array, coarray, direction, mimo\n",
      "52.325265884399414 -10.492863655090332 16\n",
      "0.00322523434344616 0.0036092940135230823 0.004840365672777021\n",
      "43.76257661999131\n",
      "\n",
      "31\n",
      "small_rna, salmonella, regulatory, hfq, typhimurium\n",
      "-12.738247871398926 49.22489929199219 25\n",
      "0.012617892109053333 0.016762394235296 0.020980683062645864\n",
      "53.56391529109098\n",
      "\n",
      "32\n",
      "education, physical, job, teacher, hospitality\n",
      "9.60732126235962 -48.11747360229492 18\n",
      "0.008654203491514302 0.05916307539251983 0.05979268122415589\n",
      "30.36039444014954\n",
      "\n",
      "33\n",
      "genetic_programming, semantic, geometric, bloat, crossover\n",
      "32.98885154724121 26.98097324371338 28\n",
      "0.008278342470590207 0.009998144752095223 0.012980518192438497\n",
      "57.01253154206819\n",
      "\n",
      "34\n",
      "femtosecond, x-ray, laser, crystallography, serial\n",
      "15.003948211669922 18.75078582763672 17\n",
      "0.004934629877934574 0.004103521473882655 0.006417901566619783\n",
      "57.557818856801575\n",
      "\n",
      "35\n",
      "genome, assembly, fungal, pathogen, wheat\n",
      "-19.6341552734375 30.890378952026367 61\n",
      "0.023643411702093966 0.01950251014999572 0.030648961141699405\n",
      "307.9701994479684\n",
      "\n",
      "36\n",
      "bats, bat, african, viruses, fruit_bats\n",
      "-31.15316390991211 -30.426525115966797 44\n",
      "0.012413065201243832 0.06357648478557758 0.06477695273306014\n",
      "119.61643928783435\n",
      "\n",
      "37\n",
      "influenza, pandemic, transmission, masks, health_care\n",
      "-36.725555419921875 -7.2529401779174805 63\n",
      "0.06420869016805651 0.10257100567722877 0.12101060738107042\n",
      "144.05935101965474\n",
      "std too big\n",
      "\n",
      "38\n",
      "bat, sars-like, coronaviruses, bats, coronavirus\n",
      "-23.95526123046875 -25.21247100830078 45\n",
      "0.03846643756515711 0.050041319664151554 0.06311735492464755\n",
      "297.4781503370295\n",
      "\n",
      "39\n",
      "kalman_filter, localization, navigation, slam, vehicle\n",
      "62.451377868652344 16.606855392456055 81\n",
      "0.050869181219318736 0.045202978225191565 0.06805132503009058\n",
      "227.79669925499306\n",
      "\n",
      "40\n",
      "microscopy, cryo-electron, cryo-em_structure, coronavirus_spike_protein, coronavirus\n",
      "5.092611789703369 -21.854299545288086 47\n",
      "0.021270742560938184 0.013731217323703523 0.025317796473715536\n",
      "287.6212671340052\n",
      "\n",
      "41\n",
      "incorporating, proteins, structure, chemistry, m2\n",
      "1.2080624103546143 8.772252082824707 7\n",
      "0.0032592656435550572 0.004244060651629076 0.00535115532852155\n",
      "19.15730922012818\n",
      "\n",
      "42\n",
      "kinetochore, toolkit, brucei, kinetoplastid, cell\n",
      "1.2453726530075073 50.057891845703125 17\n",
      "0.008976457003434376 0.05109336956314929 0.051875901858676694\n",
      "54.474819930370124\n",
      "\n",
      "43\n",
      "human, gene, htlv-1, drosophila, expression\n",
      "-10.930041313171387 29.30903434753418 70\n",
      "0.048005436688055673 0.1230850859668821 0.13211532968995565\n",
      "131.14994633318005\n",
      "std too big\n",
      "\n",
      "44\n",
      "management, conflict, organizational, multilevel, performance\n",
      "7.9133405685424805 -51.4705810546875 41\n",
      "0.017880643389326413 0.030194185134963376 0.03509139814798928\n",
      "74.68170527412403\n",
      "\n",
      "45\n",
      "ebola_virus, virus, hepatitis_e, h7n9, influenza_virus\n",
      "-15.841632843017578 -3.1135752201080322 72\n",
      "0.033029136748984285 0.03209929256732463 0.04605744736419732\n",
      "161.25490012316413\n",
      "\n",
      "46\n",
      "free, groups, laminations, ℝ-trees, complex\n",
      "25.514127731323242 7.252660036087036 10\n",
      "0.0015840254490805676 0.0018265773763482848 0.0024177513385628137\n",
      "21.628766281879642\n"
     ]
    }
   ],
   "source": [
    "##### clDict = pickle.load(open(\"clDict.p\", \"rb\"))\n",
    "\n",
    "label_source_df = pd.DataFrame(index=clDict.keys(), columns=[\"x\", \"y\", \"text\"])\n",
    "\n",
    "for cl_id, cluster in clDict.items():\n",
    "    #print(cluster)\n",
    "    cl_dois = cluster[\"dois\"]\n",
    "    x = np.zeros(len(cl_dois))\n",
    "    y = np.zeros(len(cl_dois))\n",
    "    w = np.zeros(len(cl_dois))\n",
    "    r = np.zeros(len(cl_dois))\n",
    "    g = np.zeros(len(cl_dois))\n",
    "    b = np.zeros(len(cl_dois))\n",
    "    n = 0\n",
    "    \n",
    "    \n",
    "    for i, doi in enumerate(cl_dois):\n",
    "        if doi in df.index:\n",
    "            n += 1\n",
    "            x[i] = df.at[doi, \"xval\"]\n",
    "            y[i] = df.at[doi, \"yval\"]\n",
    "            w[i] = df.at[doi, \"refCountNormalized\"]\n",
    "        else:\n",
    "            w[i] = 0\n",
    "    #if n < 3:\n",
    "    #    continue\n",
    "    \n",
    "    xx = np.median(x)\n",
    "    yy = np.median(y)\n",
    "    \n",
    "    #xx = np.average(x, weights=w)\n",
    "    #yy = np.average(y, weights=w)\n",
    "    \n",
    "    x_std = np.std(x/x_span)\n",
    "    y_std = np.std(y/y_span)\n",
    "    \n",
    "    print()\n",
    "    print(cl_id)\n",
    "    print(\", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.1 and totalFreq>=3][:5]))\n",
    "    print(xx,yy,len(cl_dois))\n",
    "    print(x_std, y_std , np.sqrt(x_std**2 + y_std**2))\n",
    "    print(np.sum(w))\n",
    "    \n",
    "    \n",
    "    if np.sum(w) < 10:\n",
    "        print(\"w too small\")\n",
    "        continue\n",
    "    \n",
    "    std_max = 0.1\n",
    "    \n",
    "    if x_std**2 + y_std**2 > std_max**2:\n",
    "        print(\"std too big\")\n",
    "        #continue\n",
    "    \n",
    "    \n",
    "    \n",
    "    label_source_df.at[cl_id, \"x\"] = xx\n",
    "    label_source_df.at[cl_id, \"y\"] = yy\n",
    "    #label_source.at[cl_id, \"text\"] = \", \".join([s for s, n in cluster[\"top_authors\"][:3]]) \\\n",
    "    #    + \" | \" + \",\\n\".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.25 and clFreq>=3][:2])\n",
    "    \n",
    "    #label_source_df.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.3])\n",
    "    label_source_df.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.1 and totalFreq>=3][:2])\n",
    "    #label_source.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.225 and clFreq>=2][:2])\n",
    "    #label_source.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=0.01 and clFreq>=1][:1])\n",
    "    #label_source.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"]][:1])\n",
    "    #label_source.at[cl_id, \"text\"] = \", \".join([s for s, totalFreq, clFreq, v in cluster[\"vec\"] if v>=500])\n",
    "    if len(label_source_df.at[cl_id, \"text\"]) == 0:\n",
    "        label_source_df.at[cl_id, \"text\"] = np.nan\n",
    "    \n",
    "    top_authors = cluster[\"top_authors\"]\n",
    "    label_source_df.at[cl_id, \"authors\"] = \", \".join([a for a,n in top_authors][:3])\n",
    "    \n",
    "    #print(xx,yy,len(cl_dois),label_source_df.at[cl_id, \"text\"])\n",
    "    \n",
    "\n",
    "                 \n",
    "    #xy = X_embedded[, dtype=int), :]\n",
    "    #x = df[\"xval\"].loc[cl_dois]\n",
    "    #print()\n",
    "\n",
    "#label_source = \n",
    "label_source_df = label_source_df[pd.notnull(label_source_df[\"x\"])]\n",
    "label_source_df = label_source_df[pd.notnull(label_source_df[\"text\"])]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 16:16:25,844 : INFO : Session output file 'test.html' already exists, will be overwritten.\n"
     ]
    }
   ],
   "source": [
    "from bokeh.plotting import figure, output_file, show, save, output_notebook\n",
    "from bokeh.models import HoverTool, Range1d, LabelSet, LassoSelectTool, CustomJS, TapTool, OpenURL\n",
    "from bokeh.models.sources import ColumnDataSource\n",
    "from bokeh.io import reset_output\n",
    "\n",
    "reset_output()\n",
    "\n",
    "\n",
    "#class hover(HoverTool):\n",
    "#    hover.tooltips = [\n",
    "#    (\"index\", \"$index\"),\n",
    "\n",
    "\n",
    "hover = HoverTool()\n",
    "hover.tooltips = [\n",
    "    (\"Title\", \"@title\"),\n",
    "    (\"Autoren\", \"@author\"),\n",
    "    (\"Veröffentlicht\", \"@issued\"),\n",
    "    (\"Journal\", \"@containerTitle\"),\n",
    "    (\"Referenzen\", \"@refCount\"),\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "s1 = ColumnDataSource(data=df)\n",
    "\n",
    "\n",
    "callback = CustomJS(code=\"\"\"\n",
    "// the event that triggered the callback is cb_obj:\n",
    "// The event type determines the relevant attributes\n",
    "console.log('Tap event occured at x-position: ' + cb_obj.x)\n",
    "\"\"\")\n",
    "\n",
    "\n",
    "def lasso_cb(x):\n",
    "    print(x)\n",
    "\n",
    "#lasso = LassoSelectTool(callback=lasso_cb)\n",
    "lasso = LassoSelectTool()\n",
    "\n",
    "\n",
    "p = figure(plot_width=1200, plot_height=1000, active_scroll = \"wheel_zoom\")\n",
    "#p = figure(plot_width=1200, plot_height=1000, output_backend=\"webgl\", active_scroll = \"wheel_zoom\")\n",
    "p.circle(x='xval', y='yval', size=\"scatterSize\", color=\"rgb\", alpha=\"alpha\", source=s1)\n",
    "\n",
    "p.add_tools(hover)\n",
    "p.add_tools(lasso)\n",
    "\n",
    "p.js_on_event('tap', callback)\n",
    "\n",
    "#p.toolbar.active_scroll = \"wheel_zoom\"\n",
    "\n",
    "df[\"xval\"].max()\n",
    "\n",
    "left, right, bottom, top = -120, 120, -90, 90\n",
    "left, right, bottom, top = -80, 80, -60, 60\n",
    "left, right, bottom, top = -100, 100, -90, 90\n",
    "p.x_range=Range1d(left, right)\n",
    "p.y_range=Range1d(bottom, top)\n",
    "\n",
    "try:\n",
    "    label_source = ColumnDataSource(label_source_df)   \n",
    "\n",
    "    p.scatter(x='x', y='y', source=label_source, size=8, color=\"black\", marker=\"diamond\")\n",
    "    \n",
    "    labels = LabelSet(x='x', y='y', x_offset=-20, y_offset=5, text='text', level='glyph', \n",
    "                  source=label_source, render_mode='canvas')\n",
    "    #author_labels = LabelSet(x='x', y='y', x_offset=10, y_offset=-15, text='authors', level='glyph', \n",
    "    #              source=label_source, render_mode='canvas')\n",
    "\n",
    "    p.add_layout(labels)\n",
    "    #p.add_layout(author_labels)\n",
    "except ValueError:\n",
    "    pass\n",
    "except NameError:\n",
    "    pass\n",
    "\n",
    "output_file(\"test.html\")\n",
    "#output_notebook()\n",
    "#show(p)\n",
    "#save(p)\n",
    "\n",
    "from bokeh.layouts import widgetbox, column, row\n",
    "from bokeh.models import ColumnDataSource\n",
    "from bokeh.models.widgets import DataTable, DateFormatter, TableColumn\n",
    "\n",
    "s2 = ColumnDataSource(df)\n",
    "\n",
    "columns = [\n",
    "        TableColumn(field=\"clusterNo\", title=\"Cluster\"),\n",
    "        TableColumn(field=\"title\", title=\"Title\"),\n",
    "        TableColumn(field=\"author\", title=\"Autoren\"),\n",
    "        TableColumn(field=\"issued\", title=\"Veröffentlichungsjahr\"),\n",
    "    ]\n",
    "data_table = DataTable(source=s1, columns=columns, width=640, height=800)\n",
    "\n",
    "#p.add_widget(widgetbox(data_table))\n",
    "\n",
    "#show(p)\n",
    "\n",
    "\n",
    "#s1.callback = \n",
    "jscode = CustomJS(args=dict(s2=s2), code=\"\"\"\n",
    "    var inds = cb_obj.get('selected')['1d'].indices;\n",
    "    var d1 = cb_obj.get('data');\n",
    "    var d2 = s2.get('data');\n",
    "    \n",
    "    \n",
    "    d2[\"title] = []\n",
    "    d2[\"author] = []\n",
    "    d2[\"issued] = []\n",
    "    for (i = 0; i < inds.length; i++) {\n",
    "        d2[\"title\"].push(d1[\"title\"][inds[i]])\n",
    "        d2[\"author\"].push(d1[\"author\"][inds[i]])\n",
    "        d2[\"issued\"].push(d1[\"issued\"][inds[i]])\n",
    "    }\n",
    "    s2.trigger('change');\n",
    "    \n",
    "\"\"\")\n",
    "\n",
    "p.add_tools(TapTool(callback=OpenURL(url=\"http://doi.org/@doi\")))\n",
    "\n",
    "# use the \"color\" column of the CDS to complete the URL\n",
    "# e.g. if the glyph at index 10 is selected, then @color\n",
    "# will be replaced with source.data['color'][10]\n",
    "#url = \"http://doi.org/@doi\"\n",
    "#taptool = p.select(type=TapTool)\n",
    "##taptool = TapTool()\n",
    "#taptool.callback = OpenURL(url=url)\n",
    "##p.add_tools(taptool)\n",
    "\n",
    "layout = row(p, widgetbox(data_table))\n",
    "#layout = column(p, widgetbox(data_table))\n",
    "\n",
    "show(layout)\n",
    "#save(layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (<ipython-input-204-6aaf1f276005>, line 4)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-204-6aaf1f276005>\"\u001b[1;36m, line \u001b[1;32m4\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sample(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vars(s1)\n",
    "#s1.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"display: table;\"><div style=\"display: table-row;\"><div style=\"display: table-cell;\"><b title=\"bokeh.models.sources.ColumnDataSource\">ColumnDataSource</b>(</div><div style=\"display: table-cell;\">id&nbsp;=&nbsp;'1942', <span id=\"2121\" style=\"cursor: pointer;\">&hellip;)</span></div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">callback&nbsp;=&nbsp;None,</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">data&nbsp;=&nbsp;{'y': [-13.080853462219238, -10.985311508178711, 39.02933120727539, -23.198823928833008, 8.040507316589355, 19.44241714477539, 24.850072860717773, 8.548147201538086, -43.940181732177734, -30.684757232666016, 18.498282432556152, -13.98934268951416, -8.146056175231934, -51.387237548828125, -2.225185990333557, -64.12828826904297, -0.45696789026260376, -28.198047637939453, 2.832048177719116, 23.98355484008789, 38.47444152832031, -49.870521545410156, 35.27365493774414, 16.320935249328613, 2.4808733463287354, 24.053661346435547, 21.02427101135254, -1.7381477355957031, -46.46993637084961, -0.706668496131897, -10.492863655090332, 49.22489929199219, -48.11747360229492, 26.98097324371338, 18.75078582763672, 30.890378952026367, -30.426525115966797, -7.2529401779174805, -25.21247100830078, 16.606855392456055, -21.854299545288086, 8.772252082824707, 50.057891845703125, 29.30903434753418, -51.4705810546875, -3.1135752201080322, 7.252660036087036], 'text': ['wireless_sensor_networks, networks', 'porcine, coronavirus', 'frankia, sp.', 'middle_east_respiratory_syndrome, coronavirus', 'toll-like, innate_immune', 'carboxylesterase_1, hereditary_angioedema', 'chlorophyll, nmda_receptor', 'indoor_positioning, indoor', 'police, effect', 'middle_east_respiratory_syndrome, mers-cov', 'mhc_class, birds', 'china, 2019-ncov', 'rna, sars-coronavirus', 'policy, systems:', 'patients, lung', 'welfare, europe', 'coronavirus, protein', 'mers-cov, middle_east_respiratory_syndrome', 'zika_virus, concentrates', 'acute_lymphoblastic_leukemia, sequencing', 'genome, genomes', 'public, performance', 'thermobifida, polyethylene', 'society, research', 'inertial, gait', 'phylogenomic, phylogeny', 'prediction, peptide', 'localization, localization_wireless_sensor', 'leisure, constraints', 'underwater_wireless_sensor_networks, underwater_sensor_networks', 'doa_estimation, array', 'small_rna, salmonella', 'education, physical', 'genetic_programming, semantic', 'femtosecond, x-ray', 'genome, assembly', 'bats, bat', 'influenza, pandemic', 'bat, sars-like', 'kalman_filter, localization', 'microscopy, cryo-electron', 'incorporating, proteins', 'kinetochore, toolkit', 'human, gene', 'management, conflict', 'ebola_virus, virus', 'free, groups'], 'index': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46], 'x': [40.25177001953125, -1.8652465343475342, -32.83207130432129, -9.132807731628418, -9.87115478515625, -22.789487838745117, 9.903870582580566, 52.99466133117676, 9.93327522277832, -9.291117191314697, -41.36316108703613, -30.508604049682617, 3.439861536026001, 26.29372501373291, -25.563318252563477, 26.817556381225586, -2.059544086456299, -1.4798554182052612, -21.88584041595459, -14.739650249481201, -24.3460054397583, 22.1043758392334, 4.819597005844116, -28.18621253967285, 70.88712310791016, -43.82094955444336, -0.11516466736793518, 46.96911811828613, -7.542612552642822, 29.72447395324707, 52.325265884399414, -12.738247871398926, 9.60732126235962, 32.98885154724121, 15.003948211669922, -19.6341552734375, -31.15316390991211, -36.725555419921875, -23.95526123046875, 62.451377868652344, 5.092611789703369, 1.2080624103546143, 1.2453726530075073, -10.930041313171387, 7.9133405685424805, -15.841632843017578, 25.514127731323242], 'authors': ['Liu, Wang, Kim', 'Chen, Huang, Zhang', 'Chan, Yin, Gtari', 'Chan, Yuen, Li', 'Chen, Wang, Zhang', 'Zhu, Markowitz, Bork', 'Reynisson, Chen, Hansen', 'Liu, Zhang, Wang', 'Grisolía, Willis, Terrill', 'Kim, Lee, Drosten', 'Braun, Sheldon, Murrell', 'Wang, Li, Zhang', 'Snijder, Decroly, Denison', 'Howlett, Migone', 'Li, Liu, Zhao', 'Marx, Chung, Soroka', 'Enjuanes, DeDiego, Nieto-Torres', 'Wang, Jiang, Li', 'Zhang, Chen, Seltsam', 'Chen, Ma, Zhang', 'Li, Chen, Wu', 'Lapsley, Mitchell, Steccolini', 'Zimmermann, Yang, Wei', 'Kim, Lee, Zhang', 'Aminian, Sabatini, Chardonnens', 'Li, Wang, Giribet', 'Wang, Nielsen, Li', 'Tomic, Beko, Zhang', 'Oh, Lyu', 'Khan, Javaid, Wang', 'Wang, Chen, Li', 'Vogel, Lee, Luisi', 'Richards, Templin, Hsieh', 'Silva, O’Neill, Castelli', 'Fromme, Boutet, Wang', 'Li, Kim, Wang', 'Wang, Hayman, Cunningham', 'Nishiura, Ferguson, Colizza', 'Zhang, Wang, Li', 'Liu, Li, Wang', 'Li, Wang, Bosch', 'Huang, Zhou, Chou', 'Akiyoshi, Gull, Hayashi', 'Li, Wang, Nguyen', 'Liu, Tay, Li', 'Wang, Zhang, Liu', 'Lustig, Bestvina, Coulbois']},</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">js_event_callbacks&nbsp;=&nbsp;{},</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">js_property_callbacks&nbsp;=&nbsp;{},</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">name&nbsp;=&nbsp;None,</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">selected&nbsp;=&nbsp;Selection(id='1974', ...),</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">selection_policy&nbsp;=&nbsp;UnionRenderers(id='1973', ...),</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">subscribed_events&nbsp;=&nbsp;[],</div></div><div class=\"2120\" style=\"display: none;\"><div style=\"display: table-cell;\"></div><div style=\"display: table-cell;\">tags&nbsp;=&nbsp;[])</div></div></div>\n",
       "<script>\n",
       "(function() {\n",
       "  var expanded = false;\n",
       "  var ellipsis = document.getElementById(\"2121\");\n",
       "  ellipsis.addEventListener(\"click\", function() {\n",
       "    var rows = document.getElementsByClassName(\"2120\");\n",
       "    for (var i = 0; i < rows.length; i++) {\n",
       "      var el = rows[i];\n",
       "      el.style.display = expanded ? \"none\" : \"table-row\";\n",
       "    }\n",
       "    ellipsis.innerHTML = expanded ? \"&hellip;)\" : \"&lsaquo;&lsaquo;&lsaquo;\";\n",
       "    expanded = !expanded;\n",
       "  });\n",
       "})();\n",
       "</script>\n"
      ],
      "text/plain": [
       "ColumnDataSource(id='1942', ...)"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "label_source\n",
    "#type(label_source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\n",
    "        TableColumn(field=\"title\", title=\"Title\"),\n",
    "        TableColumn(field=\"author\", title=\"Autoren\"),\n",
    "        TableColumn(field=\"issued\", title=\"Jahr\", formatter=DateFormatter()),\n",
    "    ]\n",
    "data_table = DataTable(source=ColumnDataSource(df), columns=columns, width=400, height=280)\n",
    "\n",
    "show(widgetbox(data_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 31261\n",
      "1000 31261\n",
      "2000 31261\n",
      "3000 31261\n",
      "4000 31261\n",
      "5000 31261\n",
      "6000 31261\n",
      "7000 31261\n",
      "8000 31261\n",
      "9000 31261\n",
      "10000 31261\n",
      "11000 31261\n",
      "12000 31261\n",
      "13000 31261\n",
      "14000 31261\n",
      "15000 31261\n",
      "16000 31261\n",
      "17000 31261\n",
      "18000 31261\n",
      "19000 31261\n",
      "20000 31261\n",
      "21000 31261\n",
      "22000 31261\n",
      "23000 31261\n",
      "24000 31261\n",
      "25000 31261\n",
      "26000 31261\n",
      "27000 31261\n",
      "28000 31261\n",
      "29000 31261\n",
      "30000 31261\n",
      "31000 31261\n"
     ]
    }
   ],
   "source": [
    "#df_all = pd.DataFrame(index=set(k for k in work_dict.keys() if k in df), columns=[\"doi\", \"clusterID\", \"issued\", \"title\", \"authors\", \"containerTitles\", \"refCount\", \"link\", \"inPlot\"])\n",
    "#for i, doi in enumerate(work_dict.keys()):\n",
    "\n",
    "#df_dois = set(dois_to_plot & work_dict.keys())\n",
    "df_dois = set(reference_vectors.vocab.keys() & work_dict.keys())\n",
    "\n",
    "df_all = pd.DataFrame(index=df_dois, columns=[\"doi\", \"issued\", \"issuedYear\", \"title\", \"authors\", \"containerTitles\", \"refCount\", \"refCountNormalized\", \"link\", \"clusterID\", \"clusterKWs\", \"inPlot\"])\n",
    "for i, doi in enumerate(df_dois):\n",
    "    \n",
    "    if i % 1000 == 0:\n",
    "        print(i, len(df_dois))\n",
    "    \n",
    "    pub = work_dict[doi]\n",
    "    #print(pub)\n",
    "    df_all.at[doi, \"doi\"] = doi\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    try:\n",
    "        df_all.at[doi, \"title\"] = pub[\"title\"][0]\n",
    "    except IndexError:\n",
    "        pass\n",
    "        #df_all.at[doi, \"title\"] = None\n",
    "    except KeyError:\n",
    "        pass\n",
    "    df_all.at[doi, \"issued\"] = \"-\".join([str(x) for x in pub[\"issued\"][\"date-parts\"][0]])\n",
    "    df_all.at[doi, \"issuedYear\"] = pub[\"issued\"][\"date-parts\"][0][0]\n",
    "    df_all.at[doi, \"refCount\"] = ref_counts[doi]\n",
    "    try:\n",
    "        pub_authors = [a[\"family\"] for a in pub[\"author\"]]\n",
    "        df_all.at[doi, \"authors\"] = \", \".join(pub_authors)\n",
    "    except KeyError:\n",
    "        df_all.at[doi, \"authors\"] = \"unavailable\"\n",
    "    try:\n",
    "        df_all.at[doi, \"containerTitles\"] = pub[\"container-title\"]\n",
    "    except KeyError:\n",
    "        pass\n",
    "    try:\n",
    "        df_all.at[doi, \"link\"] = \"=Hyperlink(\\\"https://doi.org/{}\\\")\".format(doi)\n",
    "    except KeyError:\n",
    "        pass\n",
    "    \n",
    "    if doi in dois_to_plot:\n",
    "        df_all.at[doi, \"inPlot\"] = True\n",
    "        clid = df.at[doi, \"clusterNo\"]\n",
    "        df_all.at[doi, \"clusterID\"] = clid\n",
    "        if clid in clDict:\n",
    "            df_all.at[doi, \"clusterKWs\"] = \", \".join([s for s, totalFreq, clFreq, v in clDict[clid][\"vec\"] if v>=0.1 and totalFreq>=3])\n",
    "    else:\n",
    "        df_all.at[doi, \"inPlot\"] = False\n",
    "        \n",
    "    try:\n",
    "        year = pub[\"issued\"][\"date-parts\"][0][0]\n",
    "        x = ref_counts[doi]\n",
    "        if year in lambda_mle_ref_count:\n",
    "            l = lambda_mle_ref_count[year]\n",
    "        else:\n",
    "            l = lambda_mle_ref_count_default\n",
    "        #df_all.at[doi, \"expRefCountPercentile\"] = 1. - np.exp(-l*x)\n",
    "        df_all.at[doi, \"refCountNormalized\"] = l*x\n",
    "    except KeyError:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.sort_values(by=\"refCount\", ascending=False)\n",
    "df_all.sample(10)\n",
    "df_all.to_excel(\"df_all.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2303"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_inPlot = df_all[df_all[\"inPlot\"]==True]\n",
    "len(df_all_inPlot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_inPlot.to_excel(\"df_all_inPlot.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>doi</th>\n",
       "      <th>issued</th>\n",
       "      <th>issuedYear</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>containerTitles</th>\n",
       "      <th>refCount</th>\n",
       "      <th>refCountNormalized</th>\n",
       "      <th>link</th>\n",
       "      <th>clusterID</th>\n",
       "      <th>clusterKWs</th>\n",
       "      <th>inPlot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10.1086/155591</th>\n",
       "      <td>10.1086/155591</td>\n",
       "      <td>1977-10</td>\n",
       "      <td>1977</td>\n",
       "      <td>The size distribution of interstellar grains</td>\n",
       "      <td>Mathis, Rumpl, Nordsieck</td>\n",
       "      <td>[The Astrophysical Journal]</td>\n",
       "      <td>10</td>\n",
       "      <td>0.96164</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1086/155591\")</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1080/00336297.2016.1180308</th>\n",
       "      <td>10.1080/00336297.2016.1180308</td>\n",
       "      <td>2016-6-9</td>\n",
       "      <td>2016</td>\n",
       "      <td>Social Justice and the Future of Higher Educat...</td>\n",
       "      <td>Culp</td>\n",
       "      <td>[Quest]</td>\n",
       "      <td>4</td>\n",
       "      <td>1.63233</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1080/00336297.2...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1177/1075547010378658</th>\n",
       "      <td>10.1177/1075547010378658</td>\n",
       "      <td>2010-8-18</td>\n",
       "      <td>2010</td>\n",
       "      <td>Gender Differences in Scientists’ Public Outre...</td>\n",
       "      <td>Crettaz von Roten</td>\n",
       "      <td>[Science Communication]</td>\n",
       "      <td>5</td>\n",
       "      <td>1.25598</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1177/1075547010...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1063/1.445869</th>\n",
       "      <td>10.1063/1.445869</td>\n",
       "      <td>1983-7-15</td>\n",
       "      <td>1983</td>\n",
       "      <td>Comparison of simple potential functions for s...</td>\n",
       "      <td>Jorgensen, Chandrasekhar, Madura, Impey, Klein</td>\n",
       "      <td>[The Journal of Chemical Physics]</td>\n",
       "      <td>56</td>\n",
       "      <td>7.32954</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1063/1.445869\")</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1007/s10648-013-9223-5</th>\n",
       "      <td>10.1007/s10648-013-9223-5</td>\n",
       "      <td>2013-4-12</td>\n",
       "      <td>2013</td>\n",
       "      <td>On Going (Way) Beyond One’s Data: A Proposal t...</td>\n",
       "      <td>Robinson, Levin, Schraw, Patall, Hunt</td>\n",
       "      <td>[Educational Psychology Review]</td>\n",
       "      <td>10</td>\n",
       "      <td>3.00607</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1007/s10648-013...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.5194/acp-7-6131-2007</th>\n",
       "      <td>10.5194/acp-7-6131-2007</td>\n",
       "      <td>2007-12-14</td>\n",
       "      <td>2007</td>\n",
       "      <td>Closure study between chemical composition and...</td>\n",
       "      <td>Gysel, Crosier, Topping, Whitehead, Bower, Cub...</td>\n",
       "      <td>[Atmospheric Chemistry and Physics]</td>\n",
       "      <td>21</td>\n",
       "      <td>3.99416</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.5194/acp-7-6131...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.3390/s111009887</th>\n",
       "      <td>10.3390/s111009887</td>\n",
       "      <td>2011-10-20</td>\n",
       "      <td>2011</td>\n",
       "      <td>Cooperative MIMO Communication at Wireless Sen...</td>\n",
       "      <td>Islam, Han</td>\n",
       "      <td>[Sensors]</td>\n",
       "      <td>4</td>\n",
       "      <td>1.08477</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.3390/s111009887\")</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1086/652373</th>\n",
       "      <td>10.1086/652373</td>\n",
       "      <td>2010-6</td>\n",
       "      <td>2010</td>\n",
       "      <td>Conceptual Synthesis in Community Ecology</td>\n",
       "      <td>Vellend</td>\n",
       "      <td>[The Quarterly Review of Biology]</td>\n",
       "      <td>17</td>\n",
       "      <td>4.27033</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1086/652373\")</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1021/nl500515q</th>\n",
       "      <td>10.1021/nl500515q</td>\n",
       "      <td>2014-5-27</td>\n",
       "      <td>2014</td>\n",
       "      <td>Tuning Interlayer Coupling in Large-Area Heter...</td>\n",
       "      <td>Tongay, Fan, Kang, Park, Koldemir, Suh, Narang...</td>\n",
       "      <td>[Nano Letters]</td>\n",
       "      <td>6</td>\n",
       "      <td>1.87038</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1021/nl500515q\")</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10.1006/mssp.2002.1489</th>\n",
       "      <td>10.1006/mssp.2002.1489</td>\n",
       "      <td>2002-7</td>\n",
       "      <td>2002</td>\n",
       "      <td>BLADE-TIP TIMING MEASUREMENT OF SYNCHRONOUS VI...</td>\n",
       "      <td>DIMITRIADIS, CARRINGTON, WRIGHT, COOPER</td>\n",
       "      <td>[Mechanical Systems and Signal Processing]</td>\n",
       "      <td>10</td>\n",
       "      <td>1.67915</td>\n",
       "      <td>=Hyperlink(\"https://doi.org/10.1006/mssp.2002....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         doi      issued  \\\n",
       "10.1086/155591                                10.1086/155591     1977-10   \n",
       "10.1080/00336297.2016.1180308  10.1080/00336297.2016.1180308    2016-6-9   \n",
       "10.1177/1075547010378658            10.1177/1075547010378658   2010-8-18   \n",
       "10.1063/1.445869                            10.1063/1.445869   1983-7-15   \n",
       "10.1007/s10648-013-9223-5          10.1007/s10648-013-9223-5   2013-4-12   \n",
       "10.5194/acp-7-6131-2007              10.5194/acp-7-6131-2007  2007-12-14   \n",
       "10.3390/s111009887                        10.3390/s111009887  2011-10-20   \n",
       "10.1086/652373                                10.1086/652373      2010-6   \n",
       "10.1021/nl500515q                          10.1021/nl500515q   2014-5-27   \n",
       "10.1006/mssp.2002.1489                10.1006/mssp.2002.1489      2002-7   \n",
       "\n",
       "                              issuedYear  \\\n",
       "10.1086/155591                      1977   \n",
       "10.1080/00336297.2016.1180308       2016   \n",
       "10.1177/1075547010378658            2010   \n",
       "10.1063/1.445869                    1983   \n",
       "10.1007/s10648-013-9223-5           2013   \n",
       "10.5194/acp-7-6131-2007             2007   \n",
       "10.3390/s111009887                  2011   \n",
       "10.1086/652373                      2010   \n",
       "10.1021/nl500515q                   2014   \n",
       "10.1006/mssp.2002.1489              2002   \n",
       "\n",
       "                                                                           title  \\\n",
       "10.1086/155591                      The size distribution of interstellar grains   \n",
       "10.1080/00336297.2016.1180308  Social Justice and the Future of Higher Educat...   \n",
       "10.1177/1075547010378658       Gender Differences in Scientists’ Public Outre...   \n",
       "10.1063/1.445869               Comparison of simple potential functions for s...   \n",
       "10.1007/s10648-013-9223-5      On Going (Way) Beyond One’s Data: A Proposal t...   \n",
       "10.5194/acp-7-6131-2007        Closure study between chemical composition and...   \n",
       "10.3390/s111009887             Cooperative MIMO Communication at Wireless Sen...   \n",
       "10.1086/652373                         Conceptual Synthesis in Community Ecology   \n",
       "10.1021/nl500515q              Tuning Interlayer Coupling in Large-Area Heter...   \n",
       "10.1006/mssp.2002.1489         BLADE-TIP TIMING MEASUREMENT OF SYNCHRONOUS VI...   \n",
       "\n",
       "                                                                         authors  \\\n",
       "10.1086/155591                                          Mathis, Rumpl, Nordsieck   \n",
       "10.1080/00336297.2016.1180308                                               Culp   \n",
       "10.1177/1075547010378658                                       Crettaz von Roten   \n",
       "10.1063/1.445869                  Jorgensen, Chandrasekhar, Madura, Impey, Klein   \n",
       "10.1007/s10648-013-9223-5                  Robinson, Levin, Schraw, Patall, Hunt   \n",
       "10.5194/acp-7-6131-2007        Gysel, Crosier, Topping, Whitehead, Bower, Cub...   \n",
       "10.3390/s111009887                                                    Islam, Han   \n",
       "10.1086/652373                                                           Vellend   \n",
       "10.1021/nl500515q              Tongay, Fan, Kang, Park, Koldemir, Suh, Narang...   \n",
       "10.1006/mssp.2002.1489                   DIMITRIADIS, CARRINGTON, WRIGHT, COOPER   \n",
       "\n",
       "                                                          containerTitles  \\\n",
       "10.1086/155591                                [The Astrophysical Journal]   \n",
       "10.1080/00336297.2016.1180308                                     [Quest]   \n",
       "10.1177/1075547010378658                          [Science Communication]   \n",
       "10.1063/1.445869                        [The Journal of Chemical Physics]   \n",
       "10.1007/s10648-013-9223-5                 [Educational Psychology Review]   \n",
       "10.5194/acp-7-6131-2007               [Atmospheric Chemistry and Physics]   \n",
       "10.3390/s111009887                                              [Sensors]   \n",
       "10.1086/652373                          [The Quarterly Review of Biology]   \n",
       "10.1021/nl500515q                                          [Nano Letters]   \n",
       "10.1006/mssp.2002.1489         [Mechanical Systems and Signal Processing]   \n",
       "\n",
       "                              refCount refCountNormalized  \\\n",
       "10.1086/155591                      10            0.96164   \n",
       "10.1080/00336297.2016.1180308        4            1.63233   \n",
       "10.1177/1075547010378658             5            1.25598   \n",
       "10.1063/1.445869                    56            7.32954   \n",
       "10.1007/s10648-013-9223-5           10            3.00607   \n",
       "10.5194/acp-7-6131-2007             21            3.99416   \n",
       "10.3390/s111009887                   4            1.08477   \n",
       "10.1086/652373                      17            4.27033   \n",
       "10.1021/nl500515q                    6            1.87038   \n",
       "10.1006/mssp.2002.1489              10            1.67915   \n",
       "\n",
       "                                                                            link  \\\n",
       "10.1086/155591                      =Hyperlink(\"https://doi.org/10.1086/155591\")   \n",
       "10.1080/00336297.2016.1180308  =Hyperlink(\"https://doi.org/10.1080/00336297.2...   \n",
       "10.1177/1075547010378658       =Hyperlink(\"https://doi.org/10.1177/1075547010...   \n",
       "10.1063/1.445869                  =Hyperlink(\"https://doi.org/10.1063/1.445869\")   \n",
       "10.1007/s10648-013-9223-5      =Hyperlink(\"https://doi.org/10.1007/s10648-013...   \n",
       "10.5194/acp-7-6131-2007        =Hyperlink(\"https://doi.org/10.5194/acp-7-6131...   \n",
       "10.3390/s111009887              =Hyperlink(\"https://doi.org/10.3390/s111009887\")   \n",
       "10.1086/652373                      =Hyperlink(\"https://doi.org/10.1086/652373\")   \n",
       "10.1021/nl500515q                =Hyperlink(\"https://doi.org/10.1021/nl500515q\")   \n",
       "10.1006/mssp.2002.1489         =Hyperlink(\"https://doi.org/10.1006/mssp.2002....   \n",
       "\n",
       "                              clusterID clusterKWs inPlot  \n",
       "10.1086/155591                      NaN        NaN  False  \n",
       "10.1080/00336297.2016.1180308       NaN        NaN  False  \n",
       "10.1177/1075547010378658            NaN        NaN  False  \n",
       "10.1063/1.445869                    NaN        NaN  False  \n",
       "10.1007/s10648-013-9223-5           NaN        NaN  False  \n",
       "10.5194/acp-7-6131-2007             NaN        NaN  False  \n",
       "10.3390/s111009887                  NaN        NaN  False  \n",
       "10.1086/652373                      NaN        NaN  False  \n",
       "10.1021/nl500515q                   NaN        NaN  False  \n",
       "10.1006/mssp.2002.1489              NaN        NaN  False  "
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'10.1021/op200097d' is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-213-676aaa0251f0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mdois_to_plot\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"10.1021/op200097d\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m: '10.1021/op200097d' is not in list"
     ]
    }
   ],
   "source": [
    "dois_to_plot.index(\"10.1021/op200097d\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Idee:\n",
    "\n",
    "für Bestimmung der Wahrscheinlichkeit für Referenz eines Papers bei einer vorhandenen Menge an bekannten Referenzen\n",
    "\n",
    "Kreuzvalidierung möglicher Metriken mit Subset der Referenz-Relationen"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
