{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import spacy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "nlp = spacy.load('en_core_web_lg')\n",
    "stopwords = nlp.Defaults.stop_words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from requestWorksV3 import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_engine = create_engine('sqlite:///cov/COV.sqllite', echo=False)\n",
    "Session.configure(bind=db_engine)\n",
    "db_session = Session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gensim.models import Word2Vec, KeyedVectors\n",
    "import logging\n",
    "logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 15:14:33,532 : INFO : loading Word2VecKeyedVectors object from cov/vectors_db.rv\n",
      "2020-03-18 15:14:33,844 : INFO : loading vectors from cov/vectors_db.rv.vectors.npy with mmap=None\n",
      "2020-03-18 15:14:33,938 : INFO : setting ignored attribute vectors_norm to None\n",
      "2020-03-18 15:14:33,938 : INFO : loaded cov/vectors_db.rv\n"
     ]
    }
   ],
   "source": [
    "fname = \"cov/vectors_db.rv\"\n",
    "#fname = \"testRVnew.rv\"\n",
    "\n",
    "reference_vectors = KeyedVectors.load(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164535"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reference_vectors.vocab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "53629"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doi2vec = {}\n",
    "doi2str = {}\n",
    "doi2title = {}\n",
    "doi2year = {}\n",
    "\n",
    "\n",
    "for work_db in db_session.query(CrossRefWork).filter(CrossRefWork.date_issued >= date(2015, 1, 1)).yield_per(1000):\n",
    "    work = json.loads(work_db.json_content)\n",
    "    doi2str[work_db.doi] = str(work_db)\n",
    "    doi2title[work_db.doi] = work_db.title\n",
    "    doi2year[work_db.doi] = work_db.date_issued.year\n",
    "    \n",
    "    if work_db.doi in reference_vectors:\n",
    "        doi2vec[work_db.doi] = reference_vectors[work_db.doi]\n",
    "    else:\n",
    "        work_refs = get_reference_dois(work)\n",
    "        ref_dois = [doi for doi in work_refs if doi in reference_vectors]\n",
    "        if len(ref_dois) >= 5:\n",
    "            #print()\n",
    "            vec_mean = np.mean([reference_vectors[doi] for doi in ref_dois],axis=0)\n",
    "            doi2vec[work_db.doi] = vec_mean\n",
    "\n",
    "\n",
    "titles = list(doi2title.values())\n",
    "\n",
    "len(doi2vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "136838"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db_session.query(CrossRefWork).filter(CrossRefWork.date_issued >= date(2015, 1, 1)).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump(doi2vec, open(\"cov/doi2vec.p\", \"wb\"))\n",
    "pickle.dump(doi2str, open(\"cov/doi2str.p\", \"wb\"))\n",
    "pickle.dump(doi2title, open(\"cov/doi2title.p\", \"wb\"))\n",
    "pickle.dump(doi2year, open(\"cov/doi2year.p\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "doi2vec = pickle.load(open(\"cov/doi2vec.p\", \"rb\"))\n",
    "doi2str = pickle.load(open(\"cov/doi2str.p\", \"rb\"))\n",
    "doi2title = pickle.load(open(\"cov/doi2title.p\", \"rb\"))\n",
    "doi2year = pickle.load(open(\"cov/doi2year.p\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = np.zeros((len(doi2vec), 300))\n",
    "doi2index = {}\n",
    "\n",
    "for i, (doi, vec) in enumerate(doi2vec.items()):\n",
    "    X[i, :] = vec\n",
    "    doi2index[doi] = i\n",
    "\n",
    "index2doi = {i: doi for doi, i in doi2index.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics.pairwise import cosine_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.1038/s41586-020-2012-7 in reference_vectors\n",
      "4904 4904\n"
     ]
    }
   ],
   "source": [
    "ref_dois = [\"10.1016/j.addma.2016.06.010\",\n",
    "            \"10.1103/PhysRevFluids.3.074602\",\n",
    "            \"10.1016/j.jcp.2016.05.003\",\n",
    "           ]\n",
    "\n",
    "ref_dois = [\"10.1038/s41586-020-2012-7\"] #cov\n",
    "\n",
    "ref_vecs = []\n",
    "\n",
    "for doi in ref_dois:\n",
    "    if doi in reference_vectors:\n",
    "        ref_vecs.append(reference_vectors[doi])\n",
    "        print(doi, \"in reference_vectors\")\n",
    "    elif doi in doi2vec:\n",
    "        ref_vecs.append(doi2vec[doi].resize(300))\n",
    "        print(doi, \"in doi2vec\")\n",
    "    else:\n",
    "        print(doi, \"vector not found\")\n",
    "\n",
    "ref_vecs = np.stack(ref_vecs, axis=1).T\n",
    "#ref_vecs\n",
    "\n",
    "\n",
    "doi2index = {}\n",
    "\n",
    "i = 0\n",
    "\n",
    "for doi, vec in doi2vec.items():\n",
    "    if cosine_similarity(ref_vecs, vec.reshape(1, 300)).max() >= 0.5:\n",
    "        doi2index[doi] = i\n",
    "        i += 1\n",
    "\n",
    "X = np.zeros((len(doi2index), 300))\n",
    "for doi, i in doi2index.items():\n",
    "    X[i, :] = doi2vec[doi]\n",
    "\n",
    "index2doi = {i: doi for doi, i in doi2index.items()}\n",
    "print(len(doi2index), len(index2doi))\n",
    "\n",
    "titles = list([title for doi, title in doi2title.items() if doi in doi2index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.1038/s41586-020-2012-7 in reference_vectors\n",
      "[[1.0000001]]\n",
      "4904 4904\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#ref_vec1 = reference_vectors[\"10.1016/j.addma.2016.06.010\"]\n",
    "#ref_vec2 = reference_vectors[\"10.1103/PhysRevFluids.3.074602\"]\n",
    "#ref_vec3 = reference_vectors[\"10.1016/j.jcp.2016.05.003\"]\n",
    "\n",
    "ref_dois = [\"10.1016/j.addma.2016.06.010\",\n",
    "            \"10.1103/PhysRevFluids.3.074602\",\n",
    "            \"10.1016/j.jcp.2016.05.003\",\n",
    "            \"10.1080/13588265.2019.1582577\",\n",
    "           ]\n",
    "\n",
    "ref_dois = [\"10.1038/s41586-020-2012-7\"] #cov\n",
    "\n",
    "\n",
    "ref_vecs = []\n",
    "\n",
    "for doi in ref_dois:\n",
    "    if doi in reference_vectors:\n",
    "        ref_vecs.append(reference_vectors[doi])\n",
    "        print(doi, \"in reference_vectors\")\n",
    "    elif doi in doi2vec:\n",
    "        ref_vecs.append(doi2vec[doi].reshape(300))\n",
    "        print(doi, \"in doi2vec\")\n",
    "    else:\n",
    "        print(doi, \"vector not found\")\n",
    "\n",
    "print(cosine_similarity(ref_vecs))\n",
    "\n",
    "doi2index = {}\n",
    "\n",
    "i = 0\n",
    "for doi, vec in doi2vec.items():\n",
    "    #for ref_vec in [ref_vec1, ref_vec2,  ref_vec3]:\n",
    "    for ref_vec in ref_vecs:\n",
    "        vec = vec.reshape(300)\n",
    "        #print(ref_vec.shape, vec.shape)\n",
    "        cos_sim = np.dot(ref_vec, vec) / np.linalg.norm(ref_vec) / np.linalg.norm(vec)\n",
    "        if cos_sim >= 0.5:\n",
    "            doi2index[doi] = i\n",
    "            i += 1\n",
    "            break\n",
    "\n",
    "X = np.zeros((len(doi2index), 300))\n",
    "for doi, i in doi2index.items():\n",
    "    X[i, :] = doi2vec[doi]\n",
    "\n",
    "index2doi = {i: doi for doi, i in doi2index.items()}\n",
    "print(len(doi2index), len(index2doi))\n",
    "\n",
    "titles = list([title for doi, title in doi2title.items() if doi in doi2index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gensim.models.phrases import *\n",
    "from pprint import pprint\n",
    "import operator\n",
    "\n",
    "def phrase_lines(texts_temp):\n",
    "\n",
    "    phrases = Phrases(texts_temp, min_count=10, threshold=10)\n",
    "    bigram = Phraser(phrases)\n",
    "    texts_phrased = [bigram[t] for t in texts_temp]\n",
    "\n",
    "    phrase_scores = {}\n",
    "    for phrase, score in phrases.export_phrases(texts_temp):\n",
    "        phrase_scores[phrase] = score\n",
    "    print(\"Bigramme\")\n",
    "    pprint(sorted(phrase_scores.items(), key=operator.itemgetter(1), reverse=True)[:50])\n",
    "\n",
    "    texts_temp = texts_phrased\n",
    "    phrases = Phrases(texts_temp, min_count=5, threshold=7)\n",
    "    trigram = Phraser(phrases)\n",
    "    texts_phrased = [trigram[t] for t in texts_temp]\n",
    "\n",
    "    phrase_scores = {}\n",
    "    for phrase, score in phrases.export_phrases(texts_temp):\n",
    "        phrase_scores[phrase] = score\n",
    "    print(\"Trigramme\")\n",
    "    pprint(sorted(phrase_scores.items(), key=operator.itemgetter(1), reverse=True)[:50])\n",
    "\n",
    "    return texts_phrased, bigram, trigram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 15:22:29,034 : INFO : collecting all words and their counts\n",
      "2020-03-18 15:22:29,034 : INFO : PROGRESS: at sentence #0, processed 0 words and 0 word types\n",
      "2020-03-18 15:22:29,112 : INFO : collected 45863 word types from a corpus of 44449 words (unigram + bigrams) and 4904 sentences\n",
      "2020-03-18 15:22:29,112 : INFO : using 45863 counts as vocab in Phrases<0 vocab, min_count=10, threshold=10, max_vocab_size=40000000>\n",
      "2020-03-18 15:22:29,112 : INFO : source_vocab length 45863\n",
      "2020-03-18 15:22:29,424 : INFO : Phraser built with 78 phrasegrams\n",
      "2020-03-18 15:22:29,659 : INFO : collecting all words and their counts\n",
      "2020-03-18 15:22:29,659 : INFO : PROGRESS: at sentence #0, processed 0 words and 0 word types\n",
      "2020-03-18 15:22:29,721 : INFO : collected 46162 word types from a corpus of 42756 words (unigram + bigrams) and 4904 sentences\n",
      "2020-03-18 15:22:29,721 : INFO : using 46162 counts as vocab in Phrases<0 vocab, min_count=5, threshold=7, max_vocab_size=40000000>\n",
      "2020-03-18 15:22:29,721 : INFO : source_vocab length 46162\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bigramme\n",
      "[(b'escherichia coli', 1091.9761904761906),\n",
      " (b'mycobacterium tuberculosis', 810.709595959596),\n",
      " (b'body area', 686.8882488479263),\n",
      " (b'kalman filter', 524.1485714285715),\n",
      " (b'machine learning', 357.6062378167641),\n",
      " (b'cryo-electron microscopy', 352.7923076923077),\n",
      " (b'porcine deltacoronavirus', 318.49305555555554),\n",
      " (b'hepatitis c', 317.2341897233202),\n",
      " (b'systematic review', 307.64654282765736),\n",
      " (b'middle east', 297.17352040816326),\n",
      " (b'lung injury', 262.07428571428574),\n",
      " (b'severe acute', 239.33724722765817),\n",
      " (b'amino acid', 235.1948717948718),\n",
      " (b'saudi arabia', 224.8186274509804),\n",
      " (b'virtual screening', 224.8186274509804),\n",
      " (b'routing protocol', 207.6102880658436),\n",
      " (b'underwater acoustic', 184.93145161290323),\n",
      " (b'east respiratory', 183.452),\n",
      " (b'hepatitis e', 173.72348484848484),\n",
      " (b'sp. nov.,', 173.72348484848484),\n",
      " (b'respiratory syndrome', 171.71601346801347),\n",
      " (b'crystal structures', 147.3301209372638),\n",
      " (b'structural basis', 145.23821252638987),\n",
      " (b'doa estimation', 128.82865168539325),\n",
      " (b'sensor networks', 127.17446440240907),\n",
      " (b'united states', 115.81565656565655),\n",
      " (b'wireless sensor', 114.44006387863061),\n",
      " (b'zika virus', 112.84221223791187),\n",
      " (b'indoor positioning', 103.94773351648351),\n",
      " (b'crystal structure', 103.8564311594203),\n",
      " (b'monoclonal antibody', 102.37276785714285),\n",
      " (b'2019 novel', 99.19899062725305),\n",
      " (b'acute respiratory', 89.35257229832573),\n",
      " (b'ebola virus', 88.50754385964913),\n",
      " (b't cell', 87.86015325670498),\n",
      " (b'wuhan, china', 84.30698529411765),\n",
      " (b'genome sequence', 82.10347296813462),\n",
      " (b'wireless body', 77.41316579144787),\n",
      " (b'syndrome coronavirus', 74.60966507177032),\n",
      " (b'draft genome', 71.99843014128729),\n",
      " (b'acoustic sensor', 69.84535511560294),\n",
      " (b'public health', 68.72577422577423),\n",
      " (b'spike protein', 65.33190883190883),\n",
      " (b'indoor localization', 64.53236256544503),\n",
      " (b'pedestrian navigation', 60.66534391534391),\n",
      " (b'structure determination', 59.034181922196794),\n",
      " (b'coronavirus (2019-ncov)', 58.265154264972765),\n",
      " (b'neural network', 56.97267080745342),\n",
      " (b'data collection', 53.64093567251462),\n",
      " (b'novel coronavirus', 49.22967451523546)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 15:22:30,034 : INFO : Phraser built with 264 phrasegrams\n",
      "2020-03-18 15:22:30,284 : INFO : adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2020-03-18 15:22:30,346 : INFO : built Dictionary(13326 unique tokens: ['(covid‐19)', 'clinical', 'corona', 'follow?', 'novel']...) from 4904 documents (total 41336 corpus positions)\n",
      "2020-03-18 15:22:30,393 : INFO : collecting document frequencies\n",
      "2020-03-18 15:22:30,393 : INFO : PROGRESS: processing document #0\n",
      "2020-03-18 15:22:30,409 : INFO : calculating IDF weights for 4904 documents and 13325 features (40994 matrix non-zeros)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trigramme\n",
      "[(b'critically ill', 2098.2727272727275),\n",
      " (b'zymoseptoria tritici', 1923.4166666666665),\n",
      " (b'platelet concentrates', 1884.1632653061224),\n",
      " (b'acute_respiratory distress_syndrome', 1583.9901960784312),\n",
      " (b'hereditary angioedema', 1538.7333333333333),\n",
      " (b'de novo', 1465.4603174603174),\n",
      " (b'wireless_body area_networks', 1417.2543859649122),\n",
      " (b'incubation period', 1282.2777777777776),\n",
      " (b'visually impaired', 1282.2777777777776),\n",
      " (b'web server', 1234.2780748663101),\n",
      " (b'dromedary camels', 1154.05),\n",
      " (b'salmonella enterica', 1131.421568627451),\n",
      " (b'adipose tissue', 1086.164705882353),\n",
      " (b'sad phasing', 887.7307692307693),\n",
      " (b'copy number', 802.8173913043478),\n",
      " (b'rhesus macaques', 769.3666666666667),\n",
      " (b'acute lymphoblastic', 699.4242424242424),\n",
      " (b'mac protocol', 686.9345238095239),\n",
      " (b'nucleic acid', 665.7980769230769),\n",
      " (b'enterica serovar', 641.1388888888888),\n",
      " (b'bacillus cereus', 599.5064935064936),\n",
      " (b'pedestrian dead', 573.4409937888198),\n",
      " (b'coronavirus_disease 2019', 554.8317307692308),\n",
      " (b'\\xc3\\xa5 resolution', 539.906432748538),\n",
      " (b'life cycle', 524.5681818181819),\n",
      " (b'electron microscopy', 480.85416666666663),\n",
      " (b'underwater_wireless sensor_networks', 441.74162679425837),\n",
      " (b'respiratory syncytial', 384.68333333333334),\n",
      " (b'novel_coronavirus (2019-ncov)', 366.36507936507934),\n",
      " (b'healthcare workers', 366.36507936507934),\n",
      " (b'codon usage', 349.71212121212125),\n",
      " (b'labor market', 349.71212121212125),\n",
      " (b'south korea', 335.7236363636364),\n",
      " (b'2019_novel coronavirus_(2019-ncov)', 334.5072463768116),\n",
      " (b'carboxylesterase 1', 334.50724637681157),\n",
      " (b'job satisfaction', 322.8111888111888),\n",
      " (b'underwater_acoustic sensor_networks', 303.6973684210526),\n",
      " (b'ultraviolet c', 299.75324675324674),\n",
      " (b'reproduction number', 286.7204968944099),\n",
      " (b'korea, 2015', 284.9506172839506),\n",
      " (b'middle_east respiratory_syndrome', 281.9368389691643),\n",
      " (b'wuhan, china:', 279.76969696969695),\n",
      " (b'whole-genome sequencing', 274.7738095238095),\n",
      " (b'infectious diseases', 270.7448680351906),\n",
      " (b'substrate specificity', 256.4555555555556),\n",
      " (b'mhc class', 256.4555555555555),\n",
      " (b'main protease', 256.4555555555555),\n",
      " (b'genetic programming', 249.52432432432434),\n",
      " (b'wireless_sensor networks', 243.67038123167154),\n",
      " (b'complete_genome sequence', 237.5403087478559)]\n"
     ]
    }
   ],
   "source": [
    "from gensim import corpora, models, similarities\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "title_tokens = []\n",
    "for title in titles:\n",
    "    if len(title) >= 1:\n",
    "        title_tokens.append([s for s in title.lower().split() if not s in stopwords])\n",
    "\n",
    "\n",
    "phrased_lines, bigram, trigram = phrase_lines(title_tokens)\n",
    "\n",
    "frequency = defaultdict(int)\n",
    "\n",
    "for line in phrased_lines:\n",
    "    for term in line:\n",
    "        frequency[term] += 1\n",
    "\n",
    "dictionary = corpora.Dictionary(phrased_lines)\n",
    "\n",
    "corpus = [dictionary.doc2bow(line) for line in phrased_lines]\n",
    "\n",
    "tfidf = models.TfidfModel(corpus, normalize=True) # step 1 -- initialize a model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import Memory\n",
    "location = './cachedir'\n",
    "memory = Memory(location, verbose=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "________________________________________________________________________________\n",
      "[Memory] Calling sklearn.cluster._agglomerative._average_linkage...\n",
      "_average_linkage(array([[-0.319997, ..., -0.177637],\n",
      "       ...,\n",
      "       [ 0.0142  , ...,  0.021458]]), \n",
      "None, n_clusters=None, return_distance=True, linkage='average', affinity='cosine')\n",
      "__________________________________________________average_linkage - 4.2s, 0.1min\n"
     ]
    }
   ],
   "source": [
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "#ac = AgglomerativeClustering(n_clusters=None, affinity=\"cosine\", linkage=\"average\", compute_full_tree=True, distance_threshold=0.2)\n",
    "#ac = AgglomerativeClustering(affinity=\"cosine\", linkage=\"average\", distance_threshold=0.2)\n",
    "#ac = AgglomerativeClustering(n_clusters=400, affinity=\"cosine\", linkage=\"average\")\n",
    "#ac = AgglomerativeClustering(n_clusters=400, affinity=\"cosine\", linkage=\"average\")\n",
    "#ac = AgglomerativeClustering(n_clusters=400, affinity=\"cosine\", linkage=\"average\", memory=memory)\n",
    "ac = AgglomerativeClustering(n_clusters=None, affinity=\"cosine\", linkage=\"average\", distance_threshold=0.2, memory=memory)\n",
    "\n",
    "\n",
    "clLabels = ac.fit_predict(X)\n",
    "cluster_ids = sorted(np.unique(clLabels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 0\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 1\n",
      "Size: 30\n",
      "\n",
      "Cluster No. 2\n",
      "Size: 125\n",
      "\n",
      "Cluster No. 3\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 4\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 5\n",
      "Size: 32\n",
      "\n",
      "Cluster No. 6\n",
      "Size: 14\n",
      "\n",
      "Cluster No. 7\n",
      "Size: 497\n",
      "\n",
      "Cluster No. 8\n",
      "Size: 498\n",
      "\n",
      "Cluster No. 9\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 10\n",
      "Size: 13\n",
      "\n",
      "Cluster No. 11\n",
      "Size: 10\n",
      "\n",
      "Cluster No. 12\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 13\n",
      "Size: 776\n",
      "\n",
      "Cluster No. 14\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 15\n",
      "Size: 91\n",
      "\n",
      "Cluster No. 16\n",
      "Size: 257\n",
      "\n",
      "Cluster No. 17\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 18\n",
      "Size: 48\n",
      "\n",
      "Cluster No. 19\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 20\n",
      "Size: 300\n",
      "\n",
      "Cluster No. 21\n",
      "Size: 364\n",
      "\n",
      "Cluster No. 22\n",
      "Size: 34\n",
      "\n",
      "Cluster No. 23\n",
      "Size: 87\n",
      "\n",
      "Cluster No. 24\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 25\n",
      "Size: 23\n",
      "\n",
      "Cluster No. 26\n",
      "Size: 20\n",
      "\n",
      "Cluster No. 27\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 28\n",
      "Size: 5\n",
      "\n",
      "Cluster No. 29\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 30\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 31\n",
      "Size: 47\n",
      "\n",
      "Cluster No. 32\n",
      "Size: 50\n",
      "\n",
      "Cluster No. 33\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 34\n",
      "Size: 43\n",
      "\n",
      "Cluster No. 35\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 36\n",
      "Size: 20\n",
      "\n",
      "Cluster No. 37\n",
      "Size: 151\n",
      "\n",
      "Cluster No. 38\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 39\n",
      "Size: 25\n",
      "\n",
      "Cluster No. 40\n",
      "Size: 42\n",
      "\n",
      "Cluster No. 41\n",
      "Size: 41\n",
      "\n",
      "Cluster No. 42\n",
      "Size: 31\n",
      "\n",
      "Cluster No. 43\n",
      "Size: 14\n",
      "\n",
      "Cluster No. 44\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 45\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 46\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 47\n",
      "Size: 48\n",
      "\n",
      "Cluster No. 48\n",
      "Size: 8\n",
      "\n",
      "Cluster No. 49\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 50\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 51\n",
      "Size: 71\n",
      "\n",
      "Cluster No. 52\n",
      "Size: 23\n",
      "\n",
      "Cluster No. 53\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 54\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 55\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 56\n",
      "Size: 24\n",
      "\n",
      "Cluster No. 57\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 58\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 59\n",
      "Size: 49\n",
      "\n",
      "Cluster No. 60\n",
      "Size: 47\n",
      "\n",
      "Cluster No. 61\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 62\n",
      "Size: 11\n",
      "\n",
      "Cluster No. 63\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 64\n",
      "Size: 13\n",
      "\n",
      "Cluster No. 65\n",
      "Size: 8\n",
      "\n",
      "Cluster No. 66\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 67\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 68\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 69\n",
      "Size: 129\n",
      "\n",
      "Cluster No. 70\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 71\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 72\n",
      "Size: 67\n",
      "\n",
      "Cluster No. 73\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 74\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 75\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 76\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 77\n",
      "Size: 19\n",
      "\n",
      "Cluster No. 78\n",
      "Size: 19\n",
      "\n",
      "Cluster No. 79\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 80\n",
      "Size: 5\n",
      "\n",
      "Cluster No. 81\n",
      "Size: 29\n",
      "\n",
      "Cluster No. 82\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 83\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 84\n",
      "Size: 26\n",
      "\n",
      "Cluster No. 85\n",
      "Size: 120\n",
      "\n",
      "Cluster No. 86\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 87\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 88\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 89\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 90\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 91\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 92\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 93\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 94\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 95\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 96\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 97\n",
      "Size: 12\n",
      "\n",
      "Cluster No. 98\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 99\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 100\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 101\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 102\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 103\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 104\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 105\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 106\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 107\n",
      "Size: 9\n",
      "\n",
      "Cluster No. 108\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 109\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 110\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 111\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 112\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 113\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 114\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 115\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 116\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 117\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 118\n",
      "Size: 11\n",
      "\n",
      "Cluster No. 119\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 120\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 121\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 122\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 123\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 124\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 125\n",
      "Size: 14\n",
      "\n",
      "Cluster No. 126\n",
      "Size: 22\n",
      "\n",
      "Cluster No. 127\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 128\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 129\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 130\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 131\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 132\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 133\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 134\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 135\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 136\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 137\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 138\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 139\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 140\n",
      "Size: 17\n",
      "\n",
      "Cluster No. 141\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 142\n",
      "Size: 46\n",
      "\n",
      "Cluster No. 143\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 144\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 145\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 146\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 147\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 148\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 149\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 150\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 151\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 152\n",
      "Size: 7\n",
      "\n",
      "Cluster No. 153\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 154\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 155\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 156\n",
      "Size: 4\n",
      "\n",
      "Cluster No. 157\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 158\n",
      "Size: 30\n",
      "\n",
      "Cluster No. 159\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 160\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 161\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 162\n",
      "Size: 3\n",
      "\n",
      "Cluster No. 163\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 164\n",
      "Size: 6\n",
      "\n",
      "Cluster No. 165\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 166\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 167\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 168\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 169\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 170\n",
      "Size: 22\n",
      "\n",
      "Cluster No. 171\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 172\n",
      "Size: 11\n",
      "\n",
      "Cluster No. 173\n",
      "Size: 16\n",
      "\n",
      "Cluster No. 174\n",
      "Size: 2\n",
      "\n",
      "Cluster No. 175\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 176\n",
      "Size: 1\n",
      "\n",
      "Cluster No. 177\n",
      "Size: 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clSizes = []\n",
    "clTitles = defaultdict(list)\n",
    "\n",
    "for cl in cluster_ids:    \n",
    "    print(\"Cluster No.\", cl)\n",
    "    #cl_dois = [doi for i, doi in enumerate(dois_to_plot) if clLabels[i] == cl]\n",
    "    cl_dois = [doi for i, doi in index2doi.items() if clLabels[i] == cl]\n",
    "    print(\"Size:\", len(cl_dois))\n",
    "    for doi in cl_dois:\n",
    "        #pprint(work_dict[doi][\"title\"])\n",
    "        title = doi2title[doi]\n",
    "        if len(title) >= 1:\n",
    "            pub_title = [s for s in title.lower().split() if not s in stopwords]\n",
    "            clTitles[cl].append(pub_title)    \n",
    "        \n",
    "    print()\n",
    "    clSizes.append(len(cl_dois))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 0\n",
      "Size: 2\n",
      "Terms: \n",
      "family (33, 3, 0.41), management (47, 2, 0.25), instrument (1, 1, 0.23), medium-sized (1, 1, 0.23), sustainability (1, 1, 0.23), affective (1, 1, 0.23), interests (1, 1, 0.23), managers' (1, 1, 0.23), non-family (1, 1, 0.23), stakeholders: (1, 1, 0.23), enterprises (2, 1, 0.21), firms (2, 1, 0.21), salience (2, 1, 0.21), terms (3, 1, 0.20), commitment (7, 1, 0.18), strategic (9, 1, 0.17), power (13, 1, 0.16), resource (14, 1, 0.16), influence (21, 1, 0.15), small (23, 1, 0.14)\n",
      "\n",
      "Cluster No. 1\n",
      "Size: 30\n",
      "Terms: \n",
      "policy (27, 6, 0.24), blame (4, 4, 0.22), administrative (8, 4, 0.20), politics (11, 4, 0.19), explaining (5, 3, 0.16), avoidance (9, 3, 0.14), political (12, 3, 0.14), accountability (15, 3, 0.13), gewaltenteilung (2, 2, 0.13), european (21, 3, 0.13), sweden (2, 2, 0.12), ministers (2, 2, 0.12), policies: (2, 2, 0.12), health (28, 3, 0.12), delegation (3, 2, 0.11), bureaucratic (3, 2, 0.11), germany (5, 2, 0.10), spain (5, 2, 0.10), democratic (5, 2, 0.10), crisis (7, 2, 0.10)\n",
      "\n",
      "Cluster No. 2\n",
      "Size: 125\n",
      "Terms: \n",
      "antiviral (30, 20, 0.30), hepatitis_c_virus (15, 14, 0.24), replication (35, 12, 0.18), hepatitis_c (9, 9, 0.17), rna (60, 13, 0.17), virus (89, 14, 0.17), inhibits (20, 10, 0.16), viral (72, 12, 0.15), treatment (51, 10, 0.14), broad-spectrum (11, 7, 0.13), inhibitors (35, 8, 0.12), agents (11, 6, 0.11), virus_infection (15, 6, 0.10), antiviral_activity (6, 5, 0.10), sofosbuvir (6, 5, 0.10), host (46, 7, 0.10), protein (155, 9, 0.09), translation (11, 5, 0.09), resistance (37, 6, 0.09), nucleoside (4, 4, 0.08)\n",
      "\n",
      "Cluster No. 3\n",
      "Size: 3\n",
      "Terms: \n",
      "peptides (10, 2, 0.38), chlorotoxin-like (1, 1, 0.26), structure-activity (1, 1, 0.26), chlorotoxin: (1, 1, 0.26), diagnose (1, 1, 0.26), helpful (1, 1, 0.26), scorpion (1, 1, 0.26), cross‐linked (1, 1, 0.26), plink (1, 1, 0.26), glioma (2, 1, 0.24), fight (3, 1, 0.23), analyze (4, 1, 0.22), invasion (4, 1, 0.22), tumor (8, 1, 0.20), peptide (20, 1, 0.17), natural (25, 1, 0.16), relationship (28, 1, 0.16)\n",
      "\n",
      "Cluster No. 4\n",
      "Size: 2\n",
      "Terms: \n",
      "immunity–fecundity (1, 1, 0.29), tradeoffs (1, 1, 0.29), graft (1, 1, 0.29), immunology (2, 1, 0.26), transplant (2, 1, 0.26), years: (2, 1, 0.26), 20 (3, 1, 0.25), dysfunction (3, 1, 0.25), generate (3, 1, 0.25), revolution (4, 1, 0.24), renal (6, 1, 0.23), improvement (6, 1, 0.23), drosophila (7, 1, 0.22), survival (12, 1, 0.20), patient (15, 1, 0.20), mitochondrial (16, 1, 0.19), infection (79, 1, 0.14)\n",
      "\n",
      "Cluster No. 5\n",
      "Size: 32\n",
      "Terms: \n",
      "toll-like_receptor (7, 7, 0.33), activation (27, 7, 0.26), 8 (6, 4, 0.19), cells (43, 5, 0.17), degradation (8, 3, 0.14), signaling (11, 3, 0.13), immunity (20, 3, 0.12), unc93b1 (2, 2, 0.11), senses (2, 2, 0.11), ubiquitin (2, 2, 0.11), autoimmunity (2, 2, 0.11), tlr7 (2, 2, 0.11), sting (2, 2, 0.11), agonist (3, 2, 0.11), cell (46, 3, 0.10), inflammation (5, 2, 0.10), signalling (5, 2, 0.10), attenuation (5, 2, 0.10), inflammasome (6, 2, 0.10), innate (6, 2, 0.10)\n",
      "\n",
      "Cluster No. 6\n",
      "Size: 14\n",
      "Terms: \n",
      "library (8, 3, 0.24), kinase (14, 3, 0.22), anchoring (2, 2, 0.21), rac1 (2, 2, 0.21), encoded, (2, 2, 0.20), targeting (31, 3, 0.19), genetically (4, 2, 0.18), know (7, 2, 0.17), antibody (17, 2, 0.14), interaction (32, 2, 0.13), proteins (51, 2, 0.11), cyclic-peptide (1, 1, 0.11), phage-displayed (1, 1, 0.11), anti-hiv-1 (1, 1, 0.11), boost (1, 1, 0.11), dissect (1, 1, 0.11), hycosul (1, 1, 0.11), target, (1, 1, 0.11), a-kinase (1, 1, 0.11), cyclic‐peptide (1, 1, 0.11)\n",
      "\n",
      "Cluster No. 7\n",
      "Size: 497\n",
      "Terms: \n",
      "genome (71, 47, 0.24), genomes (40, 28, 0.16), genomic (58, 30, 0.16), microbial (29, 24, 0.15), analysis (196, 37, 0.14), sequencing (41, 22, 0.13), reveals (77, 24, 0.12), macrococcus (15, 15, 0.11), gene (59, 21, 0.11), species (82, 22, 0.11), evolution (99, 23, 0.11), microbiome (18, 16, 0.11), genes (48, 19, 0.11), resistance (37, 18, 0.10), gut (14, 14, 0.10), strain (22, 15, 0.10), sequence (37, 16, 0.09), assembly (27, 15, 0.09), diversity (54, 17, 0.09), fungal (23, 14, 0.09)\n",
      "\n",
      "Cluster No. 8\n",
      "Size: 498\n",
      "Terms: \n",
      "protein (155, 62, 0.25), structure (105, 53, 0.24), crystal_structure (45, 37, 0.20), structural (55, 31, 0.16), crystal_structures (27, 26, 0.16), binding (58, 29, 0.15), structures (36, 25, 0.14), refinement (22, 20, 0.13), domain (57, 23, 0.12), crystallographic (18, 18, 0.12), x-ray (23, 18, 0.11), structure_determination (17, 17, 0.11), electron (17, 16, 0.11), neutron (15, 15, 0.10), complex (84, 20, 0.10), crystallography (15, 14, 0.09), crystals (15, 14, 0.09), diffraction (16, 14, 0.09), macromolecular (13, 13, 0.09), data (134, 20, 0.08)\n",
      "\n",
      "Cluster No. 9\n",
      "Size: 2\n",
      "Terms: \n",
      "acquisition: (1, 1, 0.29), contraception (1, 1, 0.29), hormonal (1, 1, 0.29), low- (1, 1, 0.29), middle-income (1, 1, 0.29), workers’ (1, 1, 0.29), provision (2, 1, 0.26), countries: (2, 1, 0.26), planning (9, 1, 0.21), services (12, 1, 0.21), hiv-1 (14, 1, 0.20), biological (16, 1, 0.19), systematic_review (17, 1, 0.19), effectiveness (19, 1, 0.19), community (21, 1, 0.19), health (28, 1, 0.18), mechanisms (28, 1, 0.17), family (33, 1, 0.17)\n",
      "\n",
      "Cluster No. 10\n",
      "Size: 13\n",
      "Terms: \n",
      "rhinovirus (6, 5, 0.36), asthma (7, 5, 0.36), wheezing (3, 3, 0.24), cadherin-related (2, 2, 0.17), respiratory (29, 3, 0.17), infections (35, 3, 0.16), children (6, 2, 0.15), respiratory_tract (6, 2, 0.15), member (8, 2, 0.14), life (9, 2, 0.14), c (16, 2, 0.12), microbiome (18, 2, 0.12), lung (21, 2, 0.12), early (26, 2, 0.11), family (33, 2, 0.11), function (42, 2, 0.10), cdhr3 (1, 1, 0.09), asthma: (1, 1, 0.09), narrative (1, 1, 0.09), 3, (1, 1, 0.09)\n",
      "\n",
      "Cluster No. 11\n",
      "Size: 10\n",
      "Terms: \n",
      "bacillus (15, 4, 0.29), operon (4, 3, 0.25), heat (4, 3, 0.25), germination (2, 2, 0.19), spore (2, 2, 0.19), spova (2, 2, 0.19), spores (3, 2, 0.18), resistance (37, 3, 0.17), transposon (7, 2, 0.16), specific (8, 2, 0.15), presence (17, 2, 0.13), receptors (19, 2, 0.13), strains (24, 2, 0.13), bacterial (37, 2, 0.12), endospores (1, 1, 0.10), spova2mob (1, 1, 0.10), dodecylamine-induced (1, 1, 0.10), nutrient- (1, 1, 0.10), subtilisnegatively (1, 1, 0.10), bacilli (1, 1, 0.10)\n",
      "\n",
      "Cluster No. 12\n",
      "Size: 4\n",
      "Terms: \n",
      "trust (4, 2, 0.34), intra-party (1, 1, 0.21), parties? (1, 1, 0.21), britain (1, 1, 0.21), democracies: (1, 1, 0.21), agency’s (1, 1, 0.21), bashing (1, 1, 0.21), corruption (1, 1, 0.21), bureaucracy (2, 1, 0.19), israel (2, 1, 0.19), priming (2, 1, 0.19), democracy (2, 1, 0.19), belgium (2, 1, 0.19), insecurity (3, 1, 0.18), integrity (4, 1, 0.17), citizens’ (4, 1, 0.17), advanced (5, 1, 0.17), electoral (5, 1, 0.17), economic (6, 1, 0.16), levels (7, 1, 0.16)\n",
      "\n",
      "Cluster No. 13\n",
      "Size: 776\n",
      "Terms: \n",
      "coronavirus (161, 154, 0.38), middle_east_respiratory_syndrome (123, 117, 0.31), mers-cov (54, 54, 0.17), novel_coronavirus (48, 45, 0.15), outbreak (67, 48, 0.15), bats (44, 37, 0.13), china (49, 38, 0.12), coronaviruses (35, 34, 0.12), 2019-ncov (35, 33, 0.12), transmission (69, 36, 0.11), covid-19 (35, 31, 0.11), human (178, 45, 0.11), virus (89, 35, 0.10), potential (71, 33, 0.10), emerging (43, 29, 0.10), coronavirus_infection (27, 26, 0.10), vaccine (42, 28, 0.10), infection (79, 32, 0.09), bat (26, 25, 0.09), 2019_novel_coronavirus (26, 25, 0.09)\n",
      "\n",
      "Cluster No. 14\n",
      "Size: 2\n",
      "Terms: \n",
      "coumarin (1, 1, 0.28), domino (1, 1, 0.28), pyrimidine-fused (1, 1, 0.28), triazole (1, 1, 0.28), neurotoxins: (1, 1, 0.28), pharmacology, (1, 1, 0.28), toxicology (1, 1, 0.28), benign (2, 1, 0.26), botulinum (2, 1, 0.26), environmentally (3, 1, 0.25), motifs (3, 1, 0.25), biology, (3, 1, 0.25), catalytic (6, 1, 0.22), synthesis (12, 1, 0.20), reaction (15, 1, 0.19)\n",
      "\n",
      "Cluster No. 15\n",
      "Size: 91\n",
      "Terms: \n",
      "virtual_screening (14, 11, 0.26), docking (13, 7, 0.17), inhibitors (35, 8, 0.16), molecular (95, 10, 0.16), prediction (43, 8, 0.15), methods (20, 6, 0.13), protein (155, 9, 0.13), binding (58, 7, 0.13), structure (105, 8, 0.13), silico (12, 5, 0.12), function (42, 6, 0.12), thieno[2,3-b]pyridine (4, 4, 0.12), pharmacophore (4, 4, 0.12), screening (20, 5, 0.11), search (22, 5, 0.11), simulations (7, 4, 0.11), compounds (7, 4, 0.11), virtual (10, 4, 0.10), web_server (10, 4, 0.10), computational (17, 4, 0.09)\n",
      "\n",
      "Cluster No. 16\n",
      "Size: 257\n",
      "Terms: \n",
      "species (82, 40, 0.28), phylogeny (32, 29, 0.25), genus (26, 23, 0.21), evolution (99, 30, 0.20), molecular (95, 23, 0.16), diversity (54, 18, 0.14), morphological (15, 13, 0.13), evolutionary (47, 14, 0.11), cryptic (13, 11, 0.11), phylogenetic (18, 11, 0.11), diversification (15, 10, 0.10), new (99, 14, 0.09), reveals (77, 13, 0.09), origin (23, 10, 0.09), integrative (15, 9, 0.09), population (18, 9, 0.09), complex (84, 12, 0.08), data (134, 13, 0.08), genes (48, 10, 0.08), taxonomy (19, 8, 0.08)\n",
      "\n",
      "Cluster No. 17\n",
      "Size: 7\n",
      "Terms: \n",
      "target (31, 2, 0.20), inaccurate (1, 1, 0.17), principals (1, 1, 0.17), tolerate (1, 1, 0.17), civil (1, 1, 0.17), reform, (1, 1, 0.17), self-selection, (1, 1, 0.17), arts (1, 1, 0.17), results? (1, 1, 0.17), business: (1, 1, 0.17), smes (1, 1, 0.17), asymmetry (1, 1, 0.17), incentives (1, 1, 0.17), ratcheting (1, 1, 0.17), orchestras (1, 1, 0.17), symphony (1, 1, 0.17), performance (78, 2, 0.17), score: (2, 1, 0.16), donors (2, 1, 0.16), dependence (2, 1, 0.16)\n",
      "\n",
      "Cluster No. 18\n",
      "Size: 48\n",
      "Terms: \n",
      "splicing (10, 8, 0.30), genes (48, 6, 0.17), chromatin (6, 4, 0.16), regulation (23, 5, 0.16), rna (60, 6, 0.16), drosophila (7, 4, 0.16), landscape (16, 4, 0.14), gene (59, 5, 0.14), ribosomal (6, 3, 0.12), human (178, 6, 0.12), long (10, 3, 0.11), resource (14, 3, 0.11), regulatory (15, 3, 0.11), profiling (18, 3, 0.10), dmd (2, 2, 0.10), repeats (2, 2, 0.10), nucleolar (2, 2, 0.10), melanogaster (3, 2, 0.09), polyadenylation (3, 2, 0.09), curation (4, 2, 0.09)\n",
      "\n",
      "Cluster No. 19\n",
      "Size: 3\n",
      "Terms: \n",
      "cnv-based (1, 1, 0.25), interpretable (1, 1, 0.25), teach-discover-treat (1, 1, 0.25), tutorials (1, 1, 0.25), virtual-screening (1, 1, 0.25), classifiers (2, 1, 0.23), tumour (2, 1, 0.23), prospective (3, 1, 0.22), workflow (3, 1, 0.22), malaria (4, 1, 0.21), rule (5, 1, 0.20), fuzzy (5, 1, 0.20), bacterial_species (6, 1, 0.20), 2014 (7, 1, 0.19), competition (8, 1, 0.19), acquired (9, 1, 0.19), neural (11, 1, 0.18), antimicrobial (14, 1, 0.18), results (14, 1, 0.17), classification (23, 1, 0.16)\n",
      "\n",
      "Cluster No. 20\n",
      "Size: 300\n",
      "Terms: \n",
      "wireless_sensor_networks (62, 44, 0.25), scheme (43, 35, 0.21), routing (33, 33, 0.21), based (180, 46, 0.20), underwater_wireless_sensor_networks (29, 29, 0.19), underwater_acoustic_sensor_networks (27, 27, 0.18), networks (64, 31, 0.17), energy (46, 26, 0.16), underwater_sensor_networks (22, 22, 0.15), routing_protocol (21, 21, 0.15), mobile (51, 24, 0.14), energy-efficient (21, 20, 0.14), algorithm (74, 26, 0.14), underwater (28, 20, 0.13), node (21, 18, 0.13), communication (30, 19, 0.12), clustering (23, 17, 0.12), cooperative (35, 18, 0.11), underwater_acoustic (17, 15, 0.11), sensor_networks (20, 15, 0.11)\n",
      "\n",
      "Cluster No. 21\n",
      "Size: 364\n",
      "Terms: \n",
      "localization (98, 60, 0.27), indoor (58, 51, 0.26), based (180, 64, 0.24), indoor_localization (39, 39, 0.22), positioning (41, 38, 0.21), indoor_positioning (35, 35, 0.20), method (60, 32, 0.16), robust (46, 27, 0.15), estimation (75, 28, 0.14), algorithm (74, 25, 0.12), wi-fi (18, 18, 0.12), fingerprinting (18, 18, 0.12), wireless (36, 19, 0.11), tracking (37, 19, 0.11), navigation (39, 19, 0.11), pedestrian (16, 16, 0.11), measurements (21, 15, 0.09), sensor (46, 17, 0.09), location (17, 14, 0.09), system (77, 19, 0.09)\n",
      "\n",
      "Cluster No. 22\n",
      "Size: 34\n",
      "Terms: \n",
      "police (27, 20, 0.63), policing (6, 5, 0.20), citizen (6, 4, 0.16), crime (6, 3, 0.13), perceptions (8, 3, 0.12), analysts (2, 2, 0.10), recruits’ (2, 2, 0.09), washington (2, 2, 0.09), female (2, 2, 0.09), mental (2, 2, 0.09), encounters (2, 2, 0.09), volunteer (3, 2, 0.09), violence (3, 2, 0.09), discretion (3, 2, 0.09), law (3, 2, 0.09), arrest (4, 2, 0.09), career (4, 2, 0.09), officer (4, 2, 0.09), use (48, 3, 0.08), example (5, 2, 0.08)\n",
      "\n",
      "Cluster No. 23\n",
      "Size: 87\n",
      "Terms: \n",
      "leukemia (15, 11, 0.24), acute_lymphoblastic (10, 10, 0.23), single-cell (14, 9, 0.20), cell (46, 10, 0.18), mutations (27, 8, 0.16), genomic (58, 9, 0.15), tumor (8, 6, 0.14), recurrent (8, 6, 0.14), t_cells (11, 6, 0.14), somatic (6, 5, 0.13), t_cell (14, 5, 0.11), pediatric (13, 5, 0.11), fusions (4, 4, 0.11), cancer (21, 5, 0.10), leukaemia (6, 4, 0.10), patients (57, 6, 0.10), adipose_tissue (9, 4, 0.10), rna-seq (11, 4, 0.09), acute (14, 4, 0.09), human (178, 7, 0.09)\n",
      "\n",
      "Cluster No. 24\n",
      "Size: 2\n",
      "Terms: \n",
      "electromyography (2, 2, 0.54), muscle-computer (1, 1, 0.30), ankles (1, 1, 0.30), prosthetic (1, 1, 0.30), high-density (2, 1, 0.27), advancing (3, 1, 0.26), interfaces (3, 1, 0.26), data-driven (5, 1, 0.24), powered (5, 1, 0.24), smart (9, 1, 0.22), surface (20, 1, 0.19), optimization (32, 1, 0.18)\n",
      "\n",
      "Cluster No. 25\n",
      "Size: 23\n",
      "Terms: \n",
      "rna (60, 7, 0.31), codon_usage (6, 4, 0.27), viruses (45, 5, 0.24), bias (8, 3, 0.19), evolution (99, 4, 0.16), synonymous (3, 2, 0.15), gene (59, 3, 0.13), origins (7, 2, 0.13), dengue_virus (8, 2, 0.13), viral (72, 3, 0.13), genome-wide (13, 2, 0.12), infectious (15, 2, 0.12), family (33, 2, 0.10), analysis (196, 3, 0.10), genomes (40, 2, 0.10), sequencing (41, 2, 0.10), function (42, 2, 0.10), genes (48, 2, 0.09), diversity (54, 2, 0.09), silence: (1, 1, 0.09)\n",
      "\n",
      "Cluster No. 26\n",
      "Size: 20\n",
      "Terms: \n",
      "epitopes (5, 4, 0.27), peptide_vaccine (7, 4, 0.26), t_cell (14, 4, 0.24), designing (6, 3, 0.20), class (14, 3, 0.18), peptide–mhc (2, 2, 0.16), immunoinformatics (3, 2, 0.15), prediction (43, 3, 0.14), subunit_vaccines (6, 2, 0.13), epitope (6, 2, 0.13), b-cell (6, 2, 0.13), approach (81, 3, 0.12), ebola (18, 2, 0.11), peptide (20, 2, 0.11), novel_coronavirus (48, 2, 0.09), binding (58, 2, 0.09), identification (70, 2, 0.08), immunoinformatics‐aided (1, 1, 0.08), (sars‐cov‐2): (1, 1, 0.08), neoant-hill: (1, 1, 0.08)\n",
      "\n",
      "Cluster No. 27\n",
      "Size: 3\n",
      "Terms: \n",
      "a–associated (1, 1, 0.23), cue (1, 1, 0.23), vitellogenin-like (1, 1, 0.23), fittest (1, 1, 0.23), flattest (1, 1, 0.23), peaks: (1, 1, 0.23), ancestry-based (1, 1, 0.23), interpreting (1, 1, 0.23), life: (1, 1, 0.23), tape (1, 1, 0.23), intuition (2, 1, 0.21), behavioral (2, 1, 0.21), landscapes (2, 1, 0.21), shifts (3, 1, 0.20), responsiveness (4, 1, 0.19), regulate (5, 1, 0.18), specialization (5, 1, 0.18), fitness (7, 1, 0.17), task (7, 1, 0.17), provide (9, 1, 0.17)\n",
      "\n",
      "Cluster No. 28\n",
      "Size: 5\n",
      "Terms: \n",
      "nerve (4, 2, 0.28), injury (6, 2, 0.26), peripheral (7, 2, 0.26), cell (46, 2, 0.19), injury: (1, 1, 0.17), junction (1, 1, 0.17), tight (1, 1, 0.17), actin (1, 1, 0.17), cytoskeleton (1, 1, 0.17), schwann (1, 1, 0.17), claudin-19 (1, 1, 0.17), disrupt (1, 1, 0.17), crush (1, 1, 0.17), lesioned (1, 1, 0.17), nerves (1, 1, 0.17), sciatic (1, 1, 0.17), analyses, (2, 1, 0.15), neurogenesis (2, 1, 0.15), retinal (2, 1, 0.15), bioinformatic (2, 1, 0.15)\n",
      "\n",
      "Cluster No. 29\n",
      "Size: 4\n",
      "Terms: \n",
      "dorset (1, 1, 0.20), fourteenth-century (1, 1, 0.20), occupation (1, 1, 0.20), radiocarbon (1, 1, 0.20), denisova (1, 1, 0.20), hominin (1, 1, 0.20), palaeolithic (1, 1, 0.20), archaic (1, 1, 0.20), introgression (1, 1, 0.20), oceania (1, 1, 0.20), microsatellites (1, 1, 0.20), parentage (1, 1, 0.20), snps (1, 1, 0.20), onset (2, 1, 0.18), fossils (2, 1, 0.18), approximate (2, 1, 0.18), analysis: (2, 1, 0.18), arctic (3, 1, 0.17), cave (3, 1, 0.17), asia (5, 1, 0.16)\n",
      "\n",
      "Cluster No. 30\n",
      "Size: 6\n",
      "Terms: \n",
      "transduction (2, 2, 0.33), energy (46, 2, 0.20), proteins (51, 2, 0.19), back: (1, 1, 0.18), heterodimer (1, 1, 0.18), homodimer (1, 1, 0.18), tonb (1, 1, 0.18), 5.0 (1, 1, 0.18), signalp (1, 1, 0.18), biosilica-targeted (1, 1, 0.18), ton (2, 1, 0.16), elucidating (2, 1, 0.16), enterococcus (2, 1, 0.16), hypothetical (2, 1, 0.16), faecalis (3, 1, 0.16), diatom (3, 1, 0.16), optimizing (4, 1, 0.15), cycle (4, 1, 0.15), biosensor (4, 1, 0.15), anthracis (5, 1, 0.15)\n",
      "\n",
      "Cluster No. 31\n",
      "Size: 47\n",
      "Terms: \n",
      "array (22, 16, 0.37), doa_estimation (14, 14, 0.35), estimation (75, 14, 0.25), coprime (8, 8, 0.22), arrays (11, 8, 0.21), based (180, 13, 0.18), direction-of-arrival (6, 6, 0.17), radar (11, 6, 0.15), noncircular (5, 5, 0.14), signals (17, 6, 0.14), two-dimensional (7, 5, 0.14), direction (7, 5, 0.14), sparse (10, 5, 0.13), modulation (15, 5, 0.12), arrival (6, 4, 0.11), method (60, 6, 0.11), linear (12, 4, 0.10), coarray (3, 3, 0.09), nested (4, 3, 0.09), mimo (4, 3, 0.09)\n",
      "\n",
      "Cluster No. 32\n",
      "Size: 50\n",
      "Terms: \n",
      "influenza (39, 18, 0.42), influenza_virus (17, 9, 0.24), h7n9 (6, 6, 0.19), avian_influenza (6, 5, 0.16), infection (79, 8, 0.16), influenza_viruses (7, 5, 0.16), a(h7n9) (4, 4, 0.14), hemagglutinin (5, 4, 0.13), patients (57, 6, 0.13), respiratory (29, 5, 0.12), fatal (10, 4, 0.12), virus (89, 6, 0.12), viruses (45, 5, 0.11), severe (16, 4, 0.11), corticosteroid (3, 3, 0.11), treatment (51, 4, 0.09), avian (12, 3, 0.09), report (14, 3, 0.08), community-acquired (2, 2, 0.07), score (2, 2, 0.07)\n",
      "\n",
      "Cluster No. 33\n",
      "Size: 6\n",
      "Terms: \n",
      "organoids (2, 2, 0.31), brain (5, 2, 0.27), neural (11, 2, 0.24), zika_virus (33, 2, 0.20), human (178, 3, 0.20), cell (46, 2, 0.19), cortex (1, 1, 0.17), reproducibly (1, 1, 0.17), 3d-organoid (1, 1, 0.17), dish: (1, 1, 0.17), adherence (1, 1, 0.17), enteroaggregative (1, 1, 0.17), enteroids (1, 1, 0.17), segment- (1, 1, 0.17), organoid (1, 1, 0.17), host-specific (2, 1, 0.15), cerebral (2, 1, 0.15), alters (4, 1, 0.14), intestinal (4, 1, 0.14), form (5, 1, 0.14)\n",
      "\n",
      "Cluster No. 34\n",
      "Size: 43\n",
      "Terms: \n",
      "joint (19, 11, 0.30), inertial (29, 10, 0.25), running (9, 7, 0.22), kinematics (5, 5, 0.17), sensors: (6, 5, 0.16), muscle (8, 5, 0.16), motion (29, 6, 0.15), lower (12, 5, 0.15), magnetic (14, 5, 0.15), systematic_review (17, 5, 0.14), stiffness (6, 4, 0.13), inertial_sensors (6, 4, 0.13), hand (7, 4, 0.13), walking (8, 4, 0.12), sensor (46, 5, 0.11), human (178, 7, 0.11), gastrocnemius (3, 3, 0.11), muscle-tendon (3, 3, 0.11), sensor-based (3, 3, 0.11), limb (4, 3, 0.10)\n",
      "\n",
      "Cluster No. 35\n",
      "Size: 3\n",
      "Terms: \n",
      "classification (23, 2, 0.30), eliminating (1, 1, 0.24), quest: (1, 1, 0.24), supervised (1, 1, 0.24), biometric (1, 1, 0.24), narrowband (1, 1, 0.24), rf (1, 1, 0.24), walker (1, 1, 0.24), image: (1, 1, 0.24), links (3, 1, 0.21), lightweight (3, 1, 0.21), gesture (3, 1, 0.21), enabling (5, 1, 0.20), fine-grained (5, 1, 0.20), wi-fi_devices (6, 1, 0.19), commercial (7, 1, 0.19), online (8, 1, 0.18), signal (19, 1, 0.16), sensing (23, 1, 0.15), algorithms (24, 1, 0.15)\n",
      "\n",
      "Cluster No. 36\n",
      "Size: 20\n",
      "Terms: \n",
      "mutation (22, 7, 0.36), spontaneous (7, 5, 0.31), escherichia_coli (21, 5, 0.26), rates (6, 3, 0.19), base-pair (2, 2, 0.15), bacterial (37, 3, 0.14), substitution (4, 2, 0.14), mismatch (4, 2, 0.14), repair (5, 2, 0.13), dna (51, 3, 0.13), chromosome (8, 2, 0.12), whole-genome_sequencing (9, 2, 0.12), rate (10, 2, 0.12), long-term (13, 2, 0.11), evolution (99, 3, 0.11), bacteria (18, 2, 0.11), determinants (24, 2, 0.10), mutations (27, 2, 0.10), revealed (29, 2, 0.10), replication (35, 2, 0.09)\n",
      "\n",
      "Cluster No. 37\n",
      "Size: 151\n",
      "Terms: \n",
      "public (55, 26, 0.29), performance (78, 22, 0.22), audit (14, 14, 0.20), performance_management (14, 14, 0.20), local (36, 15, 0.18), accountability (15, 12, 0.17), government (24, 13, 0.17), nonprofit (15, 11, 0.15), public_sector (11, 10, 0.15), management (47, 13, 0.15), use (48, 13, 0.15), accounting (12, 9, 0.13), organizational (49, 11, 0.12), performance_information (9, 8, 0.12), policy (27, 9, 0.11), auditing (7, 7, 0.11), case (31, 9, 0.11), performance_measurement (7, 6, 0.10), government: (8, 6, 0.09), study (75, 9, 0.09)\n",
      "\n",
      "Cluster No. 38\n",
      "Size: 6\n",
      "Terms: \n",
      "role? (1, 1, 0.19), reca: (1, 1, 0.19), aaa+ (1, 1, 0.19), atpases (1, 1, 0.19), torsins: (1, 1, 0.19), glass (1, 1, 0.19), magnifying (1, 1, 0.19), bonds (1, 1, 0.19), novoprotein (1, 1, 0.19), amorphous (1, 1, 0.19), colloquium (1, 1, 0.19), folding: (2, 1, 0.17), stretched (2, 1, 0.17), looking (2, 1, 0.17), interspecies (2, 1, 0.17), typical (2, 1, 0.17), forde (2, 1, 0.17), driving (2, 1, 0.17), physics (2, 1, 0.17), folding (3, 1, 0.17)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 39\n",
      "Size: 25\n",
      "Terms: \n",
      "ambrosia (4, 4, 0.21), metabarcoding (6, 4, 0.19), fungal (23, 5, 0.19), edna (3, 3, 0.16), beetle (4, 3, 0.15), species (82, 5, 0.15), associates (5, 3, 0.15), dna (51, 4, 0.13), community (21, 3, 0.12), environmental (24, 3, 0.12), volvulus (2, 2, 0.11), xyleborus (2, 2, 0.11), curculionidae) (2, 2, 0.11), laurel (2, 2, 0.11), lauricola (2, 2, 0.11), raffaelea (2, 2, 0.11), euwallacea (2, 2, 0.11), fornicatus (2, 2, 0.11), nr. (2, 2, 0.11), avocado (2, 2, 0.11)\n",
      "\n",
      "Cluster No. 40\n",
      "Size: 42\n",
      "Terms: \n",
      "schizophrenia (6, 6, 0.27), variant (9, 5, 0.21), genetic (56, 7, 0.21), disease (54, 6, 0.18), association (23, 4, 0.14), variation (28, 4, 0.14), rare (6, 3, 0.13), variations (9, 3, 0.13), sequencing (41, 4, 0.13), copy_number (9, 3, 0.12), data (134, 5, 0.12), genome-wide (13, 3, 0.12), cows (2, 2, 0.11), common (19, 3, 0.11), human (178, 5, 0.11), disorders (2, 2, 0.10), mendelian (2, 2, 0.10), mutations (27, 3, 0.10), prioritization (3, 2, 0.10), analysis (196, 4, 0.08)\n",
      "\n",
      "Cluster No. 41\n",
      "Size: 41\n",
      "Terms: \n",
      "acute_respiratory_distress_syndrome (12, 11, 0.36), patients (57, 10, 0.24), septic (6, 6, 0.22), sepsis (5, 5, 0.19), shock: (5, 5, 0.19), mechanical (8, 5, 0.17), mortality (9, 5, 0.17), pressure (7, 4, 0.14), lung_injury (8, 4, 0.14), associated (42, 5, 0.13), end-expiratory (3, 3, 0.12), lung (21, 4, 0.12), ards (4, 3, 0.12), endothelial (4, 3, 0.12), acute_lung_injury (6, 3, 0.11), ventilation (6, 3, 0.11), systematic_review_meta-analysis (6, 3, 0.11), positive (8, 3, 0.10), critically_ill (10, 3, 0.10), vasodilatory (2, 2, 0.08)\n",
      "\n",
      "Cluster No. 42\n",
      "Size: 31\n",
      "Terms: \n",
      "ebola_virus (22, 14, 0.50), ebola (18, 8, 0.30), ebola_virus_disease (10, 5, 0.20), west (12, 5, 0.20), primates (6, 3, 0.13), africa (9, 3, 0.12), drc (2, 2, 0.11), infection (79, 4, 0.11), sudan (2, 2, 0.10), plant-made (4, 2, 0.09), nucleocapsid (5, 2, 0.09), convalescent (6, 2, 0.09), nonhuman (6, 2, 0.09), glycoprotein (13, 2, 0.08), 2 (15, 2, 0.08), implications (20, 2, 0.07), immune (22, 2, 0.07), – (25, 2, 0.07), human (178, 3, 0.07), vaccine (42, 2, 0.06)\n",
      "\n",
      "Cluster No. 43\n",
      "Size: 14\n",
      "Terms: \n",
      "faking (5, 5, 0.41), scores (4, 3, 0.24), response (40, 4, 0.22), personality (11, 3, 0.21), interview (2, 2, 0.19), forced-choice (2, 2, 0.18), equivalence (2, 2, 0.18), item (3, 2, 0.17), teacher (6, 2, 0.16), measures (14, 2, 0.14), time (26, 2, 0.12), effects (41, 2, 0.11), organizational (49, 2, 0.11), research (49, 2, 0.11), self-presentation (1, 1, 0.10), latencies (1, 1, 0.10), slow: (1, 1, 0.10), test-taker (1, 1, 0.10), interviews (1, 1, 0.10), lies: (1, 1, 0.10)\n",
      "\n",
      "Cluster No. 44\n",
      "Size: 6\n",
      "Terms: \n",
      "annual (6, 3, 0.31), diseases (13, 3, 0.28), national (14, 3, 0.27), notifiable (2, 2, 0.26), surveillance (19, 3, 0.26), australia, (2, 2, 0.24), preventable (2, 2, 0.24), immunisation (3, 2, 0.23), 2017 (11, 2, 0.19), report (14, 2, 0.18), coverage (18, 2, 0.17), vaccine (42, 2, 0.15), 2000–2014 (1, 1, 0.13), program: (1, 1, 0.13), summary (1, 1, 0.13), 2011–2015 (1, 1, 0.13), aboriginal (1, 1, 0.13), islander (1, 1, 0.13), people, (1, 1, 0.13), strait (1, 1, 0.13)\n",
      "\n",
      "Cluster No. 45\n",
      "Size: 2\n",
      "Terms: \n",
      "changes? (1, 1, 0.28), epochal (1, 1, 0.28), fourth (1, 1, 0.28), revolution: (1, 1, 0.28), collaboration (4, 1, 0.23), collective (4, 1, 0.23), institutional (4, 1, 0.23), korea: (4, 1, 0.23), digital (5, 1, 0.22), industrial (5, 1, 0.22), innovation (9, 1, 0.21), action (10, 1, 0.20), 2015 (12, 1, 0.20), public_health (16, 1, 0.19), south (16, 1, 0.19), mers (18, 1, 0.18), effectiveness (19, 1, 0.18), framework (19, 1, 0.18), social (40, 1, 0.16), network (42, 1, 0.16)\n",
      "\n",
      "Cluster No. 46\n",
      "Size: 4\n",
      "Terms: \n",
      "serotonin (3, 3, 0.56), transporter (8, 2, 0.32), antidepressants: (1, 1, 0.21), ssris (1, 1, 0.21), nanopharmacological (1, 1, 0.21), transporterin (1, 1, 0.21), vivoandin (1, 1, 0.21), alter (2, 1, 0.20), kinases (4, 1, 0.18), aspects (5, 1, 0.17), sites (8, 1, 0.16), ligand (10, 1, 0.16), vitro (11, 1, 0.15), allosteric (11, 1, 0.15), force (11, 1, 0.15), modulation (15, 1, 0.15), x-ray (23, 1, 0.14), sensing (23, 1, 0.13), revealed (29, 1, 0.13), structures (36, 1, 0.12)\n",
      "\n",
      "Cluster No. 47\n",
      "Size: 48\n",
      "Terms: \n",
      "welfare (14, 13, 0.38), welfare_state (8, 7, 0.22), employment (10, 7, 0.22), social (40, 8, 0.19), policies (11, 6, 0.18), labor_market (6, 5, 0.17), opinion (4, 4, 0.14), labour (4, 4, 0.14), political (12, 4, 0.12), public (55, 5, 0.11), union (3, 3, 0.11), equality (3, 3, 0.11), european (21, 4, 0.11), policy (27, 4, 0.10), preferences (6, 3, 0.10), gender (6, 3, 0.10), work: (8, 3, 0.10), education (15, 3, 0.09), new (99, 4, 0.08), dualization (2, 2, 0.08)\n",
      "\n",
      "Cluster No. 48\n",
      "Size: 8\n",
      "Terms: \n",
      "c (16, 3, 0.30), claw (2, 2, 0.30), facility (5, 2, 0.25), patterns (37, 2, 0.17), bed (1, 1, 0.15), iterator (1, 1, 0.15), idioms (1, 1, 0.15), additives (1, 1, 0.15), asphalt (1, 1, 0.15), clinoptilolite (1, 1, 0.15), mix (1, 1, 0.15), na-p1 (1, 1, 0.15), warm (1, 1, 0.15), zeolite (1, 1, 0.15), angiography (1, 1, 0.15), coherence (1, 1, 0.15), glaucoma (1, 1, 0.15), peripapillary (1, 1, 0.15), retina (1, 1, 0.15), api (1, 1, 0.15)\n",
      "\n",
      "Cluster No. 49\n",
      "Size: 6\n",
      "Terms: \n",
      "replication (35, 5, 0.41), initiation (5, 3, 0.34), origins (7, 3, 0.33), dna (51, 4, 0.30), genome-wide (13, 2, 0.20), regulation (23, 2, 0.18), mechanisms (28, 2, 0.17), cells (43, 2, 0.16), firing (1, 1, 0.14), assembly, (1, 1, 0.14), loading, (1, 1, 0.14), (ini-seq) (1, 1, 0.14), dna-binding (1, 1, 0.14), mtbp, (1, 1, 0.14), partner (1, 1, 0.14), treslin, (1, 1, 0.14), unwinding (2, 1, 0.13), dormant (2, 1, 0.13), hexameric (2, 1, 0.13), allele-specific (2, 1, 0.13)\n",
      "\n",
      "Cluster No. 50\n",
      "Size: 3\n",
      "Terms: \n",
      "zinc (7, 2, 0.29), agrobacterium (1, 1, 0.19), high-affinity (1, 1, 0.19), trocba (1, 1, 0.19), tumefaciens (1, 1, 0.19), ycic, (1, 1, 0.19), zinc-limiting (1, 1, 0.19), zint (1, 1, 0.19), znuabc, (1, 1, 0.19), cupriavidus (1, 1, 0.19), transposition (1, 1, 0.19), zinc-induced (1, 1, 0.19), cation (1, 1, 0.19), facilitator (1, 1, 0.19), transporters (2, 1, 0.18), adaptability (2, 1, 0.18), metallidurans (2, 1, 0.18), zur (4, 1, 0.16), insertion (4, 1, 0.16), chaperone (4, 1, 0.16)\n",
      "\n",
      "Cluster No. 51\n",
      "Size: 71\n",
      "Terms: \n",
      "bacterial (37, 9, 0.20), pseudomonas (15, 6, 0.16), biofilm (9, 5, 0.14), regulatory (15, 5, 0.13), rnas (5, 4, 0.13), escherichia_coli (21, 5, 0.12), symbiosis (6, 4, 0.12), antibiotic (9, 4, 0.11), response (40, 5, 0.11), host (46, 5, 0.11), global (47, 5, 0.11), bacteria (18, 4, 0.10), transcriptional (19, 4, 0.10), putida (3, 3, 0.10), k-12 (3, 3, 0.10), lactones (3, 3, 0.10), yersinia (4, 3, 0.10), small (23, 4, 0.10), enteritidis (4, 3, 0.10), gram-negative (4, 3, 0.10)\n",
      "\n",
      "Cluster No. 52\n",
      "Size: 23\n",
      "Terms: \n",
      "humanoid (4, 4, 0.24), hands (4, 4, 0.24), design (48, 6, 0.23), anthropomorphic (3, 3, 0.18), dexterous (3, 3, 0.18), control (64, 5, 0.18), soft (4, 3, 0.18), motion (29, 4, 0.17), robotic (6, 3, 0.17), hand (7, 3, 0.16), robots (12, 3, 0.15), compliant (2, 2, 0.13), walk (3, 2, 0.13), robot: (2, 2, 0.13), manipulation (5, 2, 0.11), stiffness (6, 2, 0.11), artificial (8, 2, 0.11), robot (19, 2, 0.09), optimization (32, 2, 0.08), 3d (43, 2, 0.08)\n",
      "\n",
      "Cluster No. 53\n",
      "Size: 2\n",
      "Terms: \n",
      "electrical (1, 1, 0.34), conditions, (1, 1, 0.34), technologies, (1, 1, 0.34), glioma (2, 1, 0.31), synaptic (2, 1, 0.31), circuits (3, 1, 0.30), neural (11, 1, 0.25), integrating (12, 1, 0.24), single-cell (14, 1, 0.24), transcriptomic (15, 1, 0.23), integration (22, 1, 0.22), different (32, 1, 0.20), species (82, 1, 0.17), data (134, 1, 0.15)\n",
      "\n",
      "Cluster No. 54\n",
      "Size: 4\n",
      "Terms: \n",
      "vitro (11, 2, 0.28), different (32, 2, 0.24), disassembly (1, 1, 0.20), giardia (1, 1, 0.20), length-dependent (1, 1, 0.20), lengths (1, 1, 0.20), myosin-2b (1, 1, 0.20), nonmuscle (1, 1, 0.20), tropomyosin (1, 1, 0.20), β-actinin (1, 1, 0.20), nitrilase1 (1, 1, 0.20), wnt/β-catenin (1, 1, 0.20), cryo-electron (1, 1, 0.20), isoforms (2, 1, 0.18), suppressor (2, 1, 0.18), tumour (2, 1, 0.18), microtubule (2, 1, 0.18), maintains (2, 1, 0.18), modulating (3, 1, 0.17), flagellar (4, 1, 0.17)\n",
      "\n",
      "Cluster No. 55\n",
      "Size: 2\n",
      "Terms: \n",
      "glycobiology (1, 1, 0.33), simplified: (1, 1, 0.33), organs? (1, 1, 0.33), polysialylated (1, 1, 0.33), ncam (2, 1, 0.30), glycan (3, 1, 0.28), brain (5, 1, 0.27), inflammation (5, 1, 0.27), regulator (11, 1, 0.23), roles (15, 1, 0.22), formation (15, 1, 0.22), diverse (29, 1, 0.20), recognition (32, 1, 0.19), development (55, 1, 0.17)\n",
      "\n",
      "Cluster No. 56\n",
      "Size: 24\n",
      "Terms: \n",
      "chlorophyll (11, 10, 0.42), photosystem (8, 8, 0.35), supercomplex (6, 6, 0.28), halomicronema (4, 4, 0.20), energy (46, 5, 0.16), transfer (16, 4, 0.16), cyanobacterium (3, 3, 0.15), far-red (3, 3, 0.15), light (22, 4, 0.15), f (4, 3, 0.15), algal (4, 3, 0.15), light-harvesting (4, 3, 0.15), green (8, 3, 0.13), structural_basis (19, 3, 0.11), f-producing (2, 2, 0.11), cyanobacteria (2, 2, 0.11), hongdechloris, (2, 2, 0.11), f-containing (2, 2, 0.11), photosynthesis (2, 2, 0.11), structure (105, 4, 0.11)\n",
      "\n",
      "Cluster No. 57\n",
      "Size: 3\n",
      "Terms: \n",
      "air (8, 2, 0.35), elevation (1, 1, 0.23), fine (1, 1, 0.23), particulate (1, 1, 0.23), fiberglass (1, 1, 0.23), magnetometer-based (1, 1, 0.23), pollution (2, 1, 0.21), frequencies (2, 1, 0.21), adenovirus (2, 1, 0.21), filters (2, 1, 0.21), short-term (4, 1, 0.19), matter (5, 1, 0.19), continuous (5, 1, 0.19), unit (6, 1, 0.18), handling (6, 1, 0.18), contact (7, 1, 0.18), empirical (10, 1, 0.17), lower (12, 1, 0.16), monitoring (12, 1, 0.16), behavior (13, 1, 0.16)\n",
      "\n",
      "Cluster No. 58\n",
      "Size: 2\n",
      "Terms: \n",
      "childcare: (1, 1, 0.33), out-of-school (1, 1, 0.33), income (2, 1, 0.31), availability (2, 1, 0.31), eu (2, 1, 0.31), distributional (3, 1, 0.29), inequality (4, 1, 0.28), member (8, 1, 0.25), evaluating (9, 1, 0.25), states (11, 1, 0.24), differences (14, 1, 0.23), exploring (14, 1, 0.23), quality (18, 1, 0.22)\n",
      "\n",
      "Cluster No. 59\n",
      "Size: 49\n",
      "Terms: \n",
      "protein (155, 10, 0.22), yeast (16, 4, 0.14), metabolic (21, 4, 0.14), analysis (196, 6, 0.12), mrna (8, 3, 0.12), translation (11, 3, 0.12), virulence (16, 3, 0.11), post‐translational (2, 2, 0.10), interactome (2, 2, 0.10), interaction (32, 3, 0.09), interactions (34, 3, 0.09), network (42, 3, 0.09), screens (4, 2, 0.09), proteomics (4, 2, 0.09), codon (5, 2, 0.09), absolute (4, 2, 0.09), biofilms (6, 2, 0.09), visualizing (5, 2, 0.09), proteins (51, 3, 0.09), protein-protein_interactions (6, 2, 0.08)\n",
      "\n",
      "Cluster No. 60\n",
      "Size: 47\n",
      "Terms: \n",
      "zika_virus (33, 24, 0.55), platelet_concentrates (7, 7, 0.21), zika (8, 7, 0.21), light (22, 8, 0.20), ultraviolet_c (6, 6, 0.19), inactivation (8, 6, 0.18), dengue (8, 5, 0.15), brazil (7, 4, 0.12), 2016 (10, 4, 0.11), blue (3, 3, 0.10), methylene (3, 3, 0.10), (zikv) (3, 3, 0.10), arboviruses (3, 3, 0.10), visible (8, 3, 0.09), dengue_virus (8, 3, 0.09), plasma (10, 3, 0.09), detection (64, 4, 0.08), public_health (16, 3, 0.08), quality (18, 3, 0.08), virus (89, 4, 0.07)\n",
      "\n",
      "Cluster No. 61\n",
      "Size: 3\n",
      "Terms: \n",
      "chromatin (6, 2, 0.41), nurd (1, 1, 0.26), copies (1, 1, 0.26), deacetylase (1, 1, 0.26), mta1 (1, 1, 0.26), rbbp4/7 (1, 1, 0.26), complex (84, 2, 0.25), nucleosome (2, 1, 0.24), recruit (2, 1, 0.24), repression (3, 1, 0.23), remodeling (3, 1, 0.23), subunit (9, 1, 0.19), cryo-electron_microscopy (12, 1, 0.18), provides (13, 1, 0.18), core (14, 1, 0.18), biology (16, 1, 0.17), interaction (32, 1, 0.15), insights (43, 1, 0.14), structure (105, 1, 0.12)\n",
      "\n",
      "Cluster No. 62\n",
      "Size: 11\n",
      "Terms: \n",
      "mycobacterium_tuberculosis (17, 4, 0.29), essential (16, 3, 0.22), rnha (2, 2, 0.22), rnase (2, 2, 0.20), mycobacterium (5, 2, 0.18), growth (13, 2, 0.15), synthase (13, 2, 0.15), transcriptional (19, 2, 0.14), assessment (25, 2, 0.13), microbe-enriched (1, 1, 0.11), genotoxic (1, 1, 0.11), hi (1, 1, 0.11), ber (1, 1, 0.11), human-adapted (1, 1, 0.11), necessities: (1, 1, 0.11), anti-tb (1, 1, 0.11), bedaquiline (1, 1, 0.11), forotor (1, 1, 0.11), f1fo-atp (1, 1, 0.11), β (1, 1, 0.11)\n",
      "\n",
      "Cluster No. 63\n",
      "Size: 6\n",
      "Terms: \n",
      "htlv-1 (4, 4, 0.39), t-cell (7, 3, 0.28), leukemia (15, 3, 0.24), 1 (17, 3, 0.24), virus (89, 4, 0.22), (htlv-1) (2, 2, 0.21), cell-to-cell (2, 2, 0.21), type (37, 3, 0.20), crucial (4, 2, 0.20), human (178, 4, 0.18), expression (35, 2, 0.14), cells (43, 2, 0.13), transmission (69, 2, 0.12), viral (72, 2, 0.12), bzip (1, 1, 0.12), icam-1 (1, 1, 0.12), upregulates (1, 1, 0.12), bursts (1, 1, 0.12), cell-cycle-related (1, 1, 0.12), leukemic (1, 1, 0.12)\n",
      "\n",
      "Cluster No. 64\n",
      "Size: 13\n",
      "Terms: \n",
      "directed (5, 3, 0.27), biocatalytic (2, 2, 0.20), biotechnological (3, 2, 0.19), sequence (37, 3, 0.19), synthetic (12, 2, 0.16), enzyme (18, 2, 0.15), biology (16, 2, 0.15), evolutionary (47, 2, 0.12), alkene (1, 1, 0.11), anti-markovnikov (1, 1, 0.11), metal-oxo–mediated (1, 1, 0.11), biocatalysts: (1, 1, 0.11), intelligently (1, 1, 0.11), aminase (1, 1, 0.11), pertucinogena (1, 1, 0.11), dehalogenases—from (1, 1, 0.11), halohydrin (1, 1, 0.11), learning-assisted (1, 1, 0.11), fad-dependent (1, 1, 0.11), monooxygenases: (1, 1, 0.11)\n",
      "\n",
      "Cluster No. 65\n",
      "Size: 8\n",
      "Terms: \n",
      "autophagy (6, 3, 0.36), proteins (51, 2, 0.16), gastroenteritis (1, 1, 0.15), chaperones (1, 1, 0.15), hsp70 (1, 1, 0.15), multivalency (1, 1, 0.15), p62/sqstm1 (1, 1, 0.15), zz-dependent (1, 1, 0.15), (hcc): (1, 1, 0.15), deleterious (1, 1, 0.15), hepatocellular (1, 1, 0.15), informatics (1, 1, 0.15), missense (1, 1, 0.15), mutation-structure-function (1, 1, 0.15), amyloidosis (1, 1, 0.15), lesson (1, 1, 0.15), longevity: (1, 1, 0.15), quiescence (1, 1, 0.15), tpr-containing (1, 1, 0.15), pp2a (1, 1, 0.15)\n",
      "\n",
      "Cluster No. 66\n",
      "Size: 7\n",
      "Terms: \n",
      "research (49, 4, 0.38), moderation (3, 2, 0.31), mediation (5, 2, 0.28), understanding (34, 2, 0.21), management (47, 2, 0.19), organizational (49, 2, 0.19), basics (1, 1, 0.18), mixed-effects (1, 1, 0.18), take-away (1, 1, 0.18), abc (1, 1, 0.18), too-much-of-a-good-thing (1, 1, 0.18), multiple-membership (1, 1, 0.18), nuisance (1, 1, 0.18), questions: (1, 1, 0.18), variances (1, 1, 0.18), depends” (1, 1, 0.18), “it (1, 1, 0.18), models: (3, 1, 0.15), studying (5, 1, 0.14), strategic (9, 1, 0.13)\n",
      "\n",
      "Cluster No. 67\n",
      "Size: 3\n",
      "Terms: \n",
      "acceptable (1, 1, 0.24), crises (1, 1, 0.24), ethics, (1, 1, 0.24), risk, (1, 1, 0.24), vaccinate (1, 1, 0.24), community-centered (1, 1, 0.24), liberia: (1, 1, 0.24), post-sars (1, 1, 0.24), racialization (1, 1, 0.24), women? (2, 1, 0.22), h1n1 (2, 1, 0.22), pregnant (4, 1, 0.20), view (6, 1, 0.19), people (8, 1, 0.18), virus: (8, 1, 0.18), governance (9, 1, 0.18), urban (11, 1, 0.17), politics (11, 1, 0.17), chinese (14, 1, 0.17), ebola (18, 1, 0.16)\n",
      "\n",
      "Cluster No. 68\n",
      "Size: 3\n",
      "Terms: \n",
      "noise (6, 2, 0.40), environmental (24, 2, 0.31), annoyance (1, 1, 0.24), ulcer (1, 1, 0.24), typhoid (1, 1, 0.24), region: (2, 1, 0.22), prophylaxis (2, 1, 0.22), trials (5, 1, 0.20), fever (5, 1, 0.20), randomized (5, 1, 0.20), patients: (6, 1, 0.19), preventing (7, 1, 0.19), guidelines (7, 1, 0.19), meta-analysis (8, 1, 0.18), critically_ill (10, 1, 0.18), safety (14, 1, 0.17), vaccines (16, 1, 0.16), efficacy (17, 1, 0.16), systematic_review (17, 1, 0.16), stress (19, 1, 0.16)\n",
      "\n",
      "Cluster No. 69\n",
      "Size: 129\n",
      "Terms: \n",
      "slam (22, 20, 0.27), navigation (39, 18, 0.22), monocular (15, 15, 0.22), odometry (17, 15, 0.21), mapping (35, 17, 0.21), visual (27, 15, 0.19), autonomous (30, 14, 0.18), visual-inertial (10, 10, 0.15), robust (46, 13, 0.15), robot (19, 10, 0.14), estimation (75, 12, 0.12), localization (98, 12, 0.12), system (77, 11, 0.11), mobile (51, 10, 0.11), environments (32, 9, 0.11), tightly-coupled (8, 7, 0.11), rgb-d (9, 7, 0.11), vision (9, 7, 0.11), vehicle (20, 8, 0.11), based (180, 13, 0.11)\n",
      "\n",
      "Cluster No. 70\n",
      "Size: 6\n",
      "Terms: \n",
      "neutrophil (7, 4, 0.40), l-selectin (2, 2, 0.24), traps (2, 2, 0.24), extracellular (4, 2, 0.22), behaviour (6, 2, 0.21), binding (58, 3, 0.21), catch-bonds (1, 1, 0.13), endothelium (1, 1, 0.13), integrin (1, 1, 0.13), mechanotransduce (1, 1, 0.13), selectin (1, 1, 0.13), (psgl-1) (1, 1, 0.13), e-selectin (1, 1, 0.13), immunoprecipitation-based (1, 1, 0.13), ligand-1 (1, 1, 0.13), p-selectin (1, 1, 0.13), head-to-tail (1, 1, 0.13), ezrin (1, 1, 0.13), moesin (1, 1, 0.13), protrusive (1, 1, 0.13)\n",
      "\n",
      "Cluster No. 71\n",
      "Size: 6\n",
      "Terms: \n",
      "autoinflammatory (3, 3, 0.35), early-onset (2, 2, 0.24), immunodeficiency, (2, 2, 0.24), periodic (3, 2, 0.23), mutations (27, 2, 0.16), disease (54, 2, 0.14), gene (59, 2, 0.14), biallelic (1, 1, 0.13), deubiquitinase (1, 1, 0.13), hypomorphic (1, 1, 0.13), otulipenia, (1, 1, 0.13), amylopectinosis, (1, 1, 0.13), autoinflammation, (1, 1, 0.13), hoip (1, 1, 0.13), lubac (1, 1, 0.13), lymphangiectasia (1, 1, 0.13), underlies (1, 1, 0.13), (pfit) (1, 1, 0.13), actin-regulatory (1, 1, 0.13), fever, (1, 1, 0.13)\n",
      "\n",
      "Cluster No. 72\n",
      "Size: 67\n",
      "Terms: \n",
      "influenza (39, 10, 0.23), masks (8, 7, 0.21), pandemic (10, 7, 0.20), epidemic (31, 7, 0.16), respirators (5, 5, 0.16), healthcare_workers (7, 5, 0.15), infectious_disease (7, 5, 0.15), epidemics (7, 5, 0.15), respiratory (29, 6, 0.14), outbreak (67, 7, 0.14), medical (13, 5, 0.14), infection (79, 7, 0.13), outbreaks (18, 5, 0.13), effectiveness (19, 5, 0.13), transmission (69, 6, 0.12), models (42, 5, 0.11), healthcare (14, 4, 0.11), ebola (18, 4, 0.10), n95 (3, 3, 0.10), spread (20, 4, 0.10)\n",
      "\n",
      "Cluster No. 73\n",
      "Size: 2\n",
      "Terms: \n",
      "arrestin (1, 1, 0.38), rhodopsin (2, 1, 0.35), biased (3, 1, 0.33), ligands (5, 1, 0.31), femtosecond (8, 1, 0.29), drug_discovery (9, 1, 0.28), bound (10, 1, 0.28), opportunities (11, 1, 0.27), laser (11, 1, 0.27), x-ray (23, 1, 0.24), challenges (29, 1, 0.23), crystal_structure (45, 1, 0.21)\n",
      "\n",
      "Cluster No. 74\n",
      "Size: 3\n",
      "Terms: \n",
      "68 (1, 1, 0.23), stat1 (1, 1, 0.23), eosinophilia (1, 1, 0.23), hyperreactivity (1, 1, 0.23), il‐33 (1, 1, 0.23), orchestrates (1, 1, 0.23), rsv‐evoked (1, 1, 0.23), gain-of-function (2, 1, 0.21), gammaherpesvirus (2, 1, 0.21), impairs (2, 1, 0.21), mediate (4, 1, 0.19), airway (4, 1, 0.19), advanced (5, 1, 0.19), innate (6, 1, 0.18), cd8+ (7, 1, 0.18), pulmonary (7, 1, 0.18), candida (7, 1, 0.18), complexity (7, 1, 0.18), need (8, 1, 0.17), delivery (13, 1, 0.16)\n",
      "\n",
      "Cluster No. 75\n",
      "Size: 7\n",
      "Terms: \n",
      "nmda (6, 6, 0.59), receptors (19, 4, 0.33), allosteric (11, 3, 0.27), triheteromeric (2, 2, 0.23), modulation (15, 2, 0.17), receptor (43, 2, 0.14), cell-surface (1, 1, 0.12), glun2c-selective (1, 1, 0.12), protein-ligand (1, 1, 0.12), output (1, 1, 0.12), receptors: (1, 1, 0.12), glun2d-containing (1, 1, 0.12), hippocampal (1, 1, 0.12), interneuron (1, 1, 0.12), n-methyl-d-aspartate (1, 1, 0.12), operation (2, 1, 0.11), synaptic (2, 1, 0.11), interneurons (2, 1, 0.11), biophysical (2, 1, 0.11), bioactive (2, 1, 0.11)\n",
      "\n",
      "Cluster No. 76\n",
      "Size: 4\n",
      "Terms: \n",
      "capital (1, 1, 0.23), gap: (1, 1, 0.23), perceive (1, 1, 0.23), circumstances (1, 1, 0.23), money, (1, 1, 0.23), mo’ (1, 1, 0.23), problem: (1, 1, 0.23), women (2, 1, 0.21), shocks (2, 1, 0.21), manager (2, 1, 0.21), minority (2, 1, 0.21), findings (4, 1, 0.19), responsiveness (4, 1, 0.19), experiment (4, 1, 0.19), uncovering (5, 1, 0.19), gender (6, 1, 0.18), financial (6, 1, 0.18), economic (6, 1, 0.18), strategic (9, 1, 0.17), representation (9, 1, 0.17)\n",
      "\n",
      "Cluster No. 77\n",
      "Size: 19\n",
      "Terms: \n",
      "kinetochore (11, 11, 0.56), unconventional (4, 4, 0.23), kinetoplastid (4, 4, 0.23), brucei (6, 4, 0.22), spindle (3, 3, 0.18), trypanosoma (5, 3, 0.17), anaphase (2, 2, 0.13), metaphase (2, 2, 0.13), assembly (27, 3, 0.13), machinery (3, 2, 0.12), trypanosome (3, 2, 0.12), protein (155, 4, 0.11), eukaryotes (5, 2, 0.11), outer (5, 2, 0.11), chromosome (8, 2, 0.11), kinase (14, 2, 0.10), inhibitors (35, 2, 0.08), proteins (51, 2, 0.08), identification (70, 2, 0.07), kkt10/19 (1, 1, 0.07)\n",
      "\n",
      "Cluster No. 78\n",
      "Size: 19\n",
      "Terms: \n",
      "emotional (17, 9, 0.38), burnout (10, 8, 0.37), police (27, 7, 0.27), employees (11, 5, 0.23), job (25, 5, 0.20), hotel (8, 3, 0.15), job_satisfaction (8, 3, 0.14), luxury (2, 2, 0.12), satisfaction: (2, 2, 0.12), labor, (2, 2, 0.12), frontline (3, 2, 0.11), officers (3, 2, 0.11), effect (36, 3, 0.11), officer (4, 2, 0.11), organizational (49, 3, 0.10), occupational (5, 2, 0.10), labor (5, 2, 0.10), public_service (9, 2, 0.09), south (16, 2, 0.08), strategies (23, 2, 0.08)\n",
      "\n",
      "Cluster No. 79\n",
      "Size: 2\n",
      "Terms: \n",
      "expert-coded (1, 1, 0.35), disagree: (1, 1, 0.35), irt (2, 1, 0.33), experts (2, 1, 0.33), surveys (2, 1, 0.33), expert (3, 1, 0.31), panel (4, 1, 0.30), aggregation (7, 1, 0.27), consequences (9, 1, 0.26), response (40, 1, 0.20), models (42, 1, 0.20), data (134, 1, 0.15)\n",
      "\n",
      "Cluster No. 80\n",
      "Size: 5\n",
      "Terms: \n",
      "antimicrobial (14, 4, 0.51), peptide (20, 3, 0.36), machine (5, 2, 0.30), activity (35, 2, 0.21), learning? (1, 1, 0.18), peptides, (1, 1, 0.18), peptides: (1, 1, 0.18), therapeutics: (1, 1, 0.18), apd3: (1, 1, 0.18), defence (2, 1, 0.17), immunology (2, 1, 0.17), learning-based (3, 1, 0.16), directions (4, 1, 0.15), drug_discovery (9, 1, 0.14), peptides (10, 1, 0.13), education (15, 1, 0.13), progress (14, 1, 0.13), status (14, 1, 0.13), future (19, 1, 0.12), recent (21, 1, 0.12)\n",
      "\n",
      "Cluster No. 81\n",
      "Size: 29\n",
      "Terms: \n",
      "frankia (18, 17, 0.52), sp._nov., (12, 9, 0.30), isolated (27, 9, 0.25), root (7, 6, 0.21), actinorhizal (5, 5, 0.18), actinobacterium (5, 5, 0.18), nodules (4, 4, 0.15), plants (14, 4, 0.12), sp. (21, 4, 0.12), rhizobium (3, 3, 0.12), nitrogen-fixing (3, 3, 0.12), strains (24, 3, 0.08), promotion (2, 2, 0.08), alni (2, 2, 0.08), infective (2, 2, 0.08), microsymbiont (2, 2, 0.08), nodule (2, 2, 0.08), nodulate (2, 2, 0.08), endophytic (3, 2, 0.08), proposal (5, 2, 0.07)\n",
      "\n",
      "Cluster No. 82\n",
      "Size: 2\n",
      "Terms: \n",
      "australiensis (1, 1, 0.23), enriched (1, 1, 0.23), oryza (1, 1, 0.23), salt‐treated (1, 1, 0.23), seedlings (1, 1, 0.23), (dclcyb1) (1, 1, 0.23), bydaucus (1, 1, 0.23), carotalycopene (1, 1, 0.23), carotenoid, (1, 1, 0.23), gibberellin (1, 1, 0.23), increasednicotiana (1, 1, 0.23), promoted (1, 1, 0.23), tabacumfitness (1, 1, 0.23), β-cyclase (1, 1, 0.23), energetics (2, 1, 0.21), roots (4, 1, 0.19), positive (8, 1, 0.17), chlorophyll (11, 1, 0.16), involved (12, 1, 0.16), transport (12, 1, 0.16)\n",
      "\n",
      "Cluster No. 83\n",
      "Size: 7\n",
      "Terms: \n",
      "drug (29, 3, 0.34), silicomethods (1, 1, 0.19), extendable (1, 1, 0.19), pathvisio (1, 1, 0.19), toolbox (1, 1, 0.19), multi-source (1, 1, 0.19), repositioning (1, 1, 0.19), ddx (1, 1, 0.19), impacted (1, 1, 0.19), ipathwayguide (1, 1, 0.19), significantly (1, 1, 0.19), 3: (2, 1, 0.17), pharmacology (2, 1, 0.17), melanoma (4, 1, 0.16), putative (6, 1, 0.15), molecule (7, 1, 0.15), repurposing (8, 1, 0.14), 3 (8, 1, 0.14), probabilistic (8, 1, 0.14), graph (9, 1, 0.14)\n",
      "\n",
      "Cluster No. 84\n",
      "Size: 26\n",
      "Terms: \n",
      "applications (24, 5, 0.25), nucleic_acid (8, 3, 0.18), crispr-cas9 (2, 2, 0.15), beauty (2, 2, 0.15), sleeping (3, 2, 0.14), precision (4, 2, 0.14), stem (5, 2, 0.14), crispr (5, 2, 0.13), gene (59, 3, 0.13), detection (64, 3, 0.12), technologies (8, 2, 0.12), system (77, 3, 0.12), generation (14, 2, 0.11), health (28, 2, 0.10), targeting (31, 2, 0.10), cell (46, 2, 0.09), research (49, 2, 0.09), clinical (51, 2, 0.09), dna (51, 2, 0.09), crispr-based (1, 1, 0.08)\n",
      "\n",
      "Cluster No. 85\n",
      "Size: 120\n",
      "Terms: \n",
      "team (22, 16, 0.24), job (25, 16, 0.22), organizational (49, 18, 0.22), conflict (14, 14, 0.22), performance (78, 18, 0.20), role (83, 16, 0.17), relationship (28, 12, 0.16), climate (10, 10, 0.16), work (20, 11, 0.16), moderating (11, 9, 0.14), effects (41, 11, 0.14), impact (54, 11, 0.13), employee (11, 8, 0.13), emotional (17, 8, 0.12), behaviors (7, 6, 0.10), team_performance (6, 6, 0.10), conflict_management (6, 6, 0.10), effect (36, 8, 0.10), outcomes (20, 7, 0.10), turnover (12, 6, 0.09)\n",
      "\n",
      "Cluster No. 86\n",
      "Size: 7\n",
      "Terms: \n",
      "secretion_system (6, 4, 0.39), ix (4, 3, 0.31), genitalium (2, 2, 0.22), porphyromonas (3, 2, 0.21), type (37, 3, 0.21), human (178, 3, 0.14), characterization (48, 2, 0.13), (t9ss) (1, 1, 0.12), gingivalis (1, 1, 0.12), outer-membrane (1, 1, 0.12), β-sandwich (1, 1, 0.12), gingivalis. (1, 1, 0.12), oral-microbiomic (1, 1, 0.12), porz, (1, 1, 0.12), type-ix (1, 1, 0.12), arches (1, 1, 0.12), gldm (1, 1, 0.12), gliding (1, 1, 0.12), porm (1, 1, 0.12), adhesin (1, 1, 0.12)\n",
      "\n",
      "Cluster No. 87\n",
      "Size: 2\n",
      "Terms: \n",
      "betalain (2, 2, 0.49), biosynthesis (4, 2, 0.41), committed (1, 1, 0.24), pigments (1, 1, 0.24), tilting (1, 1, 0.24), benefit (2, 1, 0.22), metabolite (3, 1, 0.21), elucidation (4, 1, 0.20), heterologous (4, 1, 0.20), step (6, 1, 0.19), engineering (10, 1, 0.18), enables (11, 1, 0.18), metabolism (12, 1, 0.17), plants (14, 1, 0.17), plant (20, 1, 0.16), enhanced (20, 1, 0.16), improved (47, 1, 0.13), human (178, 1, 0.10)\n",
      "\n",
      "Cluster No. 88\n",
      "Size: 1\n",
      "Terms: \n",
      "restricted (2, 1, 0.42), research: (4, 1, 0.38), restriction (4, 1, 0.38), indirect (5, 1, 0.37), range (13, 1, 0.32), accounting (12, 1, 0.32), selection (35, 1, 0.26), organizational (49, 1, 0.25), research (49, 1, 0.25)\n",
      "\n",
      "Cluster No. 89\n",
      "Size: 1\n",
      "Terms: \n",
      "array-based (1, 1, 0.43), foodborne-associated (1, 1, 0.43), noroviruses (2, 1, 0.39), hepatitis (7, 1, 0.33), genotyping (7, 1, 0.33), sensitive (11, 1, 0.31), platform (12, 1, 0.30), virus (89, 1, 0.20), human (178, 1, 0.17)\n",
      "\n",
      "Cluster No. 90\n",
      "Size: 1\n",
      "Terms: \n",
      "artistic (1, 1, 0.31), developed (1, 1, 0.31), expressions (1, 1, 0.31), globalized (1, 1, 0.31), northeast (1, 1, 0.31), world: (2, 1, 0.29), project (3, 1, 0.27), contributions (3, 1, 0.27), traditional (4, 1, 0.26), exploratory (7, 1, 0.24), brazil (7, 1, 0.24), science (8, 1, 0.24), communication (30, 1, 0.19)\n",
      "\n",
      "Cluster No. 91\n",
      "Size: 3\n",
      "Terms: \n",
      "education (15, 7, 0.67), teacher (6, 4, 0.45), physical (11, 4, 0.39), 30-year (1, 1, 0.13), standards-based (1, 1, 0.13), doctoral (1, 1, 0.13), socialization (2, 1, 0.12), policies, (2, 1, 0.12), scoping (3, 1, 0.12), agendas, (3, 1, 0.12), literature (4, 1, 0.11), addressing (4, 1, 0.11), reform (4, 1, 0.11), satisfaction (5, 1, 0.11), influencing (6, 1, 0.11), practices (8, 1, 0.10), review (34, 1, 0.08)\n",
      "\n",
      "Cluster No. 92\n",
      "Size: 1\n",
      "Terms: \n",
      "kpi-based (1, 1, 0.43), maximizes (1, 1, 0.43), coefficient (2, 1, 0.40), soft (4, 1, 0.36), probabilistic (8, 1, 0.33), determination (21, 1, 0.28), sensor (46, 1, 0.24), development (55, 1, 0.23), approach (81, 1, 0.21)\n",
      "\n",
      "Cluster No. 93\n",
      "Size: 1\n",
      "Terms: \n",
      "swiss-model: (1, 1, 0.56), homology (4, 1, 0.47), modelling (15, 1, 0.39), complexes (15, 1, 0.38), structures (36, 1, 0.33), protein (155, 1, 0.23)\n",
      "\n",
      "Cluster No. 94\n",
      "Size: 1\n",
      "Terms: \n",
      "perspective: (1, 1, 0.35), statistics (1, 1, 0.35), eu (2, 1, 0.32), condition (2, 1, 0.32), cultural (3, 1, 0.31), index (4, 1, 0.30), economic (6, 1, 0.28), countries (7, 1, 0.27), culture (10, 1, 0.26), european (21, 1, 0.23), comparative (31, 1, 0.21), social (40, 1, 0.20)\n",
      "\n",
      "Cluster No. 95\n",
      "Size: 1\n",
      "Terms: \n",
      "liberia (2, 1, 0.50), rural (2, 1, 0.50), vulnerability (2, 1, 0.50), ebola_virus_disease (10, 1, 0.40), social (40, 1, 0.31)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster No. 96\n",
      "Size: 1\n",
      "Terms: \n",
      "agoodbayesian (1, 1, 1.00)\n",
      "\n",
      "Cluster No. 97\n",
      "Size: 12\n",
      "Terms: \n",
      "angiotensin (11, 7, 0.50), receptor (43, 6, 0.33), type_2 (6, 3, 0.23), adipose_tissue (9, 3, 0.22), adipocyte (2, 2, 0.18), browning (2, 2, 0.18), white (4, 2, 0.16), brown (7, 2, 0.15), ii (18, 2, 0.13), mechanisms (28, 2, 0.12), management (47, 2, 0.11), remodelling (1, 1, 0.10), renewal (1, 1, 0.10), cardio-immunology (1, 1, 0.10), myocarditis: (1, 1, 0.10), pathophysiology (1, 1, 0.10), physiology (1, 1, 0.10), transduction: (1, 1, 0.10), c21 (1, 1, 0.10), c57bl/6 (1, 1, 0.10)\n",
      "\n",
      "Cluster No. 98\n",
      "Size: 7\n",
      "Terms: \n",
      "tissue (8, 3, 0.35), translational (8, 3, 0.35), research (49, 3, 0.25), scoring (7, 2, 0.24), lung (21, 2, 0.20), approaches (30, 2, 0.19), knockin (1, 1, 0.15), mouse-adapted (1, 1, 0.15), pitfalls (1, 1, 0.15), principles (1, 1, 0.15), stains (1, 1, 0.15), evaluate (1, 1, 0.15), semiquantitative (1, 1, 0.15), reproducibility (1, 1, 0.15), reproducible (2, 1, 0.14), vital (3, 1, 0.13), fundamental (3, 1, 0.13), dpp4 (4, 1, 0.13), scores (4, 1, 0.13), aging (4, 1, 0.13)\n",
      "\n",
      "Cluster No. 99\n",
      "Size: 1\n",
      "Terms: \n",
      "multidimensional (1, 1, 0.69), scaling (3, 1, 0.60), survey (39, 1, 0.39)\n",
      "\n",
      "Cluster No. 100\n",
      "Size: 2\n",
      "Terms: \n",
      "truant (2, 2, 0.42), use (48, 2, 0.25), impact (54, 2, 0.24), youths’ (1, 1, 0.23), charges: (1, 1, 0.23), delinquency (1, 1, 0.23), drug-using (1, 1, 0.23), self-reported (1, 1, 0.23), youths' (1, 1, 0.23), alcohol (2, 1, 0.21), marijuana (2, 1, 0.21), taking (2, 1, 0.21), brief (4, 1, 0.19), arrest (4, 1, 0.19), intervention (4, 1, 0.19), sexual (4, 1, 0.19), validity (6, 1, 0.18), services (12, 1, 0.16), longitudinal (12, 1, 0.16), risk (44, 1, 0.13)\n",
      "\n",
      "Cluster No. 101\n",
      "Size: 1\n",
      "Terms: \n",
      "4-methylumbelliferone (1, 1, 0.47), thermogenic (1, 1, 0.47), capacity (4, 1, 0.40), brown (7, 1, 0.37), improves (7, 1, 0.36), adipose_tissue (9, 1, 0.36)\n",
      "\n",
      "Cluster No. 102\n",
      "Size: 1\n",
      "Terms: \n",
      "constraint-negotiation (1, 1, 0.33), equation (1, 1, 0.33), leisure (1, 1, 0.33), leisure-time (1, 1, 0.33), context: (2, 1, 0.30), predictive (3, 1, 0.28), squares (5, 1, 0.26), partial (6, 1, 0.26), physical (11, 1, 0.24), power (13, 1, 0.23), modeling (30, 1, 0.20), activity (35, 1, 0.19), models (42, 1, 0.18), structural (55, 1, 0.17), approach (81, 1, 0.16)\n",
      "\n",
      "Cluster No. 103\n",
      "Size: 2\n",
      "Terms: \n",
      "externalities (1, 1, 0.44), absence-proofness: (1, 1, 0.44), fair (3, 1, 0.39), division (10, 1, 0.32), stability (11, 1, 0.32), core (14, 1, 0.30), presence (17, 1, 0.29), group (26, 1, 0.27)\n",
      "\n",
      "Cluster No. 104\n",
      "Size: 1\n",
      "Terms: \n",
      "behaviours (1, 1, 0.39), professionals (1, 1, 0.39), trainees (1, 1, 0.39), aviation (2, 1, 0.36), perception (5, 1, 0.32), engineering (10, 1, 0.29), differences (14, 1, 0.27), factors (29, 1, 0.24), risk (44, 1, 0.22), domain (57, 1, 0.21)\n",
      "\n",
      "Cluster No. 105\n",
      "Size: 1\n",
      "Terms: \n",
      "blocking (1, 1, 0.33), isobutanol (1, 1, 0.33), successively (1, 1, 0.33), ethanol (2, 1, 0.30), competing (2, 1, 0.30), glycerol (2, 1, 0.30), saccharomyces (7, 1, 0.25), cerevisiae (8, 1, 0.25), formation (15, 1, 0.22), yeast (16, 1, 0.22), pathways (17, 1, 0.22), metabolic (21, 1, 0.21), improving (23, 1, 0.21), production (24, 1, 0.21)\n",
      "\n",
      "Cluster No. 106\n",
      "Size: 1\n",
      "Terms: \n",
      "(myxozoa:myxosporea) (1, 1, 0.33), cnidarian (1, 1, 0.33), molnari (1, 1, 0.33), motility (1, 1, 0.33), sphaerospora (1, 1, 0.33), observed (3, 1, 0.29), parasitic (4, 1, 0.28), stages (6, 1, 0.26), fish (8, 1, 0.25), blood (10, 1, 0.24), cell (46, 1, 0.19), model (81, 1, 0.16), new (99, 1, 0.16)\n",
      "\n",
      "Cluster No. 107\n",
      "Size: 9\n",
      "Terms: \n",
      "influenza (39, 3, 0.23), data (134, 4, 0.23), disease (54, 3, 0.21), forecasting (6, 2, 0.21), seasonal (6, 2, 0.21), dengue (8, 2, 0.21), social (40, 2, 0.15), nonmechanistic (1, 1, 0.13), one-week-ahead (1, 1, 0.13), streams: (1, 1, 0.13), city (1, 1, 0.13), tweets (1, 1, 0.13), web: (1, 1, 0.13), 2014: (1, 1, 0.13), august (1, 1, 0.13), ebola, (1, 1, 0.13), media's (1, 1, 0.13), rumors (1, 1, 0.13), engine (1, 1, 0.13), prevention’s (1, 1, 0.13)\n",
      "\n",
      "Cluster No. 108\n",
      "Size: 1\n",
      "Terms: \n",
      "participation? (1, 1, 0.55), consultation (2, 1, 0.51), encourage (2, 1, 0.51), social (40, 1, 0.32), public (55, 1, 0.29)\n",
      "\n",
      "Cluster No. 109\n",
      "Size: 1\n",
      "Terms: \n",
      "autoregressive (2, 1, 0.38), conditionally (2, 1, 0.38), time: (2, 1, 0.38), applying (3, 1, 0.36), 50 (4, 1, 0.35), american (7, 1, 0.32), states (11, 1, 0.30), space (18, 1, 0.27), models (42, 1, 0.23)\n",
      "\n",
      "Cluster No. 110\n",
      "Size: 2\n",
      "Terms: \n",
      "renal (6, 2, 0.39), administered (1, 1, 0.24), hypertensive (1, 1, 0.24), losartan (1, 1, 0.24), mas (1, 1, 0.24), perinatally (1, 1, 0.24), rats (1, 1, 0.24), spontaneously (1, 1, 0.24), infancy (1, 1, 0.24), potentiates (1, 1, 0.24), iga (2, 1, 0.22), augments (2, 1, 0.22), cardiac (3, 1, 0.21), interferon (9, 1, 0.18), respiratory_syncytial (10, 1, 0.17), ace2 (13, 1, 0.17), virus_infection (15, 1, 0.16), immunity (20, 1, 0.15), expression (35, 1, 0.14), type (37, 1, 0.14)\n",
      "\n",
      "Cluster No. 111\n",
      "Size: 1\n",
      "Terms: \n",
      "civilizations (1, 1, 0.56), technical (2, 1, 0.52), biotechnology (4, 1, 0.47), lifetime (7, 1, 0.44)\n",
      "\n",
      "Cluster No. 112\n",
      "Size: 1\n",
      "Terms: \n",
      "flavonoids: (1, 1, 0.83), overview (18, 1, 0.55)\n",
      "\n",
      "Cluster No. 113\n",
      "Size: 1\n",
      "Terms: \n",
      "classroom (2, 2, 0.70), genome: (2, 1, 0.32), illuminate (2, 1, 0.32), observation (4, 1, 0.29), bioinformatics (7, 1, 0.27), genomics (12, 1, 0.25), tools (22, 1, 0.22), data (134, 1, 0.15)\n",
      "\n",
      "Cluster No. 114\n",
      "Size: 2\n",
      "Terms: \n",
      "equilibrium (1, 1, 0.27), hyaluronan-binding (1, 1, 0.27), two-state (1, 1, 0.27), alpha-v–containing (1, 1, 0.27), falciparum (1, 1, 0.27), integrins (1, 1, 0.27), sporozoite (1, 1, 0.27), trap (1, 1, 0.27), cd44 (2, 1, 0.24), rolling (2, 1, 0.24), protein, (3, 1, 0.23), plasmodium (8, 1, 0.20), mechanical (8, 1, 0.20), force (11, 1, 0.19), receptors (19, 1, 0.17), surface (20, 1, 0.17), effect (36, 1, 0.15), cell (46, 1, 0.15), host (46, 1, 0.15), domain (57, 1, 0.14)\n",
      "\n",
      "Cluster No. 115\n",
      "Size: 1\n",
      "Terms: \n",
      "bruijn (1, 1, 0.35), megahit: (1, 1, 0.35), single-node (1, 1, 0.35), succinct (1, 1, 0.35), ultra-fast (1, 1, 0.35), de (2, 1, 0.32), graph (9, 1, 0.26), metagenomics (16, 1, 0.23), large (19, 1, 0.23), solution (20, 1, 0.22), assembly (27, 1, 0.21), complex (84, 1, 0.17)\n",
      "\n",
      "Cluster No. 116\n",
      "Size: 7\n",
      "Terms: \n",
      "inverse (3, 2, 0.27), statistical (7, 2, 0.23), fitness (7, 2, 0.23), landscape (16, 2, 0.20), targets (19, 2, 0.19), antibodies (23, 2, 0.19), vaccine (42, 2, 0.17), inform (1, 1, 0.15), principal (1, 1, 0.15), sub-dominant (1, 1, 0.15), co-evolution (1, 1, 0.15), hiv/hcv (1, 1, 0.15), e2 (1, 1, 0.15), immunologically-vulnerable (1, 1, 0.15), deconvolving (1, 1, 0.15), crossroad (1, 1, 0.15), hiv: (1, 1, 0.15), physics, (1, 1, 0.15), ising (1, 1, 0.15), problems: (1, 1, 0.15)\n",
      "\n",
      "Cluster No. 117\n",
      "Size: 1\n",
      "Terms: \n",
      "560 (1, 1, 0.38), insurance (1, 1, 0.38), claims (2, 1, 0.35), contributions (3, 1, 0.33), estimate (7, 1, 0.29), phenotypes (7, 1, 0.29), repurposing (8, 1, 0.28), large (19, 1, 0.25), environmental (24, 1, 0.24), health (28, 1, 0.23), genetic (56, 1, 0.20), data (134, 1, 0.16)\n",
      "\n",
      "Cluster No. 118\n",
      "Size: 11\n",
      "Terms: \n",
      "latent (5, 5, 0.49), eldercare (2, 2, 0.24), person-centered (2, 2, 0.22), approaches (30, 3, 0.22), profiles (9, 2, 0.18), multilevel (10, 2, 0.18), profile (11, 2, 0.17), class (14, 2, 0.17), analysis (196, 3, 0.14), models (42, 2, 0.14), constructs: (1, 1, 0.12), variable- (1, 1, 0.12), nuance (1, 1, 0.12), person-centered, (1, 1, 0.12), person-specific (1, 1, 0.12), variable-centered, (1, 1, 0.12), sciences (1, 1, 0.12), auxiliary (1, 1, 0.12), extensions (1, 1, 0.12), mediation, (1, 1, 0.12)\n",
      "\n",
      "Cluster No. 119\n",
      "Size: 2\n",
      "Terms: \n",
      "markov (4, 2, 0.44), chain (12, 2, 0.37), divergence-based (1, 1, 0.26), kullback-leibler (1, 1, 0.26), carlo (3, 1, 0.23), monte (3, 1, 0.23), mechanistic (4, 1, 0.22), data: (5, 1, 0.21), simple (7, 1, 0.20), fitting (7, 1, 0.20), robots (12, 1, 0.18), differential (17, 1, 0.17), filter (17, 1, 0.17), approaches (30, 1, 0.16), epidemic (31, 1, 0.16), comparison (31, 1, 0.16), models (42, 1, 0.15), global (47, 1, 0.14), mobile (51, 1, 0.14), localization (98, 1, 0.12)\n",
      "\n",
      "Cluster No. 120\n",
      "Size: 1\n",
      "Terms: \n",
      "neuroscience (1, 1, 0.47), race? (1, 1, 0.47), rodent (3, 1, 0.41), research: (4, 1, 0.40), rat (5, 1, 0.39), models (42, 1, 0.27)\n",
      "\n",
      "Cluster No. 121\n",
      "Size: 3\n",
      "Terms: \n",
      "astronomical (1, 1, 0.32), genomical? (1, 1, 0.32), 21st (1, 1, 0.32), theoretic (1, 1, 0.32), calling (2, 1, 0.29), century (2, 1, 0.29), free (4, 1, 0.26), data: (5, 1, 0.26), big (7, 1, 0.24), alignment (9, 1, 0.24), variant (9, 1, 0.23), biology (16, 1, 0.21), computational (17, 1, 0.21), information (40, 1, 0.18)\n",
      "\n",
      "Cluster No. 122\n",
      "Size: 2\n",
      "Terms: \n",
      "locally (2, 1, 0.35), globally (3, 1, 0.34), completion (3, 1, 0.34), consistent (6, 1, 0.30), convolutional (8, 1, 0.29), image (8, 1, 0.29), single-particle (8, 1, 0.29), neural_network (12, 1, 0.27), cryo-electron_microscopy (12, 1, 0.27), deep (16, 1, 0.26), recognition (32, 1, 0.23), approach (81, 1, 0.19)\n",
      "\n",
      "Cluster No. 123\n",
      "Size: 1\n",
      "Terms: \n",
      "(3d (1, 1, 0.31), pdf) (1, 1, 0.31), tools, (1, 1, 0.31), document (2, 1, 0.29), portable (2, 1, 0.29), biomedical (2, 1, 0.29), format (2, 1, 0.29), sciences: (2, 1, 0.29), applications, (4, 1, 0.26), protocols (9, 1, 0.23), three-dimensional (10, 1, 0.23), systematic_review (17, 1, 0.21), communication (30, 1, 0.19), clinical (51, 1, 0.17)\n",
      "\n",
      "Cluster No. 124\n",
      "Size: 1\n",
      "Terms: \n",
      "cost-outcomes (1, 1, 0.35), infarction? (1, 1, 0.35), myocardial (1, 1, 0.35), covariates (2, 1, 0.32), spending (2, 1, 0.32), time-varying (4, 1, 0.29), examining (11, 1, 0.25), higher (12, 1, 0.25), survival (12, 1, 0.25), improve (12, 1, 0.25), outcomes (20, 1, 0.23), relationship (28, 1, 0.21)\n",
      "\n",
      "Cluster No. 125\n",
      "Size: 14\n",
      "Terms: \n",
      "hereditary_angioedema (8, 8, 0.58), deficiency (4, 3, 0.24), c1 (2, 2, 0.18), angioedema: (2, 2, 0.18), hereditary (2, 2, 0.18), management (47, 3, 0.16), contact (7, 2, 0.15), diagnosis (19, 2, 0.13), inhibitor (20, 2, 0.13), mutation (22, 2, 0.12), international (23, 2, 0.12), treatment (51, 2, 0.10), patients (57, 2, 0.10), gene (59, 2, 0.10), angioödem (1, 1, 0.10), hereditäres (1, 1, 0.10), angioedema (1, 1, 0.10), department: (1, 1, 0.10), aged (1, 1, 0.10), c1-inhibitor (1, 1, 0.10)\n",
      "\n",
      "Cluster No. 126\n",
      "Size: 22\n",
      "Terms: \n",
      "genetic_programming (9, 9, 0.55), semantic (6, 4, 0.26), grammatical (3, 3, 0.22), geometric (12, 3, 0.18), cartesian (2, 2, 0.15), genotype (3, 2, 0.14), evolutionary (47, 3, 0.14), genetic (56, 3, 0.13), search (22, 2, 0.10), algorithms (24, 2, 0.10), diversity (54, 2, 0.09), operators (1, 1, 0.08), bananas, (1, 1, 0.08), barbed (1, 1, 0.08), easy (1, 1, 0.08), envelopes (1, 1, 0.08), lenses, (1, 1, 0.08), wire (1, 1, 0.08), redundancy (1, 1, 0.08), generalised (1, 1, 0.08)\n",
      "\n",
      "Cluster No. 127\n",
      "Size: 1\n",
      "Terms: \n",
      "tissue-based (1, 1, 0.67), proteome (4, 1, 0.56), map (26, 1, 0.41), human (178, 1, 0.26)\n",
      "\n",
      "Cluster No. 128\n",
      "Size: 1\n",
      "Terms: \n",
      "finish (1, 1, 0.42), guys (1, 1, 0.42), last: (1, 1, 0.42), nice (1, 1, 0.42), success (6, 1, 0.33), personality (11, 1, 0.30), political (12, 1, 0.30)\n",
      "\n",
      "Cluster No. 129\n",
      "Size: 1\n",
      "Terms: \n",
      "near-optimal (1, 1, 0.62), probabilistic (8, 1, 0.47), quantification (10, 1, 0.45), rna-seq (11, 1, 0.44)\n",
      "\n",
      "Cluster No. 130\n",
      "Size: 2\n",
      "Terms: \n",
      "redox-active (1, 1, 0.40), selenium, (1, 1, 0.40), selenoproteins (1, 1, 0.40), selenium (2, 1, 0.37), disease: (5, 1, 0.32), conceptual (7, 1, 0.31), health (28, 1, 0.24), review (34, 1, 0.23), viral (72, 1, 0.20), infection (79, 1, 0.19)\n",
      "\n",
      "Cluster No. 131\n",
      "Size: 1\n",
      "Terms: \n",
      "safety (14, 1, 1.00)\n",
      "\n",
      "Cluster No. 132\n",
      "Size: 1\n",
      "Terms: \n",
      "siderophore (2, 2, 0.62), (nrps)-independent (1, 1, 0.31), ironing (1, 1, 0.31), non-ribosomal (1, 1, 0.31), synthetases (1, 1, 0.31), biosynthesis: (2, 1, 0.29), synthetase (5, 1, 0.25), peptide (20, 1, 0.20), review (34, 1, 0.18)\n",
      "\n",
      "Cluster No. 133\n",
      "Size: 1\n",
      "Terms: \n",
      "dispelling (1, 1, 0.49), myths (1, 1, 0.49), hypothesis (4, 1, 0.41), systematics (5, 1, 0.39), biological (16, 1, 0.33), testing (19, 1, 0.32)\n",
      "\n",
      "Cluster No. 134\n",
      "Size: 2\n",
      "Terms: \n",
      "21st-century (1, 1, 0.34), organizational, (1, 1, 0.34), societal: (1, 1, 0.34), workplace, (1, 1, 0.34), thinking (2, 1, 0.31), cities (5, 1, 0.28), federal (7, 1, 0.26), u.s. (9, 1, 0.25), domains (13, 1, 0.24), government (24, 1, 0.21), learning (26, 1, 0.21), use (48, 1, 0.19), design (48, 1, 0.19)\n",
      "\n",
      "Cluster No. 135\n",
      "Size: 1\n",
      "Terms: \n",
      "construct (2, 1, 0.54), tackling (2, 1, 0.54), proliferation (6, 1, 0.47), problem (10, 1, 0.43)\n",
      "\n",
      "Cluster No. 136\n",
      "Size: 3\n",
      "Terms: \n",
      "3/7 (1, 1, 0.24), caspase (1, 1, 0.24), kaempferol (1, 1, 0.24), mitigates (1, 1, 0.24), drugs: (1, 1, 0.24), dibenzocyclooctadiene (1, 1, 0.24), lignan (1, 1, 0.24), endoplasmic (2, 1, 0.22), reticulum (2, 1, 0.22), non-nucleoside (2, 1, 0.22), transcriptase (2, 1, 0.22), fda-approved (3, 1, 0.21), derivatives (4, 1, 0.20), products (7, 1, 0.19), reverse (10, 1, 0.18), induced (12, 1, 0.17), hiv-1 (14, 1, 0.17), death (16, 1, 0.16), inhibitor (20, 1, 0.16), stress (19, 1, 0.16)\n",
      "\n",
      "Cluster No. 137\n",
      "Size: 2\n",
      "Terms: \n",
      "maize (1, 1, 0.32), q-body (1, 1, 0.32), seeds (2, 1, 0.29), h3n2 (2, 1, 0.29), incorporation (2, 1, 0.29), unnatural (2, 1, 0.29), site-specific (3, 1, 0.28), immunogenicity (7, 1, 0.25), nucleoprotein (7, 1, 0.25), amino_acid (13, 1, 0.22), vivo (13, 1, 0.22), mice (15, 1, 0.22), expression (35, 1, 0.19), development (55, 1, 0.17), system (77, 1, 0.16), novel (119, 1, 0.14)\n",
      "\n",
      "Cluster No. 138\n",
      "Size: 1\n",
      "Terms: \n",
      "performance-turnover (1, 1, 0.46), top: (1, 1, 0.46), investigating (4, 1, 0.38), sensitivity (6, 1, 0.36), turnover (12, 1, 0.33), organizations (11, 1, 0.33), nonprofit (15, 1, 0.31)\n",
      "\n",
      "Cluster No. 139\n",
      "Size: 2\n",
      "Terms: \n",
      "eutectic (1, 1, 0.23), solvents (1, 1, 0.23), valuable (1, 1, 0.23), (cnsl)‐based (1, 1, 0.23), cashew (1, 1, 0.23), nut (1, 1, 0.23), sustainable (1, 1, 0.23), hybrids (2, 1, 0.21), neglected (2, 1, 0.21), tropical (2, 1, 0.21), liquid (3, 1, 0.20), shell (3, 1, 0.20), flavonoids (4, 1, 0.19), diseases: (4, 1, 0.19), extraction (5, 1, 0.18), trypanosoma (5, 1, 0.18), atp (6, 1, 0.18), brucei (6, 1, 0.18), media (9, 1, 0.17), drugs (14, 1, 0.16)\n",
      "\n",
      "Cluster No. 140\n",
      "Size: 17\n",
      "Terms: \n",
      "hepatitis_e_virus (10, 9, 0.56), hepatitis_e (3, 3, 0.22), orf3 (3, 3, 0.22), plasmodium (8, 3, 0.19), required (9, 3, 0.19), replication (35, 3, 0.15), stage (4, 2, 0.14), liver (8, 2, 0.13), candidate (12, 2, 0.12), infectious (15, 2, 0.12), protein (155, 3, 0.10), development (55, 2, 0.09), identification (70, 2, 0.09), viral (72, 2, 0.09), permissiveness (1, 1, 0.09), simian (1, 1, 0.09), harboring (1, 1, 0.09), orf1 (1, 1, 0.09), macro (1, 1, 0.09), (lisp2) (1, 1, 0.09)\n",
      "\n",
      "Cluster No. 141\n",
      "Size: 1\n",
      "Terms: \n",
      "adolescent (2, 2, 0.60), abuse (1, 1, 0.30), indicators: (1, 1, 0.30), substance (3, 1, 0.28), consumer (2, 1, 0.28), updated (4, 1, 0.25), programs (6, 1, 0.24), guide (7, 1, 0.23), assessing (13, 1, 0.21), quality (18, 1, 0.20), treatment (51, 1, 0.16), development (55, 1, 0.16)\n",
      "\n",
      "Cluster No. 142\n",
      "Size: 46\n",
      "Terms: \n",
      "party (8, 7, 0.28), legislative (5, 5, 0.21), policy (27, 6, 0.20), u.s. (9, 4, 0.16), politics (11, 4, 0.15), congress (3, 3, 0.14), voting (4, 3, 0.13), court (4, 3, 0.13), electoral (5, 3, 0.13), representation (9, 3, 0.12), national (14, 3, 0.11), unilateral (2, 2, 0.11), presidential (2, 2, 0.10), parliamentary (2, 2, 0.10), powers (3, 2, 0.10), dáil (2, 2, 0.10), éireann (2, 2, 0.10), trade (3, 2, 0.09), election (3, 2, 0.09), congressional (3, 2, 0.09)\n",
      "\n",
      "Cluster No. 143\n",
      "Size: 1\n",
      "Terms: \n",
      "7.0 (1, 1, 0.42), mega7: (1, 1, 0.42), version (1, 1, 0.42), bigger (2, 1, 0.39), datasets (6, 1, 0.33), genetics (14, 1, 0.29), evolutionary (47, 1, 0.23), molecular (95, 1, 0.20), analysis (196, 1, 0.16)\n",
      "\n",
      "Cluster No. 144\n",
      "Size: 1\n",
      "Terms: \n",
      "completely (1, 1, 0.37), hhpred (1, 1, 0.37), mpi (1, 1, 0.37), reimplemented (1, 1, 0.37), server (1, 1, 0.37), toolkit (2, 1, 0.34), bioinformatics (7, 1, 0.29), core (14, 1, 0.26), new (99, 1, 0.17)\n",
      "\n",
      "Cluster No. 145\n",
      "Size: 3\n",
      "Terms: \n",
      "media (9, 2, 0.48), news (1, 1, 0.32), organisations (1, 1, 0.32), producers (1, 1, 0.32), hyperlocal (1, 1, 0.32), hyperlocals (1, 1, 0.32), legacy (1, 1, 0.32), nordic (3, 1, 0.28), region (13, 1, 0.22), public (55, 1, 0.17)\n",
      "\n",
      "Cluster No. 146\n",
      "Size: 1\n",
      "Terms: \n",
      "(rcgp (1, 1, 0.32), network: (1, 1, 0.32), royal (1, 1, 0.32), rsc) (1, 1, 0.32), practitioners (2, 1, 0.30), sentinel (2, 1, 0.30), college (4, 1, 0.27), cohort (4, 1, 0.27), centre (4, 1, 0.27), profile (11, 1, 0.23), general (15, 1, 0.22), surveillance (19, 1, 0.21), research (49, 1, 0.18)\n",
      "\n",
      "Cluster No. 147\n",
      "Size: 1\n",
      "Terms: \n",
      "investigate (1, 1, 0.42), researchers (1, 1, 0.42), outcomes: (2, 1, 0.38), evaluations (6, 1, 0.33), statistical (7, 1, 0.32), guide (7, 1, 0.32), power (13, 1, 0.29), effects (41, 1, 0.24), multiple (59, 1, 0.22)\n",
      "\n",
      "Cluster No. 148\n",
      "Size: 2\n",
      "Terms: \n",
      "19 (5, 2, 0.39), peptides (10, 2, 0.34), binding (58, 2, 0.24), (pkmsp‐1 (1, 1, 0.23), knowlesi (1, 1, 0.23), merozoite (1, 1, 0.23), protein‐1 (1, 1, 0.23), cabs-dock (1, 1, 0.23), prior (2, 1, 0.21), ) (4, 1, 0.19), display (4, 1, 0.19), phage (5, 1, 0.19), plasmodium (8, 1, 0.17), library (8, 1, 0.17), flexible (10, 1, 0.17), web_server (10, 1, 0.17), knowledge (12, 1, 0.16), docking (13, 1, 0.16), site (20, 1, 0.15), surface (20, 1, 0.15)\n",
      "\n",
      "Cluster No. 149\n",
      "Size: 1\n",
      "Terms: \n",
      "cultural (3, 1, 0.55), allocation (9, 1, 0.46), services (12, 1, 0.45), government (24, 1, 0.39), local (36, 1, 0.36)\n",
      "\n",
      "Cluster No. 150\n",
      "Size: 1\n",
      "Terms: \n",
      "dipeptides (1, 1, 0.35), n-glyde: (1, 1, 0.35), pattern-based (1, 1, 0.35), gapped (2, 1, 0.32), n-linked (2, 1, 0.32), incorporating (3, 1, 0.30), encoding (3, 1, 0.30), glycosylation (5, 1, 0.28), two-stage (6, 1, 0.28), site (20, 1, 0.23), prediction (43, 1, 0.19)\n",
      "\n",
      "Cluster No. 151\n",
      "Size: 1\n",
      "Terms: \n",
      "breast (2, 1, 0.41), settings: (2, 1, 0.41), computer-aided (2, 1, 0.41), scoping (3, 1, 0.39), screening (20, 1, 0.29), cancer (21, 1, 0.29), review (34, 1, 0.26), clinical (51, 1, 0.24), detection (64, 1, 0.23)\n",
      "\n",
      "Cluster No. 152\n",
      "Size: 7\n",
      "Terms: \n",
      "carboxylesterase_1 (6, 6, 0.61), oseltamivir (3, 2, 0.22), polymorphisms (5, 2, 0.21), variant (9, 2, 0.19), activation (27, 2, 0.16), functional (37, 2, 0.15), genetic (56, 2, 0.14), ces1: (1, 1, 0.13), individualization (1, 1, 0.13), metabolized (1, 1, 0.13), (ces1) (1, 1, 0.13), ces1 (1, 1, 0.13), sacubitril (1, 1, 0.13), antihypertensive (1, 1, 0.13), ces1p1 (1, 1, 0.13), hepatic (1, 1, 0.13), trandolapril (1, 1, 0.13), −816a>c (1, 1, 0.13), bioactivation (1, 1, 0.13), c.662a>g (1, 1, 0.13)\n",
      "\n",
      "Cluster No. 153\n",
      "Size: 1\n",
      "Terms: \n",
      "facial (1, 1, 0.41), multi-point (1, 1, 0.41), warping (2, 1, 0.37), homologous (3, 1, 0.35), landmark-based (4, 1, 0.34), datasets (6, 1, 0.32), recognition (32, 1, 0.24), 3d (43, 1, 0.23), multiple (59, 1, 0.21), approach (81, 1, 0.20)\n",
      "\n",
      "Cluster No. 154\n",
      "Size: 1\n",
      "Terms: \n",
      "data-sets (1, 1, 0.40), explore (1, 1, 0.40), unfolding (2, 1, 0.37), data: (5, 1, 0.33), economic (6, 1, 0.32), big (7, 1, 0.31), change (12, 1, 0.28), small (23, 1, 0.25), social (40, 1, 0.23), multiple (59, 1, 0.21)\n",
      "\n",
      "Cluster No. 155\n",
      "Size: 4\n",
      "Terms: \n",
      "children’s (1, 1, 0.23), language (1, 1, 0.23), environment: (1, 1, 0.23), federated (1, 1, 0.23), notes (2, 1, 0.21), medication (2, 1, 0.21), shanghai, (2, 1, 0.21), obstacles (3, 1, 0.20), privacy-preserving (3, 1, 0.20), diseases: (4, 1, 0.19), childhood (5, 1, 0.19), analysis (196, 2, 0.17), medicine (8, 1, 0.17), china: (9, 1, 0.17), hospital (9, 1, 0.17), deep_learning (9, 1, 0.17), opportunities (11, 1, 0.17), mining (11, 1, 0.17), similarity (12, 1, 0.16), chronic (13, 1, 0.16)\n",
      "\n",
      "Cluster No. 156\n",
      "Size: 4\n",
      "Terms: \n",
      "battery-free (2, 2, 0.47), visible (8, 2, 0.39), light (22, 2, 0.33), sensing (23, 2, 0.32), cellphone (1, 1, 0.26), rubble: (1, 1, 0.26), trapped (1, 1, 0.26), pilot (4, 1, 0.21), detecting (8, 1, 0.19), humans (21, 1, 0.16), communication (30, 1, 0.15), evaluation (39, 1, 0.15), sensor (46, 1, 0.14), study (75, 1, 0.13), system (77, 1, 0.13), human (178, 1, 0.10)\n",
      "\n",
      "Cluster No. 157\n",
      "Size: 1\n",
      "Terms: \n",
      "arrhythmias (1, 1, 0.43), litenet: (1, 1, 0.43), resource-constrained (2, 1, 0.39), lightweight (3, 1, 0.37), detecting (8, 1, 0.32), devices (10, 1, 0.31), neural_network (12, 1, 0.30), mobile (51, 1, 0.23)\n",
      "\n",
      "Cluster No. 158\n",
      "Size: 30\n",
      "Terms: \n",
      "leadership (16, 8, 0.31), organizational (49, 10, 0.30), public (55, 10, 0.29), team (22, 5, 0.19), leadership, (8, 4, 0.17), performance (78, 6, 0.16), roles (15, 4, 0.15), motivation (5, 3, 0.13), public_service (9, 3, 0.12), performance: (9, 3, 0.12), perceived (9, 3, 0.12), agencies (3, 2, 0.10), collective (4, 2, 0.09), networking (5, 2, 0.09), research (49, 3, 0.09), private (5, 2, 0.09), satisfaction (5, 2, 0.09), transformational (7, 2, 0.08), job_satisfaction (8, 2, 0.08), employee (11, 2, 0.08)\n",
      "\n",
      "Cluster No. 159\n",
      "Size: 1\n",
      "Terms: \n",
      "auxetic, (1, 1, 0.45), coherently (1, 1, 0.45), self-assembly (1, 1, 0.45), dynamic, (3, 1, 0.39), two-dimensional (7, 1, 0.35), crystals (15, 1, 0.31), protein (155, 1, 0.18)\n",
      "\n",
      "Cluster No. 160\n",
      "Size: 1\n",
      "Terms: \n",
      "knox (1, 1, 0.37), near-repeats (1, 1, 0.37), spatio-temporal (1, 1, 0.37), adjusting (2, 1, 0.34), analyse (2, 1, 0.34), crime (6, 1, 0.31), test (9, 1, 0.28), heterogeneity (12, 1, 0.26), accounting (12, 1, 0.26), risk (44, 1, 0.21)\n",
      "\n",
      "Cluster No. 161\n",
      "Size: 1\n",
      "Terms: \n",
      "anti-diabetic (1, 1, 0.33), ptp1b (1, 1, 0.33), sideritis (1, 1, 0.33), l. (3, 1, 0.29), inhibitory (3, 1, 0.29), probable (4, 1, 0.28), assisted (8, 1, 0.25), components (8, 1, 0.25), properties (9, 1, 0.25), action (10, 1, 0.24), docking (13, 1, 0.24), biological (16, 1, 0.22), evaluation (39, 1, 0.19), prediction (43, 1, 0.19)\n",
      "\n",
      "Cluster No. 162\n",
      "Size: 3\n",
      "Terms: \n",
      "phosphodiesterases (1, 1, 0.26), 6 (1, 1, 0.26), cgmp-phosphodiesterase (1, 1, 0.26), rods (1, 1, 0.26), transducins (1, 1, 0.26), brownian (1, 1, 0.26), electrostatic (1, 1, 0.26), phosphodiesterase: (1, 1, 0.26), retinal (2, 1, 0.24), takes (2, 1, 0.24), camp (2, 1, 0.24), activate (3, 1, 0.23), steering (3, 1, 0.23), enhances (5, 1, 0.21), protein-protein_interactions (6, 1, 0.20), rate (10, 1, 0.19), modeling (30, 1, 0.16), dynamics (37, 1, 0.15), binding (58, 1, 0.14)\n",
      "\n",
      "Cluster No. 163\n",
      "Size: 1\n",
      "Terms: \n",
      "crime (6, 2, 0.61), environmental (24, 2, 0.46), crimes: (1, 1, 0.36), continuum (2, 1, 0.33), subtypes (4, 1, 0.30), convergence (7, 1, 0.28)\n",
      "\n",
      "Cluster No. 164\n",
      "Size: 6\n",
      "Terms: \n",
      "polyethylene (3, 3, 0.35), terephthalate (3, 3, 0.35), polyester (2, 2, 0.25), degrade (1, 1, 0.13), hydrolases (1, 1, 0.13), relieved (1, 1, 0.13), ideonella (1, 1, 0.13), mhetase (1, 1, 0.13), plastic-degrading (1, 1, 0.13), sakaiensis (1, 1, 0.13), isotopic (1, 1, 0.13), mealworms: (1, 1, 0.13), mineralization (1, 1, 0.13), plastic-eating (1, 1, 0.13), (pet)-degrading (1, 1, 0.13), impranil (1, 1, 0.13), pf-5 (1, 1, 0.13), polyurethane (1, 1, 0.13), protegens (1, 1, 0.13), engineered (2, 1, 0.12)\n",
      "\n",
      "Cluster No. 165\n",
      "Size: 1\n",
      "Terms: \n",
      "meta-regression (1, 1, 0.38), regions: (1, 1, 0.38), roost (1, 1, 0.38), tree-dwelling (1, 1, 0.38), biogeographical (3, 1, 0.33), updated (4, 1, 0.31), meta-analysis (8, 1, 0.28), site (20, 1, 0.25), selection (35, 1, 0.22), bats (44, 1, 0.21)\n",
      "\n",
      "Cluster No. 166\n",
      "Size: 1\n",
      "Terms: \n",
      "wcet (1, 1, 0.48), compute (2, 1, 0.44), fitness (7, 1, 0.37), estimating (21, 1, 0.31), models (42, 1, 0.27), function (42, 1, 0.27), prediction (43, 1, 0.27), genetic (56, 1, 0.25), algorithm (74, 1, 0.24)\n",
      "\n",
      "Cluster No. 167\n",
      "Size: 2\n",
      "Terms: \n",
      "data-intensive (1, 1, 0.28), networked, (1, 1, 0.28), portal: (1, 1, 0.28), (fair) (1, 1, 0.28), coming (1, 1, 0.28), findable (1, 1, 0.28), interoperable (1, 1, 0.28), re-usable (1, 1, 0.28), accessible (2, 1, 0.26), diffraction_data (6, 1, 0.22), science (8, 1, 0.21), modern (8, 1, 0.21), crystallography (15, 1, 0.19), pattern (17, 1, 0.19), design (48, 1, 0.15), research (49, 1, 0.15), data (134, 1, 0.12), protein (155, 1, 0.12)\n",
      "\n",
      "Cluster No. 168\n",
      "Size: 1\n",
      "Terms: \n",
      "crm1 (1, 1, 0.40), juxta-nuclear (1, 1, 0.40), microtubule-dependent (1, 1, 0.40), export (2, 1, 0.37), controls (4, 1, 0.34), transport (12, 1, 0.29), nuclear (13, 1, 0.28), factor (22, 1, 0.26), virus (89, 1, 0.19)\n",
      "\n",
      "Cluster No. 169\n",
      "Size: 1\n",
      "Terms: \n",
      "ff14sb: (1, 1, 0.45), ff99sb (1, 1, 0.45), backbone (3, 1, 0.40), parameters (7, 1, 0.35), chain (12, 1, 0.32), accuracy (19, 1, 0.30), improving (23, 1, 0.29), protein (155, 1, 0.19)\n",
      "\n",
      "Cluster No. 170\n",
      "Size: 22\n",
      "Terms: \n",
      "sequence (37, 5, 0.25), interactive (10, 3, 0.19), taxonomy (19, 3, 0.17), international (23, 3, 0.17), sequences (23, 3, 0.16), alignment-free (3, 2, 0.15), committee (3, 2, 0.15), epidemic: (5, 2, 0.14), genomic (58, 3, 0.14), visualization (7, 2, 0.13), signatures (8, 2, 0.13), virus (89, 3, 0.12), bayesian (14, 2, 0.12), taxonomic (17, 2, 0.12), machine_learning (22, 2, 0.11), tool (23, 2, 0.11), classification (23, 2, 0.11), coronavirus (161, 3, 0.11), evidence (51, 2, 0.09), dna (51, 2, 0.09)\n",
      "\n",
      "Cluster No. 171\n",
      "Size: 1\n",
      "Terms: \n",
      "capsaicin (1, 1, 0.45), spiciness: (1, 1, 0.45), trpv1 (1, 1, 0.45), understand (4, 1, 0.38), channel (14, 1, 0.31), activation (27, 1, 0.28), mechanism (50, 1, 0.25)\n",
      "\n",
      "Cluster No. 172\n",
      "Size: 11\n",
      "Terms: \n",
      "harmonia (8, 8, 0.46), ladybird (6, 6, 0.37), harlequin (5, 5, 0.31), axyridis (5, 5, 0.31), axyridis: (2, 2, 0.14), coccinellidae) (2, 2, 0.14), hesperomyces (2, 2, 0.14), laboulbeniales) (2, 2, 0.14), virescens (2, 2, 0.14), ladybird, (2, 2, 0.14), color (4, 2, 0.13), perspectives (15, 2, 0.10), native (15, 2, 0.10), shades (1, 1, 0.08), (ascomycota: (1, 1, 0.08), allantonematidae), (1, 1, 0.08), bifurcatus (1, 1, 0.08), caucasus (1, 1, 0.08), coinvasion (1, 1, 0.08), parasites, (1, 1, 0.08)\n",
      "\n",
      "Cluster No. 173\n",
      "Size: 16\n",
      "Terms: \n",
      "5-steps (3, 3, 0.23), medicinal (4, 3, 0.22), rule (5, 3, 0.22), proteins (51, 4, 0.19), general (15, 3, 0.18), pseudo (2, 2, 0.16), chou’s (2, 2, 0.16), pseaac (2, 2, 0.16), st8sia (2, 2, 0.16), incorporating (3, 2, 0.15), subcellular (5, 2, 0.14), chemistry (8, 2, 0.14), components (8, 2, 0.13), progress (14, 2, 0.12), predict (14, 2, 0.12), biological (16, 2, 0.12), cancer (21, 2, 0.11), interactions (34, 2, 0.10), treatment (51, 2, 0.10), mechanism (50, 2, 0.10)\n",
      "\n",
      "Cluster No. 174\n",
      "Size: 2\n",
      "Terms: \n",
      "galectin-1 (2, 2, 0.56), privileged (1, 1, 0.31), cancer: (1, 1, 0.31), prognostic (1, 1, 0.31), inhibitors: (2, 1, 0.28), significance (4, 1, 0.26), hiv (8, 1, 0.23), meta-analysis (8, 1, 0.23), cancer (21, 1, 0.20), target (31, 1, 0.18), expression (35, 1, 0.18), patients (57, 1, 0.16), human (178, 1, 0.12)\n",
      "\n",
      "Cluster No. 175\n",
      "Size: 1\n",
      "Terms: \n",
      "derived (4, 1, 0.43), shared (6, 1, 0.40), diseases (13, 1, 0.36), common (19, 1, 0.33), environmental (24, 1, 0.32), classification (23, 1, 0.32), determinants (24, 1, 0.32), genetic (56, 1, 0.27), human (178, 1, 0.20)\n",
      "\n",
      "Cluster No. 176\n",
      "Size: 1\n",
      "Terms: \n",
      "microscopy, (1, 1, 0.42), repository (1, 1, 0.42), slide (1, 1, 0.42), analysis, (2, 1, 0.39), digital (5, 1, 0.34), image (8, 1, 0.32), virtual (10, 1, 0.31)\n",
      "\n",
      "Cluster No. 177\n",
      "Size: 1\n",
      "Terms: \n",
      "6800/8800 (1, 1, 0.43), automation (1, 1, 0.43), cobas® (1, 1, 0.43), system: (4, 1, 0.36), diagnostics (6, 1, 0.34), era (6, 1, 0.34), new (99, 1, 0.20), molecular (95, 1, 0.20)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clDict = {}\n",
    "\n",
    "for cl in cluster_ids[:]:\n",
    "    #cl_titles_phrased = []\n",
    "    clTokens = []\n",
    "    if len(clTitles[cl]) < 1:\n",
    "        continue\n",
    "    for title in clTitles[cl]:\n",
    "        title_phrased = trigram[bigram[title]]\n",
    "        #pprint(title_phrased)\n",
    "        #cl_titles_phrased.append(title_phrased)\n",
    "        for t in title_phrased:\n",
    "            clTokens.append(t)\n",
    "    #print(clTokens)\n",
    "    \n",
    "    #cl_dois = [doi for i, doi in enumerate(dois_to_plot) if clLabels[i] == cl]\n",
    "    cl_dois = [doi for i, doi in index2doi.items() if clLabels[i] == cl]\n",
    "    clX = np.zeros((1,300))\n",
    "    for doi in cl_dois:\n",
    "        clX += doi2vec[doi]\n",
    "    clX = clX / len(cl_dois)\n",
    "    \n",
    "    clBow = dictionary.doc2bow(clTokens)\n",
    "    clFrequency = {}\n",
    "    for i, freq in clBow:\n",
    "        clFrequency[i] = freq\n",
    "    #print(clBow)\n",
    "    vec = sorted(tfidf[clBow], key=itemgetter(1), reverse=True)\n",
    "    \n",
    "    print(\"Cluster No.\", cl)\n",
    "    print(\"Size:\", len(cl_dois))\n",
    "    print(\"Terms:\", \", \".join([\"{}\".format(dictionary[i]) for i, v in vec[:20] if v >= 100 and clFrequency[i] >= 5]))\n",
    "    print(\", \".join([\"{} ({}, {}, {:.2f})\".format(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec[:20]]))\n",
    "    #print([(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec])\n",
    "    \n",
    "\n",
    "    print()\n",
    "    \n",
    "    \n",
    "    clDict[cl] = {}\n",
    "    clDict[cl][\"id\"] = cl\n",
    "    clDict[cl][\"size\"] = len(cl_dois)\n",
    "    clDict[cl][\"dois\"] = cl_dois\n",
    "    clDict[cl][\"vec\"] =  [(dictionary[i], frequency[dictionary[i]], clFrequency[i], v) for i, v in vec]\n",
    "    #clDict[cl][\"top_authors\"] = top_authors\n",
    "    clDict[cl][\"X\"] = clX\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_clusters = np.zeros((len(clDict), 300))\n",
    "cl2index = {}\n",
    "\n",
    "for i, (cl, cluster) in enumerate(clDict.items()):\n",
    "    X_clusters[i, :] = cluster[\"X\"]\n",
    "    cl2index[cl] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.manifold import TSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne = TSNE(n_components=2, metric=\"cosine\")\n",
    "#X_embedded = tsne.fit_transform(X_gmm)\n",
    "X_embedded = tsne.fit_transform(X_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#C_embedded = TSNE(n_components=3, metric=\"cosine\").fit_transform(X_gmm)\n",
    "C_embedded = TSNE(n_components=3, metric=\"cosine\").fit_transform(X_clusters)\n",
    "cmin = np.min(C_embedded)\n",
    "cmax = np.max(C_embedded)\n",
    "colors = (C_embedded - cmin) / (cmax - cmin) * 256\n",
    "colors = np.floor(colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bokeh.colors import RGB\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(index=cluster_ids, columns=[\"x\", \"y\", \"keywords\", \"size\"])\n",
    "df.shape\n",
    "\n",
    "for cl in clDict.keys():    \n",
    "    df.at[cl, \"clid\"] = cl\n",
    "    df.at[cl, \"x\"] = X_embedded[cl2index[cl],0]\n",
    "    df.at[cl, \"y\"] = X_embedded[cl2index[cl],1]\n",
    "    df.at[cl, \"size\"] = clDict[cl][\"size\"]\n",
    "    #df.at[cl, \"circle_size\"] = clDict[cl][\"size\"] * 0.25\n",
    "    df.at[cl, \"circle_size\"] = np.sqrt(clDict[cl][\"size\"]) * 5\n",
    "    #keywords = [vec[0] for vec in clDict[cl][\"vec\"] if vec[3] >= 0.1 and vec[2] >= 2]\n",
    "    keywords = [\"{} ({})\".format(vec[0], vec[2]) for vec in clDict[cl][\"vec\"] if vec[3] >= 0.1 and vec[2] >= 2]\n",
    "    df.at[cl, \"keywords\"] = \", \".join(keywords)\n",
    "    df.at[cl, \"keywords_top5\"] = \", \".join(keywords[:5])\n",
    "    df.at[cl, \"keywords_top10\"] = \", \".join(keywords[5:10])\n",
    "    df.at[cl, \"keywords_top15\"] = \", \".join(keywords[10:15])\n",
    "    df.at[cl, \"keywords_top20\"] = \", \".join(keywords[15:20])\n",
    "    \n",
    "    df.at[cl, \"rgb\"] = RGB(colors[cl2index[cl],0], colors[cl2index[cl],1], colors[cl2index[cl],2])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.plotting import figure, output_file, show, save, output_notebook\n",
    "from bokeh.models import HoverTool, Range1d, LabelSet, LassoSelectTool, CustomJS, TapTool, OpenURL\n",
    "from bokeh.models.sources import ColumnDataSource\n",
    "from bokeh.io import reset_output\n",
    "\n",
    "reset_output()\n",
    "\n",
    "\n",
    "#class hover(HoverTool):\n",
    "#    hover.tooltips = [\n",
    "#    (\"index\", \"$index\"),\n",
    "\n",
    "\n",
    "hover = HoverTool()\n",
    "hover.tooltips = [\n",
    "    (\"Cluster\", \"@clid\"),\n",
    "    #(\"Keywords\", \"@keywords\"),\n",
    "    (\"keywords_top5\", \"@keywords_top5\"),\n",
    "    (\"keywords_top10\", \"@keywords_top10\"),\n",
    "    (\"keywords_top15\", \"@keywords_top15\"),\n",
    "    (\"keywords_top20\", \"@keywords_top20\"),\n",
    "    (\"Size\", \"@size\"),\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "s1 = ColumnDataSource(data=df)\n",
    "\n",
    "\n",
    "callback = CustomJS(code=\"\"\"\n",
    "// the event that triggered the callback is cb_obj:\n",
    "// The event type determines the relevant attributes\n",
    "console.log('Tap event occured at x-position: ' + cb_obj.x)\n",
    "\"\"\")\n",
    "\n",
    "\n",
    "def lasso_cb(x):\n",
    "    print(x)\n",
    "\n",
    "#lasso = LassoSelectTool(callback=lasso_cb)\n",
    "lasso = LassoSelectTool()\n",
    "\n",
    "\n",
    "p = figure(plot_width=1600, plot_height=1000, active_scroll = \"wheel_zoom\")\n",
    "#p = figure(plot_width=1200, plot_height=1000, output_backend=\"webgl\", active_scroll = \"wheel_zoom\")\n",
    "#p.circle(x='x', y='y', size=8, color=\"rgb\", alpha=\"alpha\", source=s1)\n",
    "p.circle(x='x', y='y', size=\"circle_size\", color=\"rgb\", alpha=0.75, source=s1)\n",
    "\n",
    "p.add_tools(hover)\n",
    "p.add_tools(lasso)\n",
    "\n",
    "p.js_on_event('tap', callback)\n",
    "\n",
    "#p.toolbar.active_scroll = \"wheel_zoom\"\n",
    "\n",
    "#df[\"xval\"].max()\n",
    "\n",
    "left, right, bottom, top = -120, 120, -90, 90\n",
    "left, right, bottom, top = -80, 80, -60, 60\n",
    "left, right, bottom, top = -120, 120, -90, 90\n",
    "p.x_range=Range1d(left, right)\n",
    "p.y_range=Range1d(bottom, top)\n",
    "\n",
    "\n",
    "#layout = row(p, widgetbox(data_table))\n",
    "#layout = column(p, widgetbox(data_table))\n",
    "\n",
    "show(p)\n",
    "#save(layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (<ipython-input-44-6aaf1f276005>, line 4)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-44-6aaf1f276005>\"\u001b[1;36m, line \u001b[1;32m4\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(\"cluster_refvecs.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster 13\n",
      "Count: 464\n",
      "Size: 776\n",
      "coronavirus (161, 154, 0.38), middle_east_respiratory_syndrome (123, 117, 0.31), mers-cov (54, 54, 0.17), novel_coronavirus (48, 45, 0.15), outbreak (67, 48, 0.15), bats (44, 37, 0.13), china (49, 38, 0.12), coronaviruses (35, 34, 0.12), 2019-ncov (35, 33, 0.12), transmission (69, 36, 0.11), covid-19 (35, 31, 0.11), human (178, 45, 0.11), virus (89, 35, 0.10)\n",
      "[(2015, 126), (2016, 111), (2017, 92), (2018, 65), (2019, 82), (2020, 300)]\n",
      "[-0.41370717 -0.5733684 ]\n",
      "\n",
      "Cluster 72\n",
      "Count: 4\n",
      "Size: 67\n",
      "influenza (39, 10, 0.23), masks (8, 7, 0.21), pandemic (10, 7, 0.20), epidemic (31, 7, 0.16), respirators (5, 5, 0.16), healthcare_workers (7, 5, 0.15), infectious_disease (7, 5, 0.15), epidemics (7, 5, 0.15), respiratory (29, 6, 0.14), outbreak (67, 7, 0.14), medical (13, 5, 0.14), infection (79, 7, 0.13), outbreaks (18, 5, 0.13), effectiveness (19, 5, 0.13), transmission (69, 6, 0.12), models (42, 5, 0.11), healthcare (14, 4, 0.11), ebola (18, 4, 0.10), n95 (3, 3, 0.10), spread (20, 4, 0.10)\n",
      "[(2015, 14), (2016, 10), (2017, 9), (2018, 9), (2019, 17), (2020, 8)]\n",
      "[-0.24412858  1.4369376 ]\n",
      "\n",
      "Cluster 170\n",
      "Count: 4\n",
      "Size: 22\n",
      "sequence (37, 5, 0.25), interactive (10, 3, 0.19), taxonomy (19, 3, 0.17), international (23, 3, 0.17), sequences (23, 3, 0.16), alignment-free (3, 2, 0.15), committee (3, 2, 0.15), epidemic: (5, 2, 0.14), genomic (58, 3, 0.14), visualization (7, 2, 0.13), signatures (8, 2, 0.13), virus (89, 3, 0.12), bayesian (14, 2, 0.12), taxonomic (17, 2, 0.12), machine_learning (22, 2, 0.11), tool (23, 2, 0.11), classification (23, 2, 0.11), coronavirus (161, 3, 0.11)\n",
      "[(2015, 3), (2016, 1), (2017, 4), (2018, 5), (2019, 3), (2020, 6)]\n",
      "[-2.8862436 -1.1748706]\n",
      "\n",
      "Cluster 2\n",
      "Count: 2\n",
      "Size: 125\n",
      "antiviral (30, 20, 0.30), hepatitis_c_virus (15, 14, 0.24), replication (35, 12, 0.18), hepatitis_c (9, 9, 0.17), rna (60, 13, 0.17), virus (89, 14, 0.17), inhibits (20, 10, 0.16), viral (72, 12, 0.15), treatment (51, 10, 0.14), broad-spectrum (11, 7, 0.13), inhibitors (35, 8, 0.12), agents (11, 6, 0.11), virus_infection (15, 6, 0.10), antiviral_activity (6, 5, 0.10), sofosbuvir (6, 5, 0.10)\n",
      "[(2015, 27), (2016, 22), (2017, 18), (2018, 30), (2019, 24), (2020, 4)]\n",
      "[-0.11398642 -2.8251605 ]\n",
      "\n",
      "Cluster 26\n",
      "Count: 2\n",
      "Size: 20\n",
      "epitopes (5, 4, 0.27), peptide_vaccine (7, 4, 0.26), t_cell (14, 4, 0.24), designing (6, 3, 0.20), class (14, 3, 0.18), peptide–mhc (2, 2, 0.16), immunoinformatics (3, 2, 0.15), prediction (43, 3, 0.14), subunit_vaccines (6, 2, 0.13), epitope (6, 2, 0.13), b-cell (6, 2, 0.13), approach (81, 3, 0.12), ebola (18, 2, 0.11), peptide (20, 2, 0.11)\n",
      "[(2015, 2), (2016, 4), (2017, 3), (2018, 2), (2019, 4), (2020, 5)]\n",
      "[-1.6737391 -2.3459098]\n",
      "\n",
      "Cluster 60\n",
      "Count: 2\n",
      "Size: 47\n",
      "zika_virus (33, 24, 0.55), platelet_concentrates (7, 7, 0.21), zika (8, 7, 0.21), light (22, 8, 0.20), ultraviolet_c (6, 6, 0.19), inactivation (8, 6, 0.18), dengue (8, 5, 0.15), brazil (7, 4, 0.12), 2016 (10, 4, 0.11), blue (3, 3, 0.10), methylene (3, 3, 0.10), (zikv) (3, 3, 0.10), arboviruses (3, 3, 0.10)\n",
      "[(2015, 1), (2016, 17), (2017, 15), (2018, 8), (2019, 5), (2020, 1)]\n",
      "[-0.44826517  0.5334599 ]\n",
      "\n",
      "Cluster 25\n",
      "Count: 2\n",
      "Size: 23\n",
      "rna (60, 7, 0.31), codon_usage (6, 4, 0.27), viruses (45, 5, 0.24), bias (8, 3, 0.19), evolution (99, 4, 0.16), synonymous (3, 2, 0.15), gene (59, 3, 0.13), origins (7, 2, 0.13), dengue_virus (8, 2, 0.13), viral (72, 3, 0.13), genome-wide (13, 2, 0.12), infectious (15, 2, 0.12), family (33, 2, 0.10)\n",
      "[(2015, 5), (2016, 5), (2017, 4), (2018, 8), (2019, 1)]\n",
      "[-2.6833355 -1.7338325]\n",
      "\n",
      "Cluster 98\n",
      "Count: 1\n",
      "Size: 7\n",
      "tissue (8, 3, 0.35), translational (8, 3, 0.35), research (49, 3, 0.25), scoring (7, 2, 0.24), lung (21, 2, 0.20), approaches (30, 2, 0.19), knockin (1, 1, 0.15), mouse-adapted (1, 1, 0.15), pitfalls (1, 1, 0.15), principles (1, 1, 0.15), stains (1, 1, 0.15), evaluate (1, 1, 0.15), semiquantitative (1, 1, 0.15), reproducibility (1, 1, 0.15), reproducible (2, 1, 0.14), vital (3, 1, 0.13), fundamental (3, 1, 0.13), dpp4 (4, 1, 0.13), scores (4, 1, 0.13), aging (4, 1, 0.13), concepts (4, 1, 0.13), pathology (5, 1, 0.13), inflammation (5, 1, 0.13), dipeptidyl (5, 1, 0.13), rodents (5, 1, 0.13), relevance (6, 1, 0.12), lethal (6, 1, 0.12), respiratory_tract (6, 1, 0.12), human (178, 2, 0.12), peptidase (7, 1, 0.12), mers_coronavirus (8, 1, 0.12), causes (10, 1, 0.11), 4 (13, 1, 0.11), distribution (14, 1, 0.11), mice (15, 1, 0.11), common (19, 1, 0.10)\n",
      "[(2016, 2), (2017, 2), (2018, 3)]\n",
      "[-0.07578806 -0.8440936 ]\n",
      "\n",
      "Cluster 10\n",
      "Count: 1\n",
      "Size: 13\n",
      "rhinovirus (6, 5, 0.36), asthma (7, 5, 0.36), wheezing (3, 3, 0.24), cadherin-related (2, 2, 0.17), respiratory (29, 3, 0.17), infections (35, 3, 0.16), children (6, 2, 0.15), respiratory_tract (6, 2, 0.15), member (8, 2, 0.14), life (9, 2, 0.14), c (16, 2, 0.12), microbiome (18, 2, 0.12), lung (21, 2, 0.12), early (26, 2, 0.11), family (33, 2, 0.11), function (42, 2, 0.10)\n",
      "[(2015, 2), (2016, 1), (2017, 3), (2018, 6), (2019, 1)]\n",
      "[-7.342615    0.03189061]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "search_term = \"coronavirus\"\n",
    "\n",
    "# thermal, truss, hypersonic, Kullback–Leibler \n",
    "\n",
    "cl_list = []\n",
    "for doi, title in doi2title.items():\n",
    "    if search_term.lower() in title.lower():\n",
    "        if doi in doi2index:\n",
    "            i = doi2index[doi]\n",
    "            cl_list.append(clLabels[i])\n",
    "\n",
    "#print(Counter(cl_list).most_common())\n",
    "for cl, count in Counter(cl_list).most_common():\n",
    "    print(\"Cluster\", cl)\n",
    "    print(\"Count:\", count)\n",
    "    if cl in clDict:\n",
    "        vec = clDict[cl][\"vec\"]\n",
    "        print(\"Size:\", clDict[cl][\"size\"])\n",
    "        print(\", \".join([\"{} ({}, {}, {:.2f})\".format(v[0], v[1], v[2], v[3]) for v in vec if v[3] >= 0.1]))\n",
    "        print(sorted(Counter(doi2year[doi] for doi in clDict[cl][\"dois\"]).items()))\n",
    "        print(X_embedded[cl2index[cl], :])\n",
    "    print()\n",
    "                     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size: 71\n",
      "Terms: bacterial (37, 9, 0.20), pseudomonas (15, 6, 0.16), biofilm (9, 5, 0.14), regulatory (15, 5, 0.13), rnas (5, 4, 0.13), escherichia_coli (21, 5, 0.12), symbiosis (6, 4, 0.12), antibiotic (9, 4, 0.11), response (40, 5, 0.11), host (46, 5, 0.11), global (47, 5, 0.11), bacteria (18, 4, 0.10), transcriptional (19, 4, 0.10), putida (3, 3, 0.10), k-12 (3, 3, 0.10), lactones (3, 3, 0.10), yersinia (4, 3, 0.10)\n",
      "[(2015, 5), (2016, 13), (2017, 11), (2018, 13), (2019, 25), (2020, 4)]\n",
      "[-4.356726  -5.0984683]\n",
      "\n",
      "Ainsaar, Tamman, Kasvandik, Tenson, Hõrak (2019): The TonBm-PocAB System Is Required for Maintenance of Membrane Integrity and Polar Position of Flagella in Pseudomonas putida. http://dx.doi.org/10.1128/jb.00303-19\n",
      "\n",
      "Parker, Demetci, Li (2019): Rapid Accumulation of Motility-Activating Mutations in Resting Liquid Culture of Escherichia coli. http://dx.doi.org/10.1128/jb.00259-19\n",
      "\n",
      "Bianco, Fröhlich, Vanderpool (2019): Bacterial Cyclopropane Fatty Acid Synthase mRNA Is Targeted by Activating and Repressing Small RNAs. http://dx.doi.org/10.1128/jb.00461-19\n",
      "\n",
      "He, Qin, Wong, Shi, Wang, Shi (2019): Ethanol Adaptation Strategies in Salmonella enterica Serovar Enteritidis Revealed by Global Proteomic and Mutagenic Analyses. http://dx.doi.org/10.1128/aem.01107-19\n",
      "\n",
      "Septer (2019): The Vibrio-Squid Symbiosis as a Model for Studying Interbacterial Competition. http://dx.doi.org/10.1128/msystems.00108-19\n",
      "\n",
      "Ryan, Pati, Ojha, Padhi, Ray, Jaiswal, Singh, Mannala, Schultze, Chakraborty, Suar (2015): Global Transcriptome and Mutagenic Analyses of the Acid Tolerance Response of Salmonella enterica Serovar Typhimurium. http://dx.doi.org/10.1128/aem.02172-15\n",
      "\n",
      "Wiersinga, Virk, Torres, Currie, Peacock, Dance, Limmathurotsakul (2018): Melioidosis. http://dx.doi.org/10.1038/nrdp.2017.107\n",
      "\n",
      "Lee, Caggianiello, van Swam, Taverne, Meijerink, Bron, Spano, Kleerebezem (2016): Strain-Specific Features of Extracellular Polysaccharides and Their Impact on Lactobacillus plantarum-Host Interactions. http://dx.doi.org/10.1128/aem.00306-16\n",
      "\n",
      "Du, Wang-Kan, Neuberger, van Veen, Pos, Piddock, Luisi (2018): Multidrug efflux pumps: structure, function and regulation. http://dx.doi.org/10.1038/s41579-018-0048-6\n",
      "\n",
      "Stapels, Hill, Westermann, Fisher, Thurston, Saliba, Blommestein, Vogel, Helaine (2018): Salmonella persisters undermine host immune defenses during antibiotic treatment. http://dx.doi.org/10.1126/science.aat7148\n",
      "\n",
      "Speare, Cecere, Guckes, Smith, Wollenberg, Mandel, Miyashiro, Septer (2018): Bacterial symbionts use a type VI secretion system to eliminate competitors in their natural host. http://dx.doi.org/10.1073/pnas.1808302115\n",
      "\n",
      "Holmqvist, Wagner (2017): Impact of bacterial sRNAs in stress responses. http://dx.doi.org/10.1042/bst20160363\n",
      "\n",
      "Pletzer, Mansour, Wuerth, Rahanjam, Hancock (2017): New Mouse Model for Chronic Infections by Gram-Negative Bacteria Enabling the Study of Anti-Infective Efficacy and Host-Microbe Interactions. http://dx.doi.org/10.1128/mbio.00140-17\n",
      "\n",
      "Fahrner, Berg (2015): Mutations That StimulateflhDCExpression in Escherichia coli K-12. http://dx.doi.org/10.1128/jb.00455-15\n",
      "\n",
      "Li, Zhu, Zhang, Fu, Wang, Chen, Zhao (2016): Characterization of genome-reduced Bacillus subtilis strains and their application for the production of guanosine and thymidine. http://dx.doi.org/10.1186/s12934-016-0494-7\n",
      "\n",
      "Zhang, Tang, Zhang, Nguyen, Moore (2017): Broad-Host-Range Expression Reveals Native and Host Regulatory Elements That Influence Heterologous Antibiotic Production in Gram-Negative Bacteria. http://dx.doi.org/10.1128/mbio.01291-17\n",
      "\n",
      "Winsor, Griffiths, Lo, Dhillon, Shay, Brinkman (2015): Enhanced annotations and features for comparing thousands ofPseudomonasgenomes in the Pseudomonas genome database. http://dx.doi.org/10.1093/nar/gkv1227\n",
      "\n",
      "Akarsu, Bordes, Mansour, Bigot, Genevaux, Falquet (2019): TASmania: A bacterial Toxin-Antitoxin Systems database. http://dx.doi.org/10.1371/journal.pcbi.1006946\n",
      "\n",
      "Paduszynska, Maciejewska, Greber, Sawicki, Kamysz (2019): Antibacterial Activities of Lipopeptide (C10)2-KKKK-NH2 Applied Alone and in Combination with Lens Liquids to Fight Biofilms Formed on Polystyrene Surfaces and Contact Lenses. http://dx.doi.org/10.3390/ijms20020393\n",
      "\n",
      "Felden, Cattoir (2018): Bacterial Adaptation to Antibiotics through Regulatory RNAs. http://dx.doi.org/10.1128/aac.02503-17\n",
      "\n",
      "Hospenthal, Costa, Waksman (2017): A comprehensive guide to pilus biogenesis in Gram-negative bacteria. http://dx.doi.org/10.1038/nrmicro.2017.40\n",
      "\n",
      "Marzel, Desai, Goren, Schorr, Nissan, Porwollik, Valinsky, McClelland, Rahav, Gal-Mor (2016): Persistent Infections by NontyphoidalSalmonellain Humans: Epidemiology and Genetics. http://dx.doi.org/10.1093/cid/civ1221\n",
      "\n",
      "Bernal, Allsopp, Filloux, Llamas (2017): The Pseudomonas putida T6SS is a plant warden against phytopathogens. http://dx.doi.org/10.1038/ismej.2016.169\n",
      "\n",
      "Limmathurotsakul, Golding, Dance, Messina, Pigott, Moyes, Rolim, Bertherat, Day, Peacock, Hay (2016): Predicted global distribution of Burkholderia pseudomallei and burden of melioidosis. http://dx.doi.org/10.1038/nmicrobiol.2015.8\n",
      "\n",
      "Sun, LaSota, Cecere, LaPenna, Larios-Valencia, Wollenberg, Miyashiro (2016): Intraspecific Competition Impacts Vibrio fischeri Strain Diversity during Initial Colonization of the Squid Light Organ. http://dx.doi.org/10.1128/aem.04143-15\n",
      "\n",
      "Visick, Hodge-Hanson, Tischler, Bennett, Mastrodomenico (2018): Tools for Rapid Genetic Engineering of\n",
      "                    Vibrio fischeri. http://dx.doi.org/10.1128/aem.00850-18\n",
      "\n",
      "Wu, Zaiden, Cao (2018): The Core- and Pan-Genomic Analyses of the Genus Comamonas: From Environmental Adaptation to Potential Virulence. http://dx.doi.org/10.3389/fmicb.2018.03096\n",
      "\n",
      "Ainelo, Tamman, Leppik, Remme, Hõrak (2016): The toxin GraT inhibits ribosome biogenesis. http://dx.doi.org/10.1111/mmi.13344\n",
      "\n",
      "Barroso, Previato-Mello, Batista, Batista, da Silva Neto (2018): EmrR-Dependent Upregulation of the Efflux Pump EmrCAB Contributes to Antibiotic Resistance in Chromobacterium violaceum. http://dx.doi.org/10.3389/fmicb.2018.02756\n",
      "\n",
      "Belcaid, Casaburi, McAnulty, Schmidbaur, Suria, Moriano-Gutierrez, Pankey, Oakley, Kremer, Koch, Collins, Nguyen, Lek, Goncharenko-Foster, Minx, Sodergren, Weinstock, Rokhsar, McFall-Ngai, Simakov, Foster, Nyholm (2019): Symbiotic organs shaped by distinct modes of genome evolution in cephalopods. http://dx.doi.org/10.1073/pnas.1817322116\n",
      "\n",
      "Baron, Bonnassie, Alabdeh, Cochet, Nau, Guérin-Dubiard, Gautier, Andrews, Jan (2017): Global Gene-expression Analysis of the Response of Salmonella Enteritidis to Egg White Exposure Reveals Multiple Egg White-imposed Stress Responses. http://dx.doi.org/10.3389/fmicb.2017.00829\n",
      "\n",
      "Coulthurst (2019): The Type VI secretion system: a versatile bacterial weapon. http://dx.doi.org/10.1099/mic.0.000789\n",
      "\n",
      "Hoekzema, Romilly, Holmqvist, Wagner (2019): Hfq‐dependent\n",
      "            mRNA\n",
      "            unfolding promotes\n",
      "            sRNA\n",
      "            ‐based inhibition of translation. http://dx.doi.org/10.15252/embj.2018101199\n",
      "\n",
      "Mandel, Dunn (2016): Impact and Influence of the Natural Vibrio-Squid Symbiosis in Understanding Bacterial–Animal Interactions. http://dx.doi.org/10.3389/fmicb.2016.01982\n",
      "\n",
      "Porta, Petrone, Morello, Granata, Rizzo, Memoli, Weisz, Maresca (2016): Design and expression of peptides with antimicrobial activity againstSalmonellatyphimurium. http://dx.doi.org/10.1111/cmi.12645\n",
      "\n",
      "Sanchez-Vazquez, Dewey, Kitten, Ross, Gourse (2019): Genome-wide effects onEscherichia colitranscription from ppGpp binding to its two sites on RNA polymerase. http://dx.doi.org/10.1073/pnas.1819682116\n",
      "\n",
      "Ramsey, Chavez, Johnson, Hosseinzadeh, Mahoney, Mohr, Robison, Zhong, Hall, MacCoss, Bruce, Cilia (2017): Protein interaction networks at the host–microbe interface in\n",
      "            Diaphorina citri\n",
      "            , the insect vector of the citrus greening pathogen. http://dx.doi.org/10.1098/rsos.160545\n",
      "\n",
      "Álvarez, Krogh-Andersen, Tellgren-Roth, Martínez, Günaydın, Lin, Martín, Álvarez, Hammarström, Marcotte (2015): An Exopolysaccharide-Deficient Mutant of Lactobacillus rhamnosus GG Efficiently Displays a Protective Llama Antibody Fragment against Rotavirus on Its Surface. http://dx.doi.org/10.1128/aem.00945-15\n",
      "\n",
      "Bartle, Sumby, Sundstrom, Jiranek (2019): The microbial challenge of winemaking: yeast-bacteria compatibility. http://dx.doi.org/10.1093/femsyr/foz040\n",
      "\n",
      "Vidovic, Liu, An, Mendoza, Abrahante, Johny, Reed (2019): Transcriptional Profiling and Molecular Characterization of the yccT Mutant Link: A Novel STY1099 Protein with the Peroxide Stress Response and Cell Division of Salmonella enterica Serovar Enteritidis. http://dx.doi.org/10.3390/biology8040086\n",
      "\n",
      "Cameron, Bonis, Phan, Kent, Lee, Hunter (2019): A putative enoyl-CoA hydratase contributes to biofilm formation and the antibiotic tolerance of Achromobacter xylosoxidans. http://dx.doi.org/10.1038/s41522-019-0093-6\n",
      "\n",
      "Aschtgen, Brennan, Nikolakakis, Cohen, McFall-Ngai, Ruby (2019): Insights into flagellar function and mechanism from the squid–vibrio symbiosis. http://dx.doi.org/10.1038/s41522-019-0106-5\n",
      "\n",
      "Thöming, Tomasch, Preusse, Koska, Grahl, Pohl, Willger, Kaever, Müsken, Häussler (2020): Parallel evolutionary paths to produce more than one Pseudomonas aeruginosa biofilm phenotype. http://dx.doi.org/10.1038/s41522-019-0113-6\n",
      "\n",
      "Oyama, Girdwood, Cookson, Fernandez-Fuentes, Privé, Vallin, Wilkinson, Golyshin, Golyshina, Mikut, Hilpert, Richards, Wootton, Edwards, Maresca, Perrier, Lundy, Luo, Zhou, Hess, Mantovani, Creevey, Huws (2017): The rumen microbiome: an underexplored resource for novel antimicrobial discovery. http://dx.doi.org/10.1038/s41522-017-0042-1\n",
      "\n",
      "Penesyan, Nagy, Kjelleberg, Gillings, Paulsen (2019): Rapid microevolution of biofilm cells in response to antibiotics. http://dx.doi.org/10.1038/s41522-019-0108-3\n",
      "\n",
      "Kumar, Alam, Grover, Pandey, Tripathi, Kumari, Rani, Singh, Akhter, Ehtesham, Hasnain (2019): Peptidyl-prolyl isomerase-B is involved in Mycobacterium tuberculosis biofilm formation and a generic target for drug repurposing-based intervention. http://dx.doi.org/10.1038/s41522-018-0075-0\n",
      "\n",
      "Salgado, Martínez-Flores, Bustamante, Alquicira-Hernández, García-Sotelo, García-Alonso, Collado-Vides (2018): Using RegulonDB, the Escherichia coli\n",
      " K-12 Gene Regulatory Transcriptional Network Database. http://dx.doi.org/10.1002/cpbi.43\n",
      "\n",
      "Schwartz, Pan (2017): Function and origin of mistranslation in distinct cellular contexts. http://dx.doi.org/10.1080/10409238.2016.1274284\n",
      "\n",
      "Bechhofer, Deutscher (2019): Bacterial ribonucleases and their roles in RNA metabolism. http://dx.doi.org/10.1080/10409238.2019.1651816\n",
      "\n",
      "Ye, Xia, Sun, Liu, Sheng, Chi, Liu, Gu, Wang, Zhao, Wang, Xiao, Wang, Cao (2019): Reprogramming the enzymatic assembly line for site-specific fucosylation. http://dx.doi.org/10.1038/s41929-019-0281-z\n",
      "\n",
      "Girard, Blanchet, Intertaglia, Baudart, Stien, Suzuki, Lebaron, Lami (2017): Characterization of N-Acyl Homoserine Lactones in Vibrio tasmaniensis LGP32 by a Biosensor-Based UHPLC-HRMS/MS Method. http://dx.doi.org/10.3390/s17040906\n",
      "\n",
      "Li, Cui, Bai, Zhao, Li (2016): Involvement of Acylated Homoserine Lactones (AHLs) of Aeromonas sobria in Spoilage of Refrigerated Turbot (Scophthalmus maximus L.). http://dx.doi.org/10.3390/s16071083\n",
      "\n",
      "Okutsu, Morohoshi, Xie, Kato, Ikeda (2015): Characterization of N-Acylhomoserine Lactones Produced by Bacteria Isolated from Industrial Cooling Water Systems. http://dx.doi.org/10.3390/s16010044\n",
      "\n",
      "Singh, Dwivedi, Sanket, Pati (2019): Therapeutic Potential of Endophytic Compounds: A Special Reference to Drug Transporter Inhibitors. http://dx.doi.org/10.2174/1568026619666190412095105\n",
      "\n",
      "Chavarría, Goñi-Moreno, de Lorenzo, Nikel (2016): A Metabolic Widget Adjusts the Phosphoenolpyruvate-Dependent Fructose Influx in\n",
      "                    Pseudomonas putida. http://dx.doi.org/10.1128/msystems.00154-16\n",
      "\n",
      "Urem, van Rossum, Bucca, Moolenaar, Laing, Świątek-Połatyńska, Willemse, Tenconi, Rigali, Goosen, Smith, van Wezel (2016): OsdR of\n",
      "                    Streptomyces coelicolor\n",
      "                    and the Dormancy Regulator DevR of\n",
      "                    Mycobacterium tuberculosis\n",
      "                    Control Overlapping Regulons. http://dx.doi.org/10.1128/msystems.00014-16\n",
      "\n",
      "Hazen, Daugherty, Shetty, Nataro, Rasko (2017): Transcriptional Variation of Diverse EnteropathogenicEscherichia coliIsolates under Virulence-Inducing Conditions. http://dx.doi.org/10.1128/msystems.00024-17\n",
      "\n",
      "Sontag, Nakayasu, Brown, Niemann, Sydor, Sanchez, Ansong, Lu, Choi, Valleau, Weitz, Savchenko, Cambronne, Adkins (2016): Identification of Novel Host Interactors of Effectors Secreted by\n",
      "                    Salmonella\n",
      "                    and\n",
      "                    Citrobacter. http://dx.doi.org/10.1128/msystems.00032-15\n",
      "\n",
      "Kremer, Koch, El Filali, Zhou, Heath-Heckman, Ruby, McFall-Ngai (2018): Persistent Interactions with Bacterial Symbionts Direct Mature-Host Cell Morphology and Gene Expression in the Squid-Vibrio Symbiosis. http://dx.doi.org/10.1128/msystems.00165-18\n",
      "\n",
      "Naydich, Nangle, Bues, Trivedi, Nissar, Inniss, Niederhuber, Way, Silver, Riglar (2019): Synthetic Gene Circuits Enable Systems-Level Biosensor Trigger Discovery at the Host-Microbe Interface. http://dx.doi.org/10.1128/msystems.00125-19\n",
      "\n",
      "Yamamoto, Yamanaka, Shimada, Sarkar, Yoshida, Bhardwaj, Watanabe, Taira, Chatterji, Ishihama (2018): Altered Distribution of RNA Polymerase Lacking the Omega Subunit within the Prophages along the Escherichia coli K-12 Genome. http://dx.doi.org/10.1128/msystems.00172-17\n",
      "\n",
      "Chouikha, Sturdevant, Jarrett, Sun, Hinnebusch (2019): Differential Gene Expression Patterns of Yersinia pestis and Yersinia pseudotuberculosis during Infection and Biofilm Formation in the Flea Digestive Tract. http://dx.doi.org/10.1128/msystems.00217-18\n",
      "\n",
      "Anderson, Mitchell, Sintsova, Rice, Mobley (2019): Sulfur Assimilation Alters Flagellar Function and Modulates the Gene Expression Landscape of Serratia marcescens. http://dx.doi.org/10.1128/msystems.00285-19\n",
      "\n",
      "Chihara, Bischler, Barquist, Monzon, Noda, Vogel, Tsuneda (2019): Conditional Hfq Association with Small Noncoding RNAs in Pseudomonas aeruginosa Revealed through Comparative UV Cross-Linking Immunoprecipitation Followed by High-Throughput Sequencing. http://dx.doi.org/10.1128/msystems.00590-19\n",
      "\n",
      "Quillin, Hockenberry, Jewett, Seifert (2018): Neisseria gonorrhoeae Exposed to Sublethal Levels of Hydrogen Peroxide Mounts a Complex Transcriptional Response. http://dx.doi.org/10.1128/msystems.00156-18\n",
      "\n",
      "Balasubramanian, Ragunathan, Fei, Vanderpool (2016): A Prophage-Encoded Small RNA Controls Metabolism and Cell Division in\n",
      "                    Escherichia coli. http://dx.doi.org/10.1128/msystems.00021-15\n",
      "\n",
      "Khare, Tavazoie (2020): Extreme Antibiotic Persistence via Heterogeneity-Generating Mutations Targeting Translation. http://dx.doi.org/10.1128/msystems.00847-19\n",
      "\n",
      "Bellieny-Rabelo, Nkomo, Shyntum, Moleleki (2020): Horizontally Acquired Quorum-Sensing Regulators Recruited by the PhoP Regulatory Network Expand the Host Adaptation Repertoire in the Phytopathogen Pectobacterium brasiliense. http://dx.doi.org/10.1128/msystems.00650-19\n",
      "\n",
      "Prieto, Bernabeu, Aznar, Ruiz-Cruz, Bravo, Queiroz, Juárez (2018): Evolution of Bacterial Global Modulators: Role of a Novel H-NS Paralogue in the EnteroaggregativeEscherichia coliStrain 042. http://dx.doi.org/10.1128/msystems.00220-17\n",
      "\n",
      "Gelsinger, Uritskiy, Reddy, Munn, Farney, DiRuggiero (2020): Regulatory Noncoding Small RNAs Are Diverse and Abundant in an Extremophilic Microbial Community. http://dx.doi.org/10.1128/msystems.00584-19\n",
      "\n",
      "Han, Chen, Yan, Gao, Liu, Xue, Zhang, Yang (2019): Hfq Globally Binds and Destabilizes sRNAs and mRNAs in Yersinia pestis. http://dx.doi.org/10.1128/msystems.00245-19\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cl = 51\n",
    "\n",
    "print(\"Size:\", clDict[cl][\"size\"])\n",
    "print(\"Terms:\", \", \".join([\"{} ({}, {}, {:.2f})\".format(v[0], v[1], v[2], v[3]) for v in clDict[cl][\"vec\"] if v[3] >= 0.1]))\n",
    "pubs_per_year = Counter(doi2year[doi] for doi in clDict[cl][\"dois\"])\n",
    "print(sorted(pubs_per_year.items()))\n",
    "\n",
    "print(X_embedded[cl2index[cl], :])\n",
    "print()\n",
    "\n",
    "for doi in clDict[cl][\"dois\"]:\n",
    "    print(doi2str[doi][14:-2])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(LinregressResult(slope=2.2666666666666666, intercept=15.866666666666665, rvalue=0.6393508022779113, pvalue=0.2454399201729709, stderr=1.5738605519667355),\n",
       " 17.333333333333332)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import linregress\n",
    "\n",
    "x = [2016, 2017, 2018, 2019, 2020]\n",
    "y = [pubs_per_year[y] for y in x]\n",
    "x = [-2, -1, 0, 1, 2]\n",
    "y[-1] = pubs_per_year[2020] * 52. / 12.\n",
    "\n",
    "linregress(x, y), y[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Han', 'Chen', 'Yan', 'Gao', 'Liu', 'Xue', 'Zhang', 'Yang']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "doi2str[doi][14:-2].split(\" (20\", 1)[0].split(\", \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Wang', 5),\n",
       " ('Li', 4),\n",
       " ('Zhang', 4),\n",
       " ('Liu', 4),\n",
       " ('Singh', 3),\n",
       " ('Zhao', 3),\n",
       " ('Sun', 3),\n",
       " ('McFall-Ngai', 3),\n",
       " ('Tamman', 2),\n",
       " ('Hõrak', 2),\n",
       " ('Vanderpool', 2),\n",
       " ('Shi', 2),\n",
       " ('Septer', 2),\n",
       " ('Pati', 2),\n",
       " ('Peacock', 2),\n",
       " ('Dance', 2),\n",
       " ('Limmathurotsakul', 2),\n",
       " ('Lee', 2),\n",
       " ('Vogel', 2),\n",
       " ('Cecere', 2),\n",
       " ('Smith', 2),\n",
       " ('Wollenberg', 2),\n",
       " ('Mandel', 2),\n",
       " ('Miyashiro', 2),\n",
       " ('Holmqvist', 2),\n",
       " ('Wagner', 2),\n",
       " ('Mansour', 2),\n",
       " ('Chen', 2),\n",
       " ('Nguyen', 2),\n",
       " ('Cao', 2),\n",
       " ('Batista', 2),\n",
       " ('Kremer', 2),\n",
       " ('Koch', 2),\n",
       " ('Maresca', 2),\n",
       " ('Álvarez', 2),\n",
       " ('Ruby', 2),\n",
       " ('Zhou', 2),\n",
       " ('Ainsaar', 1),\n",
       " ('Kasvandik', 1),\n",
       " ('Tenson', 1),\n",
       " ('Parker', 1),\n",
       " ('Demetci', 1),\n",
       " ('Bianco', 1),\n",
       " ('Fröhlich', 1),\n",
       " ('He', 1),\n",
       " ('Qin', 1),\n",
       " ('Wong', 1),\n",
       " ('Ryan', 1),\n",
       " ('Ojha', 1),\n",
       " ('Padhi', 1),\n",
       " ('Ray', 1),\n",
       " ('Jaiswal', 1),\n",
       " ('Mannala', 1),\n",
       " ('Schultze', 1),\n",
       " ('Chakraborty', 1),\n",
       " ('Suar', 1),\n",
       " ('Wiersinga', 1),\n",
       " ('Virk', 1),\n",
       " ('Torres', 1),\n",
       " ('Currie', 1),\n",
       " ('Caggianiello', 1),\n",
       " ('van Swam', 1),\n",
       " ('Taverne', 1),\n",
       " ('Meijerink', 1),\n",
       " ('Bron', 1),\n",
       " ('Spano', 1),\n",
       " ('Kleerebezem', 1),\n",
       " ('Du', 1),\n",
       " ('Wang-Kan', 1),\n",
       " ('Neuberger', 1),\n",
       " ('van Veen', 1),\n",
       " ('Pos', 1),\n",
       " ('Piddock', 1),\n",
       " ('Luisi', 1),\n",
       " ('Stapels', 1),\n",
       " ('Hill', 1),\n",
       " ('Westermann', 1),\n",
       " ('Fisher', 1),\n",
       " ('Thurston', 1),\n",
       " ('Saliba', 1),\n",
       " ('Blommestein', 1),\n",
       " ('Helaine', 1),\n",
       " ('Speare', 1),\n",
       " ('Guckes', 1),\n",
       " ('Pletzer', 1),\n",
       " ('Wuerth', 1),\n",
       " ('Rahanjam', 1),\n",
       " ('Hancock', 1),\n",
       " ('Fahrner', 1),\n",
       " ('Berg', 1),\n",
       " ('Zhu', 1),\n",
       " ('Fu', 1),\n",
       " ('Tang', 1),\n",
       " ('Moore', 1),\n",
       " ('Winsor', 1),\n",
       " ('Griffiths', 1),\n",
       " ('Lo', 1),\n",
       " ('Dhillon', 1),\n",
       " ('Shay', 1),\n",
       " ('Brinkman', 1),\n",
       " ('Akarsu', 1),\n",
       " ('Bordes', 1),\n",
       " ('Bigot', 1),\n",
       " ('Genevaux', 1),\n",
       " ('Falquet', 1),\n",
       " ('Paduszynska', 1),\n",
       " ('Maciejewska', 1),\n",
       " ('Greber', 1),\n",
       " ('Sawicki', 1),\n",
       " ('Kamysz', 1),\n",
       " ('Felden', 1),\n",
       " ('Cattoir', 1),\n",
       " ('Hospenthal', 1),\n",
       " ('Costa', 1),\n",
       " ('Waksman', 1),\n",
       " ('Marzel', 1),\n",
       " ('Desai', 1),\n",
       " ('Goren', 1),\n",
       " ('Schorr', 1),\n",
       " ('Nissan', 1),\n",
       " ('Porwollik', 1),\n",
       " ('Valinsky', 1),\n",
       " ('McClelland', 1),\n",
       " ('Rahav', 1),\n",
       " ('Gal-Mor', 1),\n",
       " ('Bernal', 1),\n",
       " ('Allsopp', 1),\n",
       " ('Filloux', 1),\n",
       " ('Llamas', 1),\n",
       " ('Golding', 1),\n",
       " ('Messina', 1),\n",
       " ('Pigott', 1),\n",
       " ('Moyes', 1),\n",
       " ('Rolim', 1),\n",
       " ('Bertherat', 1),\n",
       " ('Day', 1),\n",
       " ('Hay', 1),\n",
       " ('LaSota', 1),\n",
       " ('LaPenna', 1),\n",
       " ('Larios-Valencia', 1),\n",
       " ('Visick', 1),\n",
       " ('Hodge-Hanson', 1),\n",
       " ('Tischler', 1),\n",
       " ('Bennett', 1),\n",
       " ('Mastrodomenico', 1),\n",
       " ('Wu', 1),\n",
       " ('Zaiden', 1),\n",
       " ('Ainelo', 1),\n",
       " ('Leppik', 1),\n",
       " ('Remme', 1),\n",
       " ('Barroso', 1),\n",
       " ('Previato-Mello', 1),\n",
       " ('da Silva Neto', 1),\n",
       " ('Belcaid', 1),\n",
       " ('Casaburi', 1),\n",
       " ('McAnulty', 1),\n",
       " ('Schmidbaur', 1),\n",
       " ('Suria', 1),\n",
       " ('Moriano-Gutierrez', 1),\n",
       " ('Pankey', 1),\n",
       " ('Oakley', 1),\n",
       " ('Collins', 1),\n",
       " ('Lek', 1),\n",
       " ('Goncharenko-Foster', 1),\n",
       " ('Minx', 1),\n",
       " ('Sodergren', 1),\n",
       " ('Weinstock', 1),\n",
       " ('Rokhsar', 1),\n",
       " ('Simakov', 1),\n",
       " ('Foster', 1),\n",
       " ('Nyholm', 1),\n",
       " ('Baron', 1),\n",
       " ('Bonnassie', 1),\n",
       " ('Alabdeh', 1),\n",
       " ('Cochet', 1),\n",
       " ('Nau', 1),\n",
       " ('Guérin-Dubiard', 1),\n",
       " ('Gautier', 1),\n",
       " ('Andrews', 1),\n",
       " ('Jan', 1),\n",
       " ('Coulthurst', 1),\n",
       " ('Hoekzema', 1),\n",
       " ('Romilly', 1),\n",
       " ('Dunn', 1),\n",
       " ('Porta', 1),\n",
       " ('Petrone', 1),\n",
       " ('Morello', 1),\n",
       " ('Granata', 1),\n",
       " ('Rizzo', 1),\n",
       " ('Memoli', 1),\n",
       " ('Weisz', 1),\n",
       " ('Sanchez-Vazquez', 1),\n",
       " ('Dewey', 1),\n",
       " ('Kitten', 1),\n",
       " ('Ross', 1),\n",
       " ('Gourse', 1),\n",
       " ('Ramsey', 1),\n",
       " ('Chavez', 1),\n",
       " ('Johnson', 1),\n",
       " ('Hosseinzadeh', 1),\n",
       " ('Mahoney', 1),\n",
       " ('Mohr', 1),\n",
       " ('Robison', 1),\n",
       " ('Zhong', 1),\n",
       " ('Hall', 1),\n",
       " ('MacCoss', 1),\n",
       " ('Bruce', 1),\n",
       " ('Cilia', 1),\n",
       " ('Krogh-Andersen', 1),\n",
       " ('Tellgren-Roth', 1),\n",
       " ('Martínez', 1),\n",
       " ('Günaydın', 1),\n",
       " ('Lin', 1),\n",
       " ('Martín', 1),\n",
       " ('Hammarström', 1),\n",
       " ('Marcotte', 1),\n",
       " ('Bartle', 1),\n",
       " ('Sumby', 1),\n",
       " ('Sundstrom', 1),\n",
       " ('Jiranek', 1),\n",
       " ('Vidovic', 1),\n",
       " ('An', 1),\n",
       " ('Mendoza', 1),\n",
       " ('Abrahante', 1),\n",
       " ('Johny', 1),\n",
       " ('Reed', 1),\n",
       " ('Cameron', 1),\n",
       " ('Bonis', 1),\n",
       " ('Phan', 1),\n",
       " ('Kent', 1),\n",
       " ('Hunter', 1),\n",
       " ('Aschtgen', 1),\n",
       " ('Brennan', 1),\n",
       " ('Nikolakakis', 1),\n",
       " ('Cohen', 1),\n",
       " ('Thöming', 1),\n",
       " ('Tomasch', 1),\n",
       " ('Preusse', 1),\n",
       " ('Koska', 1),\n",
       " ('Grahl', 1),\n",
       " ('Pohl', 1),\n",
       " ('Willger', 1),\n",
       " ('Kaever', 1),\n",
       " ('Müsken', 1),\n",
       " ('Häussler', 1),\n",
       " ('Oyama', 1),\n",
       " ('Girdwood', 1),\n",
       " ('Cookson', 1),\n",
       " ('Fernandez-Fuentes', 1),\n",
       " ('Privé', 1),\n",
       " ('Vallin', 1),\n",
       " ('Wilkinson', 1),\n",
       " ('Golyshin', 1),\n",
       " ('Golyshina', 1),\n",
       " ('Mikut', 1),\n",
       " ('Hilpert', 1),\n",
       " ('Richards', 1),\n",
       " ('Wootton', 1),\n",
       " ('Edwards', 1),\n",
       " ('Perrier', 1),\n",
       " ('Lundy', 1),\n",
       " ('Luo', 1),\n",
       " ('Hess', 1),\n",
       " ('Mantovani', 1),\n",
       " ('Creevey', 1),\n",
       " ('Huws', 1),\n",
       " ('Penesyan', 1),\n",
       " ('Nagy', 1),\n",
       " ('Kjelleberg', 1),\n",
       " ('Gillings', 1),\n",
       " ('Paulsen', 1),\n",
       " ('Kumar', 1),\n",
       " ('Alam', 1),\n",
       " ('Grover', 1),\n",
       " ('Pandey', 1),\n",
       " ('Tripathi', 1),\n",
       " ('Kumari', 1),\n",
       " ('Rani', 1),\n",
       " ('Akhter', 1),\n",
       " ('Ehtesham', 1),\n",
       " ('Hasnain', 1),\n",
       " ('Salgado', 1),\n",
       " ('Martínez-Flores', 1),\n",
       " ('Bustamante', 1),\n",
       " ('Alquicira-Hernández', 1),\n",
       " ('García-Sotelo', 1),\n",
       " ('García-Alonso', 1),\n",
       " ('Collado-Vides', 1),\n",
       " ('Schwartz', 1),\n",
       " ('Pan', 1),\n",
       " ('Bechhofer', 1),\n",
       " ('Deutscher', 1),\n",
       " ('Ye', 1),\n",
       " ('Xia', 1),\n",
       " ('Sheng', 1),\n",
       " ('Chi', 1),\n",
       " ('Gu', 1),\n",
       " ('Xiao', 1),\n",
       " ('Girard', 1),\n",
       " ('Blanchet', 1),\n",
       " ('Intertaglia', 1),\n",
       " ('Baudart', 1),\n",
       " ('Stien', 1),\n",
       " ('Suzuki', 1),\n",
       " ('Lebaron', 1),\n",
       " ('Lami', 1),\n",
       " ('Cui', 1),\n",
       " ('Bai', 1),\n",
       " ('Okutsu', 1),\n",
       " ('Morohoshi', 1),\n",
       " ('Xie', 1),\n",
       " ('Kato', 1),\n",
       " ('Ikeda', 1),\n",
       " ('Dwivedi', 1),\n",
       " ('Sanket', 1),\n",
       " ('Chavarría', 1),\n",
       " ('Goñi-Moreno', 1),\n",
       " ('de Lorenzo', 1),\n",
       " ('Nikel', 1),\n",
       " ('Urem', 1),\n",
       " ('van Rossum', 1),\n",
       " ('Bucca', 1),\n",
       " ('Moolenaar', 1),\n",
       " ('Laing', 1),\n",
       " ('Świątek-Połatyńska', 1),\n",
       " ('Willemse', 1),\n",
       " ('Tenconi', 1),\n",
       " ('Rigali', 1),\n",
       " ('Goosen', 1),\n",
       " ('van Wezel', 1),\n",
       " ('Hazen', 1),\n",
       " ('Daugherty', 1),\n",
       " ('Shetty', 1),\n",
       " ('Nataro', 1),\n",
       " ('Rasko', 1),\n",
       " ('Sontag', 1),\n",
       " ('Nakayasu', 1),\n",
       " ('Brown', 1),\n",
       " ('Niemann', 1),\n",
       " ('Sydor', 1),\n",
       " ('Sanchez', 1),\n",
       " ('Ansong', 1),\n",
       " ('Lu', 1),\n",
       " ('Choi', 1),\n",
       " ('Valleau', 1),\n",
       " ('Weitz', 1),\n",
       " ('Savchenko', 1),\n",
       " ('Cambronne', 1),\n",
       " ('Adkins', 1),\n",
       " ('El Filali', 1),\n",
       " ('Heath-Heckman', 1),\n",
       " ('Naydich', 1),\n",
       " ('Nangle', 1),\n",
       " ('Bues', 1),\n",
       " ('Trivedi', 1),\n",
       " ('Nissar', 1),\n",
       " ('Inniss', 1),\n",
       " ('Niederhuber', 1),\n",
       " ('Way', 1),\n",
       " ('Silver', 1),\n",
       " ('Riglar', 1),\n",
       " ('Yamamoto', 1),\n",
       " ('Yamanaka', 1),\n",
       " ('Shimada', 1),\n",
       " ('Sarkar', 1),\n",
       " ('Yoshida', 1),\n",
       " ('Bhardwaj', 1),\n",
       " ('Watanabe', 1),\n",
       " ('Taira', 1),\n",
       " ('Chatterji', 1),\n",
       " ('Ishihama', 1),\n",
       " ('Chouikha', 1),\n",
       " ('Sturdevant', 1),\n",
       " ('Jarrett', 1),\n",
       " ('Hinnebusch', 1),\n",
       " ('Anderson', 1),\n",
       " ('Mitchell', 1),\n",
       " ('Sintsova', 1),\n",
       " ('Rice', 1),\n",
       " ('Mobley', 1),\n",
       " ('Chihara', 1),\n",
       " ('Bischler', 1),\n",
       " ('Barquist', 1),\n",
       " ('Monzon', 1),\n",
       " ('Noda', 1),\n",
       " ('Tsuneda', 1),\n",
       " ('Quillin', 1),\n",
       " ('Hockenberry', 1),\n",
       " ('Jewett', 1),\n",
       " ('Seifert', 1),\n",
       " ('Balasubramanian', 1),\n",
       " ('Ragunathan', 1),\n",
       " ('Fei', 1),\n",
       " ('Khare', 1),\n",
       " ('Tavazoie', 1),\n",
       " ('Bellieny-Rabelo', 1),\n",
       " ('Nkomo', 1),\n",
       " ('Shyntum', 1),\n",
       " ('Moleleki', 1),\n",
       " ('Prieto', 1),\n",
       " ('Bernabeu', 1),\n",
       " ('Aznar', 1),\n",
       " ('Ruiz-Cruz', 1),\n",
       " ('Bravo', 1),\n",
       " ('Queiroz', 1),\n",
       " ('Juárez', 1),\n",
       " ('Gelsinger', 1),\n",
       " ('Uritskiy', 1),\n",
       " ('Reddy', 1),\n",
       " ('Munn', 1),\n",
       " ('Farney', 1),\n",
       " ('DiRuggiero', 1),\n",
       " ('Han', 1),\n",
       " ('Yan', 1),\n",
       " ('Gao', 1),\n",
       " ('Xue', 1),\n",
       " ('Yang', 1)]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cl_authors = []\n",
    "for doi in clDict[cl][\"dois\"]:\n",
    "    #print(doi2str[doi][14:-2])\n",
    "    cl_authors.append(doi2str[doi][14:-2].split(\" (20\", 1)[0].split(\", \"))\n",
    "    #cl_authors.append(doi2str[doi][14:-2].split(\" (20\", 1)[0])\n",
    "\n",
    "cl_authors\n",
    "\n",
    "cl_author_counts = Counter()\n",
    "for a in cl_authors:\n",
    "    cl_author_counts.update(a)\n",
    "\n",
    "cl_author_counts.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 15:23:27,802 : INFO : adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2020-03-18 15:23:27,802 : INFO : built Dictionary(12 unique tokens: ['computer', 'human', 'interface', 'response', 'survey']...) from 9 documents (total 29 corpus positions)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(0, 1), (3, 1), (4, 1), (5, 1), (6, 1), (7, 1)]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from gensim.test.utils import common_corpus, common_dictionary\n",
    "\n",
    "common_corpus[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-03-18 15:23:27,833 : INFO : adding document #0 to Dictionary(0 unique tokens: [])\n",
      "2020-03-18 15:23:27,833 : INFO : built Dictionary(417 unique tokens: ['Ainsaar', 'Hõrak', 'Kasvandik', 'Tamman', 'Tenson']...) from 71 documents (total 467 corpus positions)\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'LsiModel' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-53-da49ff3ed214>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mcorpus\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mdictionary\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdoc2bow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mauthors\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mauthors\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mcl_authors\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 7\u001b[1;33m \u001b[0mlsi_authors\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mLsiModel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcorpus\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mid2word\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdictionary\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mid2token\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'LsiModel' is not defined"
     ]
    }
   ],
   "source": [
    "from gensim import corpora\n",
    "\n",
    "dictionary = corpora.Dictionary(cl_authors)\n",
    "\n",
    "corpus = [dictionary.doc2bow(authors) for authors in cl_authors]\n",
    "\n",
    "lsi_authors = LsiModel(corpus, id2word=dictionary.id2token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lsi_authors.print_topics()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lsi_authors = LsiModel(corpus=cl_authors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gensim.models import LsiModel\n",
    "\n",
    "i2a = {}\n",
    "for name, count in cl_author_counts.items():\n",
    "    i = len(i2a)\n",
    "    i2a[i] = name\n",
    "\n",
    "lsi_authors = LsiModel(corpus=cl_authors, id2word=i2a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([s for s in clSizes if s >= 50])  # diese Cluster-Dois per Random Forest klassifizieren"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['reverse',\n",
       " 'logistics',\n",
       " 'network',\n",
       " 'design',\n",
       " 'effective',\n",
       " 'management',\n",
       " 'medical',\n",
       " 'waste',\n",
       " 'epidemic',\n",
       " 'outbreaks:',\n",
       " 'insights',\n",
       " 'coronavirus',\n",
       " 'disease',\n",
       " '2019',\n",
       " '(covid-19)',\n",
       " 'outbreak',\n",
       " 'wuhan',\n",
       " '(china)']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "title_tokens[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "token_counter = Counter()\n",
    "\n",
    "for tt in title_tokens:\n",
    "    token_counter.update(tt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('coronavirus', 380),\n",
       " ('networks', 243),\n",
       " ('sensor', 233),\n",
       " ('novel', 228),\n",
       " ('virus', 228),\n",
       " ('respiratory', 225),\n",
       " ('based', 223),\n",
       " ('human', 213),\n",
       " ('analysis', 210),\n",
       " ('protein', 195),\n",
       " ('localization', 191),\n",
       " ('data', 190),\n",
       " ('structure', 184),\n",
       " ('wireless', 172),\n",
       " ('syndrome', 165),\n",
       " ('indoor', 160),\n",
       " ('middle', 140),\n",
       " ('east', 140),\n",
       " ('infection', 139),\n",
       " ('underwater', 136),\n",
       " ('system', 117),\n",
       " ('new', 116),\n",
       " ('performance', 114),\n",
       " ('molecular', 104),\n",
       " ('algorithm', 101),\n",
       " ('evolution', 99),\n",
       " ('genome', 98),\n",
       " ('structural', 98),\n",
       " ('disease', 93),\n",
       " ('public', 91),\n",
       " ('positioning', 91),\n",
       " ('estimation', 89),\n",
       " ('species', 88),\n",
       " ('study', 87),\n",
       " ('crystal', 84),\n",
       " ('complex', 84),\n",
       " ('role', 83),\n",
       " ('transmission', 82),\n",
       " ('model', 81),\n",
       " ('approach', 81),\n",
       " ('reveals', 77),\n",
       " ('genetic', 74),\n",
       " ('2019', 73),\n",
       " ('outbreak', 73),\n",
       " ('acute', 73),\n",
       " ('method', 73),\n",
       " ('viral', 72),\n",
       " ('cell', 72),\n",
       " ('mers-cov', 72),\n",
       " ('potential', 71),\n",
       " ('gene', 71),\n",
       " ('network', 70),\n",
       " ('identification', 70),\n",
       " ('influenza', 69),\n",
       " ('clinical', 68),\n",
       " ('china', 68),\n",
       " ('rna', 68),\n",
       " ('management', 67),\n",
       " ('binding', 65),\n",
       " ('detection', 64),\n",
       " ('control', 64),\n",
       " ('structures', 63),\n",
       " ('diversity', 63),\n",
       " ('navigation', 63),\n",
       " ('acoustic', 62),\n",
       " ('insights', 61),\n",
       " ('multiple', 59),\n",
       " ('genomic', 58),\n",
       " ('review', 57),\n",
       " ('patients', 57),\n",
       " ('receptor', 57),\n",
       " ('learning', 57),\n",
       " ('domain', 57),\n",
       " ('sequence', 57),\n",
       " ('sequencing', 56),\n",
       " ('development', 55),\n",
       " ('design', 54),\n",
       " ('evolutionary', 54),\n",
       " ('impact', 54),\n",
       " ('characterization', 54),\n",
       " ('cells', 54),\n",
       " ('spike', 54),\n",
       " ('routing', 54),\n",
       " ('energy', 53),\n",
       " ('host', 52),\n",
       " ('viruses', 52),\n",
       " ('treatment', 51),\n",
       " ('evidence', 51),\n",
       " ('proteins', 51),\n",
       " ('mobile', 51),\n",
       " ('dna', 51),\n",
       " ('function', 50),\n",
       " ('mechanism', 50),\n",
       " ('ebola', 50),\n",
       " ('coronaviruses', 49),\n",
       " ('vaccine', 49),\n",
       " ('research', 49),\n",
       " ('information', 49),\n",
       " ('organizational', 49),\n",
       " ('target', 48),\n",
       " ('genes', 48),\n",
       " ('use', 48),\n",
       " ('expression', 47),\n",
       " ('global', 47),\n",
       " ('improved', 47),\n",
       " ('functional', 46),\n",
       " ('c', 46),\n",
       " ('robust', 46),\n",
       " ('resistance', 45),\n",
       " ('protocol', 45),\n",
       " ('tracking', 45),\n",
       " ('risk', 44),\n",
       " ('health', 44),\n",
       " ('bats', 44),\n",
       " ('drug', 44),\n",
       " ('comparative', 44),\n",
       " ('hepatitis', 44),\n",
       " ('emerging', 43),\n",
       " ('dynamics', 43),\n",
       " ('case', 43),\n",
       " ('prediction', 43),\n",
       " ('type', 43),\n",
       " ('bacterial', 43),\n",
       " ('scheme', 43),\n",
       " ('adaptive', 43),\n",
       " ('3d', 43),\n",
       " ('systems', 43),\n",
       " ('associated', 42),\n",
       " ('severe', 42),\n",
       " ('models', 42),\n",
       " ('effects', 41),\n",
       " ('activity', 41),\n",
       " ('zika', 41),\n",
       " ('response', 40),\n",
       " ('social', 40),\n",
       " ('genomes', 40),\n",
       " ('interactions', 40),\n",
       " ('crystallography', 40),\n",
       " ('evaluation', 39),\n",
       " ('fusion', 39),\n",
       " ('survey', 39),\n",
       " ('acid', 39),\n",
       " ('replication', 39),\n",
       " ('bat', 38),\n",
       " ('environments', 38),\n",
       " ('determination', 38),\n",
       " ('efficient', 38),\n",
       " ('epidemic', 37),\n",
       " ('patterns', 37),\n",
       " ('sensors', 37),\n",
       " ('antiviral', 36),\n",
       " ('effect', 36),\n",
       " ('microbial', 36),\n",
       " ('mapping', 36),\n",
       " ('local', 36),\n",
       " ('pedestrian', 36),\n",
       " ('2019-ncov', 35),\n",
       " ('covid-19', 35),\n",
       " ('recent', 35),\n",
       " ('lung', 35),\n",
       " ('infections', 35),\n",
       " ('discovery', 35),\n",
       " ('selection', 35),\n",
       " ('cooperative', 35),\n",
       " ('inhibitors', 35),\n",
       " ('filter', 35),\n",
       " ('inertial', 35),\n",
       " ('understanding', 34),\n",
       " ('systematic', 34),\n",
       " ('screening', 34),\n",
       " ('family', 33),\n",
       " ('sp.', 33),\n",
       " ('job', 33),\n",
       " ('pneumonia', 32),\n",
       " ('different', 32),\n",
       " ('application', 32),\n",
       " ('recognition', 32),\n",
       " ('interaction', 32),\n",
       " ('optimization', 32),\n",
       " ('phylogeny', 32),\n",
       " ('current', 31),\n",
       " ('targeting', 31),\n",
       " ('infectious', 31),\n",
       " ('comparison', 31),\n",
       " ('body', 31),\n",
       " ('studies', 30),\n",
       " ('protease', 30),\n",
       " ('modeling', 30),\n",
       " ('communication', 30),\n",
       " ('hybrid', 30),\n",
       " ('approaches', 30),\n",
       " ('x-ray', 30),\n",
       " ('autonomous', 30),\n",
       " ('state', 30),\n",
       " ('origin', 29),\n",
       " ('(2019-ncov)', 29),\n",
       " ('characteristics', 29),\n",
       " ('features', 29),\n",
       " ('challenges', 29),\n",
       " ('t', 29),\n",
       " ('basis', 29),\n",
       " ('responses', 29),\n",
       " ('factors', 29),\n",
       " ('motion', 29),\n",
       " ('revealed', 29),\n",
       " ('diverse', 29),\n",
       " ('database', 29),\n",
       " ('therapeutic', 28),\n",
       " ('variation', 28),\n",
       " ('mechanisms', 28),\n",
       " ('dynamic', 28),\n",
       " ('antibody', 28),\n",
       " ('area', 28),\n",
       " ('relationship', 28),\n",
       " ('team', 28),\n",
       " ('isolated', 27),\n",
       " ('visual', 27),\n",
       " ('peptide', 27),\n",
       " ('machine', 27),\n",
       " ('policy', 27),\n",
       " ('mutations', 27),\n",
       " ('assembly', 27),\n",
       " ('activation', 27),\n",
       " ('police', 27),\n",
       " ('mers', 26),\n",
       " ('early', 26),\n",
       " ('group', 26),\n",
       " ('single', 26),\n",
       " ('time', 26),\n",
       " ('map', 26),\n",
       " ('genus', 26),\n",
       " ('fast', 26),\n",
       " ('pathogen', 25),\n",
       " ('assessment', 25),\n",
       " ('–', 25),\n",
       " ('phylogenetic', 25),\n",
       " ('perspective', 25),\n",
       " ('south', 25),\n",
       " ('integrated', 25),\n",
       " ('adaptation', 25),\n",
       " ('direct', 25),\n",
       " ('deep', 25),\n",
       " ('natural', 25),\n",
       " ('signal', 25),\n",
       " ('wuhan,', 24),\n",
       " ('entry', 24),\n",
       " ('porcine', 24),\n",
       " ('government', 24),\n",
       " ('virtual', 24),\n",
       " ('reveal', 24),\n",
       " ('production', 24),\n",
       " ('strains', 24),\n",
       " ('wi-fi', 24),\n",
       " ('applications', 24),\n",
       " ('electron', 24),\n",
       " ('determinants', 24),\n",
       " ('algorithms', 24),\n",
       " ('environmental', 24),\n",
       " ('macromolecular', 24),\n",
       " ('emergence', 23),\n",
       " ('strategies', 23),\n",
       " ('real-time', 23),\n",
       " ('number', 23),\n",
       " ('1', 23),\n",
       " ('antibodies', 23),\n",
       " ('analyses', 23),\n",
       " ('international', 23),\n",
       " ('tool', 23),\n",
       " ('rapid', 23),\n",
       " ('classification', 23),\n",
       " ('inhibition', 23),\n",
       " ('sequences', 23),\n",
       " ('small', 23),\n",
       " ('fungal', 23),\n",
       " ('clustering', 23),\n",
       " ('neural', 23),\n",
       " ('regulation', 23),\n",
       " ('association', 23),\n",
       " ('improving', 23),\n",
       " ('sensing', 23),\n",
       " ('korea', 22),\n",
       " ('sars', 22),\n",
       " ('states', 22),\n",
       " ('highly', 22),\n",
       " ('diseases', 22),\n",
       " ('comprehensive', 22),\n",
       " ('tools', 22),\n",
       " ('high', 22),\n",
       " ('vaccines', 22),\n",
       " ('light', 22),\n",
       " ('factor', 22),\n",
       " ('immune', 22),\n",
       " ('strain', 22),\n",
       " ('coli', 22),\n",
       " ('array', 22),\n",
       " ('mutation', 22),\n",
       " ('bacillus', 22),\n",
       " ('search', 22),\n",
       " ('support', 22),\n",
       " ('mycobacterium', 22),\n",
       " ('source', 22),\n",
       " ('diffraction', 22),\n",
       " ('integration', 22),\n",
       " ('wsns', 22),\n",
       " ('channel', 22),\n",
       " ('slam', 22),\n",
       " ('refinement', 22),\n",
       " ('welfare', 22),\n",
       " ('humans', 21),\n",
       " ('community', 21),\n",
       " ('2', 21),\n",
       " ('care', 21),\n",
       " ('estimating', 21),\n",
       " ('healthcare', 21),\n",
       " ('metabolic', 21),\n",
       " ('marine', 21),\n",
       " ('position', 21),\n",
       " ('cancer', 21),\n",
       " ('escherichia', 21),\n",
       " ('energy-efficient', 21),\n",
       " ('node', 21),\n",
       " ('measurements', 21),\n",
       " ('accurate', 21),\n",
       " ('low', 21),\n",
       " ('influence', 21),\n",
       " ('european', 21),\n",
       " ('environment', 21),\n",
       " ('experimental', 21),\n",
       " ('service', 21),\n",
       " ('effective', 20),\n",
       " ('processing', 20),\n",
       " ('epidemiology', 20),\n",
       " ('injury', 20),\n",
       " ('outcomes', 20),\n",
       " ('implications', 20),\n",
       " ('critical', 20),\n",
       " ('spread', 20),\n",
       " ('surface', 20),\n",
       " ('site', 20),\n",
       " ('populations', 20),\n",
       " ('enhanced', 20),\n",
       " ('immunity', 20),\n",
       " ('distributed', 20),\n",
       " ('microscopy', 20),\n",
       " ('class', 20),\n",
       " ('methods', 20),\n",
       " ('kalman', 20),\n",
       " ('vehicle', 20),\n",
       " ('inhibits', 20),\n",
       " ('inhibitor', 20),\n",
       " ('work', 20),\n",
       " ('plant', 20),\n",
       " ('changes', 20),\n",
       " ('conflict', 20),\n",
       " ('solution', 20),\n",
       " ('phasing', 20),\n",
       " ('(covid-19)', 19),\n",
       " ('receptors', 19),\n",
       " ('diagnosis', 19),\n",
       " ('sars-cov-2', 19),\n",
       " ('therapy', 19),\n",
       " ('future', 19),\n",
       " ('common', 19),\n",
       " ('glycoprotein', 19),\n",
       " ('large', 19),\n",
       " ('surveillance', 19),\n",
       " ('effectiveness', 19),\n",
       " ('cases', 19),\n",
       " ('targets', 19),\n",
       " ('robot', 19),\n",
       " ('serial', 19),\n",
       " ('smartphone', 19),\n",
       " ('optimal', 19),\n",
       " ('framework', 19),\n",
       " ('testing', 19),\n",
       " ('transcriptional', 19),\n",
       " ('accuracy', 19),\n",
       " ('resolution', 19),\n",
       " ('taxonomy', 19),\n",
       " ('stress', 19),\n",
       " ('joint', 19),\n",
       " ('sars-cov', 18),\n",
       " ('united', 18),\n",
       " ('outbreaks', 18),\n",
       " ('overview', 18),\n",
       " ('history', 18),\n",
       " ('enzyme', 18),\n",
       " ('update', 18),\n",
       " ('(mers-cov)', 18),\n",
       " ('population', 18),\n",
       " ('coverage', 18),\n",
       " ('e', 18),\n",
       " ('profiling', 18),\n",
       " ('bacteria', 18),\n",
       " ('membrane', 18),\n",
       " ('ii', 18),\n",
       " ('avian', 18),\n",
       " ('genomics', 18),\n",
       " ('tuberculosis', 18),\n",
       " ('2015', 18),\n",
       " ('fingerprinting', 18),\n",
       " ('feature', 18),\n",
       " ('change', 18),\n",
       " ('space', 18),\n",
       " ('frankia', 18),\n",
       " ('microbiome', 18),\n",
       " ('quality', 18),\n",
       " ('collection', 18),\n",
       " ('crystallographic', 18),\n",
       " ('wuhan', 17),\n",
       " ('pathogenic', 17),\n",
       " ('differential', 17),\n",
       " ('advances', 17),\n",
       " ('efficacy', 17),\n",
       " ('pattern', 17),\n",
       " ('saudi', 17),\n",
       " ('cluster', 17),\n",
       " ('reference', 17),\n",
       " ('identifies', 17),\n",
       " ('presence', 17),\n",
       " ('complete', 17),\n",
       " ('salmonella', 17),\n",
       " ('antibiotic', 17),\n",
       " ('tissue', 17),\n",
       " ('odometry', 17),\n",
       " ('web', 17),\n",
       " ('computational', 17),\n",
       " ('variants', 17),\n",
       " ('pathways', 17),\n",
       " ('taxonomic', 17),\n",
       " ('emotional', 17),\n",
       " ('location', 17),\n",
       " ('signals', 17),\n",
       " ('coronavirus:', 16),\n",
       " ('polymerase', 16),\n",
       " ('death', 16),\n",
       " ('cryo-em', 16),\n",
       " ('identifying', 16),\n",
       " ('mouse', 16),\n",
       " ('essential', 16),\n",
       " ('monoclonal', 16),\n",
       " ('life', 16),\n",
       " ('biological', 16),\n",
       " ('dengue', 16),\n",
       " ('biology', 16),\n",
       " ('metagenomics', 16),\n",
       " ('doa', 16),\n",
       " ('yeast', 16),\n",
       " ('sources', 16),\n",
       " ('major', 16),\n",
       " ('pathway', 16),\n",
       " ('virulence', 16),\n",
       " ('description', 16),\n",
       " ('landscape', 16),\n",
       " ('nov.,', 16),\n",
       " ('climate', 16),\n",
       " ('leadership', 16),\n",
       " ('transfer', 16),\n",
       " ('mitochondrial', 16),\n",
       " ('related', 16),\n",
       " ('devices', 16),\n",
       " ('measurement', 16),\n",
       " ('sector', 16),\n",
       " ('roles', 15),\n",
       " ('patient', 15),\n",
       " ('china:', 15),\n",
       " ('phase', 15),\n",
       " ('b', 15),\n",
       " ('animal', 15),\n",
       " ('enhancement', 15),\n",
       " ('general', 15),\n",
       " ('modelling', 15),\n",
       " ('subunit', 15),\n",
       " ('integrative', 15),\n",
       " ('complexes', 15),\n",
       " ('substrate', 15),\n",
       " ('macrococcus', 15),\n",
       " ('modulation', 15),\n",
       " ('large-scale', 15),\n",
       " ('pseudomonas', 15),\n",
       " ('amino', 15),\n",
       " ('leukemia', 15),\n",
       " ('perspectives', 15),\n",
       " ('monocular', 15),\n",
       " ('distress', 15),\n",
       " ('reaction', 15),\n",
       " ('morphological', 15),\n",
       " ('formation', 15),\n",
       " ('native', 15),\n",
       " ('mice', 15),\n",
       " ('error', 15),\n",
       " ('particle', 15),\n",
       " ('conserved', 15),\n",
       " ('transcriptomic', 15),\n",
       " ('crystals', 15),\n",
       " ('diversification', 15),\n",
       " ('validation', 15),\n",
       " ('regulatory', 15),\n",
       " ('education', 15),\n",
       " ('accountability', 15),\n",
       " ('neutron', 15),\n",
       " ('nonprofit', 15),\n",
       " ('pathogenesis', 14),\n",
       " ('report', 14),\n",
       " ('bayesian', 14),\n",
       " ('single-cell', 14),\n",
       " ('epidemiological', 14),\n",
       " ('chinese', 14),\n",
       " ('composition', 14),\n",
       " ('progress', 14),\n",
       " ('measures', 14),\n",
       " ('hiv-1', 14),\n",
       " ('predict', 14),\n",
       " ('status', 14),\n",
       " ('zoonotic', 14),\n",
       " ('drugs', 14),\n",
       " ('safety', 14),\n",
       " ('ant', 14),\n",
       " ('building', 14),\n",
       " ('vehicles', 14),\n",
       " ('functions', 14),\n",
       " ('plants', 14),\n",
       " ('mac', 14),\n",
       " ('high-throughput', 14),\n",
       " ('gut', 14),\n",
       " ('genetics', 14),\n",
       " ('passive', 14),\n",
       " ('years', 14),\n",
       " ('kinase', 14),\n",
       " ('antimicrobial', 14),\n",
       " ('core', 14),\n",
       " ('meta-analysis', 14),\n",
       " ('differences', 14),\n",
       " ('cellular', 14),\n",
       " ('exploring', 14),\n",
       " ('reliable', 14),\n",
       " ('magnetic', 14),\n",
       " ('results', 14),\n",
       " ('nucleotide', 14),\n",
       " ('central', 14),\n",
       " ('generation', 14),\n",
       " ('distribution', 14),\n",
       " ('national', 14),\n",
       " ('conservation', 14),\n",
       " ('resource', 14),\n",
       " ('audit', 14),\n",
       " ('medical', 13),\n",
       " ('infected', 13),\n",
       " ('region', 13),\n",
       " ('2020', 13),\n",
       " ('assessing', 13),\n",
       " ('ace2', 13),\n",
       " ('strategy', 13),\n",
       " ('provides', 13),\n",
       " ('pediatric', 13),\n",
       " ('domains', 13),\n",
       " ('draft', 13),\n",
       " ('growth', 13),\n",
       " ('conformational', 13),\n",
       " ('nuclear', 13),\n",
       " ('synthase', 13),\n",
       " ('technique', 13),\n",
       " ('deployment', 13),\n",
       " ('uwb', 13),\n",
       " ('docking', 13),\n",
       " ('transcriptome', 13),\n",
       " ('collaborative', 13),\n",
       " ('behavior', 13),\n",
       " ('power', 13),\n",
       " ('genome-wide', 13),\n",
       " ('chronic', 13),\n",
       " ('range', 13),\n",
       " ('vivo', 13),\n",
       " ('metagenomic', 13),\n",
       " ('mammalian', 13),\n",
       " ('long-term', 13),\n",
       " ('communities', 13),\n",
       " ('cryo-electron', 13),\n",
       " ('relationships', 13),\n",
       " ('cryptic', 13),\n",
       " ('4', 13),\n",
       " ('delivery', 13),\n",
       " ('active', 13),\n",
       " ('orientation', 13),\n",
       " ('satisfaction', 13),\n",
       " ('sad', 13),\n",
       " ('scheduling', 13),\n",
       " ('lessons', 12),\n",
       " ('china,', 12),\n",
       " ('susceptibility', 12),\n",
       " ('emergency', 12),\n",
       " ('usage', 12),\n",
       " ('lineage', 12),\n",
       " ('protective', 12),\n",
       " ('arabia', 12),\n",
       " ('moving', 12),\n",
       " ('workers', 12),\n",
       " ('deltacoronavirus', 12),\n",
       " ('challenge', 12),\n",
       " ('knowledge', 12),\n",
       " ('higher', 12),\n",
       " ('isolation', 12),\n",
       " ('receptor-binding', 12),\n",
       " ('specificity', 12),\n",
       " ('platform', 12),\n",
       " ('enterica', 12),\n",
       " ('candidate', 12),\n",
       " ('synthesis', 12),\n",
       " ('targeted', 12),\n",
       " ('involved', 12),\n",
       " ('transport', 12),\n",
       " ('recombinant', 12),\n",
       " ('survival', 12),\n",
       " ('improve', 12),\n",
       " ('multipath', 12),\n",
       " ('delay', 12),\n",
       " ('calibration', 12),\n",
       " ('whole-genome', 12),\n",
       " ('synthetic', 12),\n",
       " ('longitudinal', 12),\n",
       " ('linear', 12),\n",
       " ('chain', 12),\n",
       " ('robots', 12),\n",
       " ('dromedary', 12),\n",
       " ('geometric', 12),\n",
       " ('turnover', 12),\n",
       " ('west', 12),\n",
       " ('conditions', 12),\n",
       " ('similarity', 12),\n",
       " ('induced', 12),\n",
       " ('ecological', 12),\n",
       " ('integrating', 12),\n",
       " ('silico', 12),\n",
       " ('correction', 12),\n",
       " ('heterogeneity', 12),\n",
       " ('metabolism', 12),\n",
       " ('elements', 12),\n",
       " ('recovery', 12),\n",
       " ('lower', 12),\n",
       " ('pathogens', 12),\n",
       " ('enzymes', 12),\n",
       " ('isolates', 12),\n",
       " ('low-cost', 12),\n",
       " ('replacement', 12),\n",
       " ('chromosomal', 12),\n",
       " ('accounting', 12),\n",
       " ('market', 12),\n",
       " ('monitoring', 12),\n",
       " ('interference', 12),\n",
       " ('aggregation', 12),\n",
       " ('political', 12),\n",
       " ('services', 12),\n",
       " ('investigation', 11),\n",
       " ('distinct', 11),\n",
       " ('recombination', 11),\n",
       " ('oral', 11),\n",
       " ('rna-seq', 11),\n",
       " ('agents', 11),\n",
       " ('angiotensin', 11),\n",
       " ('caused', 11),\n",
       " ('vitro', 11),\n",
       " ('mitigation', 11),\n",
       " ('codon', 11),\n",
       " ('options', 11),\n",
       " ('protection', 11),\n",
       " ('link', 11),\n",
       " ('2017', 11),\n",
       " ('sensitive', 11),\n",
       " ('nucleocapsid', 11),\n",
       " ('cycle', 11),\n",
       " ('republic', 11),\n",
       " ('age', 11),\n",
       " ('urban', 11),\n",
       " ('access', 11),\n",
       " ('chlorophyll', 11),\n",
       " ('primary', 11),\n",
       " ('tree', 11),\n",
       " ('opportunistic', 11),\n",
       " ('field', 11),\n",
       " ('radio', 11),\n",
       " ('kinetochore', 11),\n",
       " ('server', 11),\n",
       " ('african', 11),\n",
       " ('critically', 11),\n",
       " ('path', 11),\n",
       " ('identify', 11),\n",
       " ('implementation', 11),\n",
       " ('fingerprint', 11),\n",
       " ('lymphoblastic', 11),\n",
       " ('employee', 11),\n",
       " ('processes', 11),\n",
       " ('enables', 11),\n",
       " ('allosteric', 11),\n",
       " ('opportunities', 11),\n",
       " ('regulator', 11),\n",
       " ('combining', 11),\n",
       " ('broad-spectrum', 11),\n",
       " ('protects', 11),\n",
       " ('translation', 11),\n",
       " ('profile', 11),\n",
       " ('mining', 11),\n",
       " ('signaling', 11),\n",
       " ('reconstruction', 11),\n",
       " ('stability', 11),\n",
       " ('spectrum', 11),\n",
       " ('laser', 11),\n",
       " ('force', 11),\n",
       " ('bacterium', 11),\n",
       " ('extreme', 11),\n",
       " ('politics', 11),\n",
       " ('cause', 11),\n",
       " ('transcription', 11),\n",
       " ('radar', 11),\n",
       " ('policies', 11),\n",
       " ('physical', 11),\n",
       " ('length', 11),\n",
       " ('ecology', 11),\n",
       " ('murine', 11),\n",
       " ('activities', 11),\n",
       " ('moderating', 11),\n",
       " ('nlos', 11),\n",
       " ('labor', 11),\n",
       " ('examining', 11),\n",
       " ('iot', 11),\n",
       " ('personality', 11),\n",
       " ('employees', 11),\n",
       " ('organizations', 11),\n",
       " ('arrays', 11),\n",
       " ('reverse', 10),\n",
       " ('2019‐ncov', 10),\n",
       " ('travel', 10),\n",
       " ('divergence', 10),\n",
       " ('rate', 10),\n",
       " ('causes', 10),\n",
       " ('unknown', 10),\n",
       " ('prevalence', 10),\n",
       " ('rhesus', 10),\n",
       " ('fatal', 10),\n",
       " ('plasma', 10),\n",
       " ('blood', 10),\n",
       " ('neutralizing', 10),\n",
       " ('potent', 10),\n",
       " ('engineering', 10),\n",
       " ('flexible', 10),\n",
       " ('trends', 10),\n",
       " ('sparse', 10),\n",
       " ('division', 10),\n",
       " ('decision', 10),\n",
       " ('efficiency', 10),\n",
       " ('smartphones', 10),\n",
       " ('simultaneous', 10),\n",
       " ('multilevel', 10),\n",
       " ('culture', 10),\n",
       " ('pandemic', 10),\n",
       " ('peptides', 10),\n",
       " ('copy', 10),\n",
       " ('adipose', 10),\n",
       " ('maps', 10),\n",
       " ('spatial', 10),\n",
       " ('visual-inertial', 10),\n",
       " ('vector', 10),\n",
       " ('nodes', 10),\n",
       " ('programming', 10),\n",
       " ('three-dimensional', 10),\n",
       " ('empirical', 10),\n",
       " ('dual', 10),\n",
       " ('splicing', 10),\n",
       " ('ill', 10),\n",
       " ('camels', 10),\n",
       " ('hereditary', 10),\n",
       " ('product', 10),\n",
       " ('order', 10),\n",
       " ('frequency', 10),\n",
       " ('quantification', 10),\n",
       " ('sample', 10),\n",
       " ('interactive', 10),\n",
       " ('open', 10),\n",
       " ('loci', 10),\n",
       " ('syncytial', 10),\n",
       " ('extensive', 10),\n",
       " ('regulates', 10),\n",
       " ('action', 10),\n",
       " ('loss', 10),\n",
       " ('long', 10),\n",
       " ('combined', 10),\n",
       " ('bound', 10),\n",
       " ('interface', 10),\n",
       " ('next-generation', 10),\n",
       " ('ligand', 10),\n",
       " ('samples', 10),\n",
       " ('2016', 10),\n",
       " ('tract', 10),\n",
       " ('pipeline', 10),\n",
       " ('matching', 10),\n",
       " ('problem', 10),\n",
       " ('employment', 10),\n",
       " ('attitude', 10),\n",
       " ('burnout', 10),\n",
       " ('anomalous', 10),\n",
       " ('preliminary', 9),\n",
       " ('important', 9),\n",
       " ('adults', 9),\n",
       " ('identified', 9),\n",
       " ('january', 9),\n",
       " ('organization', 9),\n",
       " ('practice', 9),\n",
       " ('test', 9),\n",
       " ('biochemical', 9),\n",
       " ('hospital', 9),\n",
       " ('nipah', 9),\n",
       " ('points', 9),\n",
       " ('media', 9),\n",
       " ('variant', 9),\n",
       " ('small-molecule', 9),\n",
       " ('affect', 9),\n",
       " ('protocols', 9),\n",
       " ('metagenomes', 9),\n",
       " ('required', 9),\n",
       " ('pneumoniae', 9),\n",
       " ('methylation', 9),\n",
       " ('de', 9),\n",
       " ('provide', 9),\n",
       " ('mhc', 9),\n",
       " ('variations', 9),\n",
       " ('wifi', 9),\n",
       " ('governance', 9),\n",
       " ('technology', 9),\n",
       " ('eastern', 9),\n",
       " ('construction', 9),\n",
       " ('tritici', 9),\n",
       " ('angioedema', 9),\n",
       " ('predicting', 9),\n",
       " ('mortality', 9),\n",
       " ('innovation', 9),\n",
       " ('alignment', 9),\n",
       " ('planning', 9),\n",
       " ('water', 9),\n",
       " ('properties', 9),\n",
       " ('n-terminal', 9),\n",
       " ('differentiation', 9),\n",
       " ('administration', 9),\n",
       " ('annotation', 9),\n",
       " ('residues', 9),\n",
       " ('korea,', 9),\n",
       " ('imu', 9),\n",
       " ('secretion', 9),\n",
       " ('c-terminal', 9),\n",
       " ('independent', 9),\n",
       " ('hidden', 9),\n",
       " ('running', 9),\n",
       " ('resources', 9),\n",
       " ('å', 9),\n",
       " ('evaluating', 9),\n",
       " ('pathogenicity', 9),\n",
       " ('biofilm', 9),\n",
       " ('heterogeneous', 9),\n",
       " ('representation', 9),\n",
       " ('combination', 9),\n",
       " ('phenotype', 9),\n",
       " ('mode', 9),\n",
       " ('void', 9),\n",
       " ('acquired', 9),\n",
       " ('interferon', 9),\n",
       " ('smart', 9),\n",
       " ('fruit', 9),\n",
       " ('automated', 9),\n",
       " ('cognitive', 9),\n",
       " ('profiles', 9),\n",
       " ('macrophage', 9),\n",
       " ('game', 9),\n",
       " ('theory', 9),\n",
       " ('graph', 9),\n",
       " ('rgb-d', 9),\n",
       " ('cameras', 9),\n",
       " ('intelligent', 9),\n",
       " ('hydrolase', 9),\n",
       " ('africa', 9),\n",
       " ('consequences', 9),\n",
       " ('vision', 9),\n",
       " ('strategic', 9),\n",
       " ('insight', 9),\n",
       " ('carbon', 9),\n",
       " ('finding', 9),\n",
       " ('performance:', 9),\n",
       " ('perceived', 9),\n",
       " ('avoidance', 9),\n",
       " ('u.s.', 9),\n",
       " ('allocation', 9),\n",
       " ('imaging', 8),\n",
       " ('temporal', 8),\n",
       " ('medicine', 8),\n",
       " ('basic', 8),\n",
       " ('-', 8),\n",
       " ('chemistry', 8),\n",
       " ('structure,', 8),\n",
       " ('developing', 8),\n",
       " ('scenario', 8),\n",
       " ('inactivation', 8),\n",
       " ('repurposing', 8),\n",
       " ('hosts', 8),\n",
       " ('infection:', 8),\n",
       " ('conformation', 8),\n",
       " ('masks', 8),\n",
       " ('air', 8),\n",
       " ('need', 8),\n",
       " ('helicase', 8),\n",
       " ('detecting', 8),\n",
       " ('individual', 8),\n",
       " ('importance', 8),\n",
       " ('visible', 8),\n",
       " ('antigen', 8),\n",
       " ('good', 8),\n",
       " ('science', 8),\n",
       " ('diagnostic', 8),\n",
       " ('modes', 8),\n",
       " ('signatures', 8),\n",
       " ('people', 8),\n",
       " ('multi-sensor', 8),\n",
       " ('techniques', 8),\n",
       " ('photosystem', 8),\n",
       " ('hiv', 8),\n",
       " ('wheat', 8),\n",
       " ('streptococcus', 8),\n",
       " ('exploration', 8),\n",
       " ('promote', 8),\n",
       " ('alternative', 8),\n",
       " ('mrna', 8),\n",
       " ('(coleoptera:', 8),\n",
       " ('temperature', 8),\n",
       " ('quantitative', 8),\n",
       " ('plasmodium', 8),\n",
       " ('library', 8),\n",
       " ('competition', 8),\n",
       " ('convolutional', 8),\n",
       " ('online', 8),\n",
       " ('cerevisiae', 8),\n",
       " ('assisted', 8),\n",
       " ('chromosome', 8),\n",
       " ('tightly-coupled', 8),\n",
       " ('gathering', 8),\n",
       " ('tumor', 8),\n",
       " ('zymoseptoria', 8),\n",
       " ('technologies', 8),\n",
       " ('3', 8),\n",
       " ('symbiotic', 8),\n",
       " ('polymorphism', 8),\n",
       " ('nucleic', 8),\n",
       " ('transcriptomics', 8),\n",
       " ('artificial', 8),\n",
       " ('mechanical', 8),\n",
       " ('green', 8),\n",
       " ('key', 8),\n",
       " ('atomic', 8),\n",
       " ('chromosomes', 8),\n",
       " ('walking', 8),\n",
       " ('size', 8),\n",
       " ('distance', 8),\n",
       " ('red', 8),\n",
       " ('degradation', 8),\n",
       " ('recurrent', 8),\n",
       " ('exchange', 8),\n",
       " ('code', 8),\n",
       " ('biases', 8),\n",
       " ('image', 8),\n",
       " ('sites', 8),\n",
       " ('structure-based', 8),\n",
       " ('mediates', 8),\n",
       " ('components', 8),\n",
       " ('trial', 8),\n",
       " ('making', 8),\n",
       " ('liver', 8),\n",
       " ('probing', 8),\n",
       " ('positive', 8),\n",
       " ('muscle', 8),\n",
       " ('bluetooth', 8),\n",
       " ('single-particle', 8),\n",
       " ('fish', 8),\n",
       " ('virus:', 8),\n",
       " ('mobility', 8),\n",
       " ('ribavirin', 8),\n",
       " ('member', 8),\n",
       " ('leadership,', 8),\n",
       " ('harmonia', 8),\n",
       " ('europe', 8),\n",
       " ('contributes', 8),\n",
       " ('bias', 8),\n",
       " ('selective', 8),\n",
       " ('shape', 8),\n",
       " ('ancient', 8),\n",
       " ('radiation', 8),\n",
       " ('specific', 8),\n",
       " ('southern', 8),\n",
       " ('transporter', 8),\n",
       " ('freshwater', 8),\n",
       " ('modern', 8),\n",
       " ('practices', 8),\n",
       " ('strength', 8),\n",
       " ('variable', 8),\n",
       " ('bank', 8),\n",
       " ('solving', 8),\n",
       " ...]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_counter.most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_tokens = [t for t, c in token_counter.most_common() if c >= 100]\n",
    "len(feature_tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3583"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dois_to_cluster = set()\n",
    "doi2cl = {}\n",
    "\n",
    "for i, cluster in clDict.items():\n",
    "    if cluster[\"size\"] >= 50:\n",
    "        dois_to_cluster.update(cluster[\"dois\"])\n",
    "        for doi in cluster[\"dois\"]:\n",
    "            doi2cl[doi] = i\n",
    "        \n",
    "len(dois_to_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "dois_to_cluster = list(dois_to_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "titles_to_cluster = [doi2title[doi] for doi in dois_to_cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "clLabels_for_clustering = [clLabels[doi2index[doi]] for doi in dois_to_cluster]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import HashingVectorizer, CountVectorizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#vectorizer = HashingVectorizer(n_features=2000, binary=True)\n",
    "#X_tokens = vectorizer.fit_transform(titles_to_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "vectorizer = CountVectorizer(max_features=10000, binary=True, stop_words=\"english\")\n",
    "X_tokens = vectorizer.fit_transform(titles_to_cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "clf = RandomForestClassifier(n_estimators=100, verbose=True, n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=8)]: Using backend ThreadingBackend with 8 concurrent workers.\n",
      "[Parallel(n_jobs=8)]: Done  34 tasks      | elapsed:    0.1s\n",
      "[Parallel(n_jobs=8)]: Done 100 out of 100 | elapsed:    0.5s finished\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,\n",
       "                       criterion='gini', max_depth=None, max_features='auto',\n",
       "                       max_leaf_nodes=None, max_samples=None,\n",
       "                       min_impurity_decrease=0.0, min_impurity_split=None,\n",
       "                       min_samples_leaf=1, min_samples_split=2,\n",
       "                       min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=8,\n",
       "                       oob_score=False, random_state=None, verbose=True,\n",
       "                       warm_start=False)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf.fit(X_tokens, clLabels_for_clustering)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "inds = np.argsort(-1*clf.feature_importances_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'analyzer': 'word',\n",
       " 'binary': True,\n",
       " 'decode_error': 'strict',\n",
       " 'dtype': numpy.int64,\n",
       " 'encoding': 'utf-8',\n",
       " 'input': 'content',\n",
       " 'lowercase': True,\n",
       " 'max_df': 1.0,\n",
       " 'max_features': 10000,\n",
       " 'min_df': 1,\n",
       " 'ngram_range': (1, 1),\n",
       " 'preprocessor': None,\n",
       " 'stop_words': 'english',\n",
       " 'strip_accents': None,\n",
       " 'token_pattern': '(?u)\\\\b\\\\w\\\\w+\\\\b',\n",
       " 'tokenizer': None,\n",
       " 'vocabulary': None}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vectorizer.get_params()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'000'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_names = vectorizer.get_feature_names()\n",
    "feature_names[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coronavirus 0.036690505230718005\n",
      "indoor 0.01587723658897665\n",
      "networks 0.014524035675938908\n",
      "localization 0.013545398003637599\n",
      "underwater 0.011155058337060092\n",
      "mers 0.01048791240272321\n",
      "structure 0.010421275707357127\n",
      "cov 0.010168780009076152\n",
      "genome 0.010104428912632128\n",
      "crystal 0.009174085348032148\n",
      "sensor 0.008724810044546715\n",
      "positioning 0.008404052345708514\n",
      "syndrome 0.007584517147880553\n",
      "based 0.007549262900292152\n",
      "sars 0.007318577394272119\n",
      "2019 0.007040544788330334\n",
      "structural 0.006892933902469009\n",
      "performance 0.006600973837533623\n",
      "wireless 0.006578006762560717\n",
      "ncov 0.006480015934453416\n",
      "covid 0.0060168118788440925\n",
      "respiratory 0.005970777729449471\n",
      "public 0.005898304441848631\n",
      "coronaviruses 0.005836212352650954\n",
      "crystallography 0.005286666667056347\n",
      "phylogeny 0.005256992983355685\n",
      "middle 0.005255241076447172\n",
      "virus 0.00525123113138278\n",
      "structures 0.005162421112288714\n",
      "19 0.005022428468777805\n",
      "protein 0.004967760198220739\n",
      "influenza 0.004530390450281038\n",
      "novel 0.004190728719165276\n",
      "hepatitis 0.003928577371041725\n",
      "slam 0.003847722715196237\n",
      "bats 0.003804805269778152\n",
      "antiviral 0.0037732971253509624\n",
      "infection 0.0036753405599633097\n",
      "cell 0.0035873131345424396\n",
      "navigation 0.003523806018731746\n",
      "genomes 0.0035108205296723173\n",
      "binding 0.003509047874322016\n",
      "electron 0.003467243976739642\n",
      "microbial 0.003424456161494352\n",
      "species 0.003272986908042583\n",
      "conflict 0.0032038444377945467\n",
      "china 0.00309078487397692\n",
      "bat 0.003066097185672072\n",
      "east 0.002989867082142802\n",
      "spike 0.0028246167609255425\n",
      "routing 0.0028031445015737626\n",
      "pedestrian 0.0027486543924690344\n",
      "using 0.002716690960629455\n",
      "genomic 0.002645455782585606\n",
      "sequence 0.0026427449483181096\n",
      "evolution 0.002603749574200905\n",
      "genus 0.002568189571798327\n",
      "human 0.002543387963104547\n",
      "accountability 0.002470594449040926\n",
      "acoustic 0.0024515264828304613\n",
      "management 0.0023913987064207052\n",
      "estimation 0.0023759917372325567\n",
      "organizational 0.002368859268990454\n",
      "visual 0.002351336897148654\n",
      "government 0.0023488523871909774\n",
      "energy 0.002342800060147356\n",
      "sequencing 0.002308148563448604\n",
      "refinement 0.002272359737988045\n",
      "phasing 0.0022551430177039453\n",
      "robust 0.0021780177809525873\n",
      "porcine 0.002157189362365174\n",
      "scheme 0.0021116983667711924\n",
      "work 0.00209462880553753\n",
      "inertial 0.002067082976193517\n",
      "data 0.002007068790010666\n",
      "job 0.0020034270506618383\n",
      "macromolecular 0.0019998363823189717\n",
      "sector 0.001994212264581942\n",
      "team 0.001847708416722527\n",
      "emerging 0.0018412694503304513\n",
      "microbiome 0.0018387323494390025\n",
      "diversity 0.0018364837270217283\n",
      "method 0.0018333491560201534\n",
      "wi 0.0018158620384324492\n",
      "tracking 0.0018121654096399592\n",
      "algorithm 0.0018063235500707848\n",
      "virtual 0.0017568900292618162\n",
      "crystallographic 0.0017528859480723988\n",
      "efficient 0.0016932116537463182\n",
      "screening 0.0016802116320076233\n",
      "analysis 0.0016604084620392922\n",
      "cryo 0.0016596110027155635\n",
      "protocol 0.0016593711112464385\n",
      "resistance 0.0016344433114774044\n",
      "ray 0.001617304952020621\n",
      "evolutionary 0.0016071730442221374\n",
      "communication 0.0016048947083261336\n",
      "gene 0.0015942020936856541\n",
      "phylogenetic 0.0015899959566080635\n",
      "molecular 0.0015682583528237194\n"
     ]
    }
   ],
   "source": [
    "for i in inds[:100]:\n",
    "    print(feature_names[i], clf.feature_importances_[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=8)]: Using backend ThreadingBackend with 8 concurrent workers.\n",
      "[Parallel(n_jobs=8)]: Done  34 tasks      | elapsed:    0.0s\n",
      "[Parallel(n_jobs=8)]: Done 100 out of 100 | elapsed:    0.0s finished\n"
     ]
    }
   ],
   "source": [
    "title = \"topology optimization and uncertainty quantification in crashworthiness\"\n",
    "\n",
    "x = vectorizer.transform([title])\n",
    "\n",
    "y = clf.predict_proba(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.01, 0.32, 0.05, 0.03, 0.07, 0.13, 0.14, 0.12, 0.03, 0.01, 0.  ,\n",
       "        0.07, 0.  , 0.  , 0.02]])"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([10, 12, 13,  0,  9, 14,  3,  8,  2,  4, 11,  7,  5,  6,  1],\n",
       "      dtype=int64)"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.argsort(y.flat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 154 is out of bounds for axis 1 with size 15",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-77-f70c04196d3c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0my\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m154\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m: index 154 is out of bounds for axis 1 with size 15"
     ]
    }
   ],
   "source": [
    "y[0,154]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clf.classes_[154]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_session.query(CrossRefWork).filter(CrossRefWork.author_names.ilike(\"%Fiebig%\")).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "title = \"A design methodology for structural and control systems of a prosthetic leg\"\n",
    "\n",
    "x = vectorizer.transform([title])\n",
    "\n",
    "y = clf.predict_proba(x)\n",
    "\n",
    "\n",
    "for i in np.argsort(y.flat)[::-1]:\n",
    "    \n",
    "    proba = y[0,i]\n",
    "    if proba >= 0.01:\n",
    "        print(\"Prob:\", proba)\n",
    "    else:\n",
    "        break\n",
    "    \n",
    "    #i = 154\n",
    "    cl = clf.classes_[i]\n",
    "\n",
    "    print(\"Cluster\", cl)\n",
    "    #print(\"Count:\", count)\n",
    "    if cl in clDict:\n",
    "        vec = clDict[cl][\"vec\"]\n",
    "        print(\"Size:\", clDict[cl][\"size\"])\n",
    "        print(\", \".join([\"{} ({}, {}, {:.2f})\".format(v[0], v[1], v[2], v[3]) for v in vec if v[3] >= 0.1]))\n",
    "        print(X_embedded[cl2index[cl], :])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A Neuro-fuzzy-Based Multi-criteria Risk Evaluation Approach DET\n",
      "A DET DT\n",
      "Neuro PROPN NNP\n",
      "- PUNCT HYPH\n",
      "fuzzy PROPN NNP\n",
      "- PUNCT HYPH\n",
      "Based VERB VBN\n",
      "Multi ADJ JJ\n",
      "- PUNCT HYPH\n",
      "criteria ADJ JJ\n",
      "Risk PROPN NNP\n",
      "Evaluation PROPN NNP\n",
      "Approach PROPN NNP\n",
      "A Case Study DET\n",
      "A DET DT\n",
      "Case PROPN NNP\n",
      "Study NOUN NN\n",
      "Underground Mining PROPN\n",
      "Underground PROPN NNP\n",
      "Mining PROPN NNP\n"
     ]
    }
   ],
   "source": [
    "doc = nlp(\"A Framework for Multi-Criteria Decision Analysis.\")\n",
    "doc = nlp(\"A Neuro-fuzzy-Based Multi-criteria Risk Evaluation Approach: A Case Study of Underground Mining.\")\n",
    "for nc in doc.noun_chunks:\n",
    "    print(nc[0], nc[1:], nc[0].pos_)\n",
    "    for t in nc:\n",
    "        print(t, t.pos_, t.tag_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "4904it [00:10, 453.45it/s]\n"
     ]
    }
   ],
   "source": [
    "docs = list(tqdm(nlp.pipe(titles)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "nc_counts = defaultdict(int)\n",
    "\n",
    "for doc in docs:\n",
    "    for nc in doc.noun_chunks:\n",
    "        if nc[0].pos_ == \"DET\":\n",
    "            nc_counts[nc[1:].text.lower()] += 1\n",
    "        else:\n",
    "            nc_counts[nc.text.lower()] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "term_tuples = sorted(nc_counts.items(), key=itemgetter(1), reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "middle east respiratory syndrome coronavirus 57\n",
      "novel coronavirus 40\n",
      "2019 novel coronavirus 25\n",
      "coronaviruses 14\n",
      "coronavirus disease 11\n",
      "coronavirus 11\n",
      "severe acute respiratory syndrome coronavirus 10\n",
      "human coronavirus 10\n",
      "middle east respiratory syndrome coronavirus infection 10\n",
      "mers coronavirus 7\n",
      "sars coronavirus 6\n",
      "porcine deltacoronavirus 6\n",
      "novel coronavirus disease 5\n",
      "coronavirus infection 4\n",
      "bat coronaviruses 4\n",
      "2019 novel coronavirus disease 4\n",
      "new coronavirus 4\n",
      "betacoronaviruses 4\n",
      "human coronaviruses 4\n",
      "novel coronavirus outbreak 3\n",
      "novel coronavirus infection 3\n",
      "coronavirus spike proteins 3\n",
      "novel human coronavirus 3\n",
      "2019 novel coronavirus outbreak 3\n",
      "coronavirus infections 2\n",
      "novel coronaviruses 2\n",
      "emerging coronaviruses 2\n",
      "wuhan 2019 novel coronavirus 2\n",
      "emerging novel coronavirus 2\n",
      "coronavirus entry 2\n",
      "novel coronavirus infections 2\n",
      "2019-novel coronavirus 2\n",
      "severe acute respiratory syndrome-related coronavirus 2\n",
      "2019 novel coronavirus pneumonia 2\n",
      "sars-related coronaviruses 2\n",
      "china coronavirus 2\n",
      "2019 novel coronavirus infection 2\n",
      "novel coronavirus pneumonia 2\n",
      "zoonotic coronaviruses 2\n",
      "covid-19 coronavirus 2\n",
      "porcine deltacoronaviruses 2\n",
      "middle east respiratory syndrome coronavirus receptor 2\n",
      "coronavirus nonstructural protein 2\n",
      "coronavirus tropism 2\n",
      "middle east respiratory syndrome coronavirus transmission 2\n",
      "porcine deltacoronavirus nsp5 2\n",
      "2019-ncov-like coronavirus 2\n",
      "porcine epidemic diarrhea coronavirus 2\n",
      "middle east respiratory syndrome coronavirus helicase 2\n",
      "new human coronavirus 1\n"
     ]
    }
   ],
   "source": [
    "search_term = \"coronavirus\"\n",
    "\n",
    "i = 0\n",
    "for term, count in term_tuples:\n",
    "    if search_term.lower() in term.lower():\n",
    "        print(term, count)\n",
    "        i += 1\n",
    "    if i >= 50:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics.pairwise import cosine_similarity, euclidean_distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "china 78 438 0.46752793\n",
      "structure 65 1307 0.4327239\n",
      "middle east respiratory syndrome coronavirus 57 102 0.914214\n",
      "evolution 52 587 0.46410134\n",
      "role 50 1064 0.4351763\n",
      "identification 46 427 0.4611907\n",
      "wireless sensor networks 45 445 0.82846636\n",
      "novel coronavirus 40 111 0.9342701\n",
      "impact 39 863 0.4195619\n",
      "analysis 38 2168 0.40766364\n",
      "mers-cov 35 93 0.9228495\n",
      "crystal structure 34 80 0.8414862\n",
      "wuhan 31 46 0.8749646\n",
      "development 30 1151 0.4110555\n",
      "insights 29 222 0.43352178\n",
      "humans 29 56 0.53116655\n",
      "characterization 29 331 0.45069885\n",
      "function 29 829 0.4352656\n",
      "evidence 27 871 0.47702244\n",
      "treatment 26 313 0.4571783\n",
      "2019 novel coronavirus 25 44 0.9311907\n",
      "use 25 1858 0.41544524\n",
      "mechanism 25 429 0.42736593\n",
      "underwater wireless sensor networks 25 36 0.93592477\n",
      "patients 24 318 0.5102235\n",
      "application 23 1406 0.4591425\n",
      "covid-19 22 70 0.8968225\n",
      "detection 22 1878 0.5069753\n",
      "what 22 376 0.4370134\n",
      "bats 22 53 0.82374734\n",
      "emergence 21 80 0.45736533\n",
      "middle east respiratory syndrome 21 136 0.9139225\n",
      "protein 20 608 0.5780601\n",
      "nov. 20 25 0.85902226\n",
      "crystal structures 20 31 0.90335983\n",
      "review 19 1205 0.42017\n",
      "evaluation 19 669 0.43043017\n",
      "structural basis 19 21 0.7753475\n",
      "mers-cov infection 19 20 0.9536989\n",
      "underwater acoustic sensor networks 19 31 0.96559054\n",
      "korea 18 159 0.4728518\n",
      "effects 18 987 0.42590865\n",
      "complex 18 581 0.4208464\n",
      "zika virus 18 38 0.8921446\n",
      "challenges 17 328 0.41810375\n",
      "systematic review 17 246 0.54153293\n",
      "discovery 17 127 0.46943447\n",
      "infection 17 278 0.6263645\n",
      "design 17 985 0.43563077\n",
      "survey 17 584 0.4769149\n",
      "localization 17 478 0.72114307\n",
      "underwater sensor networks 17 28 0.95119023\n",
      "effect 16 2041 0.4115659\n",
      "south korea 16 68 0.49793693\n",
      "saudi arabia 15 27 0.62591195\n",
      "influenza 15 79 0.7539321\n",
      "epidemiology 15 34 0.59753954\n",
      "selection 15 291 0.46204394\n",
      "transmission 15 216 0.50450736\n",
      "case 15 870 0.423723\n",
      "phylogeny 15 38 0.8908124\n",
      "arrival 15 49 0.66225743\n",
      "comparison 15 387 0.41563287\n",
      "relationship 15 475 0.46970174\n",
      "mers 14 459 0.47437164\n",
      "coronaviruses 14 60 0.8810072\n",
      "presence 14 105 0.41970477\n",
      "influence 14 525 0.42131764\n",
      "study 14 1503 0.40536523\n",
      "mapping 14 298 0.4889069\n",
      "estimation 13 632 0.5479605\n",
      "implications 13 296 0.42015705\n",
      "mechanisms 13 217 0.43447623\n",
      "diagnosis 13 261 0.5504739\n",
      "viruses 13 142 0.7444183\n",
      "indoor positioning 13 82 0.9250953\n",
      "applications 13 858 0.48016524\n",
      "overview 13 136 0.42226514\n",
      "description 13 56 0.4584213\n",
      "performance 13 1020 0.43916017\n",
      "work 13 3719 0.47237593\n",
      "pathogenesis 12 29 0.63336885\n",
      "lessons 12 97 0.4477453\n",
      "sars 12 101 0.79083174\n",
      "risk 12 508 0.43636602\n",
      "origin 12 159 0.4323945\n",
      "prediction 12 333 0.5010711\n",
      "control 12 1002 0.41981807\n",
      "indoor localization 12 98 0.9177615\n",
      "research 12 918 0.45227897\n",
      "association 12 243 0.4672164\n",
      "ebola virus 12 32 0.8081646\n",
      "coronavirus disease 11 26 0.9433944\n",
      "coronavirus 11 482 0.8554134\n",
      "outbreak 11 107 0.74134445\n",
      "sars-cov-2 11 26 0.9181806\n",
      "update 11 93 0.458431\n",
      "new insights 11 27 0.50419015\n",
      "type 11 582 0.418082\n",
      "data 11 1997 0.48019278\n",
      "diversity 11 272 0.5010834\n",
      "meta-analysis 11 182 0.6035912\n",
      "wsns 11 61 0.848093\n",
      "structures 11 449 0.4540377\n",
      "taxonomy 11 50 0.5211181\n",
      "genes 11 184 0.5834213\n",
      "accountability 11 57 0.6619785\n",
      "severe acute respiratory syndrome coronavirus 10 16 0.8508302\n",
      "united states 10 152 0.49359554\n",
      "host 10 211 0.5094266\n",
      "replication 10 102 0.52940094\n",
      "progress 10 199 0.42808175\n",
      "effectiveness 10 149 0.4356724\n",
      "who 10 240 0.4283887\n",
      "management 10 640 0.47765455\n",
      "they 10 83 0.4615618\n",
      "tool 10 298 0.42505243\n",
      "dromedary camels 10 10 0.96862036\n",
      "hereditary angioedema 10 10 0.99517614\n",
      "human coronavirus 10 35 0.89530116\n",
      "middle east respiratory syndrome coronavirus infection 10 15 0.9515793\n",
      "acute respiratory distress syndrome 10 15 0.93482393\n",
      "tools 10 108 0.44491172\n",
      "validation 10 184 0.45656225\n",
      "wireless body area networks 10 22 0.9311263\n",
      "pneumonia 9 57 0.6744686\n",
      "death 9 81 0.47096628\n",
      "isolation 9 49 0.471945\n",
      "age 9 4498 0.41178784\n",
      "expression 9 210 0.4921416\n",
      "accuracy 9 185 0.5073933\n",
      "machine learning 9 210 0.5724027\n",
      "regulation 9 239 0.44166967\n",
      "opportunities 9 133 0.41662192\n",
      "activation 9 116 0.4842284\n",
      "genomics 9 69 0.71554536\n",
      "escherichia coli 9 66 0.644004\n",
      "conservation 9 49 0.48511317\n",
      "performance management 9 21 0.8629223\n",
      "recent advances 8 99 0.49203092\n",
      "clinical characteristics 8 12 0.7469636\n",
      "understanding 8 383 0.44219327\n",
      "evolutionary history 8 9 0.8688786\n",
      "virus 8 866 0.74166536\n",
      "italy 8 80 0.458287\n",
      "interaction 8 510 0.42345703\n",
      "inhibition 8 72 0.5374154\n",
      "inhibitors 8 101 0.64347106\n",
      "complete genome sequence 8 9 0.9362481\n",
      "smartphone sensors 8 24 0.7719648\n",
      "introduction 8 183 0.4474144\n",
      "genetics 8 59 0.5637133\n",
      "investigation 8 239 0.40646836\n",
      "republic 8 44 0.54241085\n",
      "plants 8 77 0.49355733\n",
      "biogeography 8 12 0.79866385\n",
      "harmonia axyridis 8 9 0.9515703\n",
      "diversification 8 45 0.5016634\n",
      "mycobacterium tuberculosis 8 35 0.79064155\n",
      "bacteria 8 298 0.6304709\n",
      "molecular replacement 8 11 0.97040015\n",
      "smartphones 8 60 0.6578182\n",
      "pregnancy 7 41 0.6274314\n",
      "characteristics 7 238 0.40972376\n",
      "adults 7 178 0.5214263\n",
      "potential 7 387 0.41775373\n",
      "cells 7 430 0.53780776\n",
      "virtual screening 7 24 0.8341677\n",
      "prevalence 7 56 0.49456635\n",
      "recognition 7 624 0.6246206\n",
      "mers coronavirus 7 14 0.922261\n",
      "healthcare workers 7 9 0.8029966\n",
      "spike protein 7 32 0.9096098\n",
      "importance 7 92 0.43233874\n",
      "platelet 7 26 0.53572315\n",
      "vaccines 7 26 0.7695483\n",
      "efficacy 7 147 0.4980635\n",
      "drug discovery 7 20 0.6784291\n",
      "draft genome sequence 7 12 0.9253423\n",
      "strains 7 44 0.6019304\n",
      "proteins 7 140 0.57097703\n",
      "disease 7 651 0.5227336\n",
      "perspectives 7 231 0.42808452\n",
      "critically ill patients 7 9 0.80494493\n",
      "virulence 7 38 0.71484494\n",
      "ebola 7 63 0.7428912\n",
      "lack 7 246 0.472403\n",
      "factors 7 290 0.43232232\n",
      "response 7 586 0.4125761\n",
      "phylogeography 7 7 0.9084555\n",
      "ebola virus disease 7 11 0.8445225\n",
      "activity 7 560 0.47748432\n",
      "navigation 7 244 0.7197193\n",
      "outcomes 7 246 0.48626664\n",
      "time 7 1467 0.4351784\n",
      "inertial sensors 7 54 0.83095187\n",
      "case study 7 281 0.43082467\n",
      "europe 7 473 0.51403373\n",
      "macromolecular crystallography 7 14 0.9263938\n",
      "processing 6 298 0.46604884\n",
      "cell entry 6 6 0.94930255\n",
      "sars‐cov‐2 6 10 0.93840224\n",
      "binding 6 143 0.5857718\n",
      "sars coronavirus 6 11 0.87082\n",
      "human 6 1178 0.44925326\n",
      "recent developments 6 23 0.46875224\n",
      "antibodies 6 37 0.61730325\n",
      "metagenomes 6 9 0.88414377\n",
      "salmonella enterica 6 20 0.799576\n",
      "dynamics 6 417 0.42383027\n",
      "biology 6 216 0.5673833\n",
      "target localization 6 19 0.7868471\n",
      "web server 6 12 0.7425453\n",
      "determinants 6 158 0.49213603\n",
      "resistance 6 208 0.43496898\n",
      "schizophrenia 6 15 0.9037027\n",
      "mutation 6 112 0.58371395\n",
      "west africa 6 16 0.62250704\n",
      "porcine deltacoronavirus 6 16 0.9764104\n",
      "order 6 507 0.4051098\n",
      "advances 6 174 0.46947363\n",
      "spain 6 66 0.49458358\n",
      "comparative analysis 6 51 0.44102395\n",
      "recovery 6 107 0.42086166\n",
      "methods 6 418 0.4304532\n",
      "ribavirin 6 8 0.8132652\n",
      "representation 6 280 0.4424749\n",
      "combination 6 117 0.44710746\n",
      "integration 6 315 0.42966664\n",
      "patterns 6 270 0.42327023\n",
      "solution 6 572 0.44075003\n",
      "structural studies 6 9 0.78473276\n",
      "structural insights 6 8 0.82705796\n",
      "comparative genomics 6 7 0.8435341\n",
      "genetic diversity 6 12 0.7879885\n",
      "structure determination 6 24 0.8924375\n",
      "changes 6 210 0.4242262\n",
      "tracking 6 408 0.6147816\n",
      "refinement 6 43 0.69114906\n",
      "iot 6 747 0.52715665\n",
      "burnout 6 23 0.79044014\n",
      "performance analysis 6 65 0.60234904\n",
      "doa estimation 6 46 0.96535826\n",
      "(covid‐19 5 6 0.9769502\n",
      "covid‐19 5 15 0.9564402\n",
      "(covid-19 5 23 0.9151771\n",
      "clinical features 5 9 0.66305846\n",
      "antibody-dependent enhancement 5 5 0.8114058\n",
      "inactivation 5 9 0.7515886\n",
      "species 5 239 0.5822516\n",
      "hong kong 5 43 0.48535883\n",
      "children 5 423 0.54375654\n",
      "novel coronavirus disease 5 10 0.9456089\n",
      "ultraviolet c light 5 6 0.9830744\n",
      "internet 5 316 0.6504806\n",
      "screening 5 105 0.47915068\n",
      "safety 5 121 0.42241633\n",
      "immunity 5 70 0.6000067\n",
      "adaptation 5 124 0.43597806\n",
      "cancer 5 531 0.605698\n",
      "germany 5 67 0.48996335\n",
      "growth 5 365 0.46875462\n",
      "genome 5 272 0.7341544\n",
      "imu 5 812 0.4380829\n",
      "mhc class 5 9 0.8293392\n",
      "survival 5 69 0.46884158\n",
      "(nematoda 5 7 0.924724\n",
      "protein-protein interactions 5 10 0.767925\n",
      "zymoseptoria tritici 5 8 0.96544814\n",
      "whole-genome sequencing 5 10 0.8421334\n",
      "mortality 5 56 0.54015875\n",
      "life 5 573 0.420886\n",
      "sofosbuvir 5 8 0.9189286\n",
      "sources 5 208 0.41564372\n",
      "rna 5 1391 0.4261888\n",
      "vaccine development 5 5 0.93814826\n",
      "eukaryotes 5 12 0.67679816\n",
      "differences 5 204 0.47656146\n",
      "resources 5 93 0.43422788\n",
      "phylogenetic analysis 5 7 0.84820044\n",
      "assembly 5 123 0.47664282\n",
      "hepatitis e virus 5 14 0.96194786\n",
      "fusion 5 464 0.5339021\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vivo 5 121 0.48815575\n",
      "pathogenicity 5 15 0.64785475\n",
      "strategies 5 334 0.41209695\n",
      "toll-like receptor 5 8 0.89239883\n",
      "production 5 272 0.41367826\n",
      "sequencing 5 105 0.65894574\n",
      "generation 5 426 0.40977913\n",
      "members 5 77 0.44581845\n",
      "politics 5 351 0.6096154\n",
      "cell 5 1213 0.5118366\n",
      "enhancement 5 147 0.47152707\n",
      "link 5 348 0.41421077\n",
      "interactions 5 274 0.43361083\n",
      "consequences 5 101 0.48234138\n",
      "genome evolution 5 8 0.8131703\n",
      "indoor environments 5 22 0.74052656\n",
      "current status 5 26 0.42952046\n",
      "rodents 5 8 0.67003214\n",
      "assessment 5 505 0.4375784\n",
      "mice 5 99 0.536321\n",
      "brazil 5 56 0.48588082\n",
      "speciation 5 14 0.59848607\n",
      "example 5 62 0.43938148\n",
      "complexes 5 79 0.52672064\n",
      "genus 5 41 0.69988555\n",
      "job satisfaction 5 28 0.7775471\n",
      "public sector 5 35 0.72242725\n",
      "supreme audit institutions 5 5 0.9715944\n",
      "performance information 5 15 0.9044599\n",
      "underwater wsns 5 7 0.98770297\n",
      "pedestrian navigation 5 25 0.9176244\n",
      "pedestrians 5 13 0.8047252\n"
     ]
    }
   ],
   "source": [
    "term_mean_sims = defaultdict(float)\n",
    "\n",
    "for term, count in term_tuples:\n",
    "    if len(term) <= 2:\n",
    "        continue\n",
    "    if count <= 4:\n",
    "        break\n",
    "    term_dois = set()\n",
    "    term_vecs = []\n",
    "    for doi, title in doi2title.items():\n",
    "        if term.lower() in title.lower():\n",
    "            if doi in doi2vec:\n",
    "                term_dois.add(doi)\n",
    "                term_vecs.append(doi2vec[doi].reshape(300))\n",
    "    #term_X = np.zeros()\n",
    "    #if\n",
    "    sims = cosine_similarity(term_vecs)\n",
    "    print(term, count, len(term_dois), np.mean(sims))\n",
    "    term_mean_sims[term] = np.mean(sims)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13, '(covid-19) outbreak', 4)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(term_vecs), term, count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('middle east respiratory syndrome coronavirus', 57, 0.914214),\n",
       " ('wireless sensor networks', 45, 0.82846636),\n",
       " ('novel coronavirus', 40, 0.9342701),\n",
       " ('mers-cov', 35, 0.9228495),\n",
       " ('crystal structure', 34, 0.8414862),\n",
       " ('wuhan', 31, 0.8749646),\n",
       " ('2019 novel coronavirus', 25, 0.9311907),\n",
       " ('underwater wireless sensor networks', 25, 0.93592477),\n",
       " ('covid-19', 22, 0.8968225),\n",
       " ('bats', 22, 0.82374734),\n",
       " ('middle east respiratory syndrome', 21, 0.9139225),\n",
       " ('nov.', 20, 0.85902226),\n",
       " ('crystal structures', 20, 0.90335983),\n",
       " ('structural basis', 19, 0.7753475),\n",
       " ('mers-cov infection', 19, 0.9536989),\n",
       " ('underwater acoustic sensor networks', 19, 0.96559054),\n",
       " ('zika virus', 18, 0.8921446),\n",
       " ('localization', 17, 0.72114307),\n",
       " ('underwater sensor networks', 17, 0.95119023),\n",
       " ('influenza', 15, 0.7539321),\n",
       " ('phylogeny', 15, 0.8908124),\n",
       " ('coronaviruses', 14, 0.8810072),\n",
       " ('viruses', 13, 0.7444183),\n",
       " ('indoor positioning', 13, 0.9250953),\n",
       " ('sars', 12, 0.79083174),\n",
       " ('indoor localization', 12, 0.9177615),\n",
       " ('ebola virus', 12, 0.8081646),\n",
       " ('coronavirus disease', 11, 0.9433944),\n",
       " ('coronavirus', 11, 0.8554134),\n",
       " ('outbreak', 11, 0.74134445),\n",
       " ('sars-cov-2', 11, 0.9181806),\n",
       " ('wsns', 11, 0.848093),\n",
       " ('severe acute respiratory syndrome coronavirus', 10, 0.8508302),\n",
       " ('dromedary camels', 10, 0.96862036),\n",
       " ('hereditary angioedema', 10, 0.99517614),\n",
       " ('human coronavirus', 10, 0.89530116),\n",
       " ('middle east respiratory syndrome coronavirus infection', 10, 0.9515793),\n",
       " ('acute respiratory distress syndrome', 10, 0.93482393),\n",
       " ('wireless body area networks', 10, 0.9311263),\n",
       " ('genomics', 9, 0.71554536),\n",
       " ('performance management', 9, 0.8629223),\n",
       " ('clinical characteristics', 8, 0.7469636),\n",
       " ('evolutionary history', 8, 0.8688786),\n",
       " ('virus', 8, 0.74166536),\n",
       " ('complete genome sequence', 8, 0.9362481),\n",
       " ('smartphone sensors', 8, 0.7719648),\n",
       " ('biogeography', 8, 0.79866385),\n",
       " ('harmonia axyridis', 8, 0.9515703),\n",
       " ('mycobacterium tuberculosis', 8, 0.79064155),\n",
       " ('molecular replacement', 8, 0.97040015),\n",
       " ('virtual screening', 7, 0.8341677),\n",
       " ('mers coronavirus', 7, 0.922261),\n",
       " ('healthcare workers', 7, 0.8029966),\n",
       " ('spike protein', 7, 0.9096098),\n",
       " ('vaccines', 7, 0.7695483),\n",
       " ('draft genome sequence', 7, 0.9253423),\n",
       " ('critically ill patients', 7, 0.80494493),\n",
       " ('virulence', 7, 0.71484494),\n",
       " ('ebola', 7, 0.7428912),\n",
       " ('phylogeography', 7, 0.9084555),\n",
       " ('ebola virus disease', 7, 0.8445225),\n",
       " ('navigation', 7, 0.7197193),\n",
       " ('inertial sensors', 7, 0.83095187),\n",
       " ('macromolecular crystallography', 7, 0.9263938),\n",
       " ('cell entry', 6, 0.94930255),\n",
       " ('sars‐cov‐2', 6, 0.93840224),\n",
       " ('sars coronavirus', 6, 0.87082),\n",
       " ('metagenomes', 6, 0.88414377),\n",
       " ('salmonella enterica', 6, 0.799576),\n",
       " ('target localization', 6, 0.7868471),\n",
       " ('web server', 6, 0.7425453),\n",
       " ('schizophrenia', 6, 0.9037027),\n",
       " ('porcine deltacoronavirus', 6, 0.9764104),\n",
       " ('ribavirin', 6, 0.8132652),\n",
       " ('structural studies', 6, 0.78473276),\n",
       " ('structural insights', 6, 0.82705796),\n",
       " ('comparative genomics', 6, 0.8435341),\n",
       " ('genetic diversity', 6, 0.7879885),\n",
       " ('structure determination', 6, 0.8924375),\n",
       " ('burnout', 6, 0.79044014),\n",
       " ('doa estimation', 6, 0.96535826),\n",
       " ('(covid‐19', 5, 0.9769502),\n",
       " ('covid‐19', 5, 0.9564402),\n",
       " ('(covid-19', 5, 0.9151771),\n",
       " ('antibody-dependent enhancement', 5, 0.8114058),\n",
       " ('inactivation', 5, 0.7515886),\n",
       " ('novel coronavirus disease', 5, 0.9456089),\n",
       " ('ultraviolet c light', 5, 0.9830744),\n",
       " ('genome', 5, 0.7341544),\n",
       " ('mhc class', 5, 0.8293392),\n",
       " ('(nematoda', 5, 0.924724),\n",
       " ('protein-protein interactions', 5, 0.767925),\n",
       " ('zymoseptoria tritici', 5, 0.96544814),\n",
       " ('whole-genome sequencing', 5, 0.8421334),\n",
       " ('sofosbuvir', 5, 0.9189286),\n",
       " ('vaccine development', 5, 0.93814826),\n",
       " ('phylogenetic analysis', 5, 0.84820044),\n",
       " ('hepatitis e virus', 5, 0.96194786),\n",
       " ('toll-like receptor', 5, 0.89239883),\n",
       " ('genome evolution', 5, 0.8131703),\n",
       " ('indoor environments', 5, 0.74052656),\n",
       " ('job satisfaction', 5, 0.7775471),\n",
       " ('public sector', 5, 0.72242725),\n",
       " ('supreme audit institutions', 5, 0.9715944),\n",
       " ('performance information', 5, 0.9044599),\n",
       " ('underwater wsns', 5, 0.98770297),\n",
       " ('pedestrian navigation', 5, 0.9176244),\n",
       " ('pedestrians', 5, 0.8047252)]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(term, nc_counts[term], msim) for term, msim in term_mean_sims.items() if msim >= 0.7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "61"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([(term, nc_counts[term], msim) for term, msim in term_mean_sims.items() if msim >= 0.5]),\n",
    "len([(term, nc_counts[term], msim) for term, msim in term_mean_sims.items() if msim >= 0.5 and nc_counts[term] >= 10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "middle east respiratory syndrome coronavirus 0.914214 57 102\n",
      "1\n",
      "[102]\n",
      "\n",
      "wireless sensor networks 0.82846636 45 445\n",
      "8\n",
      "[435, 3]\n",
      "\n",
      "novel coronavirus 0.9342701 40 111\n",
      "1\n",
      "[111]\n",
      "\n",
      "mers-cov 0.9228495 35 93\n",
      "1\n",
      "[93]\n",
      "\n",
      "crystal structure 0.8414862 34 80\n",
      "6\n",
      "[75]\n",
      "\n",
      "wuhan 0.8749646 31 46\n",
      "2\n",
      "[44]\n",
      "\n",
      "humans 0.53116655 29 56\n",
      "21\n",
      "[23, 5, 4, 3]\n",
      "\n",
      "2019 novel coronavirus 0.9311907 25 44\n",
      "1\n",
      "[44]\n",
      "\n",
      "underwater wireless sensor networks 0.93592477 25 36\n",
      "2\n",
      "[35]\n",
      "\n",
      "patients 0.5102235 24 318\n",
      "49\n",
      "[63, 45, 38, 25, 25, 19, 11, 7, 7, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "covid-19 0.8968225 22 70\n",
      "3\n",
      "[68]\n",
      "\n",
      "detection 0.5069753 22 1878\n",
      "84\n",
      "[341, 308, 224, 108, 101, 92, 76, 53, 44, 39, 33, 26, 25, 23, 23, 21, 20, 20, 20, 18, 18, 17, 15, 13, 11, 11, 10, 10, 10, 9, 8, 8, 7, 6, 6, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "bats 0.82374734 22 53\n",
      "6\n",
      "[48]\n",
      "\n",
      "middle east respiratory syndrome 0.9139225 21 136\n",
      "1\n",
      "[136]\n",
      "\n",
      "protein 0.5780601 20 608\n",
      "48\n",
      "[168, 110, 58, 40, 40, 24, 17, 16, 12, 12, 8, 8, 7, 6, 6, 6, 5, 5, 4, 4, 4, 4, 4, 4, 4, 3, 3]\n",
      "\n",
      "nov. 0.85902226 20 25\n",
      "1\n",
      "[25]\n",
      "\n",
      "crystal structures 0.90335983 20 31\n",
      "2\n",
      "[30]\n",
      "\n",
      "structural basis 0.7753475 19 21\n",
      "3\n",
      "[19]\n",
      "\n",
      "mers-cov infection 0.9536989 19 20\n",
      "1\n",
      "[20]\n",
      "\n",
      "underwater acoustic sensor networks 0.96559054 19 31\n",
      "1\n",
      "[31]\n",
      "\n",
      "zika virus 0.8921446 18 38\n",
      "1\n",
      "[38]\n",
      "\n",
      "systematic review 0.54153293 17 246\n",
      "44\n",
      "[55, 32, 21, 15, 13, 10, 9, 9, 9, 6, 5, 4, 4, 4, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "infection 0.6263645 17 278\n",
      "22\n",
      "[166, 29, 24, 24, 5, 4, 4, 3]\n",
      "\n",
      "localization 0.72114307 17 478\n",
      "30\n",
      "[375, 27, 17, 7, 6, 5, 4, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "underwater sensor networks 0.95119023 17 28\n",
      "1\n",
      "[28]\n",
      "\n",
      "saudi arabia 0.62591195 15 27\n",
      "10\n",
      "[16]\n",
      "\n",
      "influenza 0.7539321 15 79\n",
      "6\n",
      "[67, 4, 3]\n",
      "\n",
      "epidemiology 0.59753954 15 34\n",
      "10\n",
      "[19, 4, 3]\n",
      "\n",
      "transmission 0.50450736 15 216\n",
      "29\n",
      "[73, 62, 16, 6, 6, 5, 5, 4, 4, 4, 3, 3]\n",
      "\n",
      "phylogeny 0.8908124 15 38\n",
      "1\n",
      "[38]\n",
      "\n",
      "arrival 0.66225743 15 49\n",
      "9\n",
      "[33, 6, 3]\n",
      "\n",
      "coronaviruses 0.8810072 14 60\n",
      "1\n",
      "[60]\n",
      "\n",
      "estimation 0.5479605 13 632\n",
      "50\n",
      "[111, 105, 101, 47, 46, 29, 28, 26, 18, 13, 12, 6, 6, 6, 6, 6, 5, 5, 4, 4, 4, 3, 3, 3, 3]\n",
      "\n",
      "diagnosis 0.5504739 13 261\n",
      "43\n",
      "[124, 14, 12, 11, 10, 9, 7, 5, 5, 5, 5, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "viruses 0.7444183 13 142\n",
      "11\n",
      "[120, 13]\n",
      "\n",
      "indoor positioning 0.9250953 13 82\n",
      "1\n",
      "[82]\n",
      "\n",
      "pathogenesis 0.63336885 12 29\n",
      "6\n",
      "[14, 7, 5]\n",
      "\n",
      "sars 0.79083174 12 101\n",
      "5\n",
      "[91, 6]\n",
      "\n",
      "prediction 0.5010711 12 333\n",
      "62\n",
      "[40, 33, 25, 25, 22, 22, 14, 9, 9, 8, 8, 7, 7, 6, 6, 6, 6, 4, 4, 4, 4, 3, 3, 3, 3]\n",
      "\n",
      "indoor localization 0.9177615 12 98\n",
      "1\n",
      "[98]\n",
      "\n",
      "ebola virus 0.8081646 12 32\n",
      "3\n",
      "[30]\n",
      "\n",
      "coronavirus disease 0.9433944 11 26\n",
      "1\n",
      "[26]\n",
      "\n",
      "coronavirus 0.8554134 11 482\n",
      "1\n",
      "[482]\n",
      "\n",
      "outbreak 0.74134445 11 107\n",
      "9\n",
      "[88, 7, 3]\n",
      "\n",
      "sars-cov-2 0.9181806 11 26\n",
      "2\n",
      "[25]\n",
      "\n",
      "new insights 0.50419015 11 27\n",
      "15\n",
      "[8, 3, 3]\n",
      "\n",
      "diversity 0.5010834 11 272\n",
      "46\n",
      "[100, 22, 20, 19, 11, 10, 9, 6, 5, 5, 5, 4, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "meta-analysis 0.6035912 11 182\n",
      "26\n",
      "[53, 49, 28, 8, 7, 4, 3, 3, 3]\n",
      "\n",
      "wsns 0.848093 11 61\n",
      "2\n",
      "[60]\n",
      "\n",
      "taxonomy 0.5211181 11 50\n",
      "19\n",
      "[16, 13, 3]\n",
      "\n",
      "genes 0.5834213 11 184\n",
      "28\n",
      "[53, 28, 18, 15, 11, 10, 7, 7, 4, 3, 3, 3, 3]\n",
      "\n",
      "accountability 0.6619785 11 57\n",
      "9\n",
      "[36, 9, 4, 3]\n",
      "\n",
      "severe acute respiratory syndrome coronavirus 0.8508302 10 16\n",
      "1\n",
      "[16]\n",
      "\n",
      "host 0.5094266 10 211\n",
      "36\n",
      "[68, 37, 11, 11, 9, 9, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "replication 0.52940094 10 102\n",
      "20\n",
      "[34, 30, 9, 4, 4, 3, 3]\n",
      "\n",
      "dromedary camels 0.96862036 10 10\n",
      "1\n",
      "[10]\n",
      "\n",
      "hereditary angioedema 0.99517614 10 10\n",
      "1\n",
      "[10]\n",
      "\n",
      "human coronavirus 0.89530116 10 35\n",
      "1\n",
      "[35]\n",
      "\n",
      "middle east respiratory syndrome coronavirus infection 0.9515793 10 15\n",
      "1\n",
      "[15]\n",
      "\n",
      "acute respiratory distress syndrome 0.93482393 10 15\n",
      "1\n",
      "[15]\n",
      "\n",
      "wireless body area networks 0.9311263 10 22\n",
      "1\n",
      "[22]\n",
      "\n",
      "pneumonia 0.6744686 9 57\n",
      "9\n",
      "[32, 12, 5, 3]\n",
      "\n",
      "accuracy 0.5073933 9 185\n",
      "45\n",
      "[32, 31, 21, 12, 8, 7, 6, 6, 5, 4, 3, 3, 3, 3]\n",
      "\n",
      "machine learning 0.5724027 9 210\n",
      "40\n",
      "[34, 25, 18, 17, 16, 15, 11, 8, 6, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "genomics 0.71554536 9 69\n",
      "9\n",
      "[48, 12]\n",
      "\n",
      "escherichia coli 0.644004 9 66\n",
      "6\n",
      "[38, 16, 7, 3]\n",
      "\n",
      "performance management 0.8629223 9 21\n",
      "2\n",
      "[20]\n",
      "\n",
      "clinical characteristics 0.7469636 8 12\n",
      "4\n",
      "[9]\n",
      "\n",
      "evolutionary history 0.8688786 8 9\n",
      "1\n",
      "[9]\n",
      "\n",
      "virus 0.74166536 8 866\n",
      "26\n",
      "[705, 53, 39, 12, 10, 6, 6, 6, 5, 3, 3]\n",
      "\n",
      "inhibition 0.5374154 8 72\n",
      "20\n",
      "[15, 10, 9, 6, 6, 6, 3, 3]\n",
      "\n",
      "inhibitors 0.64347106 8 101\n",
      "11\n",
      "[35, 23, 14, 11, 6, 4]\n",
      "\n",
      "complete genome sequence 0.9362481 8 9\n",
      "1\n",
      "[9]\n",
      "\n",
      "smartphone sensors 0.7719648 8 24\n",
      "3\n",
      "[21]\n",
      "\n",
      "genetics 0.5637133 8 59\n",
      "16\n",
      "[13, 12, 10, 5, 3, 3, 3]\n",
      "\n",
      "republic 0.54241085 8 44\n",
      "14\n",
      "[12, 11, 7]\n",
      "\n",
      "biogeography 0.79866385 8 12\n",
      "2\n",
      "[11]\n",
      "\n",
      "harmonia axyridis 0.9515703 8 9\n",
      "1\n",
      "[9]\n",
      "\n",
      "diversification 0.5016634 8 45\n",
      "13\n",
      "[15, 8, 6, 3, 3]\n",
      "\n",
      "mycobacterium tuberculosis 0.79064155 8 35\n",
      "2\n",
      "[34]\n",
      "\n",
      "bacteria 0.6304709 8 298\n",
      "26\n",
      "[168, 23, 20, 18, 14, 9, 7, 7, 6, 4, 3]\n",
      "\n",
      "molecular replacement 0.97040015 8 11\n",
      "1\n",
      "[11]\n",
      "\n",
      "smartphones 0.6578182 8 60\n",
      "13\n",
      "[35, 7, 5, 3]\n",
      "\n",
      "pregnancy 0.6274314 7 41\n",
      "8\n",
      "[25, 6, 3]\n",
      "\n",
      "adults 0.5214263 7 178\n",
      "30\n",
      "[38, 28, 24, 13, 8, 7, 5, 5, 5, 5, 4, 4, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "cells 0.53780776 7 430\n",
      "44\n",
      "[61, 47, 42, 40, 23, 23, 21, 21, 18, 16, 12, 11, 9, 7, 7, 7, 6, 5, 5, 4, 4, 4, 4, 4, 3, 3, 3]\n",
      "\n",
      "virtual screening 0.8341677 7 24\n",
      "2\n",
      "[22]\n",
      "\n",
      "recognition 0.6246206 7 624\n",
      "43\n",
      "[298, 171, 18, 13, 13, 9, 9, 8, 6, 6, 6, 5, 5, 4, 4, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "mers coronavirus 0.922261 7 14\n",
      "1\n",
      "[14]\n",
      "\n",
      "healthcare workers 0.8029966 7 9\n",
      "2\n",
      "[8]\n",
      "\n",
      "spike protein 0.9096098 7 32\n",
      "1\n",
      "[32]\n",
      "\n",
      "platelet 0.53572315 7 26\n",
      "7\n",
      "[8, 7, 6]\n",
      "\n",
      "vaccines 0.7695483 7 26\n",
      "4\n",
      "[22]\n",
      "\n",
      "drug discovery 0.6784291 7 20\n",
      "7\n",
      "[10, 4]\n",
      "\n",
      "draft genome sequence 0.9253423 7 12\n",
      "1\n",
      "[12]\n",
      "\n",
      "strains 0.6019304 7 44\n",
      "15\n",
      "[25, 3, 3]\n",
      "\n",
      "proteins 0.57097703 7 140\n",
      "26\n",
      "[35, 15, 13, 12, 10, 9, 5, 5, 4, 4, 4, 4, 3]\n",
      "\n",
      "disease 0.5227336 7 651\n",
      "57\n",
      "[133, 73, 71, 62, 57, 42, 23, 19, 15, 14, 12, 11, 9, 8, 8, 7, 5, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "critically ill patients 0.80494493 7 9\n",
      "2\n",
      "[8]\n",
      "\n",
      "virulence 0.71484494 7 38\n",
      "4\n",
      "[24, 5, 5, 4]\n",
      "\n",
      "ebola 0.7428912 7 63\n",
      "8\n",
      "[52, 3, 3]\n",
      "\n",
      "phylogeography 0.9084555 7 7\n",
      "1\n",
      "[7]\n",
      "\n",
      "ebola virus disease 0.8445225 7 11\n",
      "1\n",
      "[11]\n",
      "\n",
      "navigation 0.7197193 7 244\n",
      "20\n",
      "[198, 15, 7, 3, 3]\n",
      "\n",
      "inertial sensors 0.83095187 7 54\n",
      "1\n",
      "[54]\n",
      "\n",
      "europe 0.51403373 7 473\n",
      "62\n",
      "[86, 82, 50, 34, 27, 21, 13, 12, 12, 10, 9, 8, 7, 6, 5, 5, 5, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "macromolecular crystallography 0.9263938 7 14\n",
      "1\n",
      "[14]\n",
      "\n",
      "cell entry 0.94930255 6 6\n",
      "1\n",
      "[6]\n",
      "\n",
      "sars‐cov‐2 0.93840224 6 10\n",
      "1\n",
      "[10]\n",
      "\n",
      "binding 0.5857718 6 143\n",
      "26\n",
      "[51, 29, 10, 6, 6, 6, 3, 3, 3, 3]\n",
      "\n",
      "sars coronavirus 0.87082 6 11\n",
      "1\n",
      "[11]\n",
      "\n",
      "antibodies 0.61730325 6 37\n",
      "9\n",
      "[23, 3, 3]\n",
      "\n",
      "metagenomes 0.88414377 6 9\n",
      "1\n",
      "[9]\n",
      "\n",
      "salmonella enterica 0.799576 6 20\n",
      "2\n",
      "[19]\n",
      "\n",
      "biology 0.5673833 6 216\n",
      "30\n",
      "[118, 11, 10, 10, 7, 6, 6, 5, 5, 4, 3, 3, 3]\n",
      "\n",
      "target localization 0.7868471 6 19\n",
      "3\n",
      "[14, 4]\n",
      "\n",
      "web server 0.7425453 6 12\n",
      "3\n",
      "[7, 4]\n",
      "\n",
      "schizophrenia 0.9037027 6 15\n",
      "2\n",
      "[14]\n",
      "\n",
      "mutation 0.58371395 6 112\n",
      "21\n",
      "[44, 16, 13, 8, 5, 5, 4]\n",
      "\n",
      "west africa 0.62250704 6 16\n",
      "6\n",
      "[9]\n",
      "\n",
      "porcine deltacoronavirus 0.9764104 6 16\n",
      "1\n",
      "[16]\n",
      "\n",
      "ribavirin 0.8132652 6 8\n",
      "1\n",
      "[8]\n",
      "\n",
      "structural studies 0.78473276 6 9\n",
      "2\n",
      "[8]\n",
      "\n",
      "structural insights 0.82705796 6 8\n",
      "2\n",
      "[7]\n",
      "\n",
      "comparative genomics 0.8435341 6 7\n",
      "2\n",
      "[6]\n",
      "\n",
      "genetic diversity 0.7879885 6 12\n",
      "2\n",
      "[9, 3]\n",
      "\n",
      "structure determination 0.8924375 6 24\n",
      "1\n",
      "[24]\n",
      "\n",
      "tracking 0.6147816 6 408\n",
      "36\n",
      "[243, 47, 23, 11, 9, 6, 6, 6, 5, 4, 4, 4, 4, 3, 3, 3, 3]\n",
      "\n",
      "refinement 0.69114906 6 43\n",
      "9\n",
      "[30, 5]\n",
      "\n",
      "iot 0.52715665 6 747\n",
      "69\n",
      "[302, 221, 22, 14, 12, 10, 9, 9, 8, 8, 7, 5, 5, 5, 5, 5, 5, 5, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "burnout 0.79044014 6 23\n",
      "3\n",
      "[21]\n",
      "\n",
      "performance analysis 0.60234904 6 65\n",
      "14\n",
      "[25, 19, 4, 4, 3]\n",
      "\n",
      "doa estimation 0.96535826 6 46\n",
      "1\n",
      "[46]\n",
      "\n",
      "(covid‐19 0.9769502 5 6\n",
      "1\n",
      "[6]\n",
      "\n",
      "covid‐19 0.9564402 5 15\n",
      "1\n",
      "[15]\n",
      "\n",
      "(covid-19 0.9151771 5 23\n",
      "2\n",
      "[22]\n",
      "\n",
      "clinical features 0.66305846 5 9\n",
      "4\n",
      "[6]\n",
      "\n",
      "antibody-dependent enhancement 0.8114058 5 5\n",
      "2\n",
      "[3]\n",
      "\n",
      "inactivation 0.7515886 5 9\n",
      "3\n",
      "[6]\n",
      "\n",
      "species 0.5822516 5 239\n",
      "30\n",
      "[135, 17, 9, 8, 7, 6, 6, 5, 5, 5, 4, 4, 4, 3, 3]\n",
      "\n",
      "children 0.54375654 5 423\n",
      "48\n",
      "[163, 43, 20, 18, 13, 12, 11, 10, 10, 10, 9, 8, 8, 8, 6, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3]\n",
      "\n",
      "novel coronavirus disease 0.9456089 5 10\n",
      "1\n",
      "[10]\n",
      "\n",
      "ultraviolet c light 0.9830744 5 6\n",
      "1\n",
      "[6]\n",
      "\n",
      "internet 0.6504806 5 316\n",
      "30\n",
      "[233, 16, 6, 6, 6, 6, 4, 3, 3, 3]\n",
      "\n",
      "immunity 0.6000067 5 70\n",
      "15\n",
      "[14, 12, 11, 9, 8, 3, 3, 3]\n",
      "\n",
      "cancer 0.605698 5 531\n",
      "50\n",
      "[196, 72, 39, 24, 21, 21, 17, 17, 14, 9, 8, 8, 6, 6, 5, 5, 5, 5, 5, 4, 4, 4, 3, 3]\n",
      "\n",
      "genome 0.7341544 5 272\n",
      "13\n",
      "[188, 32, 26, 11, 5]\n",
      "\n",
      "mhc class 0.8293392 5 9\n",
      "2\n",
      "[6, 3]\n",
      "\n",
      "(nematoda 0.924724 5 7\n",
      "1\n",
      "[7]\n",
      "\n",
      "protein-protein interactions 0.767925 5 10\n",
      "2\n",
      "[9]\n",
      "\n",
      "zymoseptoria tritici 0.96544814 5 8\n",
      "1\n",
      "[8]\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "whole-genome sequencing 0.8421334 5 10\n",
      "1\n",
      "[10]\n",
      "\n",
      "mortality 0.54015875 5 56\n",
      "16\n",
      "[10, 10, 8, 7, 4, 4]\n",
      "\n",
      "sofosbuvir 0.9189286 5 8\n",
      "1\n",
      "[8]\n",
      "\n",
      "vaccine development 0.93814826 5 5\n",
      "1\n",
      "[5]\n",
      "\n",
      "eukaryotes 0.67679816 5 12\n",
      "5\n",
      "[5, 3]\n",
      "\n",
      "phylogenetic analysis 0.84820044 5 7\n",
      "1\n",
      "[7]\n",
      "\n",
      "hepatitis e virus 0.96194786 5 14\n",
      "1\n",
      "[14]\n",
      "\n",
      "fusion 0.5339021 5 464\n",
      "52\n",
      "[95, 85, 46, 34, 34, 20, 11, 10, 10, 8, 8, 7, 6, 6, 6, 5, 5, 5, 5, 4, 3, 3, 3, 3, 3]\n",
      "\n",
      "pathogenicity 0.64785475 5 15\n",
      "6\n",
      "[7, 3]\n",
      "\n",
      "toll-like receptor 0.89239883 5 8\n",
      "2\n",
      "[7]\n",
      "\n",
      "sequencing 0.65894574 5 105\n",
      "18\n",
      "[70, 5, 4, 4, 4, 3, 3]\n",
      "\n",
      "politics 0.6096154 5 351\n",
      "25\n",
      "[134, 41, 38, 33, 29, 19, 12, 6, 6, 5, 4, 4, 4, 3]\n",
      "\n",
      "cell 0.5118366 5 1213\n",
      "91\n",
      "[155, 85, 79, 67, 65, 63, 61, 59, 58, 33, 30, 26, 25, 25, 21, 21, 21, 20, 20, 20, 19, 12, 10, 9, 9, 8, 8, 8, 7, 7, 7, 7, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 3, 3, 3]\n",
      "\n",
      "genome evolution 0.8131703 5 8\n",
      "2\n",
      "[6]\n",
      "\n",
      "indoor environments 0.74052656 5 22\n",
      "3\n",
      "[18]\n",
      "\n",
      "rodents 0.67003214 5 8\n",
      "4\n",
      "[5]\n",
      "\n",
      "mice 0.536321 5 99\n",
      "21\n",
      "[27, 14, 9, 7, 7, 6, 5, 5, 4, 3]\n",
      "\n",
      "speciation 0.59848607 5 14\n",
      "4\n",
      "[9]\n",
      "\n",
      "complexes 0.52672064 5 79\n",
      "15\n",
      "[14, 13, 12, 11, 9, 7, 3, 3]\n",
      "\n",
      "genus 0.69988555 5 41\n",
      "4\n",
      "[29, 8]\n",
      "\n",
      "job satisfaction 0.7775471 5 28\n",
      "4\n",
      "[25]\n",
      "\n",
      "public sector 0.72242725 5 35\n",
      "6\n",
      "[22, 8]\n",
      "\n",
      "supreme audit institutions 0.9715944 5 5\n",
      "1\n",
      "[5]\n",
      "\n",
      "performance information 0.9044599 5 15\n",
      "1\n",
      "[15]\n",
      "\n",
      "underwater wsns 0.98770297 5 7\n",
      "1\n",
      "[7]\n",
      "\n",
      "pedestrian navigation 0.9176244 5 25\n",
      "1\n",
      "[25]\n",
      "\n",
      "pedestrians 0.8047252 5 13\n",
      "2\n",
      "[12]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "term_plot_tuples = []\n",
    "\n",
    "for term, msim in term_mean_sims.items():\n",
    "    if msim >= 0.5:\n",
    "        term_dois = set()\n",
    "        term_vecs = []\n",
    "        for doi, title in doi2title.items():\n",
    "            if term.lower() in title.lower():\n",
    "                if doi in doi2vec:\n",
    "                    term_dois.add(doi)\n",
    "                    term_vecs.append(doi2vec[doi])\n",
    "        \n",
    "        print(term, msim, nc_counts[term], len(term_dois))\n",
    "        \n",
    "        ac2 = AgglomerativeClustering(n_clusters=None, affinity=\"cosine\", linkage=\"average\", distance_threshold=0.3)\n",
    "        kwLabels = ac2.fit_predict(term_vecs)\n",
    "        kwLabels_unique = sorted(np.unique(kwLabels))        \n",
    "        \n",
    "        print(len(kwLabels_unique))\n",
    "        print([c for _, c in Counter(kwLabels).most_common() if c >= 3])\n",
    "        print()\n",
    "        \n",
    "        for l, c in Counter(kwLabels).most_common():\n",
    "            if c <= 2:\n",
    "                break\n",
    "            if float(c) / len(term_dois) >= 0.1:\n",
    "                mean_vec = np.zeros((1,300))\n",
    "                for i, vec in enumerate(term_vecs):\n",
    "                    if kwLabels[i] == l:\n",
    "                        mean_vec += vec\n",
    "                mean_vec = mean_vec / c\n",
    "                \n",
    "                #print((term, c, mean_vec))\n",
    "                term_plot_tuples.append((term, c, mean_vec))\n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "269"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(term_plot_tuples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "term_plot_tuples = sorted(term_plot_tuples, key=lambda t: nc_counts[t[0]], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('middle east respiratory syndrome coronavirus',\n",
       " 102,\n",
       " array([[-5.31590190e-01, -1.86610506e-01,  2.64814234e-01,\n",
       "          1.99480345e-01, -5.11750235e-02, -9.65396915e-02,\n",
       "          4.49238814e-01,  1.44566788e-01, -3.53449752e-01,\n",
       "         -1.67309835e-01, -5.43028040e-01,  5.31383475e-01,\n",
       "         -5.24092466e-02,  2.67159874e-01, -1.50493329e-01,\n",
       "          2.93785569e-02,  2.11959457e-01, -1.84202470e-01,\n",
       "         -1.79091106e-01, -4.50877383e-01, -2.07967684e-01,\n",
       "          6.60292821e-01, -3.56962250e-01, -5.41648844e-01,\n",
       "         -1.30957247e-01,  1.86663550e-01,  9.91178523e-02,\n",
       "          2.67951868e-01, -3.69076188e-01,  3.42541845e-01,\n",
       "         -1.06311148e-01, -1.80245596e-01,  3.82492825e-01,\n",
       "         -1.03822742e-01, -1.50341193e-01,  2.58981259e-02,\n",
       "         -2.07655426e-02,  3.18744321e-01, -1.58346365e-01,\n",
       "         -9.17587405e-02,  2.01180057e-01, -6.91961544e-01,\n",
       "         -3.09582624e-01,  2.11756939e-01, -2.70671195e-01,\n",
       "         -2.36620074e-01, -3.43644415e-01,  3.20804801e-01,\n",
       "         -4.41987264e-01,  2.05573321e-02, -2.84188004e-01,\n",
       "         -3.00501268e-01, -1.47829711e-01,  2.51428887e-01,\n",
       "          1.61377924e-01,  2.89280386e-01, -4.59889839e-02,\n",
       "          4.53660332e-01,  2.19161912e-02, -1.80845127e-01,\n",
       "         -2.48722915e-01, -2.44417070e-01, -1.45786955e-01,\n",
       "         -2.69328378e-01, -2.35912809e-01,  3.63771267e-01,\n",
       "          3.06370462e-01,  2.34867726e-01,  1.10692039e-01,\n",
       "         -4.18261770e-01,  1.07488333e-01, -1.30301236e-01,\n",
       "          4.67280683e-01, -1.38601828e-01,  3.79583374e-02,\n",
       "         -4.85094362e-01, -2.57187148e-01,  3.63446149e-01,\n",
       "         -1.81134907e-01,  7.65746716e-02, -2.83627029e-01,\n",
       "          9.91432946e-02,  4.33627272e-01,  2.03967394e-01,\n",
       "          2.85484705e-01, -1.44817165e-03, -1.08083618e-01,\n",
       "          2.66050413e-01, -4.98826476e-02,  3.19648620e-01,\n",
       "         -4.58925722e-01,  6.93314672e-01, -3.20833136e-03,\n",
       "          5.25045734e-01, -4.17849258e-01, -2.29631723e-01,\n",
       "         -8.00114485e-01, -2.66474234e-01, -3.45603799e-02,\n",
       "         -3.24996971e-01,  1.20776936e-01, -3.43715329e-01,\n",
       "          4.61255258e-02,  7.81085775e-02,  5.48643019e-02,\n",
       "         -4.68347024e-01,  4.00203787e-01,  1.27098160e-01,\n",
       "         -1.94656384e-02, -6.60615004e-01,  6.27129668e-02,\n",
       "          4.16997018e-01, -7.49996682e-02, -3.77736756e-02,\n",
       "          1.65030934e-01, -1.77904674e-01, -1.38780186e-01,\n",
       "          8.88197896e-02, -2.35177970e-02, -3.54413174e-01,\n",
       "          5.36340531e-01, -2.28896018e-01,  1.39123336e-01,\n",
       "         -5.72843399e-02, -6.15595263e-01,  2.47818291e-01,\n",
       "         -1.07643933e-01,  2.73337172e-01, -2.25641526e-01,\n",
       "          3.95706955e-01,  1.07793236e-01, -4.00393167e-01,\n",
       "         -6.33776309e-02, -1.12431883e-01,  4.35964044e-01,\n",
       "         -2.89512263e-01, -1.19254826e-01, -4.97718504e-01,\n",
       "          3.81445697e-01,  2.90650162e-01, -4.36302921e-01,\n",
       "          2.32285365e-01,  5.03214284e-01, -3.14635448e-01,\n",
       "         -4.69909669e-01,  1.43684952e-01,  6.09214692e-01,\n",
       "         -5.56657980e-01,  1.42697171e-01, -1.28678655e-02,\n",
       "          9.40557225e-02, -2.61548542e-01,  2.39226581e-01,\n",
       "          4.86000361e-02, -4.30637587e-02, -3.23712906e-01,\n",
       "          7.24159189e-02,  7.17516588e-02, -2.79543505e-01,\n",
       "          3.19895153e-01, -5.23838300e-01, -9.59623415e-02,\n",
       "          1.85508754e-01, -1.42303733e-01, -7.29289212e-02,\n",
       "          4.91329209e-01,  3.30089065e-01, -3.39049663e-01,\n",
       "          4.02184699e-01, -5.23681998e-01,  3.57043006e-01,\n",
       "         -1.53051296e-01, -6.25056475e-01, -2.24544150e-01,\n",
       "         -3.31184029e-01,  2.22527048e-01, -2.49182099e-01,\n",
       "          5.91791655e-02, -2.56961433e-01, -4.71796188e-01,\n",
       "          4.19143203e-01,  2.58339679e-01, -1.40962555e-01,\n",
       "          1.08829348e-02,  3.57928339e-01, -1.22023744e-01,\n",
       "         -4.43411655e-01,  1.18315794e-01, -5.32236458e-02,\n",
       "         -2.14371102e-02,  1.83194496e-01, -8.67456940e-01,\n",
       "         -1.37640300e-01,  2.50389582e-01,  9.45280233e-02,\n",
       "          6.20593480e-01, -8.95469201e-03, -4.87477939e-01,\n",
       "          4.96500873e-01, -6.68610659e-02, -8.79002615e-02,\n",
       "          4.06036258e-01,  2.14426661e-01,  3.40545752e-01,\n",
       "         -4.81229884e-01,  2.66836987e-01, -3.89585122e-01,\n",
       "         -5.33070594e-01, -8.81317995e-02,  2.77489764e-03,\n",
       "         -9.28025430e-02, -2.56530240e-01, -1.58067028e-01,\n",
       "          4.90275371e-01,  3.20974277e-02,  3.75158822e-01,\n",
       "         -8.76219962e-02,  1.49456663e-01, -5.67434535e-01,\n",
       "          1.22815467e-01, -1.01997873e-01, -1.58489330e-01,\n",
       "         -1.66800701e-01, -5.83762293e-01, -3.08961356e-01,\n",
       "          5.48336675e-01, -5.48641047e-01, -6.78995848e-02,\n",
       "         -1.91589441e-01, -1.28798190e-01,  3.25203365e-01,\n",
       "         -5.46236785e-01, -1.15194783e-01, -1.92014187e-01,\n",
       "          4.37906674e-01, -7.44371374e-02, -6.42874309e-01,\n",
       "         -2.72796704e-01, -1.88820845e-01,  2.30537304e-01,\n",
       "          2.10172807e-01,  1.42399766e-01,  3.09677535e-01,\n",
       "         -3.75442049e-01, -4.39311960e-01, -3.71982820e-01,\n",
       "          2.43580786e-01, -2.20540589e-02,  1.00390881e-01,\n",
       "          3.48205826e-03,  2.29086115e-01, -2.17138412e-01,\n",
       "          2.08181393e-01, -1.62792652e-01,  6.44339365e-02,\n",
       "          4.38377257e-01, -4.57846087e-01,  7.39488834e-01,\n",
       "          1.03987689e-01,  9.04748898e-02,  4.01123269e-04,\n",
       "         -1.80752345e-02,  1.61658699e-01,  9.81477430e-01,\n",
       "          2.18009338e-02, -4.69357742e-01,  5.29029497e-01,\n",
       "          8.20462959e-02,  2.39082939e-01, -2.98942047e-01,\n",
       "          2.39360880e-01, -4.34489287e-01, -5.23509771e-02,\n",
       "          2.65322128e-01,  9.55376632e-02,  5.88341424e-02,\n",
       "          3.41066050e-01, -7.15241465e-02,  1.08394069e-01,\n",
       "         -3.02277030e-01, -1.78213610e-01, -9.36083020e-01,\n",
       "         -1.79145602e-01,  6.76999476e-02,  7.08058911e-02,\n",
       "         -3.55963979e-01, -5.03329318e-02, -3.49205707e-02,\n",
       "         -4.35300729e-01, -5.97575210e-01,  3.52866163e-02,\n",
       "          3.36379153e-01, -3.53256586e-01, -1.95297842e-02,\n",
       "         -6.12677019e-01, -1.74955113e-02,  1.87624381e-01,\n",
       "         -1.28233833e-01, -3.96789786e-01, -1.40336904e-01]]))"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_plot_tuples[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_kw = np.zeros((len(term_plot_tuples), 300))\n",
    "for i, (_, _, vec) in enumerate(term_plot_tuples):\n",
    "    X_kw[i, :] = vec\n",
    "\n",
    "X_kw_emb = TSNE(n_components=2, metric=\"cosine\").fit_transform(X_kw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "C_kw_emb = TSNE(n_components=3, metric=\"cosine\").fit_transform(X_kw)\n",
    "cmin_kw = np.min(C_kw_emb)\n",
    "cmax_kw = np.max(C_kw_emb)\n",
    "colors_kw = (C_kw_emb - cmin_kw) / (cmax_kw - cmin_kw) * 256\n",
    "colors_kw = np.floor(colors_kw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_kw_sims = cosine_similarity(X_kw)\n",
    "X_kw_emb_dists = euclidean_distances(X_kw_emb)\n",
    "X_kw_emb_dists_x = euclidean_distances(X_kw_emb[:,0].reshape(-1, 1))\n",
    "X_kw_emb_dists_y = euclidean_distances(X_kw_emb[:,0].reshape(-1, 1))\n",
    "X_kw_emb_dists = np.sqrt(np.square(X_kw_emb_dists_y) + np.square(0.1*X_kw_emb_dists_x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_kw_sims[100,:].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>kw</th>\n",
       "      <th>size</th>\n",
       "      <th>nc_count</th>\n",
       "      <th>circle_size</th>\n",
       "      <th>rgb</th>\n",
       "      <th>max_sim</th>\n",
       "      <th>min_emb_dist</th>\n",
       "      <th>no_kw_vecs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-112.622</td>\n",
       "      <td>-51.2819</td>\n",
       "      <td>middle east respiratory syndrome coronavirus</td>\n",
       "      <td>102</td>\n",
       "      <td>57.0</td>\n",
       "      <td>50.497525</td>\n",
       "      <td>rgb(39, 88, 78)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36.789352</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23.2157</td>\n",
       "      <td>68.8196</td>\n",
       "      <td>wireless sensor networks</td>\n",
       "      <td>435</td>\n",
       "      <td>45.0</td>\n",
       "      <td>104.283268</td>\n",
       "      <td>rgb(146, 182, 221)</td>\n",
       "      <td>0.606799</td>\n",
       "      <td>10.445533</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33.8798</td>\n",
       "      <td>109.127</td>\n",
       "      <td>novel coronavirus</td>\n",
       "      <td>111</td>\n",
       "      <td>40.0</td>\n",
       "      <td>52.678269</td>\n",
       "      <td>rgb(155, 214, 100)</td>\n",
       "      <td>0.879144</td>\n",
       "      <td>5.463013</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>76.0156</td>\n",
       "      <td>-17.5924</td>\n",
       "      <td>mers-cov</td>\n",
       "      <td>93</td>\n",
       "      <td>35.0</td>\n",
       "      <td>48.218254</td>\n",
       "      <td>rgb(188, 114, 118)</td>\n",
       "      <td>0.998738</td>\n",
       "      <td>0.427948</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>78.325</td>\n",
       "      <td>-121.269</td>\n",
       "      <td>crystal structure</td>\n",
       "      <td>75</td>\n",
       "      <td>34.0</td>\n",
       "      <td>43.301270</td>\n",
       "      <td>rgb(190, 32, 127)</td>\n",
       "      <td>0.735524</td>\n",
       "      <td>4.407660</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>91.7094</td>\n",
       "      <td>-75.8707</td>\n",
       "      <td>wuhan</td>\n",
       "      <td>44</td>\n",
       "      <td>31.0</td>\n",
       "      <td>33.166248</td>\n",
       "      <td>rgb(201, 68, 100)</td>\n",
       "      <td>0.998520</td>\n",
       "      <td>0.359852</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>94.1134</td>\n",
       "      <td>33.2064</td>\n",
       "      <td>humans</td>\n",
       "      <td>23</td>\n",
       "      <td>29.0</td>\n",
       "      <td>23.979158</td>\n",
       "      <td>rgb(202, 154, 107)</td>\n",
       "      <td>0.968882</td>\n",
       "      <td>0.830428</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>31.6672</td>\n",
       "      <td>133.303</td>\n",
       "      <td>2019 novel coronavirus</td>\n",
       "      <td>44</td>\n",
       "      <td>25.0</td>\n",
       "      <td>33.166248</td>\n",
       "      <td>rgb(153, 233, 120)</td>\n",
       "      <td>0.999674</td>\n",
       "      <td>0.305976</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-57.9994</td>\n",
       "      <td>69.3501</td>\n",
       "      <td>underwater wireless sensor networks</td>\n",
       "      <td>35</td>\n",
       "      <td>25.0</td>\n",
       "      <td>29.580399</td>\n",
       "      <td>rgb(82, 183, 34)</td>\n",
       "      <td>0.924585</td>\n",
       "      <td>0.171446</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-3.09047</td>\n",
       "      <td>-111.81</td>\n",
       "      <td>patients</td>\n",
       "      <td>63</td>\n",
       "      <td>24.0</td>\n",
       "      <td>39.686270</td>\n",
       "      <td>rgb(126, 40, 184)</td>\n",
       "      <td>0.578510</td>\n",
       "      <td>4.303370</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9.79806</td>\n",
       "      <td>-54.2324</td>\n",
       "      <td>patients</td>\n",
       "      <td>45</td>\n",
       "      <td>24.0</td>\n",
       "      <td>33.541020</td>\n",
       "      <td>rgb(136, 85, 174)</td>\n",
       "      <td>0.692708</td>\n",
       "      <td>0.843866</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>53.0442</td>\n",
       "      <td>81.0132</td>\n",
       "      <td>patients</td>\n",
       "      <td>38</td>\n",
       "      <td>24.0</td>\n",
       "      <td>30.822070</td>\n",
       "      <td>rgb(170, 192, 111)</td>\n",
       "      <td>0.982306</td>\n",
       "      <td>0.852884</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>9.78586</td>\n",
       "      <td>115.466</td>\n",
       "      <td>covid-19</td>\n",
       "      <td>68</td>\n",
       "      <td>22.0</td>\n",
       "      <td>41.231056</td>\n",
       "      <td>rgb(136, 219, 111)</td>\n",
       "      <td>0.998771</td>\n",
       "      <td>0.135117</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-20.7609</td>\n",
       "      <td>-14.6939</td>\n",
       "      <td>detection</td>\n",
       "      <td>341</td>\n",
       "      <td>22.0</td>\n",
       "      <td>92.330927</td>\n",
       "      <td>rgb(112, 116, 43)</td>\n",
       "      <td>0.689144</td>\n",
       "      <td>0.057089</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-128.715</td>\n",
       "      <td>-26.0274</td>\n",
       "      <td>detection</td>\n",
       "      <td>308</td>\n",
       "      <td>22.0</td>\n",
       "      <td>87.749644</td>\n",
       "      <td>rgb(26, 108, 185)</td>\n",
       "      <td>0.591937</td>\n",
       "      <td>1.472012</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>49.864</td>\n",
       "      <td>38.2858</td>\n",
       "      <td>detection</td>\n",
       "      <td>224</td>\n",
       "      <td>22.0</td>\n",
       "      <td>74.833148</td>\n",
       "      <td>rgb(167, 158, 195)</td>\n",
       "      <td>0.950529</td>\n",
       "      <td>0.945062</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>39.7554</td>\n",
       "      <td>-10.4566</td>\n",
       "      <td>bats</td>\n",
       "      <td>48</td>\n",
       "      <td>22.0</td>\n",
       "      <td>34.641016</td>\n",
       "      <td>rgb(159, 120, 54)</td>\n",
       "      <td>0.943157</td>\n",
       "      <td>1.300673</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>43.8203</td>\n",
       "      <td>-21.7732</td>\n",
       "      <td>middle east respiratory syndrome</td>\n",
       "      <td>136</td>\n",
       "      <td>21.0</td>\n",
       "      <td>58.309519</td>\n",
       "      <td>rgb(163, 111, 122)</td>\n",
       "      <td>0.999662</td>\n",
       "      <td>0.337104</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-106.72</td>\n",
       "      <td>38.3675</td>\n",
       "      <td>protein</td>\n",
       "      <td>168</td>\n",
       "      <td>20.0</td>\n",
       "      <td>64.807407</td>\n",
       "      <td>rgb(44, 158, 104)</td>\n",
       "      <td>0.982954</td>\n",
       "      <td>1.120590</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.0017</td>\n",
       "      <td>48.9669</td>\n",
       "      <td>protein</td>\n",
       "      <td>110</td>\n",
       "      <td>20.0</td>\n",
       "      <td>52.440442</td>\n",
       "      <td>rgb(129, 167, 100)</td>\n",
       "      <td>0.959406</td>\n",
       "      <td>0.245621</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>-14.9209</td>\n",
       "      <td>-89.9211</td>\n",
       "      <td>nov.</td>\n",
       "      <td>25</td>\n",
       "      <td>20.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>rgb(116, 57, 38)</td>\n",
       "      <td>0.779219</td>\n",
       "      <td>12.334796</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-112.803</td>\n",
       "      <td>69.9097</td>\n",
       "      <td>crystal structures</td>\n",
       "      <td>30</td>\n",
       "      <td>20.0</td>\n",
       "      <td>27.386128</td>\n",
       "      <td>rgb(39, 183, 73)</td>\n",
       "      <td>0.996659</td>\n",
       "      <td>0.137975</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>-88.0328</td>\n",
       "      <td>68.683</td>\n",
       "      <td>structural basis</td>\n",
       "      <td>19</td>\n",
       "      <td>19.0</td>\n",
       "      <td>21.794495</td>\n",
       "      <td>rgb(59, 182, 109)</td>\n",
       "      <td>0.988462</td>\n",
       "      <td>0.316845</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>72.9304</td>\n",
       "      <td>9.63638</td>\n",
       "      <td>mers-cov infection</td>\n",
       "      <td>20</td>\n",
       "      <td>19.0</td>\n",
       "      <td>22.360680</td>\n",
       "      <td>rgb(186, 136, 118)</td>\n",
       "      <td>0.998422</td>\n",
       "      <td>0.004294</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>-25.097</td>\n",
       "      <td>78.5998</td>\n",
       "      <td>underwater acoustic sensor networks</td>\n",
       "      <td>31</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27.838822</td>\n",
       "      <td>rgb(108, 190, 216)</td>\n",
       "      <td>0.996747</td>\n",
       "      <td>0.030495</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>-33.0915</td>\n",
       "      <td>56.0053</td>\n",
       "      <td>zika virus</td>\n",
       "      <td>38</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.822070</td>\n",
       "      <td>rgb(102, 172, 86)</td>\n",
       "      <td>0.853774</td>\n",
       "      <td>1.963809</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>79.2218</td>\n",
       "      <td>61.9953</td>\n",
       "      <td>systematic review</td>\n",
       "      <td>55</td>\n",
       "      <td>17.0</td>\n",
       "      <td>37.080992</td>\n",
       "      <td>rgb(191, 177, 25)</td>\n",
       "      <td>0.911842</td>\n",
       "      <td>0.349941</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>8.42145</td>\n",
       "      <td>-32.1892</td>\n",
       "      <td>systematic review</td>\n",
       "      <td>32</td>\n",
       "      <td>17.0</td>\n",
       "      <td>28.284271</td>\n",
       "      <td>rgb(135, 103, 154)</td>\n",
       "      <td>0.982906</td>\n",
       "      <td>0.386113</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>80.3787</td>\n",
       "      <td>-34.5894</td>\n",
       "      <td>infection</td>\n",
       "      <td>166</td>\n",
       "      <td>17.0</td>\n",
       "      <td>64.420494</td>\n",
       "      <td>rgb(192, 101, 93)</td>\n",
       "      <td>0.985767</td>\n",
       "      <td>0.558301</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.641837</td>\n",
       "      <td>-41.9627</td>\n",
       "      <td>infection</td>\n",
       "      <td>29</td>\n",
       "      <td>17.0</td>\n",
       "      <td>26.925824</td>\n",
       "      <td>rgb(129, 95, 118)</td>\n",
       "      <td>0.695506</td>\n",
       "      <td>2.434479</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>-0.374083</td>\n",
       "      <td>5.12656</td>\n",
       "      <td>fusion</td>\n",
       "      <td>85</td>\n",
       "      <td>5.0</td>\n",
       "      <td>46.097722</td>\n",
       "      <td>rgb(128, 132, 206)</td>\n",
       "      <td>0.989754</td>\n",
       "      <td>0.106339</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>41.6152</td>\n",
       "      <td>83.3581</td>\n",
       "      <td>pathogenicity</td>\n",
       "      <td>7</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.228757</td>\n",
       "      <td>rgb(161, 194, 177)</td>\n",
       "      <td>0.988270</td>\n",
       "      <td>0.172044</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>41.8266</td>\n",
       "      <td>-102.988</td>\n",
       "      <td>pathogenicity</td>\n",
       "      <td>3</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.660254</td>\n",
       "      <td>rgb(161, 47, 90)</td>\n",
       "      <td>0.871837</td>\n",
       "      <td>0.019240</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>-30.5015</td>\n",
       "      <td>35.7539</td>\n",
       "      <td>toll-like receptor</td>\n",
       "      <td>7</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.228757</td>\n",
       "      <td>rgb(104, 156, 117)</td>\n",
       "      <td>0.915672</td>\n",
       "      <td>0.005941</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>17.1402</td>\n",
       "      <td>114.562</td>\n",
       "      <td>sequencing</td>\n",
       "      <td>70</td>\n",
       "      <td>5.0</td>\n",
       "      <td>41.833001</td>\n",
       "      <td>rgb(142, 219, 175)</td>\n",
       "      <td>0.974527</td>\n",
       "      <td>0.022586</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>38.3714</td>\n",
       "      <td>117.529</td>\n",
       "      <td>politics</td>\n",
       "      <td>134</td>\n",
       "      <td>5.0</td>\n",
       "      <td>57.879185</td>\n",
       "      <td>rgb(158, 221, 54)</td>\n",
       "      <td>0.854494</td>\n",
       "      <td>0.014411</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>95.0391</td>\n",
       "      <td>-26.5558</td>\n",
       "      <td>politics</td>\n",
       "      <td>41</td>\n",
       "      <td>5.0</td>\n",
       "      <td>32.015621</td>\n",
       "      <td>rgb(203, 107, 220)</td>\n",
       "      <td>0.962211</td>\n",
       "      <td>0.008920</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>75.6478</td>\n",
       "      <td>-13.0713</td>\n",
       "      <td>politics</td>\n",
       "      <td>38</td>\n",
       "      <td>5.0</td>\n",
       "      <td>30.822070</td>\n",
       "      <td>rgb(188, 118, 207)</td>\n",
       "      <td>0.942851</td>\n",
       "      <td>0.083256</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>-68.7191</td>\n",
       "      <td>-11.8057</td>\n",
       "      <td>cell</td>\n",
       "      <td>155</td>\n",
       "      <td>5.0</td>\n",
       "      <td>62.249498</td>\n",
       "      <td>rgb(74, 119, 158)</td>\n",
       "      <td>0.990388</td>\n",
       "      <td>0.034504</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>-42.36</td>\n",
       "      <td>-96.2237</td>\n",
       "      <td>genome evolution</td>\n",
       "      <td>6</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.247449</td>\n",
       "      <td>rgb(95, 52, 108)</td>\n",
       "      <td>0.978395</td>\n",
       "      <td>0.021899</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>-8.16832</td>\n",
       "      <td>103.133</td>\n",
       "      <td>indoor environments</td>\n",
       "      <td>18</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.213203</td>\n",
       "      <td>rgb(122, 210, 46)</td>\n",
       "      <td>0.964984</td>\n",
       "      <td>0.165660</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>21.7304</td>\n",
       "      <td>-27.8912</td>\n",
       "      <td>rodents</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.180340</td>\n",
       "      <td>rgb(145, 106, 46)</td>\n",
       "      <td>0.970087</td>\n",
       "      <td>0.015492</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>3.24872</td>\n",
       "      <td>-47.3212</td>\n",
       "      <td>mice</td>\n",
       "      <td>27</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.980762</td>\n",
       "      <td>rgb(131, 91, 96)</td>\n",
       "      <td>0.991966</td>\n",
       "      <td>0.028480</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>-98.2037</td>\n",
       "      <td>-63.3631</td>\n",
       "      <td>mice</td>\n",
       "      <td>14</td>\n",
       "      <td>5.0</td>\n",
       "      <td>18.708287</td>\n",
       "      <td>rgb(51, 78, 182)</td>\n",
       "      <td>0.524699</td>\n",
       "      <td>0.443099</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>54.6482</td>\n",
       "      <td>-75.1355</td>\n",
       "      <td>speciation</td>\n",
       "      <td>9</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>rgb(171, 69, 71)</td>\n",
       "      <td>0.991925</td>\n",
       "      <td>0.046796</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>-108.31</td>\n",
       "      <td>6.33701</td>\n",
       "      <td>complexes</td>\n",
       "      <td>14</td>\n",
       "      <td>5.0</td>\n",
       "      <td>18.708287</td>\n",
       "      <td>rgb(43, 133, 185)</td>\n",
       "      <td>0.676159</td>\n",
       "      <td>0.029222</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>-27.8831</td>\n",
       "      <td>101.549</td>\n",
       "      <td>complexes</td>\n",
       "      <td>13</td>\n",
       "      <td>5.0</td>\n",
       "      <td>18.027756</td>\n",
       "      <td>rgb(106, 208, 173)</td>\n",
       "      <td>0.633787</td>\n",
       "      <td>1.573069</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>-148.422</td>\n",
       "      <td>28.5221</td>\n",
       "      <td>complexes</td>\n",
       "      <td>12</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.320508</td>\n",
       "      <td>rgb(11, 151, 110)</td>\n",
       "      <td>0.770216</td>\n",
       "      <td>0.063596</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>-107.598</td>\n",
       "      <td>58.961</td>\n",
       "      <td>complexes</td>\n",
       "      <td>11</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.583124</td>\n",
       "      <td>rgb(43, 175, 121)</td>\n",
       "      <td>0.992044</td>\n",
       "      <td>0.019288</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>-106.098</td>\n",
       "      <td>-1.11005</td>\n",
       "      <td>complexes</td>\n",
       "      <td>9</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>rgb(44, 127, 165)</td>\n",
       "      <td>0.805343</td>\n",
       "      <td>0.015992</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>41.6955</td>\n",
       "      <td>-76.4265</td>\n",
       "      <td>genus</td>\n",
       "      <td>29</td>\n",
       "      <td>5.0</td>\n",
       "      <td>26.925824</td>\n",
       "      <td>rgb(161, 68, 47)</td>\n",
       "      <td>0.991352</td>\n",
       "      <td>0.017252</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>-43.0595</td>\n",
       "      <td>103.344</td>\n",
       "      <td>genus</td>\n",
       "      <td>8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.142136</td>\n",
       "      <td>rgb(94, 210, 182)</td>\n",
       "      <td>0.919267</td>\n",
       "      <td>0.001639</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>60.793</td>\n",
       "      <td>17.9592</td>\n",
       "      <td>job satisfaction</td>\n",
       "      <td>25</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>rgb(176, 142, 163)</td>\n",
       "      <td>0.959600</td>\n",
       "      <td>0.024172</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>133.496</td>\n",
       "      <td>14.2314</td>\n",
       "      <td>public sector</td>\n",
       "      <td>22</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23.452079</td>\n",
       "      <td>rgb(234, 139, 147)</td>\n",
       "      <td>0.977283</td>\n",
       "      <td>0.051083</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>86.643</td>\n",
       "      <td>21.5466</td>\n",
       "      <td>public sector</td>\n",
       "      <td>8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.142136</td>\n",
       "      <td>rgb(197, 145, 160)</td>\n",
       "      <td>0.910077</td>\n",
       "      <td>0.005057</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>-23.5999</td>\n",
       "      <td>10.3618</td>\n",
       "      <td>supreme audit institutions</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.180340</td>\n",
       "      <td>rgb(109, 136, 10)</td>\n",
       "      <td>0.971488</td>\n",
       "      <td>0.011979</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>109.162</td>\n",
       "      <td>36.1417</td>\n",
       "      <td>performance information</td>\n",
       "      <td>15</td>\n",
       "      <td>5.0</td>\n",
       "      <td>19.364917</td>\n",
       "      <td>rgb(214, 157, 152)</td>\n",
       "      <td>0.981452</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>-2.41774</td>\n",
       "      <td>91.287</td>\n",
       "      <td>underwater wsns</td>\n",
       "      <td>7</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.228757</td>\n",
       "      <td>rgb(126, 200, 208)</td>\n",
       "      <td>0.995398</td>\n",
       "      <td>0.074038</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>-27.3919</td>\n",
       "      <td>-15.5137</td>\n",
       "      <td>pedestrian navigation</td>\n",
       "      <td>25</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>rgb(106, 116, 176)</td>\n",
       "      <td>0.968220</td>\n",
       "      <td>0.007089</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>-17.1335</td>\n",
       "      <td>-23.9913</td>\n",
       "      <td>pedestrians</td>\n",
       "      <td>12</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.320508</td>\n",
       "      <td>rgb(115, 109, 209)</td>\n",
       "      <td>0.981975</td>\n",
       "      <td>0.021809</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>269 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            x        y                                            kw size  \\\n",
       "0    -112.622 -51.2819  middle east respiratory syndrome coronavirus  102   \n",
       "1     23.2157  68.8196                      wireless sensor networks  435   \n",
       "2     33.8798  109.127                             novel coronavirus  111   \n",
       "3     76.0156 -17.5924                                      mers-cov   93   \n",
       "4      78.325 -121.269                             crystal structure   75   \n",
       "5     91.7094 -75.8707                                         wuhan   44   \n",
       "6     94.1134  33.2064                                        humans   23   \n",
       "7     31.6672  133.303                        2019 novel coronavirus   44   \n",
       "8    -57.9994  69.3501           underwater wireless sensor networks   35   \n",
       "9    -3.09047  -111.81                                      patients   63   \n",
       "10    9.79806 -54.2324                                      patients   45   \n",
       "11    53.0442  81.0132                                      patients   38   \n",
       "12    9.78586  115.466                                      covid-19   68   \n",
       "13   -20.7609 -14.6939                                     detection  341   \n",
       "14   -128.715 -26.0274                                     detection  308   \n",
       "15     49.864  38.2858                                     detection  224   \n",
       "16    39.7554 -10.4566                                          bats   48   \n",
       "17    43.8203 -21.7732              middle east respiratory syndrome  136   \n",
       "18    -106.72  38.3675                                       protein  168   \n",
       "19     1.0017  48.9669                                       protein  110   \n",
       "20   -14.9209 -89.9211                                          nov.   25   \n",
       "21   -112.803  69.9097                            crystal structures   30   \n",
       "22   -88.0328   68.683                              structural basis   19   \n",
       "23    72.9304  9.63638                            mers-cov infection   20   \n",
       "24    -25.097  78.5998           underwater acoustic sensor networks   31   \n",
       "25   -33.0915  56.0053                                    zika virus   38   \n",
       "26    79.2218  61.9953                             systematic review   55   \n",
       "27    8.42145 -32.1892                             systematic review   32   \n",
       "28    80.3787 -34.5894                                     infection  166   \n",
       "29   0.641837 -41.9627                                     infection   29   \n",
       "..        ...      ...                                           ...  ...   \n",
       "239 -0.374083  5.12656                                        fusion   85   \n",
       "240   41.6152  83.3581                                 pathogenicity    7   \n",
       "241   41.8266 -102.988                                 pathogenicity    3   \n",
       "242  -30.5015  35.7539                            toll-like receptor    7   \n",
       "243   17.1402  114.562                                    sequencing   70   \n",
       "244   38.3714  117.529                                      politics  134   \n",
       "245   95.0391 -26.5558                                      politics   41   \n",
       "246   75.6478 -13.0713                                      politics   38   \n",
       "247  -68.7191 -11.8057                                          cell  155   \n",
       "248    -42.36 -96.2237                              genome evolution    6   \n",
       "249  -8.16832  103.133                           indoor environments   18   \n",
       "250   21.7304 -27.8912                                       rodents    5   \n",
       "251   3.24872 -47.3212                                          mice   27   \n",
       "252  -98.2037 -63.3631                                          mice   14   \n",
       "253   54.6482 -75.1355                                    speciation    9   \n",
       "254   -108.31  6.33701                                     complexes   14   \n",
       "255  -27.8831  101.549                                     complexes   13   \n",
       "256  -148.422  28.5221                                     complexes   12   \n",
       "257  -107.598   58.961                                     complexes   11   \n",
       "258  -106.098 -1.11005                                     complexes    9   \n",
       "259   41.6955 -76.4265                                         genus   29   \n",
       "260  -43.0595  103.344                                         genus    8   \n",
       "261    60.793  17.9592                              job satisfaction   25   \n",
       "262   133.496  14.2314                                 public sector   22   \n",
       "263    86.643  21.5466                                 public sector    8   \n",
       "264  -23.5999  10.3618                    supreme audit institutions    5   \n",
       "265   109.162  36.1417                       performance information   15   \n",
       "266  -2.41774   91.287                               underwater wsns    7   \n",
       "267  -27.3919 -15.5137                         pedestrian navigation   25   \n",
       "268  -17.1335 -23.9913                                   pedestrians   12   \n",
       "\n",
       "     nc_count  circle_size                 rgb   max_sim  min_emb_dist  \\\n",
       "0        57.0    50.497525     rgb(39, 88, 78)  0.000000     36.789352   \n",
       "1        45.0   104.283268  rgb(146, 182, 221)  0.606799     10.445533   \n",
       "2        40.0    52.678269  rgb(155, 214, 100)  0.879144      5.463013   \n",
       "3        35.0    48.218254  rgb(188, 114, 118)  0.998738      0.427948   \n",
       "4        34.0    43.301270   rgb(190, 32, 127)  0.735524      4.407660   \n",
       "5        31.0    33.166248   rgb(201, 68, 100)  0.998520      0.359852   \n",
       "6        29.0    23.979158  rgb(202, 154, 107)  0.968882      0.830428   \n",
       "7        25.0    33.166248  rgb(153, 233, 120)  0.999674      0.305976   \n",
       "8        25.0    29.580399    rgb(82, 183, 34)  0.924585      0.171446   \n",
       "9        24.0    39.686270   rgb(126, 40, 184)  0.578510      4.303370   \n",
       "10       24.0    33.541020   rgb(136, 85, 174)  0.692708      0.843866   \n",
       "11       24.0    30.822070  rgb(170, 192, 111)  0.982306      0.852884   \n",
       "12       22.0    41.231056  rgb(136, 219, 111)  0.998771      0.135117   \n",
       "13       22.0    92.330927   rgb(112, 116, 43)  0.689144      0.057089   \n",
       "14       22.0    87.749644   rgb(26, 108, 185)  0.591937      1.472012   \n",
       "15       22.0    74.833148  rgb(167, 158, 195)  0.950529      0.945062   \n",
       "16       22.0    34.641016   rgb(159, 120, 54)  0.943157      1.300673   \n",
       "17       21.0    58.309519  rgb(163, 111, 122)  0.999662      0.337104   \n",
       "18       20.0    64.807407   rgb(44, 158, 104)  0.982954      1.120590   \n",
       "19       20.0    52.440442  rgb(129, 167, 100)  0.959406      0.245621   \n",
       "20       20.0    25.000000    rgb(116, 57, 38)  0.779219     12.334796   \n",
       "21       20.0    27.386128    rgb(39, 183, 73)  0.996659      0.137975   \n",
       "22       19.0    21.794495   rgb(59, 182, 109)  0.988462      0.316845   \n",
       "23       19.0    22.360680  rgb(186, 136, 118)  0.998422      0.004294   \n",
       "24       19.0    27.838822  rgb(108, 190, 216)  0.996747      0.030495   \n",
       "25       18.0    30.822070   rgb(102, 172, 86)  0.853774      1.963809   \n",
       "26       17.0    37.080992   rgb(191, 177, 25)  0.911842      0.349941   \n",
       "27       17.0    28.284271  rgb(135, 103, 154)  0.982906      0.386113   \n",
       "28       17.0    64.420494   rgb(192, 101, 93)  0.985767      0.558301   \n",
       "29       17.0    26.925824   rgb(129, 95, 118)  0.695506      2.434479   \n",
       "..        ...          ...                 ...       ...           ...   \n",
       "239       5.0    46.097722  rgb(128, 132, 206)  0.989754      0.106339   \n",
       "240       5.0    13.228757  rgb(161, 194, 177)  0.988270      0.172044   \n",
       "241       5.0     8.660254    rgb(161, 47, 90)  0.871837      0.019240   \n",
       "242       5.0    13.228757  rgb(104, 156, 117)  0.915672      0.005941   \n",
       "243       5.0    41.833001  rgb(142, 219, 175)  0.974527      0.022586   \n",
       "244       5.0    57.879185   rgb(158, 221, 54)  0.854494      0.014411   \n",
       "245       5.0    32.015621  rgb(203, 107, 220)  0.962211      0.008920   \n",
       "246       5.0    30.822070  rgb(188, 118, 207)  0.942851      0.083256   \n",
       "247       5.0    62.249498   rgb(74, 119, 158)  0.990388      0.034504   \n",
       "248       5.0    12.247449    rgb(95, 52, 108)  0.978395      0.021899   \n",
       "249       5.0    21.213203   rgb(122, 210, 46)  0.964984      0.165660   \n",
       "250       5.0    11.180340   rgb(145, 106, 46)  0.970087      0.015492   \n",
       "251       5.0    25.980762    rgb(131, 91, 96)  0.991966      0.028480   \n",
       "252       5.0    18.708287    rgb(51, 78, 182)  0.524699      0.443099   \n",
       "253       5.0    15.000000    rgb(171, 69, 71)  0.991925      0.046796   \n",
       "254       5.0    18.708287   rgb(43, 133, 185)  0.676159      0.029222   \n",
       "255       5.0    18.027756  rgb(106, 208, 173)  0.633787      1.573069   \n",
       "256       5.0    17.320508   rgb(11, 151, 110)  0.770216      0.063596   \n",
       "257       5.0    16.583124   rgb(43, 175, 121)  0.992044      0.019288   \n",
       "258       5.0    15.000000   rgb(44, 127, 165)  0.805343      0.015992   \n",
       "259       5.0    26.925824    rgb(161, 68, 47)  0.991352      0.017252   \n",
       "260       5.0    14.142136   rgb(94, 210, 182)  0.919267      0.001639   \n",
       "261       5.0    25.000000  rgb(176, 142, 163)  0.959600      0.024172   \n",
       "262       5.0    23.452079  rgb(234, 139, 147)  0.977283      0.051083   \n",
       "263       5.0    14.142136  rgb(197, 145, 160)  0.910077      0.005057   \n",
       "264       5.0    11.180340   rgb(109, 136, 10)  0.971488      0.011979   \n",
       "265       5.0    19.364917  rgb(214, 157, 152)  0.981452      0.000917   \n",
       "266       5.0    13.228757  rgb(126, 200, 208)  0.995398      0.074038   \n",
       "267       5.0    25.000000  rgb(106, 116, 176)  0.968220      0.007089   \n",
       "268       5.0    17.320508  rgb(115, 109, 209)  0.981975      0.021809   \n",
       "\n",
       "     no_kw_vecs  \n",
       "0           1.0  \n",
       "1           1.0  \n",
       "2           1.0  \n",
       "3           1.0  \n",
       "4           1.0  \n",
       "5           1.0  \n",
       "6           1.0  \n",
       "7           1.0  \n",
       "8           1.0  \n",
       "9           3.0  \n",
       "10          3.0  \n",
       "11          3.0  \n",
       "12          1.0  \n",
       "13          3.0  \n",
       "14          3.0  \n",
       "15          3.0  \n",
       "16          1.0  \n",
       "17          1.0  \n",
       "18          2.0  \n",
       "19          2.0  \n",
       "20          1.0  \n",
       "21          1.0  \n",
       "22          1.0  \n",
       "23          1.0  \n",
       "24          1.0  \n",
       "25          1.0  \n",
       "26          2.0  \n",
       "27          2.0  \n",
       "28          2.0  \n",
       "29          2.0  \n",
       "..          ...  \n",
       "239         2.0  \n",
       "240         2.0  \n",
       "241         2.0  \n",
       "242         1.0  \n",
       "243         1.0  \n",
       "244         3.0  \n",
       "245         3.0  \n",
       "246         3.0  \n",
       "247         1.0  \n",
       "248         1.0  \n",
       "249         1.0  \n",
       "250         1.0  \n",
       "251         2.0  \n",
       "252         2.0  \n",
       "253         1.0  \n",
       "254         5.0  \n",
       "255         5.0  \n",
       "256         5.0  \n",
       "257         5.0  \n",
       "258         5.0  \n",
       "259         2.0  \n",
       "260         2.0  \n",
       "261         1.0  \n",
       "262         2.0  \n",
       "263         2.0  \n",
       "264         1.0  \n",
       "265         1.0  \n",
       "266         1.0  \n",
       "267         1.0  \n",
       "268         1.0  \n",
       "\n",
       "[269 rows x 10 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kw = pd.DataFrame(index=range(len(term_plot_tuples)), columns=[\"x\", \"y\", \"kw\", \"size\"])\n",
    "\n",
    "for i, (kw, size, vec) in enumerate(term_plot_tuples):\n",
    "    df_kw.at[i, \"x\"] = C_kw_emb[i, 0]\n",
    "    df_kw.at[i, \"y\"] = C_kw_emb[i, 1]\n",
    "    df_kw.at[i, \"size\"] = size\n",
    "    df_kw.at[i, \"nc_count\"] = nc_counts[kw]\n",
    "    df_kw.at[i, \"circle_size\"] = np.sqrt(size) * 5\n",
    "    #df_kw.at[i, \"circle_size\"] = np.sqrt(nc_counts[kw]) * 10\n",
    "    df_kw.at[i, \"kw\"] = kw\n",
    "    df_kw.at[i, \"rgb\"] = RGB(colors_kw[i,0], colors_kw[i,1], colors_kw[i,2])\n",
    "    if i >= 1:\n",
    "        df_kw.at[i, \"max_sim\"] = X_kw_sims[i,:i].max()\n",
    "    else:\n",
    "        df_kw.at[i, \"max_sim\"] = 0.\n",
    "    if i >= 1:\n",
    "        df_kw.at[i, \"min_emb_dist\"] = X_kw_emb_dists[i,:i].min()\n",
    "    else:\n",
    "        df_kw.at[i, \"min_emb_dist\"] = X_kw_emb_dists.max()\n",
    "    df_kw.at[i, \"no_kw_vecs\"] = len(kw2vecs[kw])\n",
    "    \n",
    "df_kw\n",
    "    \n",
    "                             "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>kw</th>\n",
       "      <th>size</th>\n",
       "      <th>nc_count</th>\n",
       "      <th>circle_size</th>\n",
       "      <th>rgb</th>\n",
       "      <th>max_sim</th>\n",
       "      <th>min_emb_dist</th>\n",
       "      <th>no_kw_vecs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [x, y, kw, size, nc_count, circle_size, rgb, max_sim, min_emb_dist, no_kw_vecs]\n",
       "Index: []"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kw[df_kw[\"kw\"]==\"energy harvesting\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>kw</th>\n",
       "      <th>size</th>\n",
       "      <th>nc_count</th>\n",
       "      <th>circle_size</th>\n",
       "      <th>rgb</th>\n",
       "      <th>max_sim</th>\n",
       "      <th>min_emb_dist</th>\n",
       "      <th>no_kw_vecs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>-63.8386</td>\n",
       "      <td>60.0347</td>\n",
       "      <td>coronavirus</td>\n",
       "      <td>482</td>\n",
       "      <td>11.0</td>\n",
       "      <td>109.772492</td>\n",
       "      <td>rgb(78, 176, 200)</td>\n",
       "      <td>0.994792</td>\n",
       "      <td>0.081993</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          x        y           kw size  nc_count  circle_size  \\\n",
       "56 -63.8386  60.0347  coronavirus  482      11.0   109.772492   \n",
       "\n",
       "                  rgb   max_sim  min_emb_dist  no_kw_vecs  \n",
       "56  rgb(78, 176, 200)  0.994792      0.081993         1.0  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kw[df_kw[\"kw\"]==\"coronavirus\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "reset_output()\n",
    "\n",
    "\n",
    "#class hover(HoverTool):\n",
    "#    hover.tooltips = [\n",
    "#    (\"index\", \"$index\"),\n",
    "\n",
    "\n",
    "hover_kw = HoverTool()\n",
    "hover_kw.tooltips = [\n",
    "    (\"Cluster\", \"@kw\"),\n",
    "    (\"Size\", \"@size\"),\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "df_to_show = df_kw[df_kw[\"no_kw_vecs\"] == 1]\n",
    "df_to_show = df_to_show[df_to_show[\"max_sim\"] <= 0.975]\n",
    "s_kw = ColumnDataSource(data=df_to_show)\n",
    "\n",
    "#s_kw = ColumnDataSource(data=df_kw)\n",
    "#s_kw = ColumnDataSource(data=df_kw[df_kw[\"max_sim\"] <= 0.95])\n",
    "#s_kw = ColumnDataSource(data=df_kw[df_kw[\"min_emb_dist\"] >= 1])\n",
    "\n",
    "#df_kw_filtered = df_kw[df_kw[\"nc_count\"] >= 0]\n",
    "#df_kw_filtered = df_kw_filtered[df_kw_filtered[\"size\"] >= 50]\n",
    "#df_kw_filtered = df_kw_filtered[df_kw_filtered[\"min_emb_dist\"] >= 2]\n",
    "df_kw_labels = df_to_show\n",
    "\n",
    "\n",
    "\n",
    "#df_kw_labels = df_kw_labels[df_kw_labels[\"min_emb_dist\"] >= 2]\n",
    "df_kw_labels = df_kw_labels[df_kw_labels[\"size\"] >= 50]\n",
    "\n",
    "s_kw_labels = ColumnDataSource(data=df_kw_labels)\n",
    "\n",
    "\n",
    "callback = CustomJS(code=\"\"\"\n",
    "// the event that triggered the callback is cb_obj:\n",
    "// The event type determines the relevant attributes\n",
    "console.log('Tap event occured at x-position: ' + cb_obj.x)\n",
    "\"\"\")\n",
    "\n",
    "\n",
    "def lasso_cb(x):\n",
    "    print(x)\n",
    "\n",
    "#lasso = LassoSelectTool(callback=lasso_cb)\n",
    "lasso = LassoSelectTool()\n",
    "\n",
    "\n",
    "p_kw = figure(plot_width=1600, plot_height=1000, active_scroll = \"wheel_zoom\")\n",
    "#p = figure(plot_width=1200, plot_height=1000, output_backend=\"webgl\", active_scroll = \"wheel_zoom\")\n",
    "#p.circle(x='x', y='y', size=8, color=\"rgb\", alpha=\"alpha\", source=s1)\n",
    "p_kw.circle(x='x', y='y', size=\"circle_size\", color=\"rgb\", alpha=0.75, source=s_kw, name=\"kw\")\n",
    "\n",
    "labels = LabelSet(x='x', y='y', text='kw', level='glyph', text_align=\"left\",\n",
    "                  x_offset=5, y_offset=5, \n",
    "                  source=s_kw_labels, render_mode='canvas')\n",
    "\n",
    "\n",
    "p_kw.add_tools(hover_kw)\n",
    "p_kw.add_tools(lasso)\n",
    "\n",
    "p_kw.js_on_event('tap', callback)\n",
    "\n",
    "#p.toolbar.active_scroll = \"wheel_zoom\"\n",
    "\n",
    "#df[\"xval\"].max()\n",
    "\n",
    "left, right, bottom, top = -120, 120, -90, 90\n",
    "left, right, bottom, top = -80, 80, -60, 60\n",
    "left, right, bottom, top = -120, 120, -90, 90\n",
    "p_kw.x_range=Range1d(left, right)\n",
    "p_kw.y_range=Range1d(bottom, top)\n",
    "\n",
    "\n",
    "#layout = row(p, widgetbox(data_table))\n",
    "#layout = column(p, widgetbox(data_table))\n",
    "\n",
    "\n",
    "p_kw.add_layout(labels)\n",
    "\n",
    "show(p_kw)\n",
    "#save(layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "kw2vecs = {}\n",
    "\n",
    "for kw, c, vec in term_plot_tuples:\n",
    "    if not kw in kw2vecs:\n",
    "        kw2vecs[kw] = []\n",
    "    kw2vecs[kw] = kw2vecs[kw] + [vec]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['(covid-19',\n",
       " '(covid‐19',\n",
       " '(nematoda',\n",
       " '2019 novel coronavirus',\n",
       " 'accountability',\n",
       " 'accuracy',\n",
       " 'acute respiratory distress syndrome',\n",
       " 'adults',\n",
       " 'antibodies',\n",
       " 'antibody-dependent enhancement',\n",
       " 'arrival',\n",
       " 'bacteria',\n",
       " 'bats',\n",
       " 'binding',\n",
       " 'biogeography',\n",
       " 'biology',\n",
       " 'burnout',\n",
       " 'cancer',\n",
       " 'cell',\n",
       " 'cell entry',\n",
       " 'cells',\n",
       " 'children',\n",
       " 'clinical characteristics',\n",
       " 'clinical features',\n",
       " 'comparative genomics',\n",
       " 'complete genome sequence',\n",
       " 'complexes',\n",
       " 'coronavirus',\n",
       " 'coronavirus disease',\n",
       " 'coronaviruses',\n",
       " 'covid-19',\n",
       " 'covid‐19',\n",
       " 'critically ill patients',\n",
       " 'crystal structure',\n",
       " 'crystal structures',\n",
       " 'detection',\n",
       " 'diagnosis',\n",
       " 'disease',\n",
       " 'diversification',\n",
       " 'diversity',\n",
       " 'doa estimation',\n",
       " 'draft genome sequence',\n",
       " 'dromedary camels',\n",
       " 'drug discovery',\n",
       " 'ebola',\n",
       " 'ebola virus',\n",
       " 'ebola virus disease',\n",
       " 'epidemiology',\n",
       " 'escherichia coli',\n",
       " 'estimation',\n",
       " 'eukaryotes',\n",
       " 'europe',\n",
       " 'evolutionary history',\n",
       " 'fusion',\n",
       " 'genes',\n",
       " 'genetic diversity',\n",
       " 'genetics',\n",
       " 'genome',\n",
       " 'genome evolution',\n",
       " 'genomics',\n",
       " 'genus',\n",
       " 'harmonia axyridis',\n",
       " 'healthcare workers',\n",
       " 'hepatitis e virus',\n",
       " 'hereditary angioedema',\n",
       " 'host',\n",
       " 'human coronavirus',\n",
       " 'humans',\n",
       " 'immunity',\n",
       " 'inactivation',\n",
       " 'indoor environments',\n",
       " 'indoor localization',\n",
       " 'indoor positioning',\n",
       " 'inertial sensors',\n",
       " 'infection',\n",
       " 'influenza',\n",
       " 'inhibition',\n",
       " 'inhibitors',\n",
       " 'internet',\n",
       " 'iot',\n",
       " 'job satisfaction',\n",
       " 'localization',\n",
       " 'machine learning',\n",
       " 'macromolecular crystallography',\n",
       " 'mers coronavirus',\n",
       " 'mers-cov',\n",
       " 'mers-cov infection',\n",
       " 'meta-analysis',\n",
       " 'metagenomes',\n",
       " 'mhc class',\n",
       " 'mice',\n",
       " 'middle east respiratory syndrome',\n",
       " 'middle east respiratory syndrome coronavirus',\n",
       " 'middle east respiratory syndrome coronavirus infection',\n",
       " 'molecular replacement',\n",
       " 'mortality',\n",
       " 'mutation',\n",
       " 'mycobacterium tuberculosis',\n",
       " 'navigation',\n",
       " 'new insights',\n",
       " 'nov.',\n",
       " 'novel coronavirus',\n",
       " 'novel coronavirus disease',\n",
       " 'outbreak',\n",
       " 'pathogenesis',\n",
       " 'pathogenicity',\n",
       " 'patients',\n",
       " 'pedestrian navigation',\n",
       " 'pedestrians',\n",
       " 'performance analysis',\n",
       " 'performance information',\n",
       " 'performance management',\n",
       " 'phylogenetic analysis',\n",
       " 'phylogeny',\n",
       " 'phylogeography',\n",
       " 'platelet',\n",
       " 'pneumonia',\n",
       " 'politics',\n",
       " 'porcine deltacoronavirus',\n",
       " 'prediction',\n",
       " 'pregnancy',\n",
       " 'protein',\n",
       " 'protein-protein interactions',\n",
       " 'proteins',\n",
       " 'public sector',\n",
       " 'recognition',\n",
       " 'refinement',\n",
       " 'replication',\n",
       " 'republic',\n",
       " 'ribavirin',\n",
       " 'rodents',\n",
       " 'salmonella enterica',\n",
       " 'sars',\n",
       " 'sars coronavirus',\n",
       " 'sars-cov-2',\n",
       " 'sars‐cov‐2',\n",
       " 'saudi arabia',\n",
       " 'schizophrenia',\n",
       " 'sequencing',\n",
       " 'severe acute respiratory syndrome coronavirus',\n",
       " 'smartphone sensors',\n",
       " 'smartphones',\n",
       " 'sofosbuvir',\n",
       " 'speciation',\n",
       " 'species',\n",
       " 'spike protein',\n",
       " 'strains',\n",
       " 'structural basis',\n",
       " 'structural insights',\n",
       " 'structural studies',\n",
       " 'structure determination',\n",
       " 'supreme audit institutions',\n",
       " 'systematic review',\n",
       " 'target localization',\n",
       " 'taxonomy',\n",
       " 'toll-like receptor',\n",
       " 'tracking',\n",
       " 'transmission',\n",
       " 'ultraviolet c light',\n",
       " 'underwater acoustic sensor networks',\n",
       " 'underwater sensor networks',\n",
       " 'underwater wireless sensor networks',\n",
       " 'underwater wsns',\n",
       " 'vaccine development',\n",
       " 'vaccines',\n",
       " 'virtual screening',\n",
       " 'virulence',\n",
       " 'virus',\n",
       " 'viruses',\n",
       " 'web server',\n",
       " 'west africa',\n",
       " 'whole-genome sequencing',\n",
       " 'wireless body area networks',\n",
       " 'wireless sensor networks',\n",
       " 'wsns',\n",
       " 'wuhan',\n",
       " 'zika virus',\n",
       " 'zymoseptoria tritici']"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(kw2vecs.keys())[-300:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "\n",
      "#############################\n",
      "####### Cluster 0 ###########\n",
      "#############################\n",
      "\n",
      "coronavirus [1.]\n",
      "virus [0.99499733]\n",
      "sars [0.99479168]\n",
      "coronaviruses [0.99236687]\n",
      "human coronavirus [0.9923146]\n",
      "infection [0.99165922]\n",
      "sars coronavirus [0.99046967]\n",
      "pathogenesis [0.98870798]\n",
      "epidemiology [0.98453255]\n",
      "vaccines [0.98288428]\n",
      "vaccine development [0.98257089]\n",
      "viruses [0.98232071]\n",
      "disease [0.98209695]\n",
      "binding [0.98146418]\n",
      "antibodies [0.98036373]\n",
      "severe acute respiratory syndrome coronavirus [0.97864354]\n",
      "patients [0.97827768]\n",
      "transmission [0.97734785]\n",
      "sars‐cov‐2 [0.97685876]\n",
      "inhibition [0.97658788]\n",
      "spike protein [0.9756481]\n",
      "humans [0.9755714]\n",
      "mers coronavirus [0.97487216]\n",
      "sars-cov-2 [0.9748482]\n",
      "protein [0.97328414]\n",
      "proteins [0.97144912]\n",
      "middle east respiratory syndrome [0.97131019]\n",
      "middle east respiratory syndrome coronavirus [0.97128649]\n",
      "pneumonia [0.97107573]\n",
      "outbreak [0.97058213]\n",
      "immunity [0.97055667]\n",
      "pregnancy [0.96829172]\n",
      "host [0.96798898]\n",
      "mers-cov [0.96724357]\n",
      "cell entry [0.96436728]\n",
      "wuhan [0.96115683]\n",
      "middle east respiratory syndrome coronavirus infection [0.95931338]\n",
      "covid-19 [0.95891862]\n",
      "mers-cov infection [0.95581777]\n",
      "novel coronavirus [0.95454562]\n",
      "new insights [0.95252787]\n",
      "dromedary camels [0.95243485]\n",
      "2019 novel coronavirus [0.95171088]\n",
      "genetic diversity [0.95088834]\n",
      "covid‐19 [0.95066764]\n",
      "saudi arabia [0.95062547]\n",
      "clinical features [0.94840707]\n",
      "inhibitors [0.94815809]\n",
      "republic [0.94743607]\n",
      "(covid‐19 [0.94548452]\n",
      "coronavirus disease [0.94530529]\n",
      "novel coronavirus disease [0.94264519]\n",
      "(covid-19 [0.94101328]\n",
      "clinical characteristics [0.9372502]\n",
      "virulence [0.93214205]\n",
      "replication [0.93181623]\n",
      "porcine deltacoronavirus [0.91294584]\n",
      "ribavirin [0.90505985]\n",
      "healthcare workers [0.90481811]\n",
      "bats [0.90481803]\n",
      "influenza [0.90032072]\n",
      "ebola [0.89422636]\n",
      "ebola virus [0.89056921]\n",
      "ebola virus disease [0.87727277]\n",
      "\n",
      "[0.98981502] 0\n",
      "<CrossRefWork Gorse, Donovan, Patel (2020): Antibodies to coronaviruses are higher in older compared with younger adults and binding antibodies are more sensitive than neutralizing antibodies in identifying coronavirus‐associated illnesses. http://dx.doi.org/10.1002/jmv.25715 >\n",
      "\n",
      "[0.98730481] 1\n",
      "<CrossRefWork Weston, Frieman (2020): COVID-19: Knowns, Unknowns, and Questions. http://dx.doi.org/10.1128/msphere.00203-20 >\n",
      "\n",
      "[0.98510198] 2\n",
      "<CrossRefWork Jiang, Shi (2020): The First Disease X is Caused by a Highly Transmissible Acute Respiratory Syndrome Coronavirus. http://dx.doi.org/10.1007/s12250-020-00206-5 >\n",
      "\n",
      "[0.98479459] 3\n",
      "<CrossRefWork She, Jiang, Ye, Hu, Bai, Song (2020): 2019 novel coronavirus of pneumonia in Wuhan, China: emerging attack and management strategies. http://dx.doi.org/10.1186/s40169-020-00271-z >\n",
      "\n",
      "[0.98388995] 4\n",
      "<CrossRefWork Xu, Zhao, Teng, Abdalla, Zhu, Xie, Wang, Guo (2020): Systematic Comparison of Two Animal-to-Human Transmitted Human Coronaviruses: SARS-CoV-2 and SARS-CoV. http://dx.doi.org/10.3390/v12020244 >\n",
      "\n",
      "[0.98375729] 5\n",
      "<CrossRefWork Huang, Herrmann (2020): Fast assessment of human receptor-binding capability of 2019 novel coronavirus (2019-nCoV). http://dx.doi.org/10.1101/2020.02.01.930537 >\n",
      "\n",
      "[0.98316143] 6\n",
      "<CrossRefWork Liu, Xiao, Wei, Li, Yang, Tan, Zhu, Zhang, Wu, Liu (2020): Composition and divergence of coronavirus spike proteins and host ACE2 receptors predict potential intermediate hosts of SARS‐CoV‐2. http://dx.doi.org/10.1002/jmv.25726 >\n",
      "\n",
      "[0.98208745] 7\n",
      "<CrossRefWork Plotkin (2020): The New Coronavirus, the Current King of China. http://dx.doi.org/10.1093/jpids/piaa018 >\n",
      "\n",
      "[0.98200952] 8\n",
      "<CrossRefWork Channappanavar, Perlman (2017): Pathogenic human coronavirus infections: causes and consequences of cytokine storm and immunopathology. http://dx.doi.org/10.1007/s00281-017-0629-x >\n",
      "\n",
      "[0.98108648] 9\n",
      "<CrossRefWork Kruse (2020): Therapeutic strategies in an outbreak scenario to treat the novel coronavirus originating in Wuhan, China. http://dx.doi.org/10.12688/f1000research.22211.2 >\n",
      "\n",
      "[0.98098568] 10\n",
      "<CrossRefWork Kruse (2020): Therapeutic strategies in an outbreak scenario to treat the novel coronavirus originating in Wuhan, China. http://dx.doi.org/10.12688/f1000research.22211.1 >\n",
      "\n",
      "[0.98045691] 11\n",
      "<CrossRefWork Wang, Shang, Jiang, Du (2020): Subunit Vaccines Against Emerging Pathogenic Human Coronaviruses. http://dx.doi.org/10.3389/fmicb.2020.00298 >\n",
      "\n",
      "[0.98029098] 12\n",
      "<CrossRefWork Willman, Kobasa, Kindrachuk (2019): A Comparative Analysis of Factors Influencing Two Outbreaks of Middle Eastern Respiratory Syndrome (MERS) in Saudi Arabia and South Korea. http://dx.doi.org/10.3390/v11121119 >\n",
      "\n",
      "[0.97977839] 13\n",
      "<CrossRefWork Muth, Corman, Roth, Binger, Dijkman, Gottula, Gloza-Rausch, Balboni, Battilani, Rihtarič, Toplak, Ameneiros, Pfeifer, Thiel, Drexler, Müller, Drosten (2018): Attenuation of replication by a 29 nucleotide deletion in SARS-coronavirus acquired during the early stages of human-to-human transmission. http://dx.doi.org/10.1038/s41598-018-33487-8 >\n",
      "\n",
      "[0.97930715] 14\n",
      "<CrossRefWork Lippi, Plebani (2020): The novel coronavirus (2019-nCoV) outbreak: think the unthinkable and be prepared to face the challenge. http://dx.doi.org/10.1515/dx-2020-0015 >\n",
      "\n",
      "[0.97897281] 15\n",
      "<CrossRefWork Seah, Agrawal (2020): Can the Coronavirus Disease 2019 (COVID-19) Affect the Eyes? A Review of Coronaviruses and Ocular Implications in Humans and Animals. http://dx.doi.org/10.1080/09273948.2020.1738501 >\n",
      "\n",
      "[0.97888198] 16\n",
      "<CrossRefWork Schwartz, Graham (2020): Potential Maternal and Infant Outcomes from Coronavirus 2019-nCoV (SARS-CoV-2) Infecting Pregnant Women: Lessons from SARS, MERS, and Other Human Coronavirus Infections. http://dx.doi.org/10.3390/v12020194 >\n",
      "\n",
      "[0.97870757] 17\n",
      "<CrossRefWork Xia, Yan, Xu, Agrawal, Algaissi, Tseng, Wang, Du, Tan, Wilson, Jiang, Yang, Lu (2019): A pan-coronavirus fusion inhibitor targeting the HR1 domain of human coronavirus spike. http://dx.doi.org/10.1126/sciadv.aav4580 >\n",
      "\n",
      "[0.97834384] 18\n",
      "<CrossRefWork Wassenaar, Zou (2020): 2019_nCoV/SARS‐CoV‐2: rapid classification of betacoronaviruses and identification of Traditional Chinese Medicine as potential origin of zoonotic coronaviruses. http://dx.doi.org/10.1111/lam.13285 >\n",
      "\n",
      "[0.9770198] 19\n",
      "<CrossRefWork Yao, Qian, Zhu, Wang, Wang (2020): A systematic review of lopinavir therapy for SARS coronavirus and MERS coronavirus—A possible reference for coronavirus disease‐19 treatment option. http://dx.doi.org/10.1002/jmv.25729 >\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "kw = \"coronavirus\"\n",
    "\n",
    "print(len(kw2vecs[kw]))\n",
    "print()\n",
    "\n",
    "for j, vec in enumerate(kw2vecs[kw]):\n",
    "    \n",
    "    print(\"#############################\")\n",
    "    print(\"####### Cluster {} ###########\".format(j))\n",
    "    print(\"#############################\")\n",
    "    print()\n",
    "    \n",
    "    kw_sims = cosine_similarity(X_kw, vec)\n",
    "    for i in np.argsort(kw_sims.flat)[::-1]:\n",
    "        if kw_sims[i] >= 0.85:\n",
    "            print(term_plot_tuples[i][0], kw_sims[i])\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    all_sims = cosine_similarity(X, vec)\n",
    "    \n",
    "    n = 0\n",
    "    for i in np.argsort(all_sims.flat)[::-1]:\n",
    "        if all_sims[i] >= 0.7:\n",
    "            doi = index2doi[i]\n",
    "            print(all_sims[i], n)\n",
    "            print(doi2str[doi])\n",
    "            print()\n",
    "        n += 1\n",
    "        if n >= 20:\n",
    "            break\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.argsort(all_sims.flat)[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_sims[2155]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
